<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>6 高通量数据中批次效应的鉴定和处理 | R Cytoscape AI生信作图分析教程</title>
  <meta name="description" content="6 高通量数据中批次效应的鉴定和处理 | R Cytoscape AI生信作图分析教程" />
  <meta name="generator" content="bookdown 0.45 and GitBook 2.6.7" />

  <meta property="og:title" content="6 高通量数据中批次效应的鉴定和处理 | R Cytoscape AI生信作图分析教程" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="6 高通量数据中批次效应的鉴定和处理 | R Cytoscape AI生信作图分析教程" />
  
  
  

<meta name="author" content="生信宝典 陈同" />
<meta name="author" content="chentong_biology@163.com" />


<meta name="date" content="2025-11-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="AI.html"/>
<link rel="next" href="questions06.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="http://www.bic.ac.cn">Hello</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Hello</a></li>
<li class="chapter" data-level="1" data-path="questions.html"><a href="questions.html"><i class="fa fa-check"></i><b>1</b> 考题</a>
<ul>
<li class="chapter" data-level="1.1" data-path="questions.html"><a href="questions.html#%E7%90%86%E8%AE%BA%E9%A2%98-50%E5%88%86"><i class="fa fa-check"></i><b>1.1</b> 理论题 (50分)</a></li>
<li class="chapter" data-level="1.2" data-path="questions.html"><a href="questions.html#%E4%BB%A3%E7%A0%81%E9%A2%98-70%E5%88%86"><i class="fa fa-check"></i><b>1.2</b> 代码题 (70分)</a></li>
<li class="chapter" data-level="1.3" data-path="questions.html"><a href="questions.html#%E9%94%99%E8%AF%AF%E8%AF%86%E5%88%AB%E9%A2%98-%E9%80%89%E7%AD%94"><i class="fa fa-check"></i><b>1.3</b> 错误识别题 (选答)</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="Rbasic.html"><a href="Rbasic.html"><i class="fa fa-check"></i><b>2</b> R基础</a>
<ul>
<li class="chapter" data-level="2.1" data-path="Rbasic.html"><a href="Rbasic.html#R_install"><i class="fa fa-check"></i><b>2.1</b> R安装</a></li>
<li class="chapter" data-level="2.2" data-path="Rbasic.html"><a href="Rbasic.html#R_basic"><i class="fa fa-check"></i><b>2.2</b> Rstudio基础</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="Rbasic.html"><a href="Rbasic.html#R_version"><i class="fa fa-check"></i><b>2.2.1</b> Rstudio版本</a></li>
<li class="chapter" data-level="2.2.2" data-path="Rbasic.html"><a href="Rbasic.html#Rstudio_install"><i class="fa fa-check"></i><b>2.2.2</b> Rstudio安装</a></li>
<li class="chapter" data-level="2.2.3" data-path="Rbasic.html"><a href="Rbasic.html#Rstudio_usage"><i class="fa fa-check"></i><b>2.2.3</b> Rstudio 使用</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="Rbasic.html"><a href="Rbasic.html#R_basic_formula"><i class="fa fa-check"></i><b>2.3</b> R基本语法</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="Rbasic.html"><a href="Rbasic.html#R_help"><i class="fa fa-check"></i><b>2.3.1</b> 获取帮助文档，查看命令或函数的使用方法、事例或适用范围</a></li>
<li class="chapter" data-level="2.3.2" data-path="Rbasic.html"><a href="Rbasic.html#R_variable"><i class="fa fa-check"></i><b>2.3.2</b> R中的变量及其初始化</a></li>
<li class="chapter" data-level="2.3.3" data-path="Rbasic.html"><a href="Rbasic.html#vairableType"><i class="fa fa-check"></i><b>2.3.3</b> 变量类型和转换</a></li>
<li class="chapter" data-level="2.3.4" data-path="Rbasic.html"><a href="Rbasic.html#Rmatrix"><i class="fa fa-check"></i><b>2.3.4</b> R中矩阵运算</a></li>
<li class="chapter" data-level="2.3.5" data-path="Rbasic.html"><a href="Rbasic.html#Rmatrixcombine"><i class="fa fa-check"></i><b>2.3.5</b> R中矩阵筛选合并</a></li>
<li class="chapter" data-level="2.3.6" data-path="Rbasic.html"><a href="Rbasic.html#str"><i class="fa fa-check"></i><b>2.3.6</b> <code>str</code>的应用</a></li>
<li class="chapter" data-level="2.3.7" data-path="Rbasic.html"><a href="Rbasic.html#Rpackage"><i class="fa fa-check"></i><b>2.3.7</b> R的包管理</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="Rbasic.html"><a href="Rbasic.html#checkinstalltpackages"><i class="fa fa-check"></i><b>2.4</b> 检查并安装本教程需要的 R 包</a></li>
<li class="chapter" data-level="2.5" data-path="Rbasic.html"><a href="Rbasic.html#ggplot2_plot"><i class="fa fa-check"></i><b>2.5</b> ggplot2绘图</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="Rbasic.html"><a href="Rbasic.html#matrixFormat"><i class="fa fa-check"></i><b>2.5.1</b> 数据格式转换和字符串处理</a></li>
<li class="chapter" data-level="2.5.2" data-path="Rbasic.html"><a href="Rbasic.html#RColorBrewer"><i class="fa fa-check"></i><b>2.5.2</b> 配色</a></li>
<li class="chapter" data-level="2.5.3" data-path="Rbasic.html"><a href="Rbasic.html#phetamap"><i class="fa fa-check"></i><b>2.5.3</b> pheatmap绘制热图</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="Rbasic.html"><a href="Rbasic.html#cheetsheets"><i class="fa fa-check"></i><b>2.6</b> CheetSheets</a></li>
<li class="chapter" data-level="2.7" data-path="Rbasic.html"><a href="Rbasic.html#%E5%8F%82%E8%80%83-ref01"><i class="fa fa-check"></i><b>2.7</b> 参考 {ref01}</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="Rplots.html"><a href="Rplots.html"><i class="fa fa-check"></i><b>3</b> R plots</a>
<ul>
<li class="chapter" data-level="3.1" data-path="Rplots.html"><a href="Rplots.html#plot"><i class="fa fa-check"></i><b>3.1</b> qplot绘制图形 (王绪宁)</a></li>
<li class="chapter" data-level="3.2" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmap"><i class="fa fa-check"></i><b>3.2</b> 热图绘制</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmap_data"><i class="fa fa-check"></i><b>3.2.1</b> 生成测试数据</a></li>
<li class="chapter" data-level="3.2.2" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmap_dataformat"><i class="fa fa-check"></i><b>3.2.2</b> 转换数据格式</a></li>
<li class="chapter" data-level="3.2.3" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmap_split"><i class="fa fa-check"></i><b>3.2.3</b> 分解绘图</a></li>
<li class="chapter" data-level="3.2.4" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmapsave"><i class="fa fa-check"></i><b>3.2.4</b> 图形存储</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmapmeihua"><i class="fa fa-check"></i><b>3.3</b> 热图美化</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmap_log"><i class="fa fa-check"></i><b>3.3.1</b> 对数转换</a></li>
<li class="chapter" data-level="3.3.2" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmapzscore"><i class="fa fa-check"></i><b>3.3.2</b> Z-score转换</a></li>
<li class="chapter" data-level="3.3.3" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmapoutlier"><i class="fa fa-check"></i><b>3.3.3</b> 抹去异常值</a></li>
<li class="chapter" data-level="3.3.4" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmapnonlinear"><i class="fa fa-check"></i><b>3.3.4</b> 非线性颜色</a></li>
<li class="chapter" data-level="3.3.5" data-path="Rplots.html"><a href="Rplots.html#ggplot2_heatmaporder"><i class="fa fa-check"></i><b>3.3.5</b> 调整行或列的顺序</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="Rplots.html"><a href="Rplots.html#pheatmap_simple"><i class="fa fa-check"></i><b>3.4</b> 热图绘制 - pheatmap</a></li>
<li class="chapter" data-level="3.5" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder"><i class="fa fa-check"></i><b>3.5</b> 聚类热图如何按自己的意愿调整分支顺序？</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorderdata"><i class="fa fa-check"></i><b>3.5.1</b> 数据示例</a></li>
<li class="chapter" data-level="3.5.2" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder1"><i class="fa fa-check"></i><b>3.5.2</b> 绘制一个聚类热图很简单</a></li>
<li class="chapter" data-level="3.5.3" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder2"><i class="fa fa-check"></i><b>3.5.3</b> 如何自定义分支顺序呢</a></li>
<li class="chapter" data-level="3.5.4" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder3"><i class="fa fa-check"></i><b>3.5.4</b> 人为指定顺序排序样品</a></li>
<li class="chapter" data-level="3.5.5" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder4"><i class="fa fa-check"></i><b>3.5.5</b> 按某个基因的表达由小到大排序</a></li>
<li class="chapter" data-level="3.5.6" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder5"><i class="fa fa-check"></i><b>3.5.6</b> 按某个基因的表达由大到小排序</a></li>
<li class="chapter" data-level="3.5.7" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder6"><i class="fa fa-check"></i><b>3.5.7</b> 按分支名字（样品名字）的字母顺序排序</a></li>
<li class="chapter" data-level="3.5.8" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder7"><i class="fa fa-check"></i><b>3.5.8</b> 梯子形排序：最小的分支在右侧</a></li>
<li class="chapter" data-level="3.5.9" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder8"><i class="fa fa-check"></i><b>3.5.9</b> 梯子形排序：最小的分支在左侧</a></li>
<li class="chapter" data-level="3.5.10" data-path="Rplots.html"><a href="Rplots.html#pheatmap_branchorder9"><i class="fa fa-check"></i><b>3.5.10</b> 按特征值排序</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="Rplots.html"><a href="Rplots.html#ggplot2boxplot"><i class="fa fa-check"></i><b>3.6</b> 箱线图</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="Rplots.html"><a href="Rplots.html#ggplot2boxplot1"><i class="fa fa-check"></i><b>3.6.1</b> 一步步解析箱线图绘制</a></li>
<li class="chapter" data-level="3.6.2" data-path="Rplots.html"><a href="Rplots.html#ggplot2boxplot2"><i class="fa fa-check"></i><b>3.6.2</b> 绘制单个基因 (A)的箱线图</a></li>
<li class="chapter" data-level="3.6.3" data-path="Rplots.html"><a href="Rplots.html#ggplot2boxplot3"><i class="fa fa-check"></i><b>3.6.3</b> 长矩阵绘制箱线图</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="Rplots.html"><a href="Rplots.html#ggplot2lineplot"><i class="fa fa-check"></i><b>3.7</b> 线图</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="Rplots.html"><a href="Rplots.html#ggplot2singleline"><i class="fa fa-check"></i><b>3.7.1</b> 单线图</a></li>
<li class="chapter" data-level="3.7.2" data-path="Rplots.html"><a href="Rplots.html#ggplot2multipleline"><i class="fa fa-check"></i><b>3.7.2</b> 多线图</a></li>
<li class="chapter" data-level="3.7.3" data-path="Rplots.html"><a href="Rplots.html#ggplot2xaxistext"><i class="fa fa-check"></i><b>3.7.3</b> 横轴文本线图</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="Rplots.html"><a href="Rplots.html#ggplot2scatter"><i class="fa fa-check"></i><b>3.8</b> 散点图</a>
<ul>
<li class="chapter" data-level="3.8.1" data-path="Rplots.html"><a href="Rplots.html#ggplot2scatter1"><i class="fa fa-check"></i><b>3.8.1</b> 横纵轴都为数字的散点图解析</a></li>
<li class="chapter" data-level="3.8.2" data-path="Rplots.html"><a href="Rplots.html#ggplot2scatter3"><i class="fa fa-check"></i><b>3.8.2</b> 横纵轴都为字符串的散点图展示</a></li>
</ul></li>
<li class="chapter" data-level="3.9" data-path="Rplots.html"><a href="Rplots.html#ggplot2enrichment"><i class="fa fa-check"></i><b>3.9</b> 功能富集泡泡图</a>
<ul>
<li class="chapter" data-level="3.9.1" data-path="Rplots.html"><a href="Rplots.html#ggplot2enric2hment"><i class="fa fa-check"></i><b>3.9.1</b> 单样品分开绘制</a></li>
<li class="chapter" data-level="3.9.2" data-path="Rplots.html"><a href="Rplots.html#ggplot2enrichmen2t"><i class="fa fa-check"></i><b>3.9.2</b> 多样品合并绘制</a></li>
</ul></li>
<li class="chapter" data-level="3.10" data-path="Rplots.html"><a href="Rplots.html#venn"><i class="fa fa-check"></i><b>3.10</b> 韦恩图</a>
<ul>
<li class="chapter" data-level="3.10.1" data-path="Rplots.html"><a href="Rplots.html#venn2"><i class="fa fa-check"></i><b>3.10.1</b> 韦恩图三个圈</a></li>
<li class="chapter" data-level="3.10.2" data-path="Rplots.html"><a href="Rplots.html#venn5"><i class="fa fa-check"></i><b>3.10.2</b> 韦恩图五个圈</a></li>
<li class="chapter" data-level="3.10.3" data-path="Rplots.html"><a href="Rplots.html#UpSetView"><i class="fa fa-check"></i><b>3.10.3</b> UpSetView展示</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="Rplots.html"><a href="Rplots.html#Barplot"><i class="fa fa-check"></i><b>3.11</b> 柱状图绘制</a>
<ul>
<li class="chapter" data-level="3.11.1" data-path="Rplots.html"><a href="Rplots.html#Barplot1"><i class="fa fa-check"></i><b>3.11.1</b> 常规矩阵柱状图绘制</a></li>
<li class="chapter" data-level="3.11.2" data-path="Rplots.html"><a href="Rplots.html#Barplot2"><i class="fa fa-check"></i><b>3.11.2</b> 长矩阵分面绘制</a></li>
</ul></li>
<li class="chapter" data-level="3.12" data-path="Rplots.html"><a href="Rplots.html#Barplot3"><i class="fa fa-check"></i><b>3.12</b> 图形支持中文字体</a>
<ul>
<li class="chapter" data-level="3.12.1" data-path="Rplots.html"><a href="Rplots.html#Barplot4"><i class="fa fa-check"></i><b>3.12.1</b> 修改图形的字体</a></li>
<li class="chapter" data-level="3.12.2" data-path="Rplots.html"><a href="Rplots.html#Barplot41"><i class="fa fa-check"></i><b>3.12.2</b> ggplot2支持中文字体输出PDF</a></li>
<li class="chapter" data-level="3.12.3" data-path="Rplots.html"><a href="Rplots.html#Barplot5"><i class="fa fa-check"></i><b>3.12.3</b> 系统可用字体</a></li>
<li class="chapter" data-level="3.12.4" data-path="Rplots.html"><a href="Rplots.html#Barplot6"><i class="fa fa-check"></i><b>3.12.4</b> 合并字体支持中英文</a></li>
<li class="chapter" data-level="3.12.5" data-path="Rplots.html"><a href="Rplots.html#Barplot7"><i class="fa fa-check"></i><b>3.12.5</b> 一个示例</a></li>
</ul></li>
<li class="chapter" data-level="3.13" data-path="Rplots.html"><a href="Rplots.html#pca"><i class="fa fa-check"></i><b>3.13</b> PCA原理解析和图形绘制</a>
<ul>
<li class="chapter" data-level="3.13.1" data-path="Rplots.html"><a href="Rplots.html#pcaintroduc"><i class="fa fa-check"></i><b>3.13.1</b> 主成分分析简介</a></li>
<li class="chapter" data-level="3.13.2" data-path="Rplots.html"><a href="Rplots.html#pcaintroduc1"><i class="fa fa-check"></i><b>3.13.2</b> 主成分分析的意义</a></li>
<li class="chapter" data-level="3.13.3" data-path="Rplots.html"><a href="Rplots.html#pcaintroduc2"><i class="fa fa-check"></i><b>3.13.3</b> 示例展示原始变量对样品的分类</a></li>
<li class="chapter" data-level="3.13.4" data-path="Rplots.html"><a href="Rplots.html#pcaintroduc3"><i class="fa fa-check"></i><b>3.13.4</b> PCA的实现原理</a></li>
<li class="chapter" data-level="3.13.5" data-path="Rplots.html"><a href="Rplots.html#pcaintroduc4"><i class="fa fa-check"></i><b>3.13.5</b> 简单的PCA实现</a></li>
<li class="chapter" data-level="3.13.6" data-path="Rplots.html"><a href="Rplots.html#pcaintroduc5"><i class="fa fa-check"></i><b>3.13.6</b> PCA结果解释</a></li>
<li class="chapter" data-level="3.13.7" data-path="Rplots.html"><a href="Rplots.html#pcaintroduc6"><i class="fa fa-check"></i><b>3.13.7</b> PCA应用于测试数据</a></li>
<li class="chapter" data-level="3.13.8" data-path="Rplots.html"><a href="Rplots.html#pcaintroduc7"><i class="fa fa-check"></i><b>3.13.8</b> PCA注意事项</a></li>
<li class="chapter" data-level="3.13.9" data-path="Rplots.html"><a href="Rplots.html#pcaintroduc8"><i class="fa fa-check"></i><b>3.13.9</b> 参考资料</a></li>
</ul></li>
<li class="chapter" data-level="3.14" data-path="Rplots.html"><a href="Rplots.html#survival"><i class="fa fa-check"></i><b>3.14</b> 生存分析</a>
<ul>
<li class="chapter" data-level="3.14.1" data-path="Rplots.html"><a href="Rplots.html#survival1"><i class="fa fa-check"></i><b>3.14.1</b> R做生存分析</a></li>
</ul></li>
<li class="chapter" data-level="3.15" data-path="Rplots.html"><a href="Rplots.html#survival2"><i class="fa fa-check"></i><b>3.15</b> 一步作图的优势</a></li>
<li class="chapter" data-level="3.16" data-path="Rplots.html"><a href="Rplots.html#survival3365"><i class="fa fa-check"></i><b>3.16</b> 不改脚本的热图绘制</a>
<ul>
<li class="chapter" data-level="3.16.1" data-path="Rplots.html"><a href="Rplots.html#survival3367"><i class="fa fa-check"></i><b>3.16.1</b> 箱线图 - 一步绘制</a></li>
<li class="chapter" data-level="3.16.2" data-path="Rplots.html"><a href="Rplots.html#survival3719"><i class="fa fa-check"></i><b>3.16.2</b> 线图 - 一步绘制</a></li>
<li class="chapter" data-level="3.16.3" data-path="Rplots.html"><a href="Rplots.html#survival4137"><i class="fa fa-check"></i><b>3.16.3</b> 一网打进散点图绘制</a></li>
</ul></li>
<li class="chapter" data-level="3.17" data-path="Rplots.html"><a href="Rplots.html#%E5%8F%82%E8%80%83%E8%B5%84%E6%96%99"><i class="fa fa-check"></i><b>3.17</b> 参考资料</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="cytoscape.html"><a href="cytoscape.html"><i class="fa fa-check"></i><b>4</b> 网络图</a>
<ul>
<li class="chapter" data-level="4.0.1" data-path="cytoscape.html"><a href="cytoscape.html#basic"><i class="fa fa-check"></i><b>4.0.1</b> 基本操作</a></li>
<li class="chapter" data-level="4.0.2" data-path="cytoscape.html"><a href="cytoscape.html#mirna"><i class="fa fa-check"></i><b>4.0.2</b> miRNA-mRNA调控网络</a></li>
<li class="chapter" data-level="4.0.3" data-path="cytoscape.html"><a href="cytoscape.html#layout"><i class="fa fa-check"></i><b>4.0.3</b> 不同的布局的调试和修改</a></li>
<li class="chapter" data-level="4.1" data-path="cytoscape.html"><a href="cytoscape.html#%E5%8F%82%E8%80%83"><i class="fa fa-check"></i><b>4.1</b> 参考</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="AI.html"><a href="AI.html"><i class="fa fa-check"></i><b>5</b> 图形排版</a>
<ul>
<li class="chapter" data-level="5.0.1" data-path="AI.html"><a href="AI.html#%E7%9F%A2%E9%87%8F%E5%9B%BE%E5%92%8C%E6%A0%87%E9%87%8F%E5%9B%BE"><i class="fa fa-check"></i><b>5.0.1</b> 矢量图和标量图</a></li>
<li class="chapter" data-level="5.0.2" data-path="AI.html"><a href="AI.html#%E7%9F%A2%E9%87%8F%E5%9B%BE%E7%9A%84%E5%88%B6%E4%BD%9C"><i class="fa fa-check"></i><b>5.0.2</b> 矢量图的制作</a></li>
<li class="chapter" data-level="5.0.3" data-path="AI.html"><a href="AI.html#%E7%9F%A2%E9%87%8F%E5%9B%BE%E7%BC%96%E8%BE%91%E5%B7%A5%E5%85%B7"><i class="fa fa-check"></i><b>5.0.3</b> 矢量图编辑工具</a></li>
<li class="chapter" data-level="5.0.4" data-path="AI.html"><a href="AI.html#%E4%BD%9C%E5%9B%BE%E5%9F%BA%E6%9C%AC%E5%8E%9F%E5%88%99"><i class="fa fa-check"></i><b>5.0.4</b> 作图基本原则</a></li>
<li class="chapter" data-level="5.0.5" data-path="AI.html"><a href="AI.html#%E4%BD%9C%E5%9B%BE%E4%B8%AD%E7%9A%84%E8%A6%81%E7%82%B9%E6%B3%A8%E6%84%8F"><i class="fa fa-check"></i><b>5.0.5</b> 作图中的要点注意</a></li>
<li class="chapter" data-level="5.0.6" data-path="AI.html"><a href="AI.html#adobe-illustrator%E4%B8%AD%E7%9A%84%E5%9F%BA%E6%9C%AC%E6%A6%82%E5%BF%B5%E5%92%8C%E6%93%8D%E4%BD%9C%E6%8F%8F%E8%BF%B0"><i class="fa fa-check"></i><b>5.0.6</b> <code>Adobe Illustrator</code>中的基本概念和操作描述</a></li>
<li class="chapter" data-level="5.0.7" data-path="AI.html"><a href="AI.html#%E8%A7%86%E9%A2%91%E6%95%99%E7%A8%8B"><i class="fa fa-check"></i><b>5.0.7</b> 视频教程</a></li>
<li class="chapter" data-level="5.0.8" data-path="AI.html"><a href="AI.html#%E5%8A%A8%E5%9B%BE%E6%95%99%E7%A8%8B"><i class="fa fa-check"></i><b>5.0.8</b> 动图教程</a></li>
<li class="chapter" data-level="5.1" data-path="AI.html"><a href="AI.html#%E5%8F%82%E8%80%83-1"><i class="fa fa-check"></i><b>5.1</b> 参考</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="batch.html"><a href="batch.html"><i class="fa fa-check"></i><b>6</b> 高通量数据中批次效应的鉴定和处理</a>
<ul>
<li class="chapter" data-level="6.1" data-path="batch.html"><a href="batch.html#batch1"><i class="fa fa-check"></i><b>6.1</b> 什么是批次效应？</a></li>
<li class="chapter" data-level="6.2" data-path="batch.html"><a href="batch.html#batch2"><i class="fa fa-check"></i><b>6.2</b> 批次效应会有什么影响？</a></li>
<li class="chapter" data-level="6.3" data-path="batch.html"><a href="batch.html#batch3"><i class="fa fa-check"></i><b>6.3</b> 怎么确认数据有无受到批次效应影响</a></li>
<li class="chapter" data-level="6.4" data-path="batch.html"><a href="batch.html#batch4"><i class="fa fa-check"></i><b>6.4</b> 怎么避免批次效应呢？</a></li>
<li class="chapter" data-level="6.5" data-path="batch.html"><a href="batch.html#batch5"><i class="fa fa-check"></i><b>6.5</b> 如何在差异基因鉴定过程中移除批次效应</a></li>
<li class="chapter" data-level="6.6" data-path="batch.html"><a href="batch.html#%E6%A3%80%E6%9F%A5%E6%9C%89%E6%97%A0%E7%BC%BA%E5%A4%B1%E7%9A%84%E5%8C%85%E5%B9%B6%E5%AE%89%E8%A3%85%E5%8A%A0%E8%BD%BD%E6%89%80%E6%9C%89%E5%8C%85"><i class="fa fa-check"></i><b>6.6</b> 检查有无缺失的包并安装，加载所有包</a></li>
<li class="chapter" data-level="6.7" data-path="batch.html"><a href="batch.html#batch6"><i class="fa fa-check"></i><b>6.7</b> 不考虑批次因素直接进行差异基因分析</a></li>
<li class="chapter" data-level="6.8" data-path="batch.html"><a href="batch.html#batch7"><i class="fa fa-check"></i><b>6.8</b> 考虑已知的批次因素进行差异基因分析</a>
<ul>
<li class="chapter" data-level="6.8.1" data-path="batch.html"><a href="batch.html#batch8"><i class="fa fa-check"></i><b>6.8.1</b> 比较批次校正前后差异基因变化</a></li>
</ul></li>
<li class="chapter" data-level="6.9" data-path="batch.html"><a href="batch.html#batch9"><i class="fa fa-check"></i><b>6.9</b> 批次效应未知时如何判断和在差异基因鉴定过程中移除批次效应</a>
<ul>
<li class="chapter" data-level="6.9.1" data-path="batch.html"><a href="batch.html#batch10"><i class="fa fa-check"></i><b>6.9.1</b> 预测混杂因素(cofounding factors)并在差异基因分析中移除这些因素</a></li>
<li class="chapter" data-level="6.9.2" data-path="batch.html"><a href="batch.html#batch11"><i class="fa fa-check"></i><b>6.9.2</b> 预测可能存在的混杂因素</a></li>
<li class="chapter" data-level="6.9.3" data-path="batch.html"><a href="batch.html#%E5%8F%AF%E8%A7%86%E5%8C%96%E5%B1%95%E7%A4%BA%E9%A2%84%E6%B5%8B%E5%87%BA%E7%9A%84surrogate-variable%E5%B1%9E%E6%80%A7%E4%B8%8E%E5%B7%B2%E7%9F%A5%E7%9A%84%E6%89%B9%E6%AC%A1%E4%BF%A1%E6%81%AF%E7%9A%84%E5%85%B3%E7%B3%BB"><i class="fa fa-check"></i><b>6.9.3</b> 可视化展示预测出的<code>Surrogate variable</code>属性与已知的批次信息的关系</a></li>
<li class="chapter" data-level="6.9.4" data-path="batch.html"><a href="batch.html#%E5%9F%BA%E4%BA%8E%E9%A2%84%E6%B5%8B%E5%87%BA%E7%9A%84%E6%B7%B7%E6%9D%82%E5%9B%A0%E7%B4%A0%E5%86%8D%E6%AC%A1%E8%BF%9B%E8%A1%8C%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90"><i class="fa fa-check"></i><b>6.9.4</b> 基于预测出的混杂因素再次进行差异分析</a></li>
<li class="chapter" data-level="6.9.5" data-path="batch.html"><a href="batch.html#batch12"><i class="fa fa-check"></i><b>6.9.5</b> 比较批次校正前、已知批次校正后和预测的批次校正后差异基因变化</a></li>
</ul></li>
<li class="chapter" data-level="6.10" data-path="batch.html"><a href="batch.html#batch13"><i class="fa fa-check"></i><b>6.10</b> 直接校正表达矩阵</a>
<ul>
<li class="chapter" data-level="6.10.1" data-path="batch.html"><a href="batch.html#%E8%AF%BB%E5%85%A5%E6%A0%87%E5%87%86%E5%8C%96%E5%90%8E%E7%9A%84%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5%E5%92%8C%E6%A0%B7%E5%93%81%E4%BF%A1%E6%81%AF%E8%A1%A8"><i class="fa fa-check"></i><b>6.10.1</b> 读入标准化后的表达矩阵和样品信息表</a></li>
<li class="chapter" data-level="6.10.2" data-path="batch.html"><a href="batch.html#batch14"><i class="fa fa-check"></i><b>6.10.2</b> 使用ComBat校正</a></li>
<li class="chapter" data-level="6.10.3" data-path="batch.html"><a href="batch.html#batch15"><i class="fa fa-check"></i><b>6.10.3</b> 使用limma校正</a></li>
<li class="chapter" data-level="6.10.4" data-path="batch.html"><a href="batch.html#%E6%9F%A5%E7%9C%8B%E6%A0%A1%E6%AD%A3%E5%AE%8C%E5%90%8E%E5%9F%BA%E5%9B%A0%E7%9A%84%E8%A1%A8%E8%BE%BE%E6%83%85%E5%86%B5"><i class="fa fa-check"></i><b>6.10.4</b> 查看校正完后基因的表达情况</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="questions06.html"><a href="questions06.html"><i class="fa fa-check"></i><b>7</b> 问题答疑</a>
<ul>
<li class="chapter" data-level="7.1" data-path="questions06.html"><a href="questions06.html#ggplot2_variable_color_saturation"><i class="fa fa-check"></i><b>7.1</b> ggplot2绘图如何根据一个变量即区分颜色又区分深浅？</a></li>
<li class="chapter" data-level="7.2" data-path="questions06.html"><a href="questions06.html#ggplot2%E7%BB%98%E5%9B%BE%E7%82%B9%E7%9A%84%E5%BD%A2%E7%8A%B6%E4%B8%8D%E5%A4%9F%E7%94%A8%E6%80%8E%E4%B9%88%E5%8A%9E"><i class="fa fa-check"></i><b>7.2</b> ggplot2绘图点的形状不够用怎么办？</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="推荐一个-r-中很好用的条件判断语句.html"><a href="推荐一个-r-中很好用的条件判断语句.html"><i class="fa fa-check"></i><b>8</b> 推荐一个 R 中很好用的条件判断语句</a>
<ul>
<li class="chapter" data-level="8.1" data-path="推荐一个-r-中很好用的条件判断语句.html"><a href="推荐一个-r-中很好用的条件判断语句.html#ifelse-%E6%A0%87%E8%AE%B0%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0"><i class="fa fa-check"></i><b>8.1</b> ifelse 标记差异基因</a></li>
<li class="chapter" data-level="8.2" data-path="推荐一个-r-中很好用的条件判断语句.html"><a href="推荐一个-r-中很好用的条件判断语句.html#dplyrcase-when-%E6%A0%87%E8%AE%B0%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0"><i class="fa fa-check"></i><b>8.2</b> dplyr::case-when 标记差异基因</a></li>
<li class="chapter" data-level="8.3" data-path="推荐一个-r-中很好用的条件判断语句.html"><a href="推荐一个-r-中很好用的条件判断语句.html#dplyrif_else-%E4%BC%9A%E6%9B%B4%E5%BF%AB%E4%B8%80%E7%82%B9"><i class="fa fa-check"></i><b>8.3</b> dplyr::if_else 会更快一点</a></li>
<li class="chapter" data-level="8.4" data-path="推荐一个-r-中很好用的条件判断语句.html"><a href="推荐一个-r-中很好用的条件判断语句.html#case-when%E5%8F%AA%E4%BF%9D%E7%95%99%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E7%9A%84%E5%90%8D%E5%AD%97"><i class="fa fa-check"></i><b>8.4</b> case-when只保留差异基因的名字</a></li>
<li class="chapter" data-level="8.5" data-path="推荐一个-r-中很好用的条件判断语句.html"><a href="推荐一个-r-中很好用的条件判断语句.html#case-when%E6%AF%8F%E9%9A%94-1-%E4%B8%AA%E5%9F%BA%E5%9B%A0%E4%BF%9D%E7%95%99-1-%E4%B8%AA"><i class="fa fa-check"></i><b>8.5</b> case-when每隔 1 个基因保留 1 个</a></li>
<li class="chapter" data-level="8.6" data-path="推荐一个-r-中很好用的条件判断语句.html"><a href="推荐一个-r-中很好用的条件判断语句.html#dplyrif_else%E9%80%9F%E5%BA%A6%E6%9C%80%E5%BF%AB"><i class="fa fa-check"></i><b>8.6</b> <code>dplyr::if_else</code>速度最快！</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="visual1.html"><a href="visual1.html"><i class="fa fa-check"></i><b>9</b> 易生信-数据可视化</a>
<ul>
<li class="chapter" data-level="9.1" data-path="visual1.html"><a href="visual1.html#visual2"><i class="fa fa-check"></i><b>9.1</b> 加载需要的包</a></li>
<li class="chapter" data-level="9.2" data-path="visual1.html"><a href="visual1.html#visual3"><i class="fa fa-check"></i><b>9.2</b> 读入数据</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="visual1.html"><a href="visual1.html#visual4"><i class="fa fa-check"></i><b>9.2.1</b> Duplicate row names</a></li>
<li class="chapter" data-level="9.2.2" data-path="visual1.html"><a href="visual1.html#visual15"><i class="fa fa-check"></i><b>9.2.2</b> 行名唯一化处理</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="visual1.html"><a href="visual1.html#visual5"><i class="fa fa-check"></i><b>9.3</b> 热图绘制</a></li>
<li class="chapter" data-level="9.4" data-path="visual1.html"><a href="visual1.html#visual7"><i class="fa fa-check"></i><b>9.4</b> 箱线图和统计比较</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="visual1.html"><a href="visual1.html#visual8"><i class="fa fa-check"></i><b>9.4.1</b> 单基因箱线图</a></li>
<li class="chapter" data-level="9.4.2" data-path="visual1.html"><a href="visual1.html#visual9"><i class="fa fa-check"></i><b>9.4.2</b> 多基因箱线图 (combine)</a></li>
<li class="chapter" data-level="9.4.3" data-path="visual1.html"><a href="visual1.html#visual10"><i class="fa fa-check"></i><b>9.4.3</b> 多基因箱线图 (merge)</a></li>
<li class="chapter" data-level="9.4.4" data-path="visual1.html"><a href="visual1.html#visual11"><i class="fa fa-check"></i><b>9.4.4</b> 数据对数转换后绘制箱线图</a></li>
<li class="chapter" data-level="9.4.5" data-path="visual1.html"><a href="visual1.html#visual12"><i class="fa fa-check"></i><b>9.4.5</b> 用ggplot2实现ggpubr</a></li>
<li class="chapter" data-level="9.4.6" data-path="visual1.html"><a href="visual1.html#visual13"><i class="fa fa-check"></i><b>9.4.6</b> 配色</a></li>
<li class="chapter" data-level="9.4.7" data-path="visual1.html"><a href="visual1.html#visual14"><i class="fa fa-check"></i><b>9.4.7</b> 箱线图加统计分析</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="visual1.html"><a href="visual1.html#visual151"><i class="fa fa-check"></i><b>9.5</b> 通路内基因的比较</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="visual1.html"><a href="visual1.html#visual161"><i class="fa fa-check"></i><b>9.5.1</b> 密度图</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="visual1.html"><a href="visual1.html#visual171"><i class="fa fa-check"></i><b>9.6</b> ggstatsplot绘图和统计分析</a>
<ul>
<li class="chapter" data-level="9.6.1" data-path="visual1.html"><a href="visual1.html#visual162"><i class="fa fa-check"></i><b>9.6.1</b> 散点图</a></li>
<li class="chapter" data-level="9.6.2" data-path="visual1.html"><a href="visual1.html#visual172"><i class="fa fa-check"></i><b>9.6.2</b> 相关性图</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="imageGP.html"><a href="imageGP.html"><i class="fa fa-check"></i><b>10</b> 高颜值免费在线绘图（提供绘图源码）</a>
<ul>
<li class="chapter" data-level="10.1" data-path="imageGP.html"><a href="imageGP.html#imagegpbasic"><i class="fa fa-check"></i><b>10.1</b> 高颜值免费在线绘图基础版视频</a></li>
<li class="chapter" data-level="10.2" data-path="imageGP.html"><a href="imageGP.html#imagepbetter"><i class="fa fa-check"></i><b>10.2</b> 高颜值免费在线绘图进阶版视频</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="ref.html"><a href="ref.html"><i class="fa fa-check"></i><b>11</b> 参考</a>
<ul>
<li class="chapter" data-level="11.1" data-path="ref.html"><a href="ref.html#illumina%E6%B5%8B%E5%BA%8F%E5%BA%94%E7%94%A8%E6%89%8B%E5%86%8C"><i class="fa fa-check"></i><b>11.1</b> Illumina测序应用手册</a></li>
<li class="chapter" data-level="11.2" data-path="ref.html"><a href="ref.html#%E7%B3%BB%E5%88%97%E6%95%99%E7%A8%8B"><i class="fa fa-check"></i><b>11.2</b> 系列教程</a></li>
<li class="chapter" data-level="11.3" data-path="ref.html"><a href="ref.html#%E8%9B%8B%E7%99%BD%E8%B4%A8%E7%BB%84%E5%AD%A6%E7%A0%94%E7%A9%B6"><i class="fa fa-check"></i><b>11.3</b> 蛋白质组学研究</a></li>
<li class="chapter" data-level="11.4" data-path="ref.html"><a href="ref.html#%E8%BD%AC%E5%BD%95%E7%BB%84%E7%A0%94%E7%A9%B6"><i class="fa fa-check"></i><b>11.4</b> 转录组研究</a></li>
<li class="chapter" data-level="11.5" data-path="ref.html"><a href="ref.html#%E5%8D%95%E7%BB%86%E8%83%9E%E7%B3%BB%E5%88%97%E6%95%99%E7%A8%8B"><i class="fa fa-check"></i><b>11.5</b> 单细胞系列教程</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="ref.html"><a href="ref.html#%E7%BB%8F%E5%85%B8%E7%BB%BC%E8%BF%B0"><i class="fa fa-check"></i><b>11.5.1</b> 经典综述</a></li>
<li class="chapter" data-level="11.5.2" data-path="ref.html"><a href="ref.html#%E6%96%87%E7%AB%A0%E8%A7%A3%E8%AF%BB"><i class="fa fa-check"></i><b>11.5.2</b> 文章解读</a></li>
<li class="chapter" data-level="11.5.3" data-path="ref.html"><a href="ref.html#%E7%B3%BB%E5%88%97%E6%95%99%E7%A8%8B-1"><i class="fa fa-check"></i><b>11.5.3</b> 系列教程</a></li>
<li class="chapter" data-level="11.5.4" data-path="ref.html"><a href="ref.html#%E7%89%B9%E8%89%B2%E5%88%86%E6%9E%90"><i class="fa fa-check"></i><b>11.5.4</b> 特色分析</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="ref.html"><a href="ref.html#geotcga%E6%95%B0%E6%8D%AE"><i class="fa fa-check"></i><b>11.6</b> GEO/TCGA数据</a></li>
<li class="chapter" data-level="11.7" data-path="ref.html"><a href="ref.html#%E6%89%A9%E5%A2%9E%E5%AD%90%E4%B8%89%E6%AD%A5%E6%9B%B2"><i class="fa fa-check"></i><b>11.7</b> 扩增子三步曲</a></li>
<li class="chapter" data-level="11.8" data-path="ref.html"><a href="ref.html#%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%95%99%E7%A8%8B"><i class="fa fa-check"></i><b>11.8</b> 宏基因组教程</a></li>
<li class="chapter" data-level="11.9" data-path="ref.html"><a href="ref.html#%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%88%86%E6%9E%90%E4%B8%93%E9%A2%98"><i class="fa fa-check"></i><b>11.9</b> 宏基因组分析专题</a></li>
<li class="chapter" data-level="11.10" data-path="ref.html"><a href="ref.html#chip-seq%E4%B8%93%E9%A2%98"><i class="fa fa-check"></i><b>11.10</b> ChIP-seq专题</a></li>
<li class="chapter" data-level="11.11" data-path="ref.html"><a href="ref.html#linux-%E5%85%A8%E4%BB%8B%E7%BB%8D"><i class="fa fa-check"></i><b>11.11</b> Linux 全介绍</a></li>
<li class="chapter" data-level="11.12" data-path="ref.html"><a href="ref.html#circos%E7%B3%BB%E5%88%97"><i class="fa fa-check"></i><b>11.12</b> CIRCOS系列</a></li>
<li class="chapter" data-level="11.13" data-path="ref.html"><a href="ref.html#r%E7%BB%9F%E8%AE%A1%E5%92%8C%E4%BD%9C%E5%9B%BE"><i class="fa fa-check"></i><b>11.13</b> R统计和作图</a></li>
<li class="chapter" data-level="11.14" data-path="ref.html"><a href="ref.html#ngs%E5%9F%BA%E7%A1%80%E5%92%8C%E8%BD%AF%E4%BB%B6%E5%BA%94%E7%94%A8"><i class="fa fa-check"></i><b>11.14</b> NGS基础和软件应用</a></li>
<li class="chapter" data-level="11.15" data-path="ref.html"><a href="ref.html#%E7%94%9F%E4%BF%A1%E5%AE%9D%E5%85%B8%E4%B9%8B%E5%82%BB%E7%93%9C%E5%BC%8F"><i class="fa fa-check"></i><b>11.15</b> 生信宝典之傻瓜式</a></li>
<li class="chapter" data-level="11.16" data-path="ref.html"><a href="ref.html#python"><i class="fa fa-check"></i><b>11.16</b> Python</a></li>
<li class="chapter" data-level="11.17" data-path="ref.html"><a href="ref.html#cytoscape%E7%BD%91%E7%BB%9C%E5%9B%BE"><i class="fa fa-check"></i><b>11.17</b> Cytoscape网络图</a></li>
<li class="chapter" data-level="11.18" data-path="ref.html"><a href="ref.html#%E5%88%86%E5%AD%90%E5%AF%B9%E6%8E%A5"><i class="fa fa-check"></i><b>11.18</b> 分子对接</a></li>
<li class="chapter" data-level="11.19" data-path="ref.html"><a href="ref.html#%E6%96%87%E7%8C%AE%E7%B2%BE%E8%AF%BB"><i class="fa fa-check"></i><b>11.19</b> 文献精读</a></li>
<li class="chapter" data-level="11.20" data-path="ref.html"><a href="ref.html#%E7%9A%84%E9%AB%98%E9%98%85%E8%AF%BB%E6%96%87%E7%AB%A0%E5%90%88%E9%9B%86"><i class="fa fa-check"></i><b>11.20</b> 2019的高阅读文章合集</a></li>
<li class="chapter" data-level="11.21" data-path="ref.html"><a href="ref.html#%E7%B2%BE%E9%80%89%E6%96%87%E7%AB%A0%E6%8E%A8%E8%8D%90"><i class="fa fa-check"></i><b>11.21</b> 精选文章推荐</a></li>
<li class="chapter" data-level="11.22" data-path="ref.html"><a href="ref.html#%E7%A7%91%E7%A0%94%E7%BB%8F%E9%AA%8C"><i class="fa fa-check"></i><b>11.22</b> 科研经验</a></li>
<li class="chapter" data-level="11.23" data-path="ref.html"><a href="ref.html#%E8%BD%AF%E4%BB%B6%E5%92%8C%E6%95%B0%E6%8D%AE%E5%BA%93"><i class="fa fa-check"></i><b>11.23</b> 软件和数据库</a></li>
<li class="chapter" data-level="11.24" data-path="ref.html"><a href="ref.html#%E6%89%A9%E5%A2%9E%E5%AD%90%E5%88%86%E6%9E%90"><i class="fa fa-check"></i><b>11.24</b> 扩增子分析</a></li>
<li class="chapter" data-level="11.25" data-path="ref.html"><a href="ref.html#%E5%AE%8F%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%88%86%E6%9E%90"><i class="fa fa-check"></i><b>11.25</b> 宏基因组分析</a></li>
<li class="chapter" data-level="11.26" data-path="ref.html"><a href="ref.html#%E5%AE%9E%E9%AA%8C%E8%AE%BE%E8%AE%A1%E4%B8%8E%E6%8A%80%E6%9C%AF"><i class="fa fa-check"></i><b>11.26</b> 实验设计与技术</a></li>
<li class="chapter" data-level="11.27" data-path="ref.html"><a href="ref.html#%E6%96%87%E7%8C%AE%E7%B2%BE%E8%AF%BB-1"><i class="fa fa-check"></i><b>11.27</b> 文献精读</a></li>
<li class="chapter" data-level="11.28" data-path="ref.html"><a href="ref.html#%E7%A7%91%E6%99%AE%E8%A7%86%E9%A2%91%E5%AF%93%E6%95%99%E4%BA%8E%E4%B9%90"><i class="fa fa-check"></i><b>11.28</b> 科普视频*寓教于乐</a></li>
<li class="chapter" data-level="11.29" data-path="ref.html"><a href="ref.html#%E7%B3%BB%E5%88%97%E5%AE%A3%E4%BC%A0"><i class="fa fa-check"></i><b>11.29</b> 系列宣传</a></li>
<li class="chapter" data-level="11.30" data-path="ref.html"><a href="ref.html#%E6%B0%B8%E4%B9%85%E9%93%BE%E6%8E%A5"><i class="fa fa-check"></i><b>11.30</b> 永久链接</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="http://www.bic.ac.cn">Hello</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R Cytoscape AI生信作图分析教程</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="batch" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">6</span> 高通量数据中批次效应的鉴定和处理<a href="batch.html#batch" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="batch1" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> 什么是批次效应？<a href="batch.html#batch1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>批次效应表示样品在不同的批次处理和测量时引入的与生物状态不相关的系统性的技术偏差。很多因素都可能导致批次效应的产生，如不同实验条件、不同操作者、不同公司的试剂、不同批的试剂、实验开展的时间、检测设备、不同的测序批次等。</p>
</div>
<div id="batch2" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> 批次效应会有什么影响？<a href="batch.html#batch2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>2014年生信领域的大牛<code>Michael P Snyder</code>在<code>PNAS</code>上发表了一篇文章<em>Comparison of the transcriptional landscapes between human and mouse tissues</em>，比较了人和小鼠不同组织和器官中表达谱的异同。研究发现不同物种之间组织特异表达的基因是一致的，但很多基因在同一物种不同组织的表达相似度大于它们在不同物种同一组织的表达相似度。“我”来引申下 （原文并没有这么直接说），大体可以理解为小鼠的脑与小鼠的肾脏的相似性大于小鼠的脑与人的脑的相似性。“我”得出的这个结论是有一些颠覆认知的，如果这样，用小鼠做为模式动物是否会对人的研究给出相似性的推导？</p>
<p>这篇<code>PNAS</code>文章发出后，芝加哥大学的<code>Yoav Gilad</code>在<code>F1000</code>上发表了一篇文章<em>A reanalysis of mouse ENCODE comparative gene expression data</em>来讨论这个不同于以往认知的研究项目的设计和分析的合理性。</p>
<p>首先作者从FASTQ数据的序列名字的ID中提取出对应测序数据来源的测序仪设备ID和测序通道信息，发现所有数据来源于5个批次，如下图所示，只有最后一个批次同时包含了人和小鼠的器官，其它批次都只包含了人的器官或小鼠的器官。</p>
<p><img src="https://f1000researchdata.s3.amazonaws.com/manuscripts/7019/9f5f4330-d81d-46b8-9a3f-d8cb7aaf577e_figure1.gif" /></p>
<p>重现者<code>Yoav Gilad</code>等通过对数据进行重分析，重现了类似于原文中的结果。不论是PCA还是Heatmap的结果，都展示出来源于同一物种的组织或器官倾向于聚类到一起。</p>
<p><img src="https://f1000researchdata.s3.amazonaws.com/manuscripts/7019/9f5f4330-d81d-46b8-9a3f-d8cb7aaf577e_figure2.gif" /></p>
<p>重现者<code>Yoav Gilad</code>等采用<code>ComBat</code>移除批次带来的影响，再次绘制PCA和Heatmap，结果显示表达谱按组织类型而非物种聚在了一起。</p>
<p><img src="https://f1000researchdata.s3.amazonaws.com/manuscripts/7019/9f5f4330-d81d-46b8-9a3f-d8cb7aaf577e_figure3.gif" /></p>
<p>大家有兴趣可以在<strong><a href="https://f1000research.com/articles/4-121" class="uri">https://f1000research.com/articles/4-121</a></strong>看看<code>Yoav Gilad</code>的具体操作和PNAS一作<code>Yoav Gilad</code>等人的讨论，通过学习双方在这段公案中辩论的出发点和落脚点，相信对数据分析也会有更多认识，这个我们后续也会涉及。</p>
</div>
<div id="batch3" class="section level2 hasAnchor" number="6.3">
<h2><span class="header-section-number">6.3</span> 怎么确认数据有无受到批次效应影响<a href="batch.html#batch3" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><em><a href="https://mp.weixin.qq.com/s/ykp1q7CiyTL2PCK5PuNyzw">通过样品的层级聚类热图+样品属性信息的注释来展示样品聚类结果有无受批次效应的影响</a></em>。如下面右图中可见<code>WT_1</code>样品在聚类分支上与其它样品处于不同的分支，而从列注释图可以看到<code>WT_1</code>的<code>seqPlatform</code>和<code>batch</code>信息与其它样品不同，这是给我们的一个提示可能存在批次效应影响。</p>
<p><img src="images/batch15.jpg" /></p>
<p>通常我们在整合多套数据集进行展示时也会加上数据来源信息以展示自己的分析结果未受批次等因素影响。如下图每一列是一个样品，每一行是一个菌群；列注释中有一行为<code>Dataset</code>指示样品来源于 2 个数据集，并且聚类结果没有明显受到数据集来源的影响（四个大的聚类分支中样品来源分布没有明显偏好性）；</p>
<p><img src="images/metaphlan2.png" /></p>
<p><em>通过<a href="https://mp.weixin.qq.com/s?__biz=MzI5MTcwNjA4NQ==&amp;mid=2247489825&amp;idx=1&amp;sn=e56d490bec2bed4068cb2f294399e675&amp;chksm=ec0ddcabdb7a55bd5ad91cc16c470250213d78c782da419e70039e851621525674203e47916f&amp;scene=158#rd">主成分分析PCA</a>查看有无批次效应的影响</em>。如下左图，样品在<code>PC1</code>和<code>PC2</code>组成的空间中按数据集而非样本类型聚在一起，表示数据来源对样本检测结果的影响超出了样本类型的影响，提示存在批次效应。如右图，批次效应移除后，在<code>PC1</code>轴上样品基本按<code>正常-癌旁-肿瘤</code>分布，表示当前样品差异的主要影响因素是样本类型。这时可以绘制样品在更多<code>PC</code>轴上的分布，如<code>PC1-PC3</code>、<code>PC1-PC4</code>等构成的空间中样品差异的主要因素是什么，也可以进一步判断批次效应移除的程度怎样。</p>
<p><img src="images/batch16.jpg" /></p>
<p><em>通过样本整体表达分布查看有无批次影响</em>。不同来源的样本一般是各自进行标准化（尤其是芯片数据），合并在一起后，可以简单的从整体表达分布来查看是否存在明显的偏移。如下左图存在明显的偏移，则提示有批次效应的存在。校正后，如右图，看上去样本的整体表达分布均一了。但是否批次影响就被移除了，却很难据此下结论。</p>
<p><img src="images/batch17.jpg" /></p>
<p><em>通过部分基因集的表达变化查看有无批次效应影响</em>。不同来源的数据一起标准化之后，如果标准化效果好的话，样品整体表达分布也会是均一的（如下面左数第二幅图）。但从中随机抽取数百基因却发现其表达收到了批次的影响（如下面左数第三幅图，只展示了数个基因），而且聚类结果也把两组正常样品分到了各自来源相对应的分支上。</p>
<p><img src="images/batch18.jpg" /></p>
</div>
<div id="batch4" class="section level2 hasAnchor" number="6.4">
<h2><span class="header-section-number">6.4</span> 怎么避免批次效应呢？<a href="batch.html#batch4" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>合理的实验设计和一致的实验操作是避免批次效应的最好方式。如下图所示如何通过合理的设计来避免检测批次带来的影响。</p>
<ol style="list-style-type: decimal">
<li>如左上角图示，若样本量数目不多，可在同一台设备（同一个芯片、同一个测序仪、同一个测序 lane）检测这是最好的方式，不存在检测批次。</li>
<li>如右上角图示，是最差的实验设计方式，不同组的样本分批检测。这种方式无法确定最终检测出的样品差异是生物差异还是设备差异还是二者共同带来的差异。这种检测方式会放大样品的差异，引入较多假阳性结果。如果只是看这种图，相信大家其实都不会这么设计。但换个角度来看，我们 <code>5</code> 月份采一批样去测序，<code>10</code> 月份采一批样去测序，在比较 <code>5</code> 月份样品与 <code>10</code> 月份样品差异时，我们实际的实验设计方式跟这个实验设计也没什么不同。或者做病例对照时，健康人样品好收集，一起测序；疾病样品一起搜集一起测序，也是类似的设计方式。<a href="https://mp.weixin.qq.com/s/l11iWr-W-mRrMUSuyeJSIg">前文所述的</a>一段公案也类似这个设计。</li>
<li>如左下角所示，平衡设计则是最好的实验设计方式。假如实验中存在三个样品组，如<code>对照组</code>、<code>基因敲除组</code>、<code>基因过表达组</code>，每组 9 个重复，则每次检测时都同时包含每组的 3 个重复，这样获得的数据则可以放在一起校正后分析。或者至少对照组的样本在每次测序中都能有 <code>2-3</code> 个重复，最后在数据校正时通过调平对照组数据的检测结果来校正其它样品的检测结果。在<a href="https://mp.weixin.qq.com/s/l11iWr-W-mRrMUSuyeJSIg">前文的留言</a>中，也确实有意识比较好的老师，做了类似设计，值得学习。</li>
<li>如右下角所示，也是芯片检测基因表达的一个常规方式，每个芯片检测一个样本。不论是生物重复还是不同生物条件之间都会受到检测芯片的影响，不同重复之间的批次影响可以评估，但不同条件下的批次影响则难以评估。芯片这么检测是可以的，但测序时每个测序批次会包含很多不同类型的样品，且测序检测影响因素更多，这一设计会引入较大技术偏差。</li>
</ol>
<p><img src="images/batch20.jpg" /></p>
<p>上面也只是从检测方式的角度设计出一个合理的试验模式降低批次效应的影响，但除了检测方式，还会有很多不可控的因素也会影响到批次，如不同操作人、不同操作时间等客观因素，还有如配对设计实验中不同的个体自身也是批次的因素。</p>
<p>所以需要有个方式去检测和尽量降低批次效应带来的数据偏差的影响。</p>
</div>
<div id="batch5" class="section level2 hasAnchor" number="6.5">
<h2><span class="header-section-number">6.5</span> 如何在差异基因鉴定过程中移除批次效应<a href="batch.html#batch5" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>在我们之前的文章<em><a href="https://mp.weixin.qq.com/s/Vmhx_TGxNkQzkekf93Xl4w">DESeq2差异基因分析和批次效应移除</a></em>中也提到了用如下方式构建设计矩阵，以便在差异基因分析过程中移除批次效应的影响。</p>
<pre><code>ddsFullCountTable &lt;- DESeqDataSetFromMatrix(countData = data,
        colData = sample,  design= ~ batch + conditions)

dds &lt;- DESeq(ddsFullCountTable)</code></pre>
<p>下面我们以一个具体例子实战（配对样品处理前后基因表达的变化）和检验下效果。为了演示批次效应的影响，大部分代码做了封装，我们只关心核心的地方。如果自己对封装的代码感兴趣，可以自行查看函数源码。</p>
</div>
<div id="检查有无缺失的包并安装加载所有包" class="section level2 hasAnchor" number="6.6">
<h2><span class="header-section-number">6.6</span> 检查有无缺失的包并安装，加载所有包<a href="batch.html#%E6%A3%80%E6%9F%A5%E6%9C%89%E6%97%A0%E7%BC%BA%E5%A4%B1%E7%9A%84%E5%8C%85%E5%B9%B6%E5%AE%89%E8%A3%85%E5%8A%A0%E8%BD%BD%E6%89%80%E6%9C%89%E5%8C%85" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb582"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb582-1"><a href="batch.html#cb582-1" tabindex="-1"></a>site <span class="ot">=</span> <span class="st">&quot;https://mirrors.tuna.tsinghua.edu.cn/CRAN&quot;</span></span>
<span id="cb582-2"><a href="batch.html#cb582-2" tabindex="-1"></a></span>
<span id="cb582-3"><a href="batch.html#cb582-3" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb582-4"><a href="batch.html#cb582-4" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">repos =</span> site)</span>
<span id="cb582-5"><a href="batch.html#cb582-5" tabindex="-1"></a></span>
<span id="cb582-6"><a href="batch.html#cb582-6" tabindex="-1"></a><span class="fu">options</span>(<span class="at">BioC_mirror=</span><span class="st">&quot;https://mirrors.tuna.tsinghua.edu.cn/bioconductor&quot;</span>)</span>
<span id="cb582-7"><a href="batch.html#cb582-7" tabindex="-1"></a>installed_packages <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">installed.packages</span>())</span>
<span id="cb582-8"><a href="batch.html#cb582-8" tabindex="-1"></a>a <span class="ot">=</span> <span class="fu">rownames</span>(installed_packages)</span>
<span id="cb582-9"><a href="batch.html#cb582-9" tabindex="-1"></a></span>
<span id="cb582-10"><a href="batch.html#cb582-10" tabindex="-1"></a><span class="co"># 安装指定版本的ggbeeswarm</span></span>
<span id="cb582-11"><a href="batch.html#cb582-11" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="st">&quot;ggbeeswarm&quot;</span> <span class="sc">%in%</span> a){</span>
<span id="cb582-12"><a href="batch.html#cb582-12" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;https://cran.r-project.org/src/contrib/Archive/ggbeeswarm/ggbeeswarm_0.6.0.tar.gz&quot;</span>, <span class="at">repos =</span> <span class="cn">NULL</span>, <span class="at">type =</span> <span class="st">&quot;source&quot;</span>)</span>
<span id="cb582-13"><a href="batch.html#cb582-13" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb582-14"><a href="batch.html#cb582-14" tabindex="-1"></a>  <span class="cf">if</span> (installed_packages[<span class="st">&quot;ggbeeswarm&quot;</span>,<span class="st">&quot;Version&quot;</span>] <span class="sc">!=</span> <span class="st">&quot;0.6.0&quot;</span>) {</span>
<span id="cb582-15"><a href="batch.html#cb582-15" tabindex="-1"></a>   <span class="fu">install.packages</span>(<span class="st">&quot;https://cran.r-project.org/src/contrib/Archive/ggbeeswarm/ggbeeswarm_0.6.0.tar.gz&quot;</span>, <span class="at">repos =</span> <span class="cn">NULL</span>, <span class="at">type =</span> <span class="st">&quot;source&quot;</span>)</span>
<span id="cb582-16"><a href="batch.html#cb582-16" tabindex="-1"></a>  }</span>
<span id="cb582-17"><a href="batch.html#cb582-17" tabindex="-1"></a>}</span>
<span id="cb582-18"><a href="batch.html#cb582-18" tabindex="-1"></a></span>
<span id="cb582-19"><a href="batch.html#cb582-19" tabindex="-1"></a>install_bioc <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;tidyverse&quot;</span>, <span class="st">&quot;DESeq2&quot;</span>, <span class="st">&quot;RColorBrewer&quot;</span>, <span class="st">&quot;ggplot2&quot;</span>, <span class="st">&quot;org.Hs.eg.db&quot;</span>,</span>
<span id="cb582-20"><a href="batch.html#cb582-20" tabindex="-1"></a>                  <span class="st">&quot;reshape2&quot;</span>, <span class="st">&quot;stringr&quot;</span>, <span class="st">&quot;beeswarm&quot;</span>, <span class="st">&quot;vipor&quot;</span>,  <span class="st">&quot;gplots&quot;</span>,<span class="st">&quot;tidyr&quot;</span>,<span class="st">&quot;amap&quot;</span>,<span class="st">&quot;BiocParallel&quot;</span>,<span class="st">&quot;sva&quot;</span>,</span>
<span id="cb582-21"><a href="batch.html#cb582-21" tabindex="-1"></a>                  <span class="st">&quot;ggfortify&quot;</span>,<span class="st">&quot;patchwork&quot;</span>, <span class="st">&quot;ggrepel&quot;</span>,</span>
<span id="cb582-22"><a href="batch.html#cb582-22" tabindex="-1"></a>                  <span class="st">&quot;VennDiagram&quot;</span>,<span class="st">&quot;grid&quot;</span>, <span class="st">&quot;limma&quot;</span>, </span>
<span id="cb582-23"><a href="batch.html#cb582-23" tabindex="-1"></a>                  <span class="st">&quot;devtools&quot;</span>,<span class="st">&quot;rmarkdown&quot;</span>,<span class="st">&quot;dplyr&quot;</span>,<span class="st">&quot;conflicted&quot;</span>)</span>
<span id="cb582-24"><a href="batch.html#cb582-24" tabindex="-1"></a></span>
<span id="cb582-25"><a href="batch.html#cb582-25" tabindex="-1"></a></span>
<span id="cb582-26"><a href="batch.html#cb582-26" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> install_bioc) {</span>
<span id="cb582-27"><a href="batch.html#cb582-27" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span>i <span class="sc">%in%</span> a){</span>
<span id="cb582-28"><a href="batch.html#cb582-28" tabindex="-1"></a>    BiocManager<span class="sc">::</span><span class="fu">install</span>(i, <span class="at">update =</span> F, <span class="at">site_repository=</span>site)</span>
<span id="cb582-29"><a href="batch.html#cb582-29" tabindex="-1"></a>    a <span class="ot">=</span> <span class="fu">rownames</span>(<span class="fu">installed.packages</span>())</span>
<span id="cb582-30"><a href="batch.html#cb582-30" tabindex="-1"></a>  }</span>
<span id="cb582-31"><a href="batch.html#cb582-31" tabindex="-1"></a>}</span>
<span id="cb582-32"><a href="batch.html#cb582-32" tabindex="-1"></a></span>
<span id="cb582-33"><a href="batch.html#cb582-33" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="st">&quot;ImageGP&quot;</span> <span class="sc">%in%</span> a){</span>
<span id="cb582-34"><a href="batch.html#cb582-34" tabindex="-1"></a>  <span class="co"># devtools::install_github(&quot;Tong-Chen/ImageGP&quot;)</span></span>
<span id="cb582-35"><a href="batch.html#cb582-35" tabindex="-1"></a>  devtools<span class="sc">::</span><span class="fu">install_git</span>(<span class="st">&quot;https://gitee.com/ct5869/ImageGP.git&quot;</span>)</span>
<span id="cb582-36"><a href="batch.html#cb582-36" tabindex="-1"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb583"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb583-1"><a href="batch.html#cb583-1" tabindex="-1"></a><span class="co"># 若缺少ImageGP包，则安装</span></span>
<span id="cb583-2"><a href="batch.html#cb583-2" tabindex="-1"></a><span class="co"># BiocManager::install(&quot;Tong-Chen/ImageGP&quot;, update=F)</span></span>
<span id="cb583-3"><a href="batch.html#cb583-3" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(DESeq2))</span>
<span id="cb583-4"><a href="batch.html#cb583-4" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">&quot;RColorBrewer&quot;</span>))</span>
<span id="cb583-5"><a href="batch.html#cb583-5" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">&quot;gplots&quot;</span>))</span>
<span id="cb583-6"><a href="batch.html#cb583-6" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">&quot;amap&quot;</span>))</span>
<span id="cb583-7"><a href="batch.html#cb583-7" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">&quot;ggplot2&quot;</span>))</span>
<span id="cb583-8"><a href="batch.html#cb583-8" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">&quot;BiocParallel&quot;</span>))</span>
<span id="cb583-9"><a href="batch.html#cb583-9" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(<span class="st">&quot;ImageGP&quot;</span>))</span>
<span id="cb583-10"><a href="batch.html#cb583-10" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(sva))</span>
<span id="cb583-11"><a href="batch.html#cb583-11" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(ggfortify))</span>
<span id="cb583-12"><a href="batch.html#cb583-12" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(patchwork))</span>
<span id="cb583-13"><a href="batch.html#cb583-13" tabindex="-1"></a><span class="co"># https://cran.r-project.org/src/contrib/Archive/ggbeeswarm/ggbeeswarm_0.6.0.tar.gz</span></span>
<span id="cb583-14"><a href="batch.html#cb583-14" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(ggbeeswarm))</span>
<span id="cb583-15"><a href="batch.html#cb583-15" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(ggrepel))</span>
<span id="cb583-16"><a href="batch.html#cb583-16" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(VennDiagram))</span>
<span id="cb583-17"><a href="batch.html#cb583-17" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(grid))</span>
<span id="cb583-18"><a href="batch.html#cb583-18" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(limma))</span>
<span id="cb583-19"><a href="batch.html#cb583-19" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(dplyr))</span>
<span id="cb583-20"><a href="batch.html#cb583-20" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(conflicted))</span>
<span id="cb583-21"><a href="batch.html#cb583-21" tabindex="-1"></a><span class="fu">conflict_prefer</span>(<span class="st">&quot;select&quot;</span>, <span class="st">&quot;dplyr&quot;</span>)</span></code></pre></div>
<pre><code>## [conflicted] Will prefer
## dplyr::select over any other package.</code></pre>
<p>输入文件1： <code>reads count</code>矩阵 (<em>Count_matrix.txt</em>)，格式如下：</p>
<pre><code>ENSG    untrt_N61311    untrt_N052611   untrt_N080611   untrt_N061011   trt_N61311  trt_N052611 trt_N080611 trt_N061011
ENSG00000223972 1   0   0   0   0   1   0   0
ENSG00000227232 13  25  23  24  12  12  22  22
ENSG00000278267 0   5   3   4   2   4   3   1</code></pre>
<p>输入文件2： <code>实验设计信息</code>表 (<em>metadata</em>): <code>conditions</code>为处理条件（<code>untrt</code>是对照, <code>trt</code>是加药处理 ），<code>individual</code>标记样品的个体来源 (4个个体：N61311、N052611、N080611、N061011)。</p>
<pre><code>Samp    conditions  individual
untrt_N61311    untrt   N61311
untrt_N052611   untrt   N052611
untrt_N080611   untrt   N080611
untrt_N061011   untrt   N061011
trt_N61311  trt N61311
trt_N052611 trt N052611
trt_N080611 trt N080611
trt_N061011 trt N061011</code></pre>
</div>
<div id="batch6" class="section level2 hasAnchor" number="6.7">
<h2><span class="header-section-number">6.7</span> 不考虑批次因素直接进行差异基因分析<a href="batch.html#batch6" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>初始化，定义输入、输出和参数</p>
<div class="sourceCode" id="cb587"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb587-1"><a href="batch.html#cb587-1" tabindex="-1"></a><span class="co"># Prefix for all output file </span></span>
<span id="cb587-2"><a href="batch.html#cb587-2" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;result/batch&quot;</span>, <span class="at">recursive =</span> T)</span>
<span id="cb587-3"><a href="batch.html#cb587-3" tabindex="-1"></a>output_prefix <span class="ot">=</span> <span class="st">&quot;result/batch/STAR.simpler&quot;</span></span>
<span id="cb587-4"><a href="batch.html#cb587-4" tabindex="-1"></a></span>
<span id="cb587-5"><a href="batch.html#cb587-5" tabindex="-1"></a><span class="co"># pipelineStar.sh或其它方式生成的reads count 文件，行为基因，列为样品</span></span>
<span id="cb587-6"><a href="batch.html#cb587-6" tabindex="-1"></a>file <span class="ot">=</span> <span class="st">&quot;Count_matrix.xls&quot;</span></span>
<span id="cb587-7"><a href="batch.html#cb587-7" tabindex="-1"></a><span class="co"># 分组信息表</span></span>
<span id="cb587-8"><a href="batch.html#cb587-8" tabindex="-1"></a>metadata <span class="ot">=</span> <span class="st">&quot;metadata.txt&quot;</span></span>
<span id="cb587-9"><a href="batch.html#cb587-9" tabindex="-1"></a><span class="co"># 分组信息所在列名字</span></span>
<span id="cb587-10"><a href="batch.html#cb587-10" tabindex="-1"></a>covariate <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb587-11"><a href="batch.html#cb587-11" tabindex="-1"></a><span class="co"># covariate = &quot;batch&quot;</span></span>
<span id="cb587-12"><a href="batch.html#cb587-12" tabindex="-1"></a>design<span class="ot">=</span><span class="st">&quot;conditions&quot;</span></span>
<span id="cb587-13"><a href="batch.html#cb587-13" tabindex="-1"></a><span class="co"># 输入数据类型，salmon结果或reads count 矩阵</span></span>
<span id="cb587-14"><a href="batch.html#cb587-14" tabindex="-1"></a>type<span class="ot">=</span><span class="st">&quot;readscount&quot;</span></span>
<span id="cb587-15"><a href="batch.html#cb587-15" tabindex="-1"></a><span class="co"># 差异基因参数</span></span>
<span id="cb587-16"><a href="batch.html#cb587-16" tabindex="-1"></a>padj<span class="ot">=</span><span class="fl">0.05</span></span>
<span id="cb587-17"><a href="batch.html#cb587-17" tabindex="-1"></a>log2FC<span class="ot">=</span><span class="dv">1</span></span></code></pre></div>
<p>数据读入和标准化</p>
<div class="sourceCode" id="cb588"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb588-1"><a href="batch.html#cb588-1" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">readscount2deseq</span>(file, metadata, <span class="at">design=</span>design, <span class="at">covariate =</span> covariate)</span></code></pre></div>
<pre><code>## [1] &quot;Read in 32799 genes&quot;
## [1] &quot;23936 genes remained after filtering of genes with all counts less than 4 in all samples.&quot;
## [1] &quot;Perform DESeq on given datasets.&quot;</code></pre>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<div class="sourceCode" id="cb596"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb596-1"><a href="batch.html#cb596-1" tabindex="-1"></a>normexpr <span class="ot">&lt;-</span> <span class="fu">deseq2normalizedExpr</span>(dds, <span class="at">output_prefix=</span>output_prefix, <span class="at">rlog=</span>F, <span class="at">vst=</span>T, <span class="at">design=</span>design)</span></code></pre></div>
<pre><code>## [1] &quot;Output normalized counts&quot;
## [1] &quot;Output vst transformed normalized counts&quot;</code></pre>
<p>检查数据标准化效果: 标准化后基因在不同样品的表达分布越均一越好。从下图看不出存在批次效应的影响。</p>
<div class="sourceCode" id="cb598"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb598-1"><a href="batch.html#cb598-1" tabindex="-1"></a><span class="co"># normalizedExpr2DistribBoxplot(normexpr, </span></span>
<span id="cb598-2"><a href="batch.html#cb598-2" tabindex="-1"></a><span class="co">#   saveplot=paste(output_prefix, &quot;DESeq2.normalizedExprDistrib.pdf&quot;, sep=&quot;.&quot;))</span></span>
<span id="cb598-3"><a href="batch.html#cb598-3" tabindex="-1"></a><span class="fu">normalizedExpr2DistribBoxplot</span>(normexpr)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-311-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><a href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzI5MTcwNjA4NQ==&amp;action=getalbum&amp;album_id=1335864997100224512&amp;subscene=38&amp;scenenote=https%3A%2F%2Fmp.weixin.qq.com%2Fs%3F__biz%3DMzI5MTcwNjA4NQ%3D%3D%26mid%3D2247495133%26idx%3D2%26sn%3D37d1572211a755566b5de5f120e476cf%26chksm%3Dec0e2857db79a1411fab8405e9f6284f10644eb9965f2d5fdfa98f0ff44a581729f990fc9c14%26scene%3D38%26key%3Dbbd9db2a3d7572cd769c7f8e4e3317fec44d11168b996d9b03e0828fc8c61abdcf8565023bdb967b8005fb7f8f9d6e2f7cb5d24289c52229bae2b2791048ea213afbf98f447947495cde5742eae678a4%26ascene%3D7%26uin%3DOTY2MDEyMzgw%26devicetype%3DWindows%2B7%2Bx64%26version%3D62090529%26lang%3Dzh_CN%26exportkey%3DAa8iMQ%252BTtWiuSpqGzIiwmvk%253D%26pass_ticket%3DDAyFVADyPalxlkE8NUofnsSAhBBJHP7CYr46pHPxHfWOVdG1RLPhHzDUX%252BmMIHmH%26winzoom%3D1#wechat_redirect">样本聚类</a>查看样品相似性，trt组和untrt组区分明显 (<em>聚类采用的不同基因数目、聚类参数都可能影响聚类结果</em>)</p>
<div class="sourceCode" id="cb599"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb599-1"><a href="batch.html#cb599-1" tabindex="-1"></a><span class="co"># clusterSampleHeatmap2(normexpr$vst, </span></span>
<span id="cb599-2"><a href="batch.html#cb599-2" tabindex="-1"></a><span class="co">#                       cor_file=paste(output_prefix, &quot;DESeq2.sampleCorrelation.txt&quot;, sep=&quot;.&quot;), </span></span>
<span id="cb599-3"><a href="batch.html#cb599-3" tabindex="-1"></a><span class="co">#                       saveplot=paste(output_prefix, &quot;DESeq2.sampleCorrelation.pdf&quot;, sep=&quot;.&quot;))</span></span>
<span id="cb599-4"><a href="batch.html#cb599-4" tabindex="-1"></a><span class="co"># 根据前5000个表达变化幅度最大的基因进行聚类分析</span></span>
<span id="cb599-5"><a href="batch.html#cb599-5" tabindex="-1"></a><span class="fu">clusterSampleHeatmap2</span>(normexpr<span class="sc">$</span>vst[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], <span class="at">cor_file=</span><span class="fu">paste</span>(output_prefix, <span class="st">&quot;DESeq2.sampleCorrelation.txt&quot;</span>, <span class="at">sep=</span><span class="st">&quot;.&quot;</span>))</span></code></pre></div>
<pre><code>## [1] &quot;Performing sample clustering&quot;</code></pre>
<div class="sourceCode" id="cb601"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb601-1"><a href="batch.html#cb601-1" tabindex="-1"></a><span class="fu">clusterSampleUpperTriPlot</span>(normexpr<span class="sc">$</span>vst[<span class="dv">1</span><span class="sc">:</span><span class="dv">500</span>,], <span class="at">cor_file=</span><span class="fu">paste</span>(output_prefix, <span class="st">&quot;DESeq2.sampleCorrelation.txt&quot;</span>, <span class="at">sep=</span><span class="st">&quot;.&quot;</span>))</span></code></pre></div>
<pre><code>## [1] &quot;Performing sample clustering&quot;</code></pre>
<p><img src="R_course_files/figure-html/unnamed-chunk-312-1.png" width="100%" style="display: block; margin: auto;" /><img src="R_course_files/figure-html/unnamed-chunk-312-2.png" width="100%" style="display: block; margin: auto;" /></p>
<p><a href="https://mp.weixin.qq.com/s?__biz=MzI5MTcwNjA4NQ==&amp;mid=2247489825&amp;idx=1&amp;sn=e56d490bec2bed4068cb2f294399e675&amp;chksm=ec0ddcabdb7a55bd5ad91cc16c470250213d78c782da419e70039e851621525674203e47916f&amp;scene=158#rd">主成分分析PCA</a>查看样品相似性，发现在<code>PC1</code>轴上，样品按<strong>处理条件</strong>区分开；在<code>PC2</code>轴上，样品按<strong>个体</strong>区分开，不同的个体是影响样品基因表达差异的一个重要因素。</p>
<div class="sourceCode" id="cb603"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb603-1"><a href="batch.html#cb603-1" tabindex="-1"></a>metadata <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(dds))</span>
<span id="cb603-2"><a href="batch.html#cb603-2" tabindex="-1"></a><span class="fu">sp_pca</span>(normexpr<span class="sc">$</span>vst[<span class="dv">1</span><span class="sc">:</span><span class="dv">2000</span>,], metadata, <span class="at">color_variable=</span><span class="st">&quot;conditions&quot;</span>, <span class="at">shape_variable =</span> <span class="st">&quot;individual&quot;</span>) <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">size =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-313-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>先鉴定出差异基因，获得差异基因文件<code>STAR.simpler.DESeq2.all.DE.txt</code>和其它可视化图表（暂时忽略）。</p>
<div class="sourceCode" id="cb604"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb604-1"><a href="batch.html#cb604-1" tabindex="-1"></a><span class="fu">multipleGroupDEgenes</span>(dds, <span class="at">design=</span>design, <span class="at">output_prefix=</span>output_prefix, <span class="at">padj=</span>padj, <span class="at">log2FC=</span>log2FC, <span class="at">normalized_counts=</span>normexpr)</span></code></pre></div>
<pre><code>## [1] &quot;DE genes between untrt trt&quot;
## [1] &quot;conditions&quot; &quot;untrt&quot;      &quot;trt&quot;</code></pre>
<pre><code>## Saving 17.8 x 12.7 cm image</code></pre>
</div>
<div id="batch7" class="section level2 hasAnchor" number="6.8">
<h2><span class="header-section-number">6.8</span> 考虑已知的批次因素进行差异基因分析<a href="batch.html#batch7" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>初始化，定义输入、输出和参数 （注意<code>covariate</code>变量使用<code>individual</code>列作为了批次信息）</p>
<div class="sourceCode" id="cb607"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb607-1"><a href="batch.html#cb607-1" tabindex="-1"></a><span class="co"># Prefix for all output file </span></span>
<span id="cb607-2"><a href="batch.html#cb607-2" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;result/batch&quot;</span>)</span>
<span id="cb607-3"><a href="batch.html#cb607-3" tabindex="-1"></a>output_prefix <span class="ot">=</span> <span class="st">&quot;result/batch/STAR.simpler.batch&quot;</span></span>
<span id="cb607-4"><a href="batch.html#cb607-4" tabindex="-1"></a></span>
<span id="cb607-5"><a href="batch.html#cb607-5" tabindex="-1"></a><span class="co"># pipelineStar.sh生成的reads count 文件，行为基因，列为样品</span></span>
<span id="cb607-6"><a href="batch.html#cb607-6" tabindex="-1"></a>file <span class="ot">=</span> <span class="st">&quot;Count_matrix.xls&quot;</span></span>
<span id="cb607-7"><a href="batch.html#cb607-7" tabindex="-1"></a><span class="co"># 分组信息表</span></span>
<span id="cb607-8"><a href="batch.html#cb607-8" tabindex="-1"></a>metadata <span class="ot">=</span> <span class="st">&quot;metadata.txt&quot;</span></span>
<span id="cb607-9"><a href="batch.html#cb607-9" tabindex="-1"></a><span class="co"># 分组信息所在列名字</span></span>
<span id="cb607-10"><a href="batch.html#cb607-10" tabindex="-1"></a><span class="co"># covariate = NULL</span></span>
<span id="cb607-11"><a href="batch.html#cb607-11" tabindex="-1"></a><span class="co"># *********</span></span>
<span id="cb607-12"><a href="batch.html#cb607-12" tabindex="-1"></a>covariate <span class="ot">=</span> <span class="st">&quot;individual&quot;</span></span>
<span id="cb607-13"><a href="batch.html#cb607-13" tabindex="-1"></a>design<span class="ot">=</span><span class="st">&quot;conditions&quot;</span></span>
<span id="cb607-14"><a href="batch.html#cb607-14" tabindex="-1"></a><span class="co"># 输入数据类型，salmon结果或reads count 矩阵</span></span>
<span id="cb607-15"><a href="batch.html#cb607-15" tabindex="-1"></a>type<span class="ot">=</span><span class="st">&quot;readscount&quot;</span></span>
<span id="cb607-16"><a href="batch.html#cb607-16" tabindex="-1"></a><span class="co"># 差异基因参数</span></span>
<span id="cb607-17"><a href="batch.html#cb607-17" tabindex="-1"></a>padj<span class="ot">=</span><span class="fl">0.05</span></span>
<span id="cb607-18"><a href="batch.html#cb607-18" tabindex="-1"></a>log2FC<span class="ot">=</span><span class="dv">1</span></span></code></pre></div>
<p>数据读入和标准化，并检查数据标准化效果: 标准化后基因在不同样品的表达分布越均一越好 （此图略过，与上面的表达分布图一致）。</p>
<div class="sourceCode" id="cb608"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb608-1"><a href="batch.html#cb608-1" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">readscount2deseq</span>(file, metadata, <span class="at">design=</span>design, <span class="at">covariate =</span> covariate)</span></code></pre></div>
<pre><code>## [1] &quot;Read in 32799 genes&quot;
## [1] &quot;23936 genes remained after filtering of genes with all counts less than 4 in all samples.&quot;
## [1] &quot;Perform DESeq on given datasets.&quot;</code></pre>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<div class="sourceCode" id="cb616"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb616-1"><a href="batch.html#cb616-1" tabindex="-1"></a>normexpr <span class="ot">&lt;-</span> <span class="fu">deseq2normalizedExpr</span>(dds, <span class="at">output_prefix=</span>output_prefix, <span class="at">rlog=</span>F, <span class="at">vst=</span>T, <span class="at">design=</span>design)</span></code></pre></div>
<pre><code>## [1] &quot;Output normalized counts&quot;
## [1] &quot;Output vst transformed normalized counts&quot;</code></pre>
<div class="sourceCode" id="cb618"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb618-1"><a href="batch.html#cb618-1" tabindex="-1"></a><span class="fu">normalizedExpr2DistribBoxplot</span>(normexpr)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-316-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>样本聚类查看样品相似性，trt组和untrt组区分明显 (此部分结果略过，与上面的聚类结果一致)</p>
<div class="sourceCode" id="cb619"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb619-1"><a href="batch.html#cb619-1" tabindex="-1"></a><span class="co"># clusterSampleHeatmap2(normexpr$vst, </span></span>
<span id="cb619-2"><a href="batch.html#cb619-2" tabindex="-1"></a><span class="co">#                       cor_file=paste(output_prefix, &quot;DESeq2.sampleCorrelation.txt&quot;, sep=&quot;.&quot;), </span></span>
<span id="cb619-3"><a href="batch.html#cb619-3" tabindex="-1"></a><span class="co">#                       saveplot=paste(output_prefix, &quot;DESeq2.sampleCorrelation.pdf&quot;, sep=&quot;.&quot;))</span></span>
<span id="cb619-4"><a href="batch.html#cb619-4" tabindex="-1"></a><span class="co"># 根据前5000个表达变化幅度最大的基因进行聚类分析</span></span>
<span id="cb619-5"><a href="batch.html#cb619-5" tabindex="-1"></a><span class="fu">clusterSampleHeatmap2</span>(normexpr<span class="sc">$</span>vst[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], <span class="at">cor_file=</span><span class="fu">paste</span>(output_prefix, <span class="st">&quot;DESeq2.sampleCorrelation.txt&quot;</span>, <span class="at">sep=</span><span class="st">&quot;.&quot;</span>))</span></code></pre></div>
<pre><code>## [1] &quot;Performing sample clustering&quot;</code></pre>
<div class="sourceCode" id="cb621"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb621-1"><a href="batch.html#cb621-1" tabindex="-1"></a><span class="fu">clusterSampleUpperTriPlot</span>(normexpr<span class="sc">$</span>vst[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], <span class="at">cor_file=</span><span class="fu">paste</span>(output_prefix, <span class="st">&quot;DESeq2.sampleCorrelation.txt&quot;</span>, <span class="at">sep=</span><span class="st">&quot;.&quot;</span>))</span></code></pre></div>
<pre><code>## [1] &quot;Performing sample clustering&quot;</code></pre>
<p><img src="R_course_files/figure-html/unnamed-chunk-317-1.png" width="100%" style="display: block; margin: auto;" /><img src="R_course_files/figure-html/unnamed-chunk-317-2.png" width="100%" style="display: block; margin: auto;" /></p>
<p>主成分分析PCA查看样品相似性，发现在<code>PC1</code>轴上，样品按<strong>处理条件</strong>区分开；在<code>PC2</code>轴上，样品按<strong>个体</strong>区分开，表明不同的个体可能会对后续的差异基因分析造成影响。这个结果也与我们前面不考虑批次因素的结果是一样的。</p>
<div class="sourceCode" id="cb623"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb623-1"><a href="batch.html#cb623-1" tabindex="-1"></a>metadata <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(dds))</span>
<span id="cb623-2"><a href="batch.html#cb623-2" tabindex="-1"></a><span class="fu">sp_pca</span>(normexpr<span class="sc">$</span>vst[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], metadata, <span class="at">color_variable=</span><span class="st">&quot;conditions&quot;</span>, <span class="at">shape_variable =</span> <span class="st">&quot;individual&quot;</span>) <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">size =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-318-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>是不是批次变量加错了呢，还是添加的批次变量未生效？可以说都不是，操作没问题，只是<code>DESeq2</code>处理时只在差异分析模型中考虑批次效应信息，而不会<code>直接校正表达矩阵</code>。那我们先看下加了批次后差异分析的结果怎样，后续我们再讲如何校正表达矩阵。</p>
<p>鉴定出差异基因，获得差异基因文件<code>STAR.simpler.batch.DESeq2.all.DE.txt</code>和其它可视化图表（暂时忽略）。</p>
<div class="sourceCode" id="cb624"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb624-1"><a href="batch.html#cb624-1" tabindex="-1"></a><span class="fu">multipleGroupDEgenes</span>(dds, <span class="at">design=</span>design, <span class="at">output_prefix=</span>output_prefix, <span class="at">padj=</span>padj, <span class="at">log2FC=</span>log2FC, <span class="at">normalized_counts=</span>normexpr)</span></code></pre></div>
<pre><code>## [1] &quot;DE genes between untrt trt&quot;
## [1] &quot;conditions&quot; &quot;untrt&quot;      &quot;trt&quot;</code></pre>
<pre><code>## Saving 17.8 x 12.7 cm image</code></pre>
<div id="batch8" class="section level3 hasAnchor" number="6.8.1">
<h3><span class="header-section-number">6.8.1</span> 比较批次校正前后差异基因变化<a href="batch.html#batch8" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>校正后，差异基因数目变多了，上调多了<em>99</em>个，下调多了<em>61</em>个。不过数目变化，也说明不了太多问题。</p>
<div class="sourceCode" id="cb627"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb627-1"><a href="batch.html#cb627-1" tabindex="-1"></a>de_before_batch <span class="ot">=</span> <span class="fu">sp_readTable</span>(<span class="st">&quot;result/batch/STAR.simpler.DESeq2.all.DE.txt&quot;</span>, <span class="at">header=</span>F)</span>
<span id="cb627-2"><a href="batch.html#cb627-2" tabindex="-1"></a>de_before_batch<span class="sc">$</span>V2 <span class="ot">=</span> <span class="fu">paste</span>(<span class="st">&quot;Before_batch&quot;</span>,de_before_batch<span class="sc">$</span>V2,<span class="at">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb627-3"><a href="batch.html#cb627-3" tabindex="-1"></a><span class="fu">table</span>(de_before_batch<span class="sc">$</span>V2)</span></code></pre></div>
<pre><code>## 
## Before_batch_untrt._higherThan_.trt  Before_batch_untrt._lowerThan_.trt 
##                                 398                                 466</code></pre>
<div class="sourceCode" id="cb629"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb629-1"><a href="batch.html#cb629-1" tabindex="-1"></a>de_after_batch <span class="ot">=</span> <span class="fu">sp_readTable</span>(<span class="st">&quot;result/batch/STAR.simpler.batch.DESeq2.all.DE.txt&quot;</span>, <span class="at">header=</span>F)</span>
<span id="cb629-2"><a href="batch.html#cb629-2" tabindex="-1"></a>de_after_batch<span class="sc">$</span>V2 <span class="ot">=</span> <span class="fu">paste</span>(<span class="st">&quot;After_known_batch&quot;</span>,de_after_batch<span class="sc">$</span>V2,<span class="at">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb629-3"><a href="batch.html#cb629-3" tabindex="-1"></a><span class="fu">table</span>(de_after_batch<span class="sc">$</span>V2)</span></code></pre></div>
<pre><code>## 
## After_known_batch_untrt._higherThan_.trt 
##                                      497 
##  After_known_batch_untrt._lowerThan_.trt 
##                                      527</code></pre>
<p>画个Venn图，看下哪些基因是新增的差异基因，哪些基因批次校正后没差异了。</p>
<div class="sourceCode" id="cb631"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb631-1"><a href="batch.html#cb631-1" tabindex="-1"></a>all_de <span class="ot">=</span> <span class="fu">rbind</span>(de_before_batch, de_after_batch)</span>
<span id="cb631-2"><a href="batch.html#cb631-2" tabindex="-1"></a><span class="co"># 随机查看6行，信息代表更全面</span></span>
<span id="cb631-3"><a href="batch.html#cb631-3" tabindex="-1"></a>all_de[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(all_de),<span class="dv">6</span>),]</span></code></pre></div>
<pre><code>##                   V1                                       V2
## 1877 ENSG00000272744  After_known_batch_untrt._lowerThan_.trt
## 777  ENSG00000197594       Before_batch_untrt._lowerThan_.trt
## 701  ENSG00000162998       Before_batch_untrt._lowerThan_.trt
## 1779 ENSG00000074660  After_known_batch_untrt._lowerThan_.trt
## 1061 ENSG00000137331 After_known_batch_untrt._higherThan_.trt
## 484  ENSG00000185813       Before_batch_untrt._lowerThan_.trt</code></pre>
<div class="sourceCode" id="cb633"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb633-1"><a href="batch.html#cb633-1" tabindex="-1"></a><span class="co"># 结果存储到文件中</span></span>
<span id="cb633-2"><a href="batch.html#cb633-2" tabindex="-1"></a><span class="fu">sp_writeTable</span>(all_de, <span class="at">file=</span><span class="st">&quot;result/batch/Compare_de_gene_beofore_and_after_batch.txt&quot;</span>, <span class="at">keep_rownames =</span> F, <span class="at">col.names =</span> F)</span></code></pre></div>
<div class="sourceCode" id="cb634"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb634-1"><a href="batch.html#cb634-1" tabindex="-1"></a><span class="fu">sp_vennDiagram</span>(all_de, <span class="at">label1=</span><span class="st">&quot;Before_batch_untrt._higherThan_.trt&quot;</span>,</span>
<span id="cb634-2"><a href="batch.html#cb634-2" tabindex="-1"></a>               <span class="at">label2=</span><span class="st">&quot;After_known_batch_untrt._higherThan_.trt&quot;</span>)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-323-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>采用在线工具<a href="https://www.bic.ac.cn/Evenn/#/" class="uri">https://www.bic.ac.cn/Evenn/#/</a> 来做，准备在线工具所需的文件，一个两列格式的文件：第一列为基因名，第二列为基因的上下调状态。拷贝文件数据到网站数据输入处：</p>
<p><img src="images/venn.gif" /></p>
<p>从Venn图可以看出，批次校正后既有新增的差异基因，又丢失了之前的一部分差异基因，那么哪个方式更合理呢？</p>
<p>选择1个批次校正后检测为上调的基因和1个批次校正后检测为下调的基因，观察下其表达模式。从下图可以看出，这些基因具有明显的个体表达一致性。<code>ENSG00000163394</code>基因在每个个体来源的样本中处理后表达都上调了近4倍，但是其本底表达在不同个体中却差异较大。如其在<code>N080611</code>个体（蓝色线）中表达整体偏低，药物处理后表达虽然有上调但表达值却低于其在<code>N061011</code>个体（绿色线）处理前的表达。从这两个例子可以看出，考虑到每个个体的基准表达水平不同，最终获得的差异倍数会有较高的方差。批次校正后解决了样品个体来源基因本底表达差异的影响，获得的差异基因倍数方差会变小，所以检测出更多差异基因，理论上也是更可靠的方式。（这个在之前文章<a href="https://mp.weixin.qq.com/s/PswNqmQ8mDleVE4yHtCvAQ">典型医学设计实验GEO数据分析 (step-by-step) - Limma差异分析、火山图、功能富集</a>也有阐述。）</p>
<div class="sourceCode" id="cb635"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb635-1"><a href="batch.html#cb635-1" tabindex="-1"></a>ENSG00000163394 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Expr=</span>normexpr<span class="sc">$</span>vst[<span class="st">&quot;ENSG00000163394&quot;</span>,])</span>
<span id="cb635-2"><a href="batch.html#cb635-2" tabindex="-1"></a><span class="fu">sp_writeTable</span>(ENSG00000163394, <span class="at">file=</span><span class="st">&quot;result/ENSG00000163394.txt&quot;</span>)</span>
<span id="cb635-3"><a href="batch.html#cb635-3" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">sp_boxplot</span>(ENSG00000163394, <span class="at">melted=</span>T, <span class="at">metadata=</span>metadata, <span class="at">xvariable =</span> <span class="st">&quot;conditions&quot;</span>, <span class="at">yvariable =</span> <span class="st">&quot;Expr&quot;</span>, <span class="at">jitter_bp =</span> T, <span class="at">group_variable_for_line =</span> <span class="st">&quot;individual&quot;</span>, <span class="at">xvariable_order =</span> <span class="fu">c</span>(<span class="st">&quot;untrt&quot;</span>,<span class="st">&quot;trt&quot;</span>), <span class="at">manual_color_vector =</span> <span class="st">&quot;Set2&quot;</span>)</span></code></pre></div>
<pre><code>## [1] 1273
## [1] &quot;#66C2A5&quot; &quot;#FC8D62&quot;</code></pre>
<div class="sourceCode" id="cb637"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb637-1"><a href="batch.html#cb637-1" tabindex="-1"></a>ENSG00000221866 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Expr=</span>normexpr<span class="sc">$</span>vst[<span class="st">&quot;ENSG00000221866&quot;</span>,])</span>
<span id="cb637-2"><a href="batch.html#cb637-2" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">sp_boxplot</span>(ENSG00000221866, <span class="at">melted=</span>T, <span class="at">metadata=</span>metadata, <span class="at">xvariable =</span> <span class="st">&quot;conditions&quot;</span>, <span class="at">yvariable =</span> <span class="st">&quot;Expr&quot;</span>, <span class="at">jitter_bp =</span> T, <span class="at">group_variable_for_line =</span> <span class="st">&quot;individual&quot;</span>, <span class="at">xvariable_order =</span> <span class="fu">c</span>(<span class="st">&quot;untrt&quot;</span>,<span class="st">&quot;trt&quot;</span>), <span class="at">manual_color_vector =</span> <span class="st">&quot;Set2&quot;</span>)</span></code></pre></div>
<pre><code>## [1] 1273
## [1] &quot;#66C2A5&quot; &quot;#FC8D62&quot;</code></pre>
<div class="sourceCode" id="cb639"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb639-1"><a href="batch.html#cb639-1" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guide =</span> <span class="st">&#39;collect&#39;</span>)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-324-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>我们再选2个批次校正前鉴定为有差异、批次校正后鉴定为无差异的基因观察下其表达模式。这两个基因的表达模式没看出存在个体本底的一致变化差异。处理前后在不同个体中变化幅度不一，可能是被动变化。但这些基因一定是没有差异吗？我个人也下不出结论，后续得结合其功能再做判断了。</p>
<div class="sourceCode" id="cb640"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb640-1"><a href="batch.html#cb640-1" tabindex="-1"></a>ENSG00000109689 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Expr=</span>normexpr<span class="sc">$</span>vst[<span class="st">&quot;ENSG00000109689&quot;</span>,])</span>
<span id="cb640-2"><a href="batch.html#cb640-2" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">sp_boxplot</span>(ENSG00000109689, <span class="at">melted=</span>T, <span class="at">metadata=</span>metadata, <span class="at">xvariable =</span> <span class="st">&quot;conditions&quot;</span>, <span class="at">yvariable =</span> <span class="st">&quot;Expr&quot;</span>, <span class="at">jitter_bp =</span> T, <span class="at">group_variable_for_line =</span> <span class="st">&quot;individual&quot;</span>, <span class="at">xvariable_order =</span> <span class="fu">c</span>(<span class="st">&quot;untrt&quot;</span>,<span class="st">&quot;trt&quot;</span>), <span class="at">manual_color_vector =</span> <span class="st">&quot;Set2&quot;</span>, <span class="at">title=</span><span class="st">&quot;ENSG00000109689&quot;</span>)</span></code></pre></div>
<pre><code>## [1] 1273
## [1] &quot;#66C2A5&quot; &quot;#FC8D62&quot;</code></pre>
<div class="sourceCode" id="cb642"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb642-1"><a href="batch.html#cb642-1" tabindex="-1"></a>ENSG00000137124 <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Expr=</span>normexpr<span class="sc">$</span>vst[<span class="st">&quot;ENSG00000137124&quot;</span>,])</span>
<span id="cb642-2"><a href="batch.html#cb642-2" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">sp_boxplot</span>(ENSG00000137124, <span class="at">melted=</span>T, <span class="at">metadata=</span>metadata, <span class="at">xvariable =</span> <span class="st">&quot;conditions&quot;</span>, <span class="at">yvariable =</span> <span class="st">&quot;Expr&quot;</span>, <span class="at">jitter_bp =</span> T, <span class="at">group_variable_for_line =</span> <span class="st">&quot;individual&quot;</span>, <span class="at">xvariable_order =</span> <span class="fu">c</span>(<span class="st">&quot;untrt&quot;</span>,<span class="st">&quot;trt&quot;</span>), <span class="at">manual_color_vector =</span> <span class="st">&quot;Set2&quot;</span>, <span class="at">title=</span><span class="st">&quot;ENSG00000137124&quot;</span>)</span></code></pre></div>
<pre><code>## [1] 1273
## [1] &quot;#66C2A5&quot; &quot;#FC8D62&quot;</code></pre>
<div class="sourceCode" id="cb644"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb644-1"><a href="batch.html#cb644-1" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guide =</span> <span class="st">&#39;collect&#39;</span>)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-325-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><code>DESeq2</code>，<code>edgeR</code>和<code>limma</code>在考虑批次因素鉴定差异基因时基本操作是一致的，上面我们也完成和比较了已知批次的数据的差异基因鉴定。</p>
<p>后续还有2个问题：</p>
<ol style="list-style-type: decimal">
<li>DESeq2不校正表达矩阵自身的值，如果需要用到批次校正后的表达矩阵怎么做？</li>
<li>如果不知道数据是否来源于同一个个体或是否有其他批次因素的影响，怎么处理？</li>
</ol>
</div>
</div>
<div id="batch9" class="section level2 hasAnchor" number="6.9">
<h2><span class="header-section-number">6.9</span> 批次效应未知时如何判断和在差异基因鉴定过程中移除批次效应<a href="batch.html#batch9" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>前面文章讲述了批次信息已知时，在差异基因分析中考虑批次效应的影响可以移除部分基因在个体中不同本底表达水平差异的影响，获得的差异基因倍数方差会变小，可以检测出更多差异基因，理论上也是更可靠的方式。那么如果批次信息未知或记录不完善时怎么处理呢？</p>
<p>这里我们就用到了另一个 R 包<code>sva</code>帮助从数据中预测可能存在的混杂因素包括但不限于批次效应的影响。下面我们实际看下这个包鉴定出的<em>混杂因素与批次效应变量</em>之间是否存在关联？利用预测出的混杂因素作为批次信息校正后结果会有什么变化？</p>
<div id="batch10" class="section level3 hasAnchor" number="6.9.1">
<h3><span class="header-section-number">6.9.1</span> 预测混杂因素(cofounding factors)并在差异基因分析中移除这些因素<a href="batch.html#batch10" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>初始化，定义输入、输出和参数</p>
<div class="sourceCode" id="cb645"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb645-1"><a href="batch.html#cb645-1" tabindex="-1"></a><span class="co"># Prefix for all output file </span></span>
<span id="cb645-2"><a href="batch.html#cb645-2" tabindex="-1"></a>output_prefix <span class="ot">=</span> <span class="st">&quot;result/batch/STAR.simpler.sva_batch&quot;</span></span>
<span id="cb645-3"><a href="batch.html#cb645-3" tabindex="-1"></a></span>
<span id="cb645-4"><a href="batch.html#cb645-4" tabindex="-1"></a><span class="co"># pipelineStar.sh或其它方式生成的reads count 文件，行为基因，列为样品</span></span>
<span id="cb645-5"><a href="batch.html#cb645-5" tabindex="-1"></a>file <span class="ot">=</span> <span class="st">&quot;Count_matrix.xls&quot;</span></span>
<span id="cb645-6"><a href="batch.html#cb645-6" tabindex="-1"></a><span class="co"># 分组信息表</span></span>
<span id="cb645-7"><a href="batch.html#cb645-7" tabindex="-1"></a>metadata <span class="ot">=</span> <span class="st">&quot;metadata.txt&quot;</span></span>
<span id="cb645-8"><a href="batch.html#cb645-8" tabindex="-1"></a><span class="co"># 分组信息所在列名字</span></span>
<span id="cb645-9"><a href="batch.html#cb645-9" tabindex="-1"></a>covariate <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb645-10"><a href="batch.html#cb645-10" tabindex="-1"></a><span class="co"># covariate = &quot;batch&quot;</span></span>
<span id="cb645-11"><a href="batch.html#cb645-11" tabindex="-1"></a>design<span class="ot">=</span><span class="st">&quot;conditions&quot;</span></span>
<span id="cb645-12"><a href="batch.html#cb645-12" tabindex="-1"></a><span class="co"># 输入数据类型，salmon结果或reads count 矩阵</span></span>
<span id="cb645-13"><a href="batch.html#cb645-13" tabindex="-1"></a>type<span class="ot">=</span><span class="st">&quot;readscount&quot;</span></span>
<span id="cb645-14"><a href="batch.html#cb645-14" tabindex="-1"></a><span class="co"># 差异基因参数</span></span>
<span id="cb645-15"><a href="batch.html#cb645-15" tabindex="-1"></a>padj<span class="ot">=</span><span class="fl">0.05</span></span>
<span id="cb645-16"><a href="batch.html#cb645-16" tabindex="-1"></a>log2FC<span class="ot">=</span><span class="dv">1</span></span></code></pre></div>
<p>数据读入和标准化</p>
<div class="sourceCode" id="cb646"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb646-1"><a href="batch.html#cb646-1" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">readscount2deseq</span>(file, metadata, <span class="at">design=</span>design, <span class="at">covariate =</span> covariate)</span></code></pre></div>
<pre><code>## [1] &quot;Read in 32799 genes&quot;
## [1] &quot;23936 genes remained after filtering of genes with all counts less than 4 in all samples.&quot;
## [1] &quot;Perform DESeq on given datasets.&quot;</code></pre>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<div class="sourceCode" id="cb654"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb654-1"><a href="batch.html#cb654-1" tabindex="-1"></a>normexpr <span class="ot">&lt;-</span> <span class="fu">deseq2normalizedExpr</span>(dds, <span class="at">output_prefix=</span>output_prefix, <span class="at">rlog=</span>F, <span class="at">vst=</span>T, <span class="at">design=</span>design)</span></code></pre></div>
<pre><code>## [1] &quot;Output normalized counts&quot;
## [1] &quot;Output vst transformed normalized counts&quot;</code></pre>
</div>
<div id="batch11" class="section level3 hasAnchor" number="6.9.2">
<h3><span class="header-section-number">6.9.2</span> 预测可能存在的混杂因素<a href="batch.html#batch11" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>下面的方式也可以 (<code>svaseq</code> 是在 <code>sva</code> 的基础上对数据做了一个 <code>log</code> 转换；如果处理的是芯片数据，通常已经做过 log 换，直接使用 <code>sva</code> 即可)。</p>
<div class="sourceCode" id="cb656"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb656-1"><a href="batch.html#cb656-1" tabindex="-1"></a><span class="co"># 获取标准化后的表达矩阵</span></span>
<span id="cb656-2"><a href="batch.html#cb656-2" tabindex="-1"></a>dat <span class="ot">&lt;-</span> normexpr<span class="sc">$</span>vst</span>
<span id="cb656-3"><a href="batch.html#cb656-3" tabindex="-1"></a><span class="co"># 根据关键生物表型构建设计矩阵</span></span>
<span id="cb656-4"><a href="batch.html#cb656-4" tabindex="-1"></a>mod  <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="fu">as.formula</span>(<span class="fu">paste0</span>(<span class="st">&quot;~ &quot;</span>, design)), <span class="fu">colData</span>(dds))</span>
<span id="cb656-5"><a href="batch.html#cb656-5" tabindex="-1"></a><span class="co"># 构建对照设计矩阵</span></span>
<span id="cb656-6"><a href="batch.html#cb656-6" tabindex="-1"></a>mod0 <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> <span class="dv">1</span>, <span class="fu">colData</span>(dds))</span>
<span id="cb656-7"><a href="batch.html#cb656-7" tabindex="-1"></a><span class="co"># 指定混杂因素的数目为 2，也可以让 sva 自己预测</span></span>
<span id="cb656-8"><a href="batch.html#cb656-8" tabindex="-1"></a>svseq2 <span class="ot">&lt;-</span> <span class="fu">sva</span>(dat, mod, mod0)</span></code></pre></div>
<pre><code>## Number of significant surrogate variables is:  3 
## Iteration (out of 5 ):1  2  3  4  5</code></pre>
<div class="sourceCode" id="cb658"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb658-1"><a href="batch.html#cb658-1" tabindex="-1"></a>svs <span class="ot">&lt;-</span> svseq2<span class="sc">$</span>sv</span>
<span id="cb658-2"><a href="batch.html#cb658-2" tabindex="-1"></a><span class="fu">colnames</span>(svs) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;SV&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(svs))</span>
<span id="cb658-3"><a href="batch.html#cb658-3" tabindex="-1"></a>svs</span></code></pre></div>
<pre><code>##               SV1        SV2         SV3
## [1,] -0.158515434 -0.4622145  0.32822946
## [2,] -0.013810298 -0.1806797 -0.55487160
## [3,] -0.381566045  0.4357423  0.08352428
## [4,]  0.567268826  0.1737819  0.16589595
## [5,] -0.181616074 -0.4695836  0.38642058
## [6,]  0.004300293 -0.1796549 -0.61360292
## [7,] -0.398059133  0.4960287  0.06242022
## [8,]  0.561997865  0.1865798  0.14198402</code></pre>
<p>添加预测出的<code>Surrogate variable</code>属性到dds对象</p>
<div class="sourceCode" id="cb660"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb660-1"><a href="batch.html#cb660-1" tabindex="-1"></a><span class="fu">colData</span>(dds) <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">colData</span>(dds), svs)</span>
<span id="cb660-2"><a href="batch.html#cb660-2" tabindex="-1"></a><span class="fu">design</span>(dds) <span class="ot">&lt;-</span> <span class="fu">as.formula</span>(<span class="fu">paste</span>(<span class="st">&quot;~&quot;</span>, <span class="fu">paste</span>(<span class="fu">colnames</span>(svs), <span class="at">collapse =</span> <span class="st">&quot;+&quot;</span>), <span class="st">&quot;+&quot;</span>, design))</span>
<span id="cb660-3"><a href="batch.html#cb660-3" tabindex="-1"></a><span class="co"># 基于预测出的混杂因素再次进行分析</span></span>
<span id="cb660-4"><a href="batch.html#cb660-4" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span></code></pre></div>
<pre><code>## using pre-existing size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## found already estimated dispersions, replacing these</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
</div>
<div id="可视化展示预测出的surrogate-variable属性与已知的批次信息的关系" class="section level3 hasAnchor" number="6.9.3">
<h3><span class="header-section-number">6.9.3</span> 可视化展示预测出的<code>Surrogate variable</code>属性与已知的批次信息的关系<a href="batch.html#%E5%8F%AF%E8%A7%86%E5%8C%96%E5%B1%95%E7%A4%BA%E9%A2%84%E6%B5%8B%E5%87%BA%E7%9A%84surrogate-variable%E5%B1%9E%E6%80%A7%E4%B8%8E%E5%B7%B2%E7%9F%A5%E7%9A%84%E6%89%B9%E6%AC%A1%E4%BF%A1%E6%81%AF%E7%9A%84%E5%85%B3%E7%B3%BB" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb668"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb668-1"><a href="batch.html#cb668-1" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(dds))</span>
<span id="cb668-2"><a href="batch.html#cb668-2" tabindex="-1"></a>plot_data<span class="sc">$</span>Sample <span class="ot">&lt;-</span> <span class="fu">rownames</span>(plot_data)</span>
<span id="cb668-3"><a href="batch.html#cb668-3" tabindex="-1"></a><span class="fu">head</span>(plot_data)</span></code></pre></div>
<pre><code>##               conditions individual sizeFactor          SV1        SV2
## untrt_N61311       untrt     N61311  1.0211325 -0.158515434 -0.4622145
## untrt_N052611      untrt    N052611  1.1803986 -0.013810298 -0.1806797
## untrt_N080611      untrt    N080611  1.1796083 -0.381566045  0.4357423
## untrt_N061011      untrt    N061011  0.9232642  0.567268826  0.1737819
## trt_N61311           trt     N61311  0.8939275 -0.181616074 -0.4695836
## trt_N052611          trt    N052611  0.6709229  0.004300293 -0.1796549
##                       SV3        Sample
## untrt_N61311   0.32822946  untrt_N61311
## untrt_N052611 -0.55487160 untrt_N052611
## untrt_N080611  0.08352428 untrt_N080611
## untrt_N061011  0.16589595 untrt_N061011
## trt_N61311     0.38642058    trt_N61311
## trt_N052611   -0.61360292   trt_N052611</code></pre>
<div class="sourceCode" id="cb670"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb670-1"><a href="batch.html#cb670-1" tabindex="-1"></a><span class="fu">sp_writeTable</span>(plot_data, <span class="at">file=</span><span class="st">&quot;result/batch/metadata2.txt&quot;</span>)</span></code></pre></div>
<p>从下图可以看出，预测出的混杂因素<code>SV1</code>, <code>SV2</code>与样品来源的个体信息 (<code>individual</code>)还是比较一致的。</p>
<div class="sourceCode" id="cb671"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb671-1"><a href="batch.html#cb671-1" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(plot_data, <span class="fu">aes</span>(<span class="at">x=</span>SV1, <span class="at">y=</span>SV2, <span class="at">color=</span>conditions, <span class="at">shape=</span>individual)) <span class="sc">+</span> </span>
<span id="cb671-2"><a href="batch.html#cb671-2" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label=</span>Sample), <span class="at">show.legend =</span> F)</span>
<span id="cb671-3"><a href="batch.html#cb671-3" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(plot_data, <span class="fu">aes</span>(<span class="at">x=</span>SV1, <span class="at">y=</span>SV3, <span class="at">color=</span>conditions, <span class="at">shape=</span>individual)) <span class="sc">+</span> </span>
<span id="cb671-4"><a href="batch.html#cb671-4" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label=</span>Sample), <span class="at">show.legend =</span> F)</span>
<span id="cb671-5"><a href="batch.html#cb671-5" tabindex="-1"></a></span>
<span id="cb671-6"><a href="batch.html#cb671-6" tabindex="-1"></a>p1<span class="sc">+</span>p2</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-331-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="基于预测出的混杂因素再次进行差异分析" class="section level3 hasAnchor" number="6.9.4">
<h3><span class="header-section-number">6.9.4</span> 基于预测出的混杂因素再次进行差异分析<a href="batch.html#%E5%9F%BA%E4%BA%8E%E9%A2%84%E6%B5%8B%E5%87%BA%E7%9A%84%E6%B7%B7%E6%9D%82%E5%9B%A0%E7%B4%A0%E5%86%8D%E6%AC%A1%E8%BF%9B%E8%A1%8C%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>获得差异基因文件<code>STAR.simpler.sva_batch.DESeq2.all.DE.txt</code>和其它可视化图表（暂时忽略）。</p>
<div class="sourceCode" id="cb672"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb672-1"><a href="batch.html#cb672-1" tabindex="-1"></a><span class="fu">multipleGroupDEgenes</span>(dds, <span class="at">design=</span>design, <span class="at">output_prefix=</span>output_prefix, <span class="at">padj=</span>padj, <span class="at">log2FC=</span>log2FC, <span class="at">normalized_counts=</span>normexpr)</span></code></pre></div>
<pre><code>## [1] &quot;DE genes between untrt trt&quot;
## [1] &quot;conditions&quot; &quot;untrt&quot;      &quot;trt&quot;</code></pre>
<pre><code>## Saving 17.8 x 12.7 cm image</code></pre>
</div>
<div id="batch12" class="section level3 hasAnchor" number="6.9.5">
<h3><span class="header-section-number">6.9.5</span> 比较批次校正前、已知批次校正后和预测的批次校正后差异基因变化<a href="batch.html#batch12" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>根据已知批次信息校正后差异基因数目变多了，上调了<em>99</em>个，下调多了<em>61</em>个。根据预测的混杂因素校正后，上调多了<em>96</em>个，下调多了<em>55</em>个。从数目上看，根据已知批次和预测的混杂因素获得的差异基因是基本已知的。</p>
<div class="sourceCode" id="cb675"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb675-1"><a href="batch.html#cb675-1" tabindex="-1"></a>de_before_batch <span class="ot">=</span> <span class="fu">sp_readTable</span>(<span class="st">&quot;result/batch/STAR.simpler.DESeq2.all.DE.txt&quot;</span>, <span class="at">header=</span>F)</span>
<span id="cb675-2"><a href="batch.html#cb675-2" tabindex="-1"></a>de_before_batch<span class="sc">$</span>V2 <span class="ot">=</span> <span class="fu">paste</span>(<span class="st">&quot;Before_batch&quot;</span>,de_before_batch<span class="sc">$</span>V2,<span class="at">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb675-3"><a href="batch.html#cb675-3" tabindex="-1"></a><span class="fu">table</span>(de_before_batch<span class="sc">$</span>V2)</span></code></pre></div>
<pre><code>## 
## Before_batch_untrt._higherThan_.trt  Before_batch_untrt._lowerThan_.trt 
##                                 398                                 466</code></pre>
<div class="sourceCode" id="cb677"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb677-1"><a href="batch.html#cb677-1" tabindex="-1"></a>de_after_known_batch <span class="ot">=</span> <span class="fu">sp_readTable</span>(<span class="st">&quot;result/batch/STAR.simpler.batch.DESeq2.all.DE.txt&quot;</span>, <span class="at">header=</span>F)</span>
<span id="cb677-2"><a href="batch.html#cb677-2" tabindex="-1"></a>de_after_known_batch<span class="sc">$</span>V2 <span class="ot">=</span> <span class="fu">paste</span>(<span class="st">&quot;After_known_batch&quot;</span>,de_after_known_batch<span class="sc">$</span>V2,<span class="at">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb677-3"><a href="batch.html#cb677-3" tabindex="-1"></a><span class="fu">table</span>(de_after_known_batch<span class="sc">$</span>V2)</span></code></pre></div>
<pre><code>## 
## After_known_batch_untrt._higherThan_.trt 
##                                      497 
##  After_known_batch_untrt._lowerThan_.trt 
##                                      527</code></pre>
<div class="sourceCode" id="cb679"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb679-1"><a href="batch.html#cb679-1" tabindex="-1"></a>de_after_sva_batch <span class="ot">=</span> <span class="fu">sp_readTable</span>(<span class="st">&quot;result/batch/STAR.simpler.sva_batch.DESeq2.all.DE.txt&quot;</span>, <span class="at">header=</span>F)</span>
<span id="cb679-2"><a href="batch.html#cb679-2" tabindex="-1"></a>de_after_sva_batch<span class="sc">$</span>V2 <span class="ot">=</span> <span class="fu">paste</span>(<span class="st">&quot;After_sva_batch&quot;</span>,de_after_sva_batch<span class="sc">$</span>V2,<span class="at">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb679-3"><a href="batch.html#cb679-3" tabindex="-1"></a><span class="fu">table</span>(de_after_sva_batch<span class="sc">$</span>V2)</span></code></pre></div>
<pre><code>## 
## After_sva_batch_untrt._higherThan_.trt  After_sva_batch_untrt._lowerThan_.trt 
##                                    500                                    516</code></pre>
<p>画个Venn图，看下哪些基因是新增的差异基因，哪些基因批次校正后没差异了。</p>
<div class="sourceCode" id="cb681"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb681-1"><a href="batch.html#cb681-1" tabindex="-1"></a>all_de <span class="ot">=</span> <span class="fu">rbind</span>(de_before_batch, de_after_known_batch, de_after_sva_batch)</span>
<span id="cb681-2"><a href="batch.html#cb681-2" tabindex="-1"></a><span class="co"># 随机查看6行，信息代表更全面</span></span>
<span id="cb681-3"><a href="batch.html#cb681-3" tabindex="-1"></a>all_de[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(all_de),<span class="dv">6</span>),]</span></code></pre></div>
<pre><code>##                   V1                                       V2
## 1255 ENSG00000278029 After_known_batch_untrt._higherThan_.trt
## 50   ENSG00000012048      Before_batch_untrt._higherThan_.trt
## 2330 ENSG00000055813   After_sva_batch_untrt._higherThan_.trt
## 1276 ENSG00000181350 After_known_batch_untrt._higherThan_.trt
## 330  ENSG00000187634      Before_batch_untrt._higherThan_.trt
## 222  ENSG00000126785      Before_batch_untrt._higherThan_.trt</code></pre>
<div class="sourceCode" id="cb683"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb683-1"><a href="batch.html#cb683-1" tabindex="-1"></a><span class="co"># 结果存储到文件中</span></span>
<span id="cb683-2"><a href="batch.html#cb683-2" tabindex="-1"></a><span class="fu">sp_writeTable</span>(all_de, <span class="at">file=</span><span class="st">&quot;result/batch/Compare_de_gene_beofore_and_after_known_sva_batch.txt&quot;</span>, <span class="at">keep_rownames =</span> F, <span class="at">col.names =</span> F)</span></code></pre></div>
<p>一个方式是采用代码，直接出图</p>
<div class="sourceCode" id="cb684"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb684-1"><a href="batch.html#cb684-1" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(VennDiagram))</span>
<span id="cb684-2"><a href="batch.html#cb684-2" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(grid))</span>
<span id="cb684-3"><a href="batch.html#cb684-3" tabindex="-1"></a><span class="fu">sp_vennDiagram</span>(all_de, <span class="at">label3=</span><span class="st">&quot;Before_batch_untrt._higherThan_.trt&quot;</span>,</span>
<span id="cb684-4"><a href="batch.html#cb684-4" tabindex="-1"></a>               <span class="at">label1=</span><span class="st">&quot;After_known_batch_untrt._higherThan_.trt&quot;</span>,</span>
<span id="cb684-5"><a href="batch.html#cb684-5" tabindex="-1"></a>               <span class="at">label2=</span><span class="st">&quot;After_sva_batch_untrt._higherThan_.trt&quot;</span>)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-337-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>这里还是采用在线工具<a href="https://www.bic.ac.cn/EVenn/#/" class="uri">https://www.bic.ac.cn/EVenn/#/</a> 来做，能直接获得每个子集的基因，准备在线工具所需的文件，一个两列格式的文件：第一列为基因名，第二列为基因的上下调状态。</p>
<p>拷贝文件数据到网站数据输入处 （操作就不演示了看上一篇文章）：</p>
<p><img src="images/batch3_untrt_higher_venn.png" /></p>
<p>从untrt上调基因Venn图可以看出，校正已知批次信息后既有新增的untrt上调差异基因，又丢失了之前的一部分untrt上调差异基因；校正预测的混杂因素后，大部分新增差异基因都与已知批次信息校正后的结果一致，但新增untrt上调差异基因少。</p>
<p><img src="images/batch3_untrt_lower_venn.png" /></p>
<p>从untrt下调基因Venn图可以看出，校正预测的混杂因素后，新增39个差异基因；批次校正前鉴定为存在差异的40个基因在校正后被认为是非差异显著基因。</p>
<p>下面还是从这些基因的表达模式上看是否可以找到一些线索？</p>
<p>下图比对绘出了<strong>7</strong>种不同类型untrt上调的差异基因中随机选取1个绘制的表达模式比较图。</p>
<div class="sourceCode" id="cb685"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb685-1"><a href="batch.html#cb685-1" tabindex="-1"></a>untrt_up_genes <span class="ot">&lt;-</span> <span class="st">&quot;Name;Type</span></span>
<span id="cb685-2"><a href="batch.html#cb685-2" tabindex="-1"></a><span class="st">ENSG00000159674;SVA_batch_specific</span></span>
<span id="cb685-3"><a href="batch.html#cb685-3" tabindex="-1"></a><span class="st">ENSG00000220563;SVA_batch_uncorrect_common</span></span>
<span id="cb685-4"><a href="batch.html#cb685-4" tabindex="-1"></a><span class="st">ENSG00000109689;Uncorrect_specific</span></span>
<span id="cb685-5"><a href="batch.html#cb685-5" tabindex="-1"></a><span class="st">ENSG00000162407;Known_batch_uncorrect_common</span></span>
<span id="cb685-6"><a href="batch.html#cb685-6" tabindex="-1"></a><span class="st">ENSG00000184254;Known_batch_specific</span></span>
<span id="cb685-7"><a href="batch.html#cb685-7" tabindex="-1"></a><span class="st">ENSG00000163394;Known_batch_SVA_batch_common</span></span>
<span id="cb685-8"><a href="batch.html#cb685-8" tabindex="-1"></a><span class="st">ENSG00000178695;All_common&quot;</span></span>
<span id="cb685-9"><a href="batch.html#cb685-9" tabindex="-1"></a></span>
<span id="cb685-10"><a href="batch.html#cb685-10" tabindex="-1"></a>untrt_up_genes <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">text=</span>untrt_up_genes, <span class="at">sep=</span><span class="st">&quot;;&quot;</span>, <span class="at">header=</span>T, <span class="at">row.names=</span><span class="cn">NULL</span>)</span>
<span id="cb685-11"><a href="batch.html#cb685-11" tabindex="-1"></a></span>
<span id="cb685-12"><a href="batch.html#cb685-12" tabindex="-1"></a>untrt_up_genes_expr <span class="ot">&lt;-</span> <span class="fu">merge</span>(untrt_up_genes, normexpr<span class="sc">$</span>vst, <span class="at">by.x=</span><span class="st">&quot;Name&quot;</span>, <span class="at">by.y=</span><span class="dv">0</span>, <span class="at">all.x=</span>T)</span>
<span id="cb685-13"><a href="batch.html#cb685-13" tabindex="-1"></a></span>
<span id="cb685-14"><a href="batch.html#cb685-14" tabindex="-1"></a>untrt_up_genes_expr_long <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(untrt_up_genes_expr, <span class="at">id_vars=</span><span class="fu">c</span>(<span class="st">&quot;Name&quot;</span>,<span class="st">&quot;Type&quot;</span>), </span>
<span id="cb685-15"><a href="batch.html#cb685-15" tabindex="-1"></a>                                 <span class="at">variable.name=</span><span class="st">&quot;Sample&quot;</span>, <span class="at">value.name =</span> <span class="st">&quot;Expr&quot;</span>)</span></code></pre></div>
<pre><code>## Using Name, Type as id variables</code></pre>
<div class="sourceCode" id="cb687"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb687-1"><a href="batch.html#cb687-1" tabindex="-1"></a><span class="fu">head</span>(untrt_up_genes_expr_long)</span></code></pre></div>
<pre><code>##              Name                         Type       Sample      Expr
## 1 ENSG00000109689           Uncorrect_specific untrt_N61311 10.591770
## 2 ENSG00000159674           SVA_batch_specific untrt_N61311 13.704637
## 3 ENSG00000162407 Known_batch_uncorrect_common untrt_N61311 14.753902
## 4 ENSG00000163394 Known_batch_SVA_batch_common untrt_N61311  9.829457
## 5 ENSG00000178695                   All_common untrt_N61311 12.205745
## 6 ENSG00000184254         Known_batch_specific untrt_N61311  9.598665</code></pre>
<div class="sourceCode" id="cb689"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb689-1"><a href="batch.html#cb689-1" tabindex="-1"></a><span class="co"># metadata$Sample = rownames(metadata)</span></span>
<span id="cb689-2"><a href="batch.html#cb689-2" tabindex="-1"></a></span>
<span id="cb689-3"><a href="batch.html#cb689-3" tabindex="-1"></a>ImageGP<span class="sc">:::</span><span class="fu">sp_writeTable</span>(untrt_up_genes_expr_long, <span class="at">file=</span><span class="st">&quot;1&quot;</span>, <span class="at">keep_rownames =</span>F)</span>
<span id="cb689-4"><a href="batch.html#cb689-4" tabindex="-1"></a></span>
<span id="cb689-5"><a href="batch.html#cb689-5" tabindex="-1"></a><span class="fu">sp_boxplot</span>(untrt_up_genes_expr_long, <span class="at">melted=</span>T, <span class="at">metadata=</span>plot_data, </span>
<span id="cb689-6"><a href="batch.html#cb689-6" tabindex="-1"></a>           <span class="at">xvariable =</span> <span class="st">&quot;conditions&quot;</span>, <span class="at">yvariable =</span> <span class="st">&quot;Expr&quot;</span>, <span class="at">jitter_bp =</span> T, </span>
<span id="cb689-7"><a href="batch.html#cb689-7" tabindex="-1"></a>           <span class="at">group_variable_for_line =</span> <span class="st">&quot;individual&quot;</span>, </span>
<span id="cb689-8"><a href="batch.html#cb689-8" tabindex="-1"></a>           <span class="at">facet_variable =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">facet_scales=</span><span class="st">&quot;free_y&quot;</span>, <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.7</span>,<span class="fl">0.15</span>), </span>
<span id="cb689-9"><a href="batch.html#cb689-9" tabindex="-1"></a>           <span class="at">x_label=</span><span class="st">&quot;&quot;</span>,<span class="at">manual_color_vector =</span> <span class="st">&quot;Set2&quot;</span>) <span class="sc">+</span> </span>
<span id="cb689-10"><a href="batch.html#cb689-10" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.direction =</span> <span class="st">&quot;horizontal&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;#66C2A5&quot; &quot;#FC8D62&quot;</code></pre>
<p><img src="R_course_files/figure-html/unnamed-chunk-338-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><code>All_common</code>代表了差异倍数特别大的基因，不论是否校正都可以检测出差异；不同类型批次信息校正后被检测视为差异的基因都有表达的本底差异；<code>Uncorrect_specific</code>的基因本底表达无固定模式。</p>
<p>下图比对绘出了<strong>7</strong>种不同类型untrt下调的差异基因表达分布，基本结论与上图类似。</p>
<div class="sourceCode" id="cb691"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb691-1"><a href="batch.html#cb691-1" tabindex="-1"></a>untrt_down_genes <span class="ot">&lt;-</span> <span class="st">&quot;Name;Type</span></span>
<span id="cb691-2"><a href="batch.html#cb691-2" tabindex="-1"></a><span class="st">ENSG00000188257;SVA_batch_specific</span></span>
<span id="cb691-3"><a href="batch.html#cb691-3" tabindex="-1"></a><span class="st">ENSG00000117479;SVA_batch_uncorrect_common</span></span>
<span id="cb691-4"><a href="batch.html#cb691-4" tabindex="-1"></a><span class="st">ENSG00000137124;Uncorrect_specific</span></span>
<span id="cb691-5"><a href="batch.html#cb691-5" tabindex="-1"></a><span class="st">ENSG00000254109;Known_batch_uncorrect_common</span></span>
<span id="cb691-6"><a href="batch.html#cb691-6" tabindex="-1"></a><span class="st">ENSG00000113578;Known_batch_specific</span></span>
<span id="cb691-7"><a href="batch.html#cb691-7" tabindex="-1"></a><span class="st">ENSG00000221866;Known_batch_SVA_batch_common</span></span>
<span id="cb691-8"><a href="batch.html#cb691-8" tabindex="-1"></a><span class="st">ENSG00000152583;All_common&quot;</span></span>
<span id="cb691-9"><a href="batch.html#cb691-9" tabindex="-1"></a></span>
<span id="cb691-10"><a href="batch.html#cb691-10" tabindex="-1"></a>untrt_down_genes <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">text=</span>untrt_down_genes, <span class="at">sep=</span><span class="st">&quot;;&quot;</span>, <span class="at">header=</span>T, <span class="at">row.names=</span><span class="cn">NULL</span>)</span>
<span id="cb691-11"><a href="batch.html#cb691-11" tabindex="-1"></a></span>
<span id="cb691-12"><a href="batch.html#cb691-12" tabindex="-1"></a>untrt_down_genes_expr <span class="ot">&lt;-</span> <span class="fu">merge</span>(untrt_down_genes, normexpr<span class="sc">$</span>vst, <span class="at">by.x=</span><span class="st">&quot;Name&quot;</span>, <span class="at">by.y=</span><span class="dv">0</span>, <span class="at">all.x=</span>T)</span>
<span id="cb691-13"><a href="batch.html#cb691-13" tabindex="-1"></a></span>
<span id="cb691-14"><a href="batch.html#cb691-14" tabindex="-1"></a>untrt_down_genes_expr_long <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(untrt_down_genes_expr, <span class="at">id_vars=</span><span class="fu">c</span>(<span class="st">&quot;Name&quot;</span>,<span class="st">&quot;Type&quot;</span>), </span>
<span id="cb691-15"><a href="batch.html#cb691-15" tabindex="-1"></a>                                 <span class="at">variable.name=</span><span class="st">&quot;Sample&quot;</span>, <span class="at">value.name =</span> <span class="st">&quot;Expr&quot;</span>)</span></code></pre></div>
<pre><code>## Using Name, Type as id variables</code></pre>
<div class="sourceCode" id="cb693"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb693-1"><a href="batch.html#cb693-1" tabindex="-1"></a><span class="fu">head</span>(untrt_down_genes_expr_long)</span></code></pre></div>
<pre><code>##              Name                         Type       Sample     Expr
## 1 ENSG00000113578         Known_batch_specific untrt_N61311 6.964600
## 2 ENSG00000117479   SVA_batch_uncorrect_common untrt_N61311 7.927545
## 3 ENSG00000137124           Uncorrect_specific untrt_N61311 8.895369
## 4 ENSG00000152583                   All_common untrt_N61311 7.026723
## 5 ENSG00000188257           SVA_batch_specific untrt_N61311 6.857589
## 6 ENSG00000221866 Known_batch_SVA_batch_common untrt_N61311 7.795699</code></pre>
<div class="sourceCode" id="cb695"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb695-1"><a href="batch.html#cb695-1" tabindex="-1"></a>metadata<span class="sc">$</span>Sample <span class="ot">=</span> <span class="fu">rownames</span>(metadata)</span>
<span id="cb695-2"><a href="batch.html#cb695-2" tabindex="-1"></a></span>
<span id="cb695-3"><a href="batch.html#cb695-3" tabindex="-1"></a><span class="fu">sp_boxplot</span>(untrt_down_genes_expr_long, <span class="at">melted=</span>T, <span class="at">metadata=</span>plot_data, </span>
<span id="cb695-4"><a href="batch.html#cb695-4" tabindex="-1"></a>           <span class="at">xvariable =</span> <span class="st">&quot;conditions&quot;</span>, <span class="at">yvariable =</span> <span class="st">&quot;Expr&quot;</span>, <span class="at">jitter_bp =</span> T, </span>
<span id="cb695-5"><a href="batch.html#cb695-5" tabindex="-1"></a>           <span class="at">group_variable_for_line =</span> <span class="st">&quot;individual&quot;</span>, </span>
<span id="cb695-6"><a href="batch.html#cb695-6" tabindex="-1"></a>           <span class="at">facet_variable =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">facet_scales=</span><span class="st">&quot;free_y&quot;</span>, <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.7</span>,<span class="fl">0.15</span>), </span>
<span id="cb695-7"><a href="batch.html#cb695-7" tabindex="-1"></a>           <span class="at">x_label=</span><span class="st">&quot;&quot;</span>,<span class="at">manual_color_vector =</span> <span class="st">&quot;Set2&quot;</span>) <span class="sc">+</span> </span>
<span id="cb695-8"><a href="batch.html#cb695-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.direction =</span> <span class="st">&quot;horizontal&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;#66C2A5&quot; &quot;#FC8D62&quot;</code></pre>
<p><img src="R_course_files/figure-html/unnamed-chunk-339-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>额外的一个信息是<code>SVA_batch_speific</code>中红色和绿色个体本底表达区分不明显。这可能是基于SVA预测的混杂因素与已知的批次因素校正后结果有差异的一个原因 (这两个个体的SV值很接近)。</p>
<p>另外一个导致SVA预测的批次与已知的批次效应校正后结果不同的原因也可能是我们只让<code>SVA</code>预测了2个混杂因素。留下2个去探索的问题，欢迎留言或投稿讨论：</p>
<ol style="list-style-type: decimal">
<li>如果不设置只返回两个混杂因素，实际SVA会判断出存在3个混杂因素，全部混杂因素都考虑进去结果会有什么变化呢？</li>
<li>上面是取了单个基因查看其表达模式，还可以进一步比较不同子集的基因表达水平、差异倍数、FDR、差异倍数方差的整体分布，分析受影响的主要是哪些类型的基因？</li>
</ol>
</div>
</div>
<div id="batch13" class="section level2 hasAnchor" number="6.10">
<h2><span class="header-section-number">6.10</span> 直接校正表达矩阵<a href="batch.html#batch13" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>处理批次因素最好的方式还是如前面所述将其整合到差异基因鉴定模型中，降低批次因素带来的模型残差的自由度。但一些下游分析，比如数据可视化，也需要直接移除效应影响的数据来展示，这时可以使用<code>ComBat</code>或<code>removeBatchEffect</code>函数来处理。</p>
<p>输入数据，标准化且log转换后的数据 <code>STAR.simpler.DESeq2.normalized.vst.txt</code></p>
<pre><code>id  untrt_N61311    untrt_N052611   untrt_N080611   untrt_N061011   trt_N61311  trt_N052611 trt_N080611 trt_N061011
ENSG00000115414 18.02   18.62   17.83   18.45   17.95   18.54   18.15   18.50 
ENSG00000011465 17.79   18.36   17.96   18.52   17.79   18.23   17.84   18.47 
ENSG00000091986 17.15   17.74   16.41   17.59   17.27   17.79   16.88   17.76 
ENSG00000103888 15.56   16.90   15.88   16.42   15.94   17.43   17.38   17.05 </code></pre>
<p>包含已知批次信息和预测的批次信息的样本属性文件 <code>metadata2</code></p>
<pre><code>Samp    conditions  individual  sizeFactor  SV1 SV2 SV3
untrt_N61311    untrt   N61311  1.0211325   -0.10060313 -0.4943517  -0.31643389
untrt_N052611   untrt   N052611 1.1803986   0.01827734  -0.1701068  0.58841464
untrt_N080611   untrt   N080611 1.1796083   -0.42949195 0.3756338   -0.08929556
untrt_N061011   untrt   N061011 0.9232642   0.53452392  0.2413738   -0.17649091
trt_N61311  trt N61311  0.8939275   -0.12535603 -0.4956603  -0.36550102
trt_N052611 trt N052611 0.6709229   0.03588273  -0.151201   0.5914179
trt_N080611 trt N080611 1.3967665   -0.46668403 0.4413431   -0.07016903
trt_N061011 trt N061011 0.9462307   0.53345114  0.2529692   -0.16194213</code></pre>
<div id="读入标准化后的表达矩阵和样品信息表" class="section level3 hasAnchor" number="6.10.1">
<h3><span class="header-section-number">6.10.1</span> 读入标准化后的表达矩阵和样品信息表<a href="batch.html#%E8%AF%BB%E5%85%A5%E6%A0%87%E5%87%86%E5%8C%96%E5%90%8E%E7%9A%84%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5%E5%92%8C%E6%A0%B7%E5%93%81%E4%BF%A1%E6%81%AF%E8%A1%A8" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb699"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb699-1"><a href="batch.html#cb699-1" tabindex="-1"></a>expr_File <span class="ot">&lt;-</span> <span class="st">&quot;result/batch/STAR.simpler.DESeq2.normalized.vst.txt&quot;</span></span>
<span id="cb699-2"><a href="batch.html#cb699-2" tabindex="-1"></a>expr_mat <span class="ot">&lt;-</span> <span class="fu">sp_readTable</span>(expr_File, <span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb699-3"><a href="batch.html#cb699-3" tabindex="-1"></a></span>
<span id="cb699-4"><a href="batch.html#cb699-4" tabindex="-1"></a><span class="fu">head</span>(expr_mat)</span></code></pre></div>
<pre><code>##                 untrt_N61311 untrt_N052611 untrt_N080611 untrt_N061011
## ENSG00000096060     8.377527      8.647174      9.193881      7.929484
## ENSG00000109906     5.835168      5.852500      5.601434      6.188280
## ENSG00000211445     9.931563     11.412991     10.507340     10.939324
## ENSG00000112936     8.351378      8.341983      9.781321      8.241310
## ENSG00000152583     7.026723      7.284537      7.285092      7.309670
## ENSG00000101347    10.705358     10.857668     10.479317     10.892392
##                 trt_N61311 trt_N052611 trt_N080611 trt_N061011
## ENSG00000096060  12.353764   12.508120   11.965273    12.10153
## ENSG00000109906   9.797577    9.461018    8.932876    10.26135
## ENSG00000211445  14.101096   15.127159   13.509829    14.68201
## ENSG00000112936  12.421820   10.489734   12.339928    11.14178
## ENSG00000152583  11.194267   10.831855   10.498043    11.18172
## ENSG00000101347  14.221389   14.836149   13.630472    15.04158</code></pre>
<div class="sourceCode" id="cb701"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb701-1"><a href="batch.html#cb701-1" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="st">&quot;result/batch/metadata2.txt&quot;</span></span>
<span id="cb701-2"><a href="batch.html#cb701-2" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">sp_readTable</span>(metadata, <span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb701-3"><a href="batch.html#cb701-3" tabindex="-1"></a><span class="fu">head</span>(metadata)</span></code></pre></div>
<pre><code>##               conditions individual sizeFactor          SV1        SV2
## untrt_N61311       untrt     N61311  1.0211325 -0.158515434 -0.4622145
## untrt_N052611      untrt    N052611  1.1803986 -0.013810298 -0.1806797
## untrt_N080611      untrt    N080611  1.1796083 -0.381566045  0.4357423
## untrt_N061011      untrt    N061011  0.9232642  0.567268826  0.1737819
## trt_N61311           trt     N61311  0.8939275 -0.181616074 -0.4695836
## trt_N052611          trt    N052611  0.6709229  0.004300293 -0.1796549
##                       SV3        Sample
## untrt_N61311   0.32822946  untrt_N61311
## untrt_N052611 -0.55487160 untrt_N052611
## untrt_N080611  0.08352428 untrt_N080611
## untrt_N061011  0.16589595 untrt_N061011
## trt_N61311     0.38642058    trt_N61311
## trt_N052611   -0.61360292   trt_N052611</code></pre>
</div>
<div id="batch14" class="section level3 hasAnchor" number="6.10.2">
<h3><span class="header-section-number">6.10.2</span> 使用ComBat校正<a href="batch.html#batch14" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>ComBat校正时考虑生物分组信息</p>
<div class="sourceCode" id="cb703"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb703-1"><a href="batch.html#cb703-1" tabindex="-1"></a>biological_group <span class="ot">=</span> <span class="st">&quot;conditions&quot;</span></span>
<span id="cb703-2"><a href="batch.html#cb703-2" tabindex="-1"></a>batch <span class="ot">=</span> <span class="st">&quot;individual&quot;</span></span>
<span id="cb703-3"><a href="batch.html#cb703-3" tabindex="-1"></a></span>
<span id="cb703-4"><a href="batch.html#cb703-4" tabindex="-1"></a>metadata[[biological_group]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(metadata[[biological_group]])</span>
<span id="cb703-5"><a href="batch.html#cb703-5" tabindex="-1"></a>metadata[[batch]] <span class="ot">&lt;-</span> <span class="fu">factor</span>(metadata[[batch]])</span>
<span id="cb703-6"><a href="batch.html#cb703-6" tabindex="-1"></a></span>
<span id="cb703-7"><a href="batch.html#cb703-7" tabindex="-1"></a><span class="co"># 模型中引入关注的生物变量和其它非批次变量，保留生物差异和非批次差异</span></span>
<span id="cb703-8"><a href="batch.html#cb703-8" tabindex="-1"></a>modcombat <span class="ot">=</span> <span class="fu">model.matrix</span>(<span class="fu">as.formula</span>(<span class="fu">paste</span>(<span class="st">&#39;~&#39;</span>, biological_group, <span class="at">sep=</span><span class="st">&quot; &quot;</span>)), <span class="at">data=</span></span>
<span id="cb703-9"><a href="batch.html#cb703-9" tabindex="-1"></a>metadata)</span>
<span id="cb703-10"><a href="batch.html#cb703-10" tabindex="-1"></a></span>
<span id="cb703-11"><a href="batch.html#cb703-11" tabindex="-1"></a><span class="co"># ComBat需要的是matrix</span></span>
<span id="cb703-12"><a href="batch.html#cb703-12" tabindex="-1"></a>expr_mat_batch_correct <span class="ot">&lt;-</span> <span class="fu">ComBat</span>(<span class="at">dat=</span><span class="fu">as.matrix</span>(expr_mat), <span class="at">batch=</span>metadata[[batch]], <span class="at">mod=</span>modcombat)</span></code></pre></div>
<pre><code>## Found 2538 genes with uniform expression within a single batch (all zeros); these will not be adjusted for batch.</code></pre>
<pre><code>## Found4batches</code></pre>
<pre><code>## Adjusting for1covariate(s) or covariate level(s)</code></pre>
<pre><code>## Standardizing Data across genes</code></pre>
<pre><code>## Fitting L/S model and finding priors</code></pre>
<pre><code>## Finding parametric adjustments</code></pre>
<pre><code>## Adjusting the Data</code></pre>
<div class="sourceCode" id="cb711"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb711-1"><a href="batch.html#cb711-1" tabindex="-1"></a>expr_mat_batch_correct <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(expr_mat_batch_correct)</span>
<span id="cb711-2"><a href="batch.html#cb711-2" tabindex="-1"></a></span>
<span id="cb711-3"><a href="batch.html#cb711-3" tabindex="-1"></a>expr_mat_batch_correct[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span></code></pre></div>
<pre><code>##                 untrt_N61311 untrt_N052611 untrt_N080611 untrt_N061011
## ENSG00000096060     8.417951      8.481407      8.847812      8.318834
## ENSG00000109906     5.784193      5.918108      5.987956      5.773905
## ENSG00000211445    10.478496     10.723636     10.890187     10.676246</code></pre>
<div id="校正后的pca" class="section level4 hasAnchor" number="6.10.2.1">
<h4><span class="header-section-number">6.10.2.1</span> 校正后的PCA<a href="batch.html#%E6%A0%A1%E6%AD%A3%E5%90%8E%E7%9A%84pca" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>PCA结果显示在<code>PC1</code>轴代表的差异变大了，<code>PC2</code>轴代表的差异变小了，不同来源的样本在<code>PC2</code>轴的分布没有规律了 (或者说成镜像分布了)。</p>
<div class="sourceCode" id="cb713"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb713-1"><a href="batch.html#cb713-1" tabindex="-1"></a><span class="fu">sp_pca</span>(expr_mat_batch_correct[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], metadata, </span>
<span id="cb713-2"><a href="batch.html#cb713-2" tabindex="-1"></a>       <span class="at">color_variable=</span><span class="st">&quot;conditions&quot;</span>, <span class="at">shape_variable =</span> <span class="st">&quot;individual&quot;</span>) <span class="sc">+</span></span>
<span id="cb713-3"><a href="batch.html#cb713-3" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">size =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-342-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>ComBat校正时不考虑分组信息，也可以获得一个合理的结果，但是一部分组间差异被抹去了。</p>
<div class="sourceCode" id="cb714"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb714-1"><a href="batch.html#cb714-1" tabindex="-1"></a><span class="co"># ComBat需要的是matrix</span></span>
<span id="cb714-2"><a href="batch.html#cb714-2" tabindex="-1"></a>expr_mat_batch_correct2 <span class="ot">&lt;-</span> <span class="fu">ComBat</span>(<span class="at">dat=</span><span class="fu">as.matrix</span>(expr_mat), <span class="at">batch=</span>metadata[[batch]], <span class="at">mod=</span><span class="cn">NULL</span>)</span></code></pre></div>
<pre><code>## Found 2538 genes with uniform expression within a single batch (all zeros); these will not be adjusted for batch.</code></pre>
<pre><code>## Found4batches</code></pre>
<pre><code>## Adjusting for0covariate(s) or covariate level(s)</code></pre>
<pre><code>## Standardizing Data across genes</code></pre>
<pre><code>## Fitting L/S model and finding priors</code></pre>
<pre><code>## Finding parametric adjustments</code></pre>
<pre><code>## Adjusting the Data</code></pre>
<div class="sourceCode" id="cb722"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb722-1"><a href="batch.html#cb722-1" tabindex="-1"></a>expr_mat_batch_correct2 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(expr_mat_batch_correct2)</span>
<span id="cb722-2"><a href="batch.html#cb722-2" tabindex="-1"></a></span>
<span id="cb722-3"><a href="batch.html#cb722-3" tabindex="-1"></a><span class="fu">sp_pca</span>(expr_mat_batch_correct2[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], metadata, </span>
<span id="cb722-4"><a href="batch.html#cb722-4" tabindex="-1"></a>       <span class="at">color_variable=</span><span class="st">&quot;conditions&quot;</span>, <span class="at">shape_variable =</span> <span class="st">&quot;individual&quot;</span>) <span class="sc">+</span></span>
<span id="cb722-5"><a href="batch.html#cb722-5" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">size =</span> <span class="st">&quot;none&quot;</span>) </span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-343-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>关于运行<code>ComBat</code>时是否应该添加关注的生物分组信息，即<code>mod</code>变量，存在一些争议。反对添加<code>mod</code>的人的担心是这么处理后，是否会强化生物分组之间的差异。支持添加<code>mod</code>的人是担心如果不添加<code>mod</code>那么去除批次时可能也会去除样本组之间的差异，尤其是实验设计不合理时。</p>
<p>如果是非平衡实验，类似我们在实验设计部分时提到的方案2，则没有办法添加<code>mod</code>变量，程序会报出<code>Design matrix is not full rank</code>类似的错误，这时是不能区分差异是来源于批次还是来源于样本，强行移除批次时，也会移除一部分或者全部样本分组带来的差异。这个在第一篇帖子处有两位朋友的留言讨论可以参考。</p>
<p><code>ComBat</code>只能处理批次信息为l离散型分组变量的数据，不能处理<code>sva</code>预测出的连续性混杂因素。</p>
</div>
</div>
<div id="batch15" class="section level3 hasAnchor" number="6.10.3">
<h3><span class="header-section-number">6.10.3</span> 使用limma校正<a href="batch.html#batch15" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>如果批次信息有多个或者不是分组变量而是类似<code>SVA</code>预测出的数值混杂因素，则需使用<code>limma</code>的<code>removeBatchEffect</code> (这里使用的是<code>SVA</code>预测出的全部3个混杂因素进行的校正。)。</p>
<p>样品在<code>PC1</code>和<code>PC2</code>组成的空间的分布与<code>ComBat</code>结果类似，只是<code>PC1</code>能解释的差异略小一些。</p>
<div class="sourceCode" id="cb723"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb723-1"><a href="batch.html#cb723-1" tabindex="-1"></a>SV <span class="ot">=</span> metadata[,<span class="fu">c</span>(<span class="st">&quot;SV1&quot;</span>,<span class="st">&quot;SV2&quot;</span>,<span class="st">&quot;SV3&quot;</span>)]</span>
<span id="cb723-2"><a href="batch.html#cb723-2" tabindex="-1"></a>expr_mat_batch_correct_limma1 <span class="ot">&lt;-</span> <span class="fu">removeBatchEffect</span>(expr_mat, <span class="at">covariates =</span> SV, <span class="at">design=</span>modcombat)</span>
<span id="cb723-3"><a href="batch.html#cb723-3" tabindex="-1"></a><span class="fu">sp_pca</span>(expr_mat_batch_correct_limma1[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], metadata, </span>
<span id="cb723-4"><a href="batch.html#cb723-4" tabindex="-1"></a>       <span class="at">color_variable=</span><span class="st">&quot;conditions&quot;</span>, <span class="at">shape_variable =</span> <span class="st">&quot;individual&quot;</span>) <span class="sc">+</span></span>
<span id="cb723-5"><a href="batch.html#cb723-5" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">size =</span> <span class="st">&quot;none&quot;</span>) </span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-344-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><code>removeBatchEffect</code>运行时也可以不提供生物分组信息。</p>
<div class="sourceCode" id="cb724"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb724-1"><a href="batch.html#cb724-1" tabindex="-1"></a>expr_mat_batch_correct_limma1 <span class="ot">&lt;-</span> <span class="fu">removeBatchEffect</span>(expr_mat, <span class="at">covariates =</span> SV)</span>
<span id="cb724-2"><a href="batch.html#cb724-2" tabindex="-1"></a><span class="fu">sp_pca</span>(expr_mat_batch_correct_limma1[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], metadata, </span>
<span id="cb724-3"><a href="batch.html#cb724-3" tabindex="-1"></a>       <span class="at">color_variable=</span><span class="st">&quot;conditions&quot;</span>, <span class="at">shape_variable =</span> <span class="st">&quot;individual&quot;</span>) <span class="sc">+</span></span>
<span id="cb724-4"><a href="batch.html#cb724-4" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">size =</span> <span class="st">&quot;none&quot;</span>) </span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-345-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><code>removeBatchEffect</code>也可以跟<code>ComBat</code>一样，对给定的已知一个或多个定性批次信息进行校正。</p>
<div class="sourceCode" id="cb725"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb725-1"><a href="batch.html#cb725-1" tabindex="-1"></a>expr_mat_batch_correct_limma2 <span class="ot">&lt;-</span> <span class="fu">removeBatchEffect</span>(expr_mat, <span class="at">batch=</span>metadata[[batch]], <span class="at">design=</span>modcombat)</span>
<span id="cb725-2"><a href="batch.html#cb725-2" tabindex="-1"></a><span class="fu">sp_pca</span>(expr_mat_batch_correct_limma2[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], metadata, </span>
<span id="cb725-3"><a href="batch.html#cb725-3" tabindex="-1"></a>       <span class="at">color_variable=</span><span class="st">&quot;conditions&quot;</span>, <span class="at">shape_variable =</span> <span class="st">&quot;individual&quot;</span>) <span class="sc">+</span></span>
<span id="cb725-4"><a href="batch.html#cb725-4" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">size =</span> <span class="st">&quot;none&quot;</span>) </span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-346-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>不指定目标分组变量，结果也不受影响。</p>
<div class="sourceCode" id="cb726"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb726-1"><a href="batch.html#cb726-1" tabindex="-1"></a>expr_mat_batch_correct_limma2 <span class="ot">&lt;-</span> <span class="fu">removeBatchEffect</span>(expr_mat, <span class="at">batch=</span>metadata[[batch]])</span>
<span id="cb726-2"><a href="batch.html#cb726-2" tabindex="-1"></a><span class="fu">sp_pca</span>(expr_mat_batch_correct_limma2[<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>,], metadata, </span>
<span id="cb726-3"><a href="batch.html#cb726-3" tabindex="-1"></a>       <span class="at">color_variable=</span><span class="st">&quot;conditions&quot;</span>, <span class="at">shape_variable =</span> <span class="st">&quot;individual&quot;</span>) <span class="sc">+</span></span>
<span id="cb726-4"><a href="batch.html#cb726-4" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">size =</span> <span class="st">&quot;none&quot;</span>) </span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-347-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="查看校正完后基因的表达情况" class="section level3 hasAnchor" number="6.10.4">
<h3><span class="header-section-number">6.10.4</span> 查看校正完后基因的表达情况<a href="batch.html#%E6%9F%A5%E7%9C%8B%E6%A0%A1%E6%AD%A3%E5%AE%8C%E5%90%8E%E5%9F%BA%E5%9B%A0%E7%9A%84%E8%A1%A8%E8%BE%BE%E6%83%85%E5%86%B5" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb727"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb727-1"><a href="batch.html#cb727-1" tabindex="-1"></a>untrt_down_genes <span class="ot">&lt;-</span> <span class="st">&quot;Name;Type</span></span>
<span id="cb727-2"><a href="batch.html#cb727-2" tabindex="-1"></a><span class="st">ENSG00000144649;SVA_batch_specific</span></span>
<span id="cb727-3"><a href="batch.html#cb727-3" tabindex="-1"></a><span class="st">ENSG00000187134;SVA_batch_uncorrect_common</span></span>
<span id="cb727-4"><a href="batch.html#cb727-4" tabindex="-1"></a><span class="st">ENSG00000137124;Uncorrect_specific</span></span>
<span id="cb727-5"><a href="batch.html#cb727-5" tabindex="-1"></a><span class="st">ENSG00000151690;Known_batch_uncorrect_common</span></span>
<span id="cb727-6"><a href="batch.html#cb727-6" tabindex="-1"></a><span class="st">ENSG00000180914;Known_batch_specific</span></span>
<span id="cb727-7"><a href="batch.html#cb727-7" tabindex="-1"></a><span class="st">ENSG00000221866;Known_batch_SVA_batch_common</span></span>
<span id="cb727-8"><a href="batch.html#cb727-8" tabindex="-1"></a><span class="st">ENSG00000152583;All_common&quot;</span></span>
<span id="cb727-9"><a href="batch.html#cb727-9" tabindex="-1"></a></span>
<span id="cb727-10"><a href="batch.html#cb727-10" tabindex="-1"></a>untrt_down_genes <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">text=</span>untrt_down_genes, <span class="at">sep=</span><span class="st">&quot;;&quot;</span>, <span class="at">header=</span>T, <span class="at">row.names=</span><span class="cn">NULL</span>)</span>
<span id="cb727-11"><a href="batch.html#cb727-11" tabindex="-1"></a></span>
<span id="cb727-12"><a href="batch.html#cb727-12" tabindex="-1"></a>untrt_down_genes_expr <span class="ot">&lt;-</span> <span class="fu">merge</span>(untrt_down_genes, expr_mat_batch_correct, <span class="at">by.x=</span><span class="st">&quot;Name&quot;</span>, <span class="at">by.y=</span><span class="dv">0</span>, <span class="at">all.x=</span>T)</span>
<span id="cb727-13"><a href="batch.html#cb727-13" tabindex="-1"></a></span>
<span id="cb727-14"><a href="batch.html#cb727-14" tabindex="-1"></a>untrt_down_genes_expr_long <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(untrt_down_genes_expr, <span class="at">id_vars=</span><span class="fu">c</span>(<span class="st">&quot;Name&quot;</span>,<span class="st">&quot;Type&quot;</span>), </span>
<span id="cb727-15"><a href="batch.html#cb727-15" tabindex="-1"></a>                                 <span class="at">variable.name=</span><span class="st">&quot;Sample&quot;</span>, <span class="at">value.name =</span> <span class="st">&quot;Expr&quot;</span>)</span></code></pre></div>
<pre><code>## Using Name, Type as id variables</code></pre>
<div class="sourceCode" id="cb729"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb729-1"><a href="batch.html#cb729-1" tabindex="-1"></a><span class="fu">head</span>(untrt_down_genes_expr_long)</span></code></pre></div>
<pre><code>##              Name                         Type       Sample      Expr
## 1 ENSG00000137124           Uncorrect_specific untrt_N61311  9.005546
## 2 ENSG00000144649           SVA_batch_specific untrt_N61311  7.067120
## 3 ENSG00000151690 Known_batch_uncorrect_common untrt_N61311  9.464342
## 4 ENSG00000152583                   All_common untrt_N61311  7.057427
## 5 ENSG00000180914         Known_batch_specific untrt_N61311  8.081169
## 6 ENSG00000187134   SVA_batch_uncorrect_common untrt_N61311 12.356100</code></pre>
<div class="sourceCode" id="cb731"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb731-1"><a href="batch.html#cb731-1" tabindex="-1"></a><span class="co">#metadata$Sample = rownames(metadata)</span></span>
<span id="cb731-2"><a href="batch.html#cb731-2" tabindex="-1"></a></span>
<span id="cb731-3"><a href="batch.html#cb731-3" tabindex="-1"></a><span class="fu">sp_boxplot</span>(untrt_down_genes_expr_long, <span class="at">melted=</span>T, <span class="at">metadata=</span>plot_data, </span>
<span id="cb731-4"><a href="batch.html#cb731-4" tabindex="-1"></a>           <span class="at">xvariable =</span> <span class="st">&quot;conditions&quot;</span>, <span class="at">yvariable =</span> <span class="st">&quot;Expr&quot;</span>, <span class="at">jitter_bp =</span> T, </span>
<span id="cb731-5"><a href="batch.html#cb731-5" tabindex="-1"></a>           <span class="at">group_variable_for_line =</span> <span class="st">&quot;individual&quot;</span>, </span>
<span id="cb731-6"><a href="batch.html#cb731-6" tabindex="-1"></a>           <span class="at">facet_variable =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">facet_scales=</span><span class="st">&quot;free_y&quot;</span>, <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.7</span>,<span class="fl">0.1</span>), </span>
<span id="cb731-7"><a href="batch.html#cb731-7" tabindex="-1"></a>           <span class="at">x_label=</span><span class="st">&quot;&quot;</span>,<span class="at">manual_color_vector =</span> <span class="st">&quot;Set2&quot;</span>) <span class="sc">+</span> </span>
<span id="cb731-8"><a href="batch.html#cb731-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.direction =</span> <span class="st">&quot;horizontal&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;#66C2A5&quot; &quot;#FC8D62&quot;</code></pre>
<p><img src="R_course_files/figure-html/unnamed-chunk-348-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb733"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb733-1"><a href="batch.html#cb733-1" tabindex="-1"></a>untrt_down_genes <span class="ot">&lt;-</span> <span class="st">&quot;Name;Type</span></span>
<span id="cb733-2"><a href="batch.html#cb733-2" tabindex="-1"></a><span class="st">ENSG00000144649;SVA_batch_specific</span></span>
<span id="cb733-3"><a href="batch.html#cb733-3" tabindex="-1"></a><span class="st">ENSG00000187134;SVA_batch_uncorrect_common</span></span>
<span id="cb733-4"><a href="batch.html#cb733-4" tabindex="-1"></a><span class="st">ENSG00000137124;Uncorrect_specific</span></span>
<span id="cb733-5"><a href="batch.html#cb733-5" tabindex="-1"></a><span class="st">ENSG00000151690;Known_batch_uncorrect_common</span></span>
<span id="cb733-6"><a href="batch.html#cb733-6" tabindex="-1"></a><span class="st">ENSG00000180914;Known_batch_specific</span></span>
<span id="cb733-7"><a href="batch.html#cb733-7" tabindex="-1"></a><span class="st">ENSG00000221866;Known_batch_SVA_batch_common</span></span>
<span id="cb733-8"><a href="batch.html#cb733-8" tabindex="-1"></a><span class="st">ENSG00000152583;All_common&quot;</span></span>
<span id="cb733-9"><a href="batch.html#cb733-9" tabindex="-1"></a></span>
<span id="cb733-10"><a href="batch.html#cb733-10" tabindex="-1"></a>untrt_down_genes <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">text=</span>untrt_down_genes, <span class="at">sep=</span><span class="st">&quot;;&quot;</span>, <span class="at">header=</span>T, <span class="at">row.names=</span><span class="cn">NULL</span>)</span>
<span id="cb733-11"><a href="batch.html#cb733-11" tabindex="-1"></a></span>
<span id="cb733-12"><a href="batch.html#cb733-12" tabindex="-1"></a>untrt_down_genes_expr <span class="ot">&lt;-</span> <span class="fu">merge</span>(untrt_down_genes, expr_mat_batch_correct_limma1, <span class="at">by.x=</span><span class="st">&quot;Name&quot;</span>, <span class="at">by.y=</span><span class="dv">0</span>, <span class="at">all.x=</span>T)</span>
<span id="cb733-13"><a href="batch.html#cb733-13" tabindex="-1"></a></span>
<span id="cb733-14"><a href="batch.html#cb733-14" tabindex="-1"></a>untrt_down_genes_expr_long <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(untrt_down_genes_expr, <span class="at">id_vars=</span><span class="fu">c</span>(<span class="st">&quot;Name&quot;</span>,<span class="st">&quot;Type&quot;</span>), </span>
<span id="cb733-15"><a href="batch.html#cb733-15" tabindex="-1"></a>                                 <span class="at">variable.name=</span><span class="st">&quot;Sample&quot;</span>, <span class="at">value.name =</span> <span class="st">&quot;Expr&quot;</span>)</span></code></pre></div>
<pre><code>## Using Name, Type as id variables</code></pre>
<div class="sourceCode" id="cb735"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb735-1"><a href="batch.html#cb735-1" tabindex="-1"></a><span class="fu">head</span>(untrt_down_genes_expr_long)</span></code></pre></div>
<pre><code>##              Name                         Type       Sample      Expr
## 1 ENSG00000137124           Uncorrect_specific untrt_N61311  9.021158
## 2 ENSG00000144649           SVA_batch_specific untrt_N61311  6.991157
## 3 ENSG00000151690 Known_batch_uncorrect_common untrt_N61311  9.527425
## 4 ENSG00000152583                   All_common untrt_N61311  7.054087
## 5 ENSG00000180914         Known_batch_specific untrt_N61311  8.214962
## 6 ENSG00000187134   SVA_batch_uncorrect_common untrt_N61311 12.282890</code></pre>
<div class="sourceCode" id="cb737"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb737-1"><a href="batch.html#cb737-1" tabindex="-1"></a><span class="co"># metadata$Sample = rownames(metadata)</span></span>
<span id="cb737-2"><a href="batch.html#cb737-2" tabindex="-1"></a></span>
<span id="cb737-3"><a href="batch.html#cb737-3" tabindex="-1"></a><span class="fu">sp_boxplot</span>(untrt_down_genes_expr_long, <span class="at">melted=</span>T, <span class="at">metadata=</span>plot_data, </span>
<span id="cb737-4"><a href="batch.html#cb737-4" tabindex="-1"></a>           <span class="at">xvariable =</span> <span class="st">&quot;conditions&quot;</span>, <span class="at">yvariable =</span> <span class="st">&quot;Expr&quot;</span>, <span class="at">jitter_bp =</span> T, </span>
<span id="cb737-5"><a href="batch.html#cb737-5" tabindex="-1"></a>           <span class="at">group_variable_for_line =</span> <span class="st">&quot;individual&quot;</span>, </span>
<span id="cb737-6"><a href="batch.html#cb737-6" tabindex="-1"></a>           <span class="at">facet_variable =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">facet_scales=</span><span class="st">&quot;free_y&quot;</span>, <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.7</span>,<span class="fl">0.1</span>), </span>
<span id="cb737-7"><a href="batch.html#cb737-7" tabindex="-1"></a>           <span class="at">x_label=</span><span class="st">&quot;&quot;</span>,<span class="at">manual_color_vector =</span> <span class="st">&quot;Set2&quot;</span>) <span class="sc">+</span> </span>
<span id="cb737-8"><a href="batch.html#cb737-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.direction =</span> <span class="st">&quot;horizontal&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;#66C2A5&quot; &quot;#FC8D62&quot;</code></pre>
<p><img src="R_course_files/figure-html/unnamed-chunk-349-1.png" width="100%" style="display: block; margin: auto;" /></p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="AI.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="questions06.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": false,
    "twitter": false,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/Tong-Chen/Bioinfo_course_R/edit/master/05-STAR.DESeq2.simpler.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["R_course.pdf"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
