<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>R, Cytoscape, AI生信作图分析培训</title>
  <meta name="description" content="R, Cytoscape, AI生信作图分析培训">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="R, Cytoscape, AI生信作图分析培训" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="R, Cytoscape, AI生信作图分析培训" />
  
  
  

<meta name="author" content="易生信培训团队">
<meta name="author" content="联系我们: train@ehbio.com">


<meta name="date" content="2018-05-09">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="Rbasic.html">
<link rel="next" href="section-4.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="http://www.ehbio.com"><img src="http://www.ehbio.com/logos/ehbio_gitbook_logo.png" width="95%"></a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>EHBIO Gene Technology</a></li>
<li class="chapter" data-level="1" data-path="questions.html"><a href="questions.html"><i class="fa fa-check"></i><b>1</b> 考题</a><ul>
<li class="chapter" data-level="1.1" data-path="questions.html"><a href="questions.html#-50"><i class="fa fa-check"></i><b>1.1</b> 理论题 (50分)</a></li>
<li class="chapter" data-level="1.2" data-path="questions.html"><a href="questions.html#-70"><i class="fa fa-check"></i><b>1.2</b> 代码题 (70分)</a></li>
<li class="chapter" data-level="1.3" data-path="questions.html"><a href="questions.html#-"><i class="fa fa-check"></i><b>1.3</b> 错误识别题 (选答)</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="Rbasic.html"><a href="Rbasic.html"><i class="fa fa-check"></i><b>2</b> R基础</a><ul>
<li class="chapter" data-level="2.1" data-path="Rbasic.html"><a href="Rbasic.html#r"><i class="fa fa-check"></i><b>2.1</b> R安装</a></li>
<li class="chapter" data-level="2.2" data-path="Rbasic.html"><a href="Rbasic.html#rstudio"><i class="fa fa-check"></i><b>2.2</b> Rstudio基础</a><ul>
<li class="chapter" data-level="2.2.1" data-path="Rbasic.html"><a href="Rbasic.html#rstudio"><i class="fa fa-check"></i><b>2.2.1</b> Rstudio版本</a></li>
<li class="chapter" data-level="2.2.2" data-path="Rbasic.html"><a href="Rbasic.html#rstudio"><i class="fa fa-check"></i><b>2.2.2</b> Rstudio安装</a></li>
<li class="chapter" data-level="2.2.3" data-path="Rbasic.html"><a href="Rbasic.html#rstudio-"><i class="fa fa-check"></i><b>2.2.3</b> Rstudio 使用</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="Rbasic.html"><a href="Rbasic.html#r"><i class="fa fa-check"></i><b>2.3</b> R基本语法</a><ul>
<li class="chapter" data-level="2.3.1" data-path="Rbasic.html"><a href="Rbasic.html#section-2.3.1"><i class="fa fa-check"></i><b>2.3.1</b> 获取帮助文档，查看命令或函数的使用方法、事例或适用范围</a></li>
<li class="chapter" data-level="2.3.2" data-path="Rbasic.html"><a href="Rbasic.html#r"><i class="fa fa-check"></i><b>2.3.2</b> R中的变量及其初始化</a></li>
<li class="chapter" data-level="2.3.3" data-path="Rbasic.html"><a href="Rbasic.html#section-2.3.3"><i class="fa fa-check"></i><b>2.3.3</b> 变量类型和转换</a></li>
<li class="chapter" data-level="2.3.4" data-path="Rbasic.html"><a href="Rbasic.html#r"><i class="fa fa-check"></i><b>2.3.4</b> R中矩阵运算</a></li>
<li class="chapter" data-level="2.3.5" data-path="Rbasic.html"><a href="Rbasic.html#r"><i class="fa fa-check"></i><b>2.3.5</b> R中矩阵筛选合并</a></li>
<li class="chapter" data-level="2.3.6" data-path="Rbasic.html"><a href="Rbasic.html#str"><i class="fa fa-check"></i><b>2.3.6</b> <code>str</code>的应用</a></li>
<li class="chapter" data-level="2.3.7" data-path="Rbasic.html"><a href="Rbasic.html#r"><i class="fa fa-check"></i><b>2.3.7</b> R的包管理</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="Rbasic.html"><a href="Rbasic.html#cheetsheets"><i class="fa fa-check"></i><b>2.4</b> CheetSheets</a></li>
<li class="chapter" data-level="2.5" data-path="Rbasic.html"><a href="Rbasic.html#section-2.5"><i class="fa fa-check"></i><b>2.5</b> 参考</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-plots.html"><a href="r-plots.html"><i class="fa fa-check"></i><b>3</b> R plots</a><ul>
<li class="chapter" data-level="3.1" data-path="r-plots.html"><a href="r-plots.html#qplot-"><i class="fa fa-check"></i><b>3.1</b> qplot绘制图形 (王绪宁)</a></li>
<li class="chapter" data-level="3.2" data-path="r-plots.html"><a href="r-plots.html#section-3.2"><i class="fa fa-check"></i><b>3.2</b> 热图绘制</a><ul>
<li class="chapter" data-level="3.2.1" data-path="r-plots.html"><a href="r-plots.html#section-3.2.1"><i class="fa fa-check"></i><b>3.2.1</b> 生成测试数据</a></li>
<li class="chapter" data-level="3.2.2" data-path="r-plots.html"><a href="r-plots.html#section-3.2.2"><i class="fa fa-check"></i><b>3.2.2</b> 转换数据格式</a></li>
<li class="chapter" data-level="3.2.3" data-path="r-plots.html"><a href="r-plots.html#section-3.2.3"><i class="fa fa-check"></i><b>3.2.3</b> 分解绘图</a></li>
<li class="chapter" data-level="3.2.4" data-path="r-plots.html"><a href="r-plots.html#section-3.2.4"><i class="fa fa-check"></i><b>3.2.4</b> 图形存储</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="r-plots.html"><a href="r-plots.html#section-3.3"><i class="fa fa-check"></i><b>3.3</b> 热图美化</a><ul>
<li class="chapter" data-level="3.3.1" data-path="r-plots.html"><a href="r-plots.html#section-3.3.1"><i class="fa fa-check"></i><b>3.3.1</b> 对数转换</a></li>
<li class="chapter" data-level="3.3.2" data-path="r-plots.html"><a href="r-plots.html#z-score"><i class="fa fa-check"></i><b>3.3.2</b> Z-score转换</a></li>
<li class="chapter" data-level="3.3.3" data-path="r-plots.html"><a href="r-plots.html#section-3.3.3"><i class="fa fa-check"></i><b>3.3.3</b> 抹去异常值</a></li>
<li class="chapter" data-level="3.3.4" data-path="r-plots.html"><a href="r-plots.html#section-3.3.4"><i class="fa fa-check"></i><b>3.3.4</b> 非线性颜色</a></li>
<li class="chapter" data-level="3.3.5" data-path="r-plots.html"><a href="r-plots.html#section-3.3.5"><i class="fa fa-check"></i><b>3.3.5</b> 调整行或列的顺序</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="r-plots.html"><a href="r-plots.html#---pheatmap"><i class="fa fa-check"></i><b>3.4</b> 热图绘制 - pheatmap</a></li>
<li class="chapter" data-level="3.5" data-path="r-plots.html"><a href="r-plots.html#section-3.5"><i class="fa fa-check"></i><b>3.5</b> 箱线图</a><ul>
<li class="chapter" data-level="3.5.1" data-path="r-plots.html"><a href="r-plots.html#section-3.5.1"><i class="fa fa-check"></i><b>3.5.1</b> 一步步解析箱线图绘制</a></li>
<li class="chapter" data-level="3.5.2" data-path="r-plots.html"><a href="r-plots.html#-a"><i class="fa fa-check"></i><b>3.5.2</b> 绘制单个基因 (A)的箱线图</a></li>
<li class="chapter" data-level="3.5.3" data-path="r-plots.html"><a href="r-plots.html#section-3.5.3"><i class="fa fa-check"></i><b>3.5.3</b> 长矩阵绘制箱线图</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="r-plots.html"><a href="r-plots.html#section-3.6"><i class="fa fa-check"></i><b>3.6</b> 线图</a><ul>
<li class="chapter" data-level="3.6.1" data-path="r-plots.html"><a href="r-plots.html#section-3.6.1"><i class="fa fa-check"></i><b>3.6.1</b> 单线图</a></li>
<li class="chapter" data-level="3.6.2" data-path="r-plots.html"><a href="r-plots.html#section-3.6.2"><i class="fa fa-check"></i><b>3.6.2</b> 多线图</a></li>
<li class="chapter" data-level="3.6.3" data-path="r-plots.html"><a href="r-plots.html#section-3.6.3"><i class="fa fa-check"></i><b>3.6.3</b> 横轴文本线图</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="r-plots.html"><a href="r-plots.html#section-3.7"><i class="fa fa-check"></i><b>3.7</b> 散点图</a><ul>
<li class="chapter" data-level="3.7.1" data-path="r-plots.html"><a href="r-plots.html#section-3.7.1"><i class="fa fa-check"></i><b>3.7.1</b> 横纵轴都为数字的散点图解析</a></li>
<li class="chapter" data-level="3.7.2" data-path="r-plots.html"><a href="r-plots.html#section-3.7.2"><i class="fa fa-check"></i><b>3.7.2</b> 横纵轴都为字符串的散点图展示</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="r-plots.html"><a href="r-plots.html#section-3.8"><i class="fa fa-check"></i><b>3.8</b> 功能富集泡泡图</a><ul>
<li class="chapter" data-level="3.8.1" data-path="r-plots.html"><a href="r-plots.html#section-3.8.1"><i class="fa fa-check"></i><b>3.8.1</b> 单样品分开绘制</a></li>
<li class="chapter" data-level="3.8.2" data-path="r-plots.html"><a href="r-plots.html#section-3.8.2"><i class="fa fa-check"></i><b>3.8.2</b> 多样品合并绘制</a></li>
</ul></li>
<li class="chapter" data-level="3.9" data-path="r-plots.html"><a href="r-plots.html#section-3.9"><i class="fa fa-check"></i><b>3.9</b> 韦恩图</a><ul>
<li class="chapter" data-level="3.9.1" data-path="r-plots.html"><a href="r-plots.html#section-3.9.1"><i class="fa fa-check"></i><b>3.9.1</b> 韦恩图三个圈</a></li>
<li class="chapter" data-level="3.9.2" data-path="r-plots.html"><a href="r-plots.html#section-3.9.2"><i class="fa fa-check"></i><b>3.9.2</b> 韦恩图五个圈</a></li>
<li class="chapter" data-level="3.9.3" data-path="r-plots.html"><a href="r-plots.html#upsetview"><i class="fa fa-check"></i><b>3.9.3</b> UpSetView展示</a></li>
</ul></li>
<li class="chapter" data-level="3.10" data-path="r-plots.html"><a href="r-plots.html#section-3.10"><i class="fa fa-check"></i><b>3.10</b> 柱状图绘制</a><ul>
<li class="chapter" data-level="3.10.1" data-path="r-plots.html"><a href="r-plots.html#section-3.10.1"><i class="fa fa-check"></i><b>3.10.1</b> 常规矩阵柱状图绘制</a></li>
<li class="chapter" data-level="3.10.2" data-path="r-plots.html"><a href="r-plots.html#section-3.10.2"><i class="fa fa-check"></i><b>3.10.2</b> 长矩阵分面绘制</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="r-plots.html"><a href="r-plots.html#section-3.11"><i class="fa fa-check"></i><b>3.11</b> 图形支持中文字体</a><ul>
<li class="chapter" data-level="3.11.1" data-path="r-plots.html"><a href="r-plots.html#section-3.11.1"><i class="fa fa-check"></i><b>3.11.1</b> 修改图形的字体</a></li>
<li class="chapter" data-level="3.11.2" data-path="r-plots.html"><a href="r-plots.html#ggplot2pdf"><i class="fa fa-check"></i><b>3.11.2</b> ggplot2支持中文字体输出PDF</a></li>
<li class="chapter" data-level="3.11.3" data-path="r-plots.html"><a href="r-plots.html#section-3.11.3"><i class="fa fa-check"></i><b>3.11.3</b> 系统可用字体</a></li>
<li class="chapter" data-level="3.11.4" data-path="r-plots.html"><a href="r-plots.html#section-3.11.4"><i class="fa fa-check"></i><b>3.11.4</b> 合并字体支持中英文</a></li>
<li class="chapter" data-level="3.11.5" data-path="r-plots.html"><a href="r-plots.html#section-3.11.5"><i class="fa fa-check"></i><b>3.11.5</b> 一个示例</a></li>
</ul></li>
<li class="chapter" data-level="3.12" data-path="r-plots.html"><a href="r-plots.html#pca"><i class="fa fa-check"></i><b>3.12</b> PCA</a><ul>
<li class="chapter" data-level="3.12.1" data-path="r-plots.html"><a href="r-plots.html#section-3.12.1"><i class="fa fa-check"></i><b>3.12.1</b> 主成分分析简介</a></li>
<li class="chapter" data-level="3.12.2" data-path="r-plots.html"><a href="r-plots.html#section-3.12.2"><i class="fa fa-check"></i><b>3.12.2</b> 主成分分析的意义</a></li>
<li class="chapter" data-level="3.12.3" data-path="r-plots.html"><a href="r-plots.html#section-3.12.3"><i class="fa fa-check"></i><b>3.12.3</b> 示例展示原始变量对样品的分类</a></li>
<li class="chapter" data-level="3.12.4" data-path="r-plots.html"><a href="r-plots.html#pca"><i class="fa fa-check"></i><b>3.12.4</b> PCA的实现原理</a></li>
<li class="chapter" data-level="3.12.5" data-path="r-plots.html"><a href="r-plots.html#pca"><i class="fa fa-check"></i><b>3.12.5</b> 简单的PCA实现</a></li>
<li class="chapter" data-level="3.12.6" data-path="r-plots.html"><a href="r-plots.html#pca"><i class="fa fa-check"></i><b>3.12.6</b> PCA结果解释</a></li>
<li class="chapter" data-level="3.12.7" data-path="r-plots.html"><a href="r-plots.html#pca"><i class="fa fa-check"></i><b>3.12.7</b> PCA应用于测试数据</a></li>
<li class="chapter" data-level="3.12.8" data-path="r-plots.html"><a href="r-plots.html#pca"><i class="fa fa-check"></i><b>3.12.8</b> PCA注意事项</a></li>
<li class="chapter" data-level="3.12.9" data-path="r-plots.html"><a href="r-plots.html#section-3.12.9"><i class="fa fa-check"></i><b>3.12.9</b> 参考资料</a></li>
</ul></li>
<li class="chapter" data-level="3.13" data-path="r-plots.html"><a href="r-plots.html#section-3.13"><i class="fa fa-check"></i><b>3.13</b> 表达聚类分析</a></li>
<li class="chapter" data-level="3.14" data-path="r-plots.html"><a href="r-plots.html#section-3.14"><i class="fa fa-check"></i><b>3.14</b> 生存分析</a><ul>
<li class="chapter" data-level="3.14.1" data-path="Rbasic.html"><a href="Rbasic.html#r"><i class="fa fa-check"></i><b>3.14.1</b> R做生存分析</a></li>
</ul></li>
<li class="chapter" data-level="3.15" data-path="r-plots.html"><a href="r-plots.html#section-3.15"><i class="fa fa-check"></i><b>3.15</b> 一步作图的优势</a></li>
<li class="chapter" data-level="3.16" data-path="r-plots.html"><a href="r-plots.html#section-3.16"><i class="fa fa-check"></i><b>3.16</b> 不改脚本的热图绘制</a><ul>
<li class="chapter" data-level="3.16.1" data-path="r-plots.html"><a href="r-plots.html#---"><i class="fa fa-check"></i><b>3.16.1</b> 箱线图 - 一步绘制</a></li>
<li class="chapter" data-level="3.16.2" data-path="r-plots.html"><a href="r-plots.html#---"><i class="fa fa-check"></i><b>3.16.2</b> 线图 - 一步绘制</a></li>
<li class="chapter" data-level="3.16.3" data-path="r-plots.html"><a href="r-plots.html#section-3.16.3"><i class="fa fa-check"></i><b>3.16.3</b> 一网打进散点图绘制</a></li>
</ul></li>
<li class="chapter" data-level="3.17" data-path="r-plots.html"><a href="r-plots.html#-1"><i class="fa fa-check"></i><b>3.17</b> 参考资料</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="section-4.html"><a href="section-4.html"><i class="fa fa-check"></i><b>4</b> 交互式图</a><ul>
<li class="chapter" data-level="4.1" data-path="section-4.html"><a href="section-4.html#section-4.1"><i class="fa fa-check"></i><b>4.1</b> 交互式网络图</a></li>
<li class="chapter" data-level="4.2" data-path="section-4.html"><a href="section-4.html#section-4.2"><i class="fa fa-check"></i><b>4.2</b> 交互式桑基图</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="imagegp.html"><a href="imagegp.html"><i class="fa fa-check"></i><b>5</b> 在线绘图</a><ul>
<li class="chapter" data-level="5.1" data-path="imagegp.html"><a href="imagegp.html#section-5.1"><i class="fa fa-check"></i><b>5.1</b> 功能和数据概览</a><ul>
<li class="chapter" data-level="5.1.1" data-path="imagegp.html"><a href="imagegp.html#section-5.1.1"><i class="fa fa-check"></i><b>5.1.1</b> 图形支持</a></li>
<li class="chapter" data-level="5.1.2" data-path="imagegp.html"><a href="imagegp.html#section-5.1.2"><i class="fa fa-check"></i><b>5.1.2</b> 整体界面</a></li>
<li class="chapter" data-level="5.1.3" data-path="imagegp.html"><a href="imagegp.html#section-5.1.3"><i class="fa fa-check"></i><b>5.1.3</b> 数据格式</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="imagegp.html"><a href="imagegp.html#section-5.2"><i class="fa fa-check"></i><b>5.2</b> 图形展示和绘图展示</a><ul>
<li class="chapter" data-level="5.2.1" data-path="r-plots.html"><a href="r-plots.html#-1"><i class="fa fa-check"></i><b>5.2.1</b> 线图</a></li>
<li class="chapter" data-level="5.2.2" data-path="imagegp.html"><a href="imagegp.html#section-5.2.2"><i class="fa fa-check"></i><b>5.2.2</b> 富集分析泡泡图</a></li>
<li class="chapter" data-level="5.2.3" data-path="imagegp.html"><a href="imagegp.html#section-5.2.3"><i class="fa fa-check"></i><b>5.2.3</b> 热图</a></li>
<li class="chapter" data-level="5.2.4" data-path="r-plots.html"><a href="r-plots.html#-1"><i class="fa fa-check"></i><b>5.2.4</b> 箱线图</a></li>
<li class="chapter" data-level="5.2.5" data-path="r-plots.html"><a href="r-plots.html#-1"><i class="fa fa-check"></i><b>5.2.5</b> 散点图</a></li>
<li class="chapter" data-level="5.2.6" data-path="imagegp.html"><a href="imagegp.html#section-5.2.6"><i class="fa fa-check"></i><b>5.2.6</b> 柱状图</a></li>
<li class="chapter" data-level="5.2.7" data-path="imagegp.html"><a href="imagegp.html#section-5.2.7"><i class="fa fa-check"></i><b>5.2.7</b> 火山图</a></li>
<li class="chapter" data-level="5.2.8" data-path="imagegp.html"><a href="imagegp.html#section-5.2.8"><i class="fa fa-check"></i><b>5.2.8</b> 维恩图</a></li>
<li class="chapter" data-level="5.2.9" data-path="imagegp.html"><a href="imagegp.html#upsetview-1"><i class="fa fa-check"></i><b>5.2.9</b> UpsetView</a></li>
<li class="chapter" data-level="5.2.10" data-path="imagegp.html"><a href="imagegp.html#section-5.2.10"><i class="fa fa-check"></i><b>5.2.10</b> <a href="http://mp.weixin.qq.com/s/ST2SAmfKOptpJOHS8podmQ">共表达密度图</a></a></li>
<li class="chapter" data-level="5.2.11" data-path="imagegp.html"><a href="imagegp.html#section-5.2.11"><i class="fa fa-check"></i><b>5.2.11</b> 直方图</a></li>
<li class="chapter" data-level="5.2.12" data-path="imagegp.html"><a href="imagegp.html#pca-1"><i class="fa fa-check"></i><b>5.2.12</b> PCA</a></li>
<li class="chapter" data-level="5.2.13" data-path="imagegp.html"><a href="imagegp.html#section-5.2.13"><i class="fa fa-check"></i><b>5.2.13</b> 曼哈顿图</a></li>
<li class="chapter" data-level="5.2.14" data-path="imagegp.html"><a href="imagegp.html#pcoa"><i class="fa fa-check"></i><b>5.2.14</b> PcOA</a></li>
<li class="chapter" data-level="5.2.15" data-path="imagegp.html"><a href="imagegp.html#cpcoa"><i class="fa fa-check"></i><b>5.2.15</b> CPcOA</a></li>
<li class="chapter" data-level="5.2.16" data-path="imagegp.html"><a href="imagegp.html#section-5.2.16"><i class="fa fa-check"></i><b>5.2.16</b> 桑基图</a></li>
<li class="chapter" data-level="5.2.17" data-path="Rbasic.html"><a href="Rbasic.html#r"><i class="fa fa-check"></i><b>5.2.17</b> R绘图文章</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="cytoscape.html"><a href="cytoscape.html"><i class="fa fa-check"></i><b>6</b> 网络图</a><ul>
<li class="chapter" data-level="6.0.1" data-path="cytoscape.html"><a href="cytoscape.html#section-6.0.1"><i class="fa fa-check"></i><b>6.0.1</b> 基本操作</a></li>
<li class="chapter" data-level="6.0.2" data-path="cytoscape.html"><a href="cytoscape.html#mirna-mrna"><i class="fa fa-check"></i><b>6.0.2</b> miRNA-mRNA调控网络</a></li>
<li class="chapter" data-level="6.0.3" data-path="cytoscape.html"><a href="cytoscape.html#section-6.0.3"><i class="fa fa-check"></i><b>6.0.3</b> 不同的布局的调试和修改</a></li>
<li class="chapter" data-level="6.1" data-path="r-plots.html"><a href="r-plots.html#-1"><i class="fa fa-check"></i><b>6.1</b> 参考</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="AI.html"><a href="AI.html"><i class="fa fa-check"></i><b>7</b> 图形排版</a><ul>
<li class="chapter" data-level="7.0.1" data-path="AI.html"><a href="AI.html#section-7.0.1"><i class="fa fa-check"></i><b>7.0.1</b> 矢量图和标量图</a></li>
<li class="chapter" data-level="7.0.2" data-path="AI.html"><a href="AI.html#section-7.0.2"><i class="fa fa-check"></i><b>7.0.2</b> 矢量图的制作</a></li>
<li class="chapter" data-level="7.0.3" data-path="AI.html"><a href="AI.html#section-7.0.3"><i class="fa fa-check"></i><b>7.0.3</b> 矢量图编辑工具</a></li>
<li class="chapter" data-level="7.0.4" data-path="AI.html"><a href="AI.html#section-7.0.4"><i class="fa fa-check"></i><b>7.0.4</b> 作图基本原则</a></li>
<li class="chapter" data-level="7.0.5" data-path="AI.html"><a href="AI.html#section-7.0.5"><i class="fa fa-check"></i><b>7.0.5</b> 作图中的要点注意</a></li>
<li class="chapter" data-level="7.0.6" data-path="AI.html"><a href="AI.html#adobe-illustrator"><i class="fa fa-check"></i><b>7.0.6</b> <code>Adobe Illustrator</code>中的基本概念和操作描述</a></li>
<li class="chapter" data-level="7.0.7" data-path="AI.html"><a href="AI.html#section-7.0.7"><i class="fa fa-check"></i><b>7.0.7</b> 视频教程</a></li>
<li class="chapter" data-level="7.0.8" data-path="AI.html"><a href="AI.html#section-7.0.8"><i class="fa fa-check"></i><b>7.0.8</b> 动图教程</a></li>
<li class="chapter" data-level="7.1" data-path="AI.html"><a href="AI.html#-2"><i class="fa fa-check"></i><b>7.1</b> 参考</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ref.html"><a href="ref.html"><i class="fa fa-check"></i><b>8</b> 参考</a><ul>
<li class="chapter" data-level="8.1" data-path="ref.html"><a href="ref.html#section-8.1"><i class="fa fa-check"></i><b>8.1</b> 程序学习心得</a></li>
<li class="chapter" data-level="8.2" data-path="ref.html"><a href="ref.html#ngs"><i class="fa fa-check"></i><b>8.2</b> NGS分析工具评估</a></li>
<li class="chapter" data-level="8.3" data-path="ref.html"><a href="ref.html#section-8.3"><i class="fa fa-check"></i><b>8.3</b> 文献精读</a></li>
<li class="chapter" data-level="8.4" data-path="ref.html"><a href="ref.html#linux-"><i class="fa fa-check"></i><b>8.4</b> Linux 学习</a></li>
<li class="chapter" data-level="8.5" data-path="Rbasic.html"><a href="Rbasic.html#r"><i class="fa fa-check"></i><b>8.5</b> R统计和作图</a></li>
<li class="chapter" data-level="8.6" data-path="ref.html"><a href="ref.html#ngs"><i class="fa fa-check"></i><b>8.6</b> NGS基础</a></li>
<li class="chapter" data-level="8.7" data-path="ref.html"><a href="ref.html#section-8.7"><i class="fa fa-check"></i><b>8.7</b> 癌症数据库</a></li>
<li class="chapter" data-level="8.8" data-path="ref.html"><a href="ref.html#python"><i class="fa fa-check"></i><b>8.8</b> Python学习</a></li>
<li class="chapter" data-level="8.9" data-path="ref.html"><a href="ref.html#ngs"><i class="fa fa-check"></i><b>8.9</b> NGS软件</a></li>
<li class="chapter" data-level="8.10" data-path="cytoscape.html"><a href="cytoscape.html#cytoscape"><i class="fa fa-check"></i><b>8.10</b> Cytoscape网络图</a></li>
<li class="chapter" data-level="8.11" data-path="ref.html"><a href="ref.html#section-8.11"><i class="fa fa-check"></i><b>8.11</b> 分子对接</a></li>
<li class="chapter" data-level="8.12" data-path="ref.html"><a href="ref.html#section-8.12"><i class="fa fa-check"></i><b>8.12</b> 生信宝典之傻瓜式</a></li>
<li class="chapter" data-level="8.13" data-path="ref.html"><a href="ref.html#section-8.13"><i class="fa fa-check"></i><b>8.13</b> 生信人写程序</a></li>
<li class="chapter" data-level="8.14" data-path="ref.html"><a href="ref.html#section-8.14"><i class="fa fa-check"></i><b>8.14</b> 小技巧系列</a></li>
<li class="chapter" data-level="8.15" data-path="ref.html"><a href="ref.html#section-8.15"><i class="fa fa-check"></i><b>8.15</b> 招聘</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="company-intro.html"><a href="company-intro.html"><i class="fa fa-check"></i><b>9</b> 公司简介</a></li>
<li class="divider"></li>
<li><a href="mailto:ct@ehbio.com" target="blank">ct@ehbio.com</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R, Cytoscape, AI生信作图分析培训</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="r-plots" class="section level1">
<h1><span class="header-section-number">3</span> R plots</h1>
<div id="qplot-" class="section level2">
<h2><span class="header-section-number">3.1</span> qplot绘制图形 (王绪宁)</h2>
<p>ggplot2中两大精髓的函数分别为 <code>qplot</code> (快速作图quick plot)，类似于R中基本的<code>plot</code>和<code>ggplot</code> (更符合ggplot2图层式绘图理念，一层层添加修改)。</p>
<p>测试数据是<code>ggplot2</code>包中自带的<code>diamond</code>数据，每一行为一种钻石，每一列为钻石不同的属性，如<code>carat</code> (克拉), <code>cut</code> (切工), <code>color</code> (色泽), <code>clarity</code> (透明度)等。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">set.seed</span>(<span class="dv">23</span>)
dat &lt;-<span class="st"> </span>diamonds[<span class="kw">sample</span>(<span class="kw">nrow</span>(diamonds), <span class="dv">1000</span>),]
<span class="kw">head</span>(dat)</code></pre></div>
<pre><code>## # A tibble: 6 x 10
##   carat cut     color clarity depth table price     x     y     z
##   &lt;dbl&gt; &lt;ord&gt;   &lt;ord&gt; &lt;ord&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 0.340 Ideal   D     VS2      61.0   58.   753  4.49  4.53  2.75
## 2 0.340 Ideal   G     VS2      61.0   57.   596  4.48  4.50  2.74
## 3 1.01  Ideal   F     VS2      61.6   56.  7229  6.48  6.45  3.98
## 4 0.320 Premium E     VVS1     61.8   59.  1020  4.39  4.35  2.70
## 5 0.510 Ideal   D     SI1      61.7   55.  1569  5.11  5.16  3.17
## 6 1.22  Ideal   G     VVS1     61.1   56. 10888  6.91  6.94  4.23</code></pre>
<p>数据读进来后，怎么绘制呢？</p>
<ol style="list-style-type: decimal">
<li>绘制散点图，横轴是克拉数，纵轴是价格 (正相关)</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(carat,price,<span class="dt">data=</span>dat)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-73-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>绘制散点图，对x,y值取log</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(<span class="kw">log</span>(carat),<span class="kw">log</span>(price),<span class="dt">data=</span>dat)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-74-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>颜色、大小、性状和其他属性的设置</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(carat,price,<span class="dt">data=</span>dat,<span class="dt">colour=</span>color)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-75-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(carat,price,<span class="dt">data=</span>dat,<span class="dt">shape=</span>cut)<span class="co">#以cut为分类依据设置不同的形状</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-76-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><strong>几何对象</strong></p>
<p><code>qplot()</code>函数配合不同的几何对象便可绘制出不同的图形:</p>
<ul>
<li>点图 <code>geom=&quot;point&quot;</code><br />
</li>
<li>平滑曲线 <code>geom=&quot;smooth&quot;</code><br />
</li>
<li>箱线图 <code>geom=&quot;boxplot&quot;</code></li>
<li>任意方向的路径性<code>geom=&quot;path&quot;</code><br />
</li>
<li>线条图 <code>geom=&quot;line&quot;</code> (从左到右连接)</li>
<li>对于连续变量，直方图<code>geom=&quot;histogram&quot;</code><br />
</li>
<li>频率多边图 <code>geom=&quot;freqpoly&quot;</code></li>
<li>绘制密度曲线 <code>geom=&quot;density&quot;</code></li>
<li>如果只有x参数传递给qplot(),那么默认是直方图</li>
<li>对于离散变量，geom=“bar”绘制条形图</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#向图形中加入平滑曲线(#从本张图片可以逐渐体会ggplot绘图的强大，</span>
<span class="co">#后期应用ggplot()函数后，可以更加自由的绘制各种组合图形)</span>
<span class="kw">qplot</span>(carat,price,<span class="dt">data=</span>dat,<span class="dt">geom=</span><span class="kw">c</span>(<span class="st">&quot;point&quot;</span>,<span class="st">&quot;smooth&quot;</span>))<span class="co">#添加了一条拟合曲线</span></code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;gam&#39;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#拟合曲线默认的方法为method=&quot;loss&quot;,程序会根据数据点的多少自动选取，</span>
<span class="co">#曲线周围的灰色部分为标准误，可以用se=FALSE曲线</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-77-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><strong>绘制其他常见图形</strong></p>
<ol style="list-style-type: decimal">
<li>箱线图</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(color,price/carat,<span class="dt">data=</span>dat,<span class="dt">geom=</span><span class="st">&quot;boxplot&quot;</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-78-1.png" width="100%" style="display: block; margin: auto;" /></p>
<ol start="2" style="list-style-type: decimal">
<li>绕动图 (抖动图)</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(color,price/carat,<span class="dt">data=</span>diamonds,<span class="dt">geom=</span><span class="st">&quot;jitter&quot;</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-79-1.png" width="100%" style="display: block; margin: auto;" /></p>
<ol start="3" style="list-style-type: decimal">
<li>直方图</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(carat,<span class="dt">data=</span>dat,<span class="dt">geom=</span><span class="st">&quot;histogram&quot;</span>)</code></pre></div>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="R_course_files/figure-html/unnamed-chunk-80-1.png" width="100%" style="display: block; margin: auto;" /></p>
<ol start="4" style="list-style-type: decimal">
<li>密度曲线图</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(carat,<span class="dt">data=</span>dat,<span class="dt">geom=</span><span class="st">&quot;density&quot;</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-81-1.png" width="100%" style="display: block; margin: auto;" /></p>
<ol start="5" style="list-style-type: decimal">
<li>条形图</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(color,<span class="dt">data=</span>dat,<span class="dt">geom=</span><span class="st">&quot;bar&quot;</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-82-1.png" width="100%" style="display: block; margin: auto;" /></p>
<ol start="6" style="list-style-type: decimal">
<li>时间序列线条图,采用另一个数据集economics</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(date,unemploy/pop,<span class="dt">data=</span>economics,<span class="dt">geom=</span><span class="st">&quot;line&quot;</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-83-1.png" width="100%" style="display: block; margin: auto;" /></p>
<ol start="7" style="list-style-type: decimal">
<li>分面绘图<code>facet</code>=分类变量</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(carat,<span class="dt">data=</span>dat,<span class="dt">facets=</span>color ~<span class="st"> </span>.,<span class="dt">geom=</span><span class="st">&quot;histogram&quot;</span>,<span class="dt">binwidth=</span><span class="fl">0.1</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">3</span>))</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-84-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>以上为ggplot2包中常见图形的快速绘制 (<code>quickplot</code>)即<code>qplot</code>函数的应用。 <code>qplot</code>函数还有很多其他的参数, 对<code>xlim</code>,<code>ylim</code>设置<code>x</code>,<code>y</code>轴的区间例如<code>xlim=c(0,20)</code>; 对轴取<code>log</code>值，<code>log=&quot;x&quot;</code>,对x轴取对数，<code>log=&quot;xy&quot;</code>表示对x轴和y轴取对数;main：图形的主题main=“qplot title”; #xlab,ylab:设置轴标签文字</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(carat,price,<span class="dt">data=</span>dat,
     <span class="dt">xlab=</span><span class="st">&quot;Price($)&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;Weight(carat)&quot;</span>,
     <span class="dt">main=</span><span class="st">&quot;Price-Weight relationship&quot;</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-85-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="section-3.2" class="section level2">
<h2><span class="header-section-number">3.2</span> 热图绘制</h2>
<p>热图是做分析时常用的展示方式，简单、直观、清晰。可以用来显示基因在不同样品中表达的高低、表观修饰水平的高低、样品之间的相关性等。任何一个数值矩阵都可以通过合适的方式用热图展示。</p>
<p>本篇使用R的<code>ggplot2</code>包实现从原始数据读入到热图输出的过程，并在教程结束后提供一份封装好的命令行绘图工具，只需要提供矩阵，即可一键绘图。</p>
<p>上一篇讲述了Rstudio的使用作为R写作和编译环境的入门，后面的命令都可以拷贝到Rstudio中运行，或写成一个R脚本，使用<code>Rscript heatmap.r</code>运行。我们还提供了Bash的封装，在不修改R脚本的情况下，改变参数绘制出不同的图形。</p>
<div id="section-3.2.1" class="section level3">
<h3><span class="header-section-number">3.2.1</span> 生成测试数据</h3>
<p>绘图首先需要数据。通过生成几组向量，转换为矩阵，得到想要的数据。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>:<span class="dv">6</span>,<span class="dv">6</span>:<span class="dv">1</span>,<span class="dv">6</span>:<span class="dv">1</span>,<span class="dv">1</span>:<span class="dv">6</span>, (<span class="dv">6</span>:<span class="dv">1</span>)/<span class="dv">10</span>,(<span class="dv">1</span>:<span class="dv">6</span>)/<span class="dv">10</span>,(<span class="dv">1</span>:<span class="dv">6</span>)/<span class="dv">10</span>,(<span class="dv">6</span>:<span class="dv">1</span>)/<span class="dv">10</span>,<span class="dv">1</span>:<span class="dv">6</span>,<span class="dv">6</span>:<span class="dv">1</span>,<span class="dv">6</span>:<span class="dv">1</span>,<span class="dv">1</span>:<span class="dv">6</span>,
        <span class="dv">6</span>:<span class="dv">1</span>,<span class="dv">1</span>:<span class="dv">6</span>,<span class="dv">1</span>:<span class="dv">6</span>,<span class="dv">6</span>:<span class="dv">1</span>)
data</code></pre></div>
<pre><code>##  [1] 1.0 2.0 3.0 4.0 5.0 6.0 6.0 5.0 4.0 3.0 2.0 1.0 6.0 5.0 4.0 3.0 2.0
## [18] 1.0 1.0 2.0 3.0 4.0 5.0 6.0 0.6 0.5 0.4 0.3 0.2 0.1 0.1 0.2 0.3 0.4
## [35] 0.5 0.6 0.1 0.2 0.3 0.4 0.5 0.6 0.6 0.5 0.4 0.3 0.2 0.1 1.0 2.0 3.0
## [52] 4.0 5.0 6.0 6.0 5.0 4.0 3.0 2.0 1.0 6.0 5.0 4.0 3.0 2.0 1.0 1.0 2.0
## [69] 3.0 4.0 5.0 6.0 6.0 5.0 4.0 3.0 2.0 1.0 1.0 2.0 3.0 4.0 5.0 6.0 1.0
## [86] 2.0 3.0 4.0 5.0 6.0 6.0 5.0 4.0 3.0 2.0 1.0</code></pre>
<p><strong>注意</strong>：运算符的优先级。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span>:<span class="dv">3+4</span></code></pre></div>
<pre><code>## [1] 5 6 7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(<span class="dv">1</span>:<span class="dv">3</span>)+<span class="dv">4</span></code></pre></div>
<pre><code>## [1] 5 6 7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span>:(<span class="dv">3+4</span>)</code></pre></div>
<pre><code>## [1] 1 2 3 4 5 6 7</code></pre>
<p>Vector转为矩阵 (matrix)，再转为数据框 (data.frame)。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># ncol: 指定列数</span>
<span class="co"># byrow: 先按行填充数据</span>
<span class="co"># ?matrix 可查看函数的使用方法</span>
<span class="co"># as.data.frame的as系列是转换用的</span>
data &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">matrix</span>(data, <span class="dt">ncol=</span><span class="dv">12</span>, <span class="dt">byrow=</span>T))
data</code></pre></div>
<pre><code>##    V1  V2  V3  V4  V5  V6  V7  V8  V9 V10 V11 V12
## 1 1.0 2.0 3.0 4.0 5.0 6.0 6.0 5.0 4.0 3.0 2.0 1.0
## 2 6.0 5.0 4.0 3.0 2.0 1.0 1.0 2.0 3.0 4.0 5.0 6.0
## 3 0.6 0.5 0.4 0.3 0.2 0.1 0.1 0.2 0.3 0.4 0.5 0.6
## 4 0.1 0.2 0.3 0.4 0.5 0.6 0.6 0.5 0.4 0.3 0.2 0.1
## 5 1.0 2.0 3.0 4.0 5.0 6.0 6.0 5.0 4.0 3.0 2.0 1.0
## 6 6.0 5.0 4.0 3.0 2.0 1.0 1.0 2.0 3.0 4.0 5.0 6.0
## 7 6.0 5.0 4.0 3.0 2.0 1.0 1.0 2.0 3.0 4.0 5.0 6.0
## 8 1.0 2.0 3.0 4.0 5.0 6.0 6.0 5.0 4.0 3.0 2.0 1.0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 增加列的名字</span>
<span class="kw">colnames</span>(data) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Zygote&quot;</span>,<span class="st">&quot;2_cell&quot;</span>,<span class="st">&quot;4_cell&quot;</span>,<span class="st">&quot;8_cell&quot;</span>,<span class="st">&quot;Morula&quot;</span>,<span class="st">&quot;ICM&quot;</span>,<span class="st">&quot;ESC&quot;</span>,
        <span class="st">&quot;4 week PGC&quot;</span>,<span class="st">&quot;7 week PGC&quot;</span>,<span class="st">&quot;10 week PGC&quot;</span>,<span class="st">&quot;17 week PGC&quot;</span>, <span class="st">&quot;OOcyte&quot;</span>)

<span class="co"># 增加行的名字</span>
<span class="co"># 注意paste和paste0的使用</span>
<span class="kw">rownames</span>(data) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;Gene&quot;</span>, <span class="dv">1</span>:<span class="dv">8</span>, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)

<span class="co"># 只显示前6行和前4列</span>
<span class="kw">head</span>(data)[,<span class="dv">1</span>:<span class="dv">4</span>]</code></pre></div>
<pre><code>##        Zygote 2_cell 4_cell 8_cell
## Gene_1    1.0    2.0    3.0    4.0
## Gene_2    6.0    5.0    4.0    3.0
## Gene_3    0.6    0.5    0.4    0.3
## Gene_4    0.1    0.2    0.3    0.4
## Gene_5    1.0    2.0    3.0    4.0
## Gene_6    6.0    5.0    4.0    3.0</code></pre>
<p>虽然方法比较繁琐，但一个数值矩阵已经获得了。</p>
<p>还有另外2种获取数值矩阵的方式。</p>
<ul>
<li>读入字符串</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 使用字符串的好处是不需要额外提供文件</span>
<span class="co"># 简单测试时可使用，写起来不繁琐，又方便重复</span>
<span class="co"># 尤其适用于在线提问时作为测试案例</span>
txt &lt;-<span class="st"> &quot;ID;Zygote;2_cell;4_cell;8_cell</span>
<span class="st">+ Gene_1;1;2;3;4</span>
<span class="st">+ Gene_2;6;5;4;5</span>
<span class="st">+ Gene_3;0.6;0.5;0.4;0.4&quot;</span>

<span class="co"># 习惯设置quote为空，避免部分基因名字或注释中存在引号，导致读入文件错误。</span>
<span class="co"># 具体错误可查看 http://blog.genesino.com/collections/R_tips/ 中的记录</span>
data2 &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>txt,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>)
<span class="kw">head</span>(data2)</code></pre></div>
<pre><code>##          Zygote X2_cell X4_cell X8_cell
## + Gene_1    1.0     2.0     3.0     4.0
## + Gene_2    6.0     5.0     4.0     5.0
## + Gene_3    0.6     0.5     0.4     0.4</code></pre>
<p>可以看到列名字中以数字开头的列都加了<strong>X</strong>。一般要尽量避免行或列名字以<strong>数字开头</strong>，会给后续分析带来匹配问题；另外名字中出现的非字母、数字、下划线、点的字符都会被转为<strong>点</strong>，也需要注意，尽量只用字母、下划线和数字。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 读入时，增加一个参数`check.names=F`也可以解决问题。</span>
<span class="co"># 这次数字前没有再加 X 了</span>
data2 &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>txt,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>, <span class="dt">check.names =</span> F)
<span class="kw">head</span>(data2)</code></pre></div>
<pre><code>##          Zygote 2_cell 4_cell 8_cell
## + Gene_1    1.0    2.0    3.0    4.0
## + Gene_2    6.0    5.0    4.0    5.0
## + Gene_3    0.6    0.5    0.4    0.4</code></pre>
<ul>
<li>读入文件</li>
</ul>
<p>与上一步类似，只是把<code>txt</code>代表的文字存到文件中，再利用文件名读取，不再赘述。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#data2 &lt;- read.table(&quot;filename&quot;,sep=&quot;;&quot;, header=T, row.names=1, quote=&quot;&quot;)</span></code></pre></div>
</div>
<div id="section-3.2.2" class="section level3">
<h3><span class="header-section-number">3.2.2</span> 转换数据格式</h3>
<p>数据读入后，还需要一步格式转换。在使用ggplot2作图时，有一种长表格模式是最为常用的，尤其是数据不规则时，更应该使用 （这点，我们在讲解箱线图时再说）。</p>
<p><code>melt</code>：把正常矩阵转换为长表格模式的函数。工作原理是把全部的非id列的<code>数值列</code>转为1列 (列名默认为<code>value</code>)；所有字符列转为一列，列名默认为<code>variable</code>。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 如果包没有安装，运行下面一句，安装包</span>
<span class="co">#install.packages(c(&quot;reshape2&quot;,&quot;ggplot2&quot;))</span>

<span class="kw">library</span>(reshape2)
<span class="kw">library</span>(ggplot2)

<span class="co"># 转换前，先增加一列ID列，保存行名字</span>
data$ID &lt;-<span class="st"> </span><span class="kw">rownames</span>(data)


<span class="co"># id.vars 列用于指定哪些列为id列；这些列不会被merge，会保留为完整一列。</span>
data_m &lt;-<span class="st"> </span><span class="kw">melt</span>(data, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>))
<span class="kw">head</span>(data_m)</code></pre></div>
<pre><code>##       ID variable value
## 1 Gene_1   Zygote   1.0
## 2 Gene_2   Zygote   6.0
## 3 Gene_3   Zygote   0.6
## 4 Gene_4   Zygote   0.1
## 5 Gene_5   Zygote   1.0
## 6 Gene_6   Zygote   6.0</code></pre>
</div>
<div id="section-3.2.3" class="section level3">
<h3><span class="header-section-number">3.2.3</span> 分解绘图</h3>
<p>数据转换后就可以画图了，分解命令如下：</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># data_m: 是前面费了九牛二虎之力得到的数据表</span>
<span class="co"># aes: aesthetic的缩写，一般指定整体的X轴、Y轴、颜色、形状、大小等。</span>
<span class="co">#      在最开始读入数据时，一般只指定x和y，其它后续指定</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable,<span class="dt">y=</span>ID)) 

<span class="co"># 热图就是一堆方块根据其值赋予不同的颜色，所以这里使用fill=value, 用数值做填充色。</span>
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>value)) 

<span class="co"># ggplot2为图层绘制，一层层添加，存储在p中，在输出p的内容时才会出图。</span>
p

## 如果你没有使用Rstudio或其它R图形版工具，而是在远程登录的服务器上运行的交互式R，
## 需要输入下面的语句，获得输出图形 （图形存储于R的工作目录下的Rplots.pdf文件中）。

## 如何指定输出，后面会讲到。
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-96-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>热图出来了，但有点不对劲，横轴重叠一起了。一个办法是调整图像的宽度，另一个是旋转横轴标记。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># theme: 是处理图美观的一个函数，可以调整横纵轴label的选择、图例的位置等。</span>
<span class="co"># 这里选择X轴标签45度。</span>
<span class="co"># hjust和vjust调整标签的相对位置，</span>
<span class="co"># 具体见下图。</span>
<span class="co"># 简单说，hjust是水平的对齐方式，0为左，1为右，0.5居中，0-1之间可以取任意值。</span>
<span class="co"># vjust是垂直对齐方式，0底对齐，1为顶对齐，0.5居中，0-1之间可以取任意值。</span>

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-97-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#knitr::include_graphics(&quot;images/hjust_vjust.png&quot;)</span>
td &lt;-<span class="st"> </span><span class="kw">expand.grid</span>(
    <span class="dt">hjust=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>),
    <span class="dt">vjust=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>),
    <span class="dt">angle=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">45</span>, <span class="dv">90</span>),
    <span class="dt">text=</span><span class="st">&quot;text&quot;</span>
)

<span class="kw">ggplot</span>(td, <span class="kw">aes</span>(<span class="dt">x=</span>hjust, <span class="dt">y=</span>vjust)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_point</span>() +
<span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>text, <span class="dt">angle=</span>angle, <span class="dt">hjust=</span>hjust, <span class="dt">vjust=</span>vjust)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">facet_grid</span>(~angle) +
<span class="st">    </span><span class="kw">scale_x_continuous</span>(<span class="dt">breaks=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="dt">expand=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.2</span>)) +
<span class="st">    </span><span class="kw">scale_y_continuous</span>(<span class="dt">breaks=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="dt">expand=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.2</span>))</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-98-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><a href="http://stackoverflow.com/questions/7263849/what-do-hjust-and-vjust-do-when-making-a-plot-using-ggplot" class="uri">http://stackoverflow.com/questions/7263849/what-do-hjust-and-vjust-do-when-making-a-plot-using-ggplot</a></p>
<p>设置想要的颜色。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 连续的数字，指定最小数值代表的颜色和最大数值赋予的颜色</span>
<span class="co"># 注意fill和color的区别，fill是填充，color只针对边缘</span>
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">scale_fill_gradient</span>(<span class="dt">low =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">high =</span> <span class="st">&quot;red&quot;</span>)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-99-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>调整legend的位置, <code>legend.position</code>, 可以接受的值有 <code>top</code>, <code>bottom</code>, <code>left</code>, <code>right</code>, 和一个坐标 <code>c(0.05,0.8)</code> (左上角，坐标是相对于图的左下角(即<strong>原点</strong>)计算的）</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;top&quot;</span>)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-100-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>调整背景和背景格线以及X轴、Y轴的标题。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;samples&quot;</span>) +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.key=</span><span class="kw">element_blank</span>())
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-101-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>为了使横轴旋转45度，需要把这句话<code>theme(axis.text.x=element_text(angle=45,hjust=1, vjust=1))</code>放在<code>theme_bw()</code>的后面。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-102-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>合并以上命令，就得到了下面这个看似复杂的绘图命令。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable,<span class="dt">y=</span>ID)) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;samples&quot;</span>) +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.key=</span><span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;top&quot;</span>) +<span class="st">  </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>value)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">scale_fill_gradient</span>(<span class="dt">low =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">high =</span> <span class="st">&quot;red&quot;</span>) +
<span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color=</span>value), <span class="dt">size=</span><span class="dv">6</span>) +
<span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>value))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-103-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>也可以只用Point</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable,<span class="dt">y=</span>ID)) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;samples&quot;</span>) +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.key=</span><span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;top&quot;</span>) +<span class="st">      </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color=</span>value), <span class="dt">size=</span><span class="dv">6</span>) +
<span class="st">  </span><span class="kw">scale_color_gradient</span>(<span class="dt">low =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">high =</span> <span class="st">&quot;red&quot;</span>) +
<span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>value))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-104-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="section-3.2.4" class="section level3">
<h3><span class="header-section-number">3.2.4</span> 图形存储</h3>
<p>图形出来了，就得考虑存储了，一般输出为<code>PDF</code>格式，方便后期的修改。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 可以跟输出文件不同的后缀，以获得不同的输出格式</span>
<span class="co"># colormode支持srgb （屏幕）和cmyk （打印，部分杂志需要，看上去有点褪色的感觉）格式</span>
<span class="kw">ggsave</span>(p, <span class="dt">filename=</span><span class="st">&quot;heatmap.pdf&quot;</span>, <span class="dt">width=</span><span class="dv">10</span>,
        <span class="dt">height=</span><span class="dv">15</span>, <span class="dt">units=</span><span class="kw">c</span>(<span class="st">&quot;cm&quot;</span>),<span class="dt">colormodel=</span><span class="st">&quot;srgb&quot;</span>)</code></pre></div>
<p>点击下载：<a href="heatmap.pdf">pdf</a></p>
<p>至此，完成了简单的heatmap的绘图。但实际绘制时，经常会碰到由于数值变化很大，导致颜色过于集中，使得图的可读性下降很多。因此需要对数据进行一些处理，具体的下次再说。</p>
</div>
</div>
<div id="section-3.3" class="section level2">
<h2><span class="header-section-number">3.3</span> 热图美化</h2>
<p>上面的测试数据，数值的分布比较均一，相差不是太大，但是<code>Gene_4</code>和<code>Gene_5</code>由于整体的值低于其它的基因，从颜色上看，不仔细看，看不出差别。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rnorm</span>(<span class="dv">5</span>,<span class="dt">mean=</span><span class="dv">5</span>), <span class="kw">rnorm</span>(<span class="dv">5</span>,<span class="dt">mean=</span><span class="dv">20</span>), <span class="kw">rnorm</span>(<span class="dv">5</span>, <span class="dt">mean=</span><span class="dv">100</span>), <span class="kw">c</span>(<span class="dv">600</span>,<span class="dv">700</span>,<span class="dv">800</span>,<span class="dv">900</span>,<span class="dv">10000</span>))
data &lt;-<span class="st"> </span><span class="kw">matrix</span>(data, <span class="dt">ncol=</span><span class="dv">5</span>, <span class="dt">byrow=</span>T)
data &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(data)
<span class="kw">rownames</span>(data) &lt;-<span class="st"> </span>letters[<span class="dv">1</span>:<span class="dv">4</span>]
<span class="kw">colnames</span>(data) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;Grp&quot;</span>, <span class="dv">1</span>:<span class="dv">5</span>, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)
data</code></pre></div>
<pre><code>##        Grp_1      Grp_2      Grp_3      Grp_4        Grp_5
## a   6.087539   5.245061   7.075796   4.820246     4.374214
## b  18.391807  18.963660  20.210934  19.701757    21.620189
## c 100.005960  99.495593  99.766925  99.985531   100.411950
## d 600.000000 700.000000 800.000000 900.000000 10000.000000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data$ID &lt;-<span class="st"> </span><span class="kw">rownames</span>(data)
data</code></pre></div>
<pre><code>##        Grp_1      Grp_2      Grp_3      Grp_4        Grp_5 ID
## a   6.087539   5.245061   7.075796   4.820246     4.374214  a
## b  18.391807  18.963660  20.210934  19.701757    21.620189  b
## c 100.005960  99.495593  99.766925  99.985531   100.411950  c
## d 600.000000 700.000000 800.000000 900.000000 10000.000000  d</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_m &lt;-<span class="st"> </span><span class="kw">melt</span>(data, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>))
<span class="kw">head</span>(data_m)</code></pre></div>
<pre><code>##   ID variable      value
## 1  a    Grp_1   6.087539
## 2  b    Grp_1  18.391807
## 3  c    Grp_1 100.005960
## 4  d    Grp_1 600.000000
## 5  a    Grp_2   5.245061
## 6  b    Grp_2  18.963660</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable,<span class="dt">y=</span>ID)) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;samples&quot;</span>) +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.key=</span><span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;top&quot;</span>) +<span class="st">  </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>value)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">scale_fill_gradient</span>(<span class="dt">low =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">high =</span> <span class="st">&quot;red&quot;</span>)
p
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-109-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>图中只有右上角可以看到红色，其他地方就没了颜色的差异。这通常不是我们想要的。为了更好的可视化效果，需要对数据做些预处理，主要有 <code>对数转换</code>，<code>Z-score转换</code>，<code>抹去异常值</code>，<code>非线性颜色</code>等方式。</p>
<div id="section-3.3.1" class="section level3">
<h3><span class="header-section-number">3.3.1</span> 对数转换</h3>
<p>假设下面的数据是基因表达数据，4个基因 (a, b, c, d)和5个样品 (Grp_1, Grp_2, Grp_3, Grp_4)，矩阵中的值代表基因表达FPKM值。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rnorm</span>(<span class="dv">5</span>,<span class="dt">mean=</span><span class="dv">5</span>), <span class="kw">rnorm</span>(<span class="dv">5</span>,<span class="dt">mean=</span><span class="dv">20</span>), <span class="kw">rnorm</span>(<span class="dv">5</span>, <span class="dt">mean=</span><span class="dv">100</span>), <span class="kw">c</span>(<span class="dv">600</span>,<span class="dv">700</span>,<span class="dv">800</span>,<span class="dv">900</span>,<span class="dv">10000</span>))
data &lt;-<span class="st"> </span><span class="kw">matrix</span>(data, <span class="dt">ncol=</span><span class="dv">5</span>, <span class="dt">byrow=</span>T)
data &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(data)
<span class="kw">rownames</span>(data) &lt;-<span class="st"> </span>letters[<span class="dv">1</span>:<span class="dv">4</span>]
<span class="kw">colnames</span>(data) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;Grp&quot;</span>, <span class="dv">1</span>:<span class="dv">5</span>, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)
data</code></pre></div>
<pre><code>##        Grp_1      Grp_2     Grp_3      Grp_4        Grp_5
## a   4.125538   4.087054   3.52572   5.749561     2.420971
## b  21.296682  19.922518  19.41533  20.510619    20.319078
## c 100.188919  99.774166  99.88726  98.414636   100.442783
## d 600.000000 700.000000 800.00000 900.000000 10000.000000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_log &lt;-<span class="st"> </span><span class="kw">log2</span>(data<span class="dv">+1</span>)
data_log</code></pre></div>
<pre><code>##      Grp_1    Grp_2    Grp_3    Grp_4     Grp_5
## a 2.357704 2.346830 2.178147 2.754794  1.774406
## b 4.478757 4.386985 4.351581 4.426977  4.414073
## c 6.660907 6.654982 6.656600 6.635386  6.664522
## d 9.231221 9.453271 9.645658 9.815383 13.287857</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_log$ID =<span class="st"> </span><span class="kw">rownames</span>(data_log)
data_log_m =<span class="st"> </span><span class="kw">melt</span>(data_log, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>))

p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_log_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable,<span class="dt">y=</span>ID)) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;samples&quot;</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="ot">NULL</span>) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme_bw</span>() +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.key=</span><span class="kw">element_blank</span>()) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;top&quot;</span>) +
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>,<span class="dt">vjust=</span><span class="dv">1</span>)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>value)) +<span class="st"> </span><span class="kw">scale_fill_gradient</span>(<span class="dt">low =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">high =</span> <span class="st">&quot;red&quot;</span>)
p
<span class="co">#ggsave(p, filename=&quot;heatmap_log.pdf&quot;, width=8, height=12, units=c(&quot;cm&quot;),colormodel=&quot;srgb&quot;)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-112-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>对数转换后的数据，看起来就清晰的多了。而且对数转换后，数据还保留着之前的变化趋势，不只是基因在不同样品之间的表达可比 (同一行的不同列)，不同基因在同一样品的值也可比 (同一列的不同行) (不同基因之间比较表达值存在理论上的问题，即便是按照长度标准化之后的FPKM也不代表基因之间是完全可比的)。</p>
</div>
<div id="z-score" class="section level3">
<h3><span class="header-section-number">3.3.2</span> Z-score转换</h3>
<p><code>Z-score</code>又称为标准分数，是一组数中的每个数减去这一组数的平均值再除以这一组数的标准差，代表的是原始分数距离原始平均值的距离，以标准差为单位。可以对不同分布的各原始分数进行比较，用来反映数据的相对变化趋势，而非绝对变化量。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_ori &lt;-<span class="st"> &quot;Grp_1;Grp_2;Grp_3;Grp_4;Grp_5</span>
<span class="st">a;6.6;20.9;100.1;600.0;5.2</span>
<span class="st">b;20.8;99.8;700.0;3.7;19.2</span>
<span class="st">c;100.0;800.0;6.2;21.4;98.6</span>
<span class="st">d;900;3.3;20.3;101.1;10000&quot;</span>

data &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>data_ori, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>)

<span class="co"># 去掉方差为0的行，也就是值全都一致的行</span>
data &lt;-<span class="st"> </span>data[<span class="kw">apply</span>(data,<span class="dv">1</span>,var)!=<span class="dv">0</span>,]

data</code></pre></div>
<pre><code>##   Grp_1 Grp_2 Grp_3 Grp_4   Grp_5
## a   6.6  20.9 100.1 600.0     5.2
## b  20.8  99.8 700.0   3.7    19.2
## c 100.0 800.0   6.2  21.4    98.6
## d 900.0   3.3  20.3 101.1 10000.0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 标准化数据，并转换为data.frame</span>
data_scale &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(<span class="kw">apply</span>(data,<span class="dv">1</span>,scale)))

<span class="co"># 重命名列</span>
<span class="kw">colnames</span>(data_scale) &lt;-<span class="st"> </span><span class="kw">colnames</span>(data)
data_scale</code></pre></div>
<pre><code>##        Grp_1      Grp_2      Grp_3      Grp_4      Grp_5
## a -0.5456953 -0.4899405 -0.1811446  1.7679341 -0.5511538
## b -0.4940465 -0.2301542  1.7747592 -0.5511674 -0.4993911
## c -0.3139042  1.7740182 -0.5936858 -0.5483481 -0.3180801
## d -0.2983707 -0.5033986 -0.4995116 -0.4810369  1.7823177</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_scale$ID =<span class="st"> </span><span class="kw">rownames</span>(data_scale)
data_scale_m =<span class="st"> </span><span class="kw">melt</span>(data_scale, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>))
data_scale_m$value &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">prettyNum</span>(data_scale_m$value, <span class="dt">digits=</span><span class="dv">2</span>))
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_scale_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable,<span class="dt">y=</span>ID)) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;samples&quot;</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="ot">NULL</span>) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme_bw</span>() +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.key=</span><span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>value)) +<span class="st"> </span><span class="kw">scale_fill_gradient</span>(<span class="dt">low =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">high =</span> <span class="st">&quot;red&quot;</span>) +
<span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>value))
p
<span class="co">#ggsave(p, filename=&quot;heatmap_scale.pdf&quot;, width=8, height=12, units=c(&quot;cm&quot;), </span>
<span class="co">#    colormodel=&quot;srgb&quot;)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-115-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><code>Z-score</code>转换后，颜色分布也相对均一了，每个基因在不同样品之间的表达的高低一目了然。但是不同基因之间就完全不可比了。</p>
</div>
<div id="section-3.3.3" class="section level3">
<h3><span class="header-section-number">3.3.3</span> 抹去异常值</h3>
<p>粗暴一点，假设检测饱和度为100，大于100的值都视为100对待。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_ori &lt;-<span class="st"> &quot;Grp_1;Grp_2;Grp_3;Grp_4;Grp_5</span>
<span class="st">a;6.6;20.9;100.1;600.0;5.2</span>
<span class="st">b;20.8;99.8;700.0;3.7;19.2</span>
<span class="st">c;100.0;800.0;6.2;21.4;98.6</span>
<span class="st">d;900;3.3;20.3;101.1;10000&quot;</span>

data &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>data_ori, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>)

data[data&gt;<span class="dv">100</span>] &lt;-<span class="st"> </span><span class="dv">100</span>
data</code></pre></div>
<pre><code>##   Grp_1 Grp_2 Grp_3 Grp_4 Grp_5
## a   6.6  20.9 100.0 100.0   5.2
## b  20.8  99.8 100.0   3.7  19.2
## c 100.0 100.0   6.2  21.4  98.6
## d 100.0   3.3  20.3 100.0 100.0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data$ID =<span class="st"> </span><span class="kw">rownames</span>(data)
data_m =<span class="st"> </span><span class="kw">melt</span>(data, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>))

p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable,<span class="dt">y=</span>ID)) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;samples&quot;</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="ot">NULL</span>) +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.key=</span><span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>value)) +<span class="st"> </span><span class="kw">scale_fill_gradient</span>(<span class="dt">low =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">high =</span> <span class="st">&quot;red&quot;</span>) +
<span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>value))
p
<span class="co">#ggsave(p, filename=&quot;heatmap_nooutlier.pdf&quot;, width=8, height=12, units=c(&quot;cm&quot;),</span>
<span class="co">#    colormodel=&quot;srgb&quot;)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-117-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>虽然损失了一部分信息，但整体模式还是出来了。<strong>但是在选择异常值标准时需要根据实际确认</strong>。</p>
</div>
<div id="section-3.3.4" class="section level3">
<h3><span class="header-section-number">3.3.4</span> 非线性颜色</h3>
<p>正常来讲，颜色的赋予在最小值到最大值之间是均匀分布的。如果最小值到最大值之间用100个颜色区分，则其中每一个<code>bin</code>，不论其大小、有没有值都会赋予一个颜色。非线性颜色则是对数据比较小但密集的地方赋予更多颜色，数据大但分布散的地方赋予更少颜色，这样既能加大区分度，又最小的影响原始数值。通常可以根据数据模式，手动设置颜色区间。为了方便自动化处理，也可选择用<strong>四分位数</strong>的方式设置颜色区间。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_ori &lt;-<span class="st"> &quot;Grp_1;Grp_2;Grp_3;Grp_4;Grp_5</span>
<span class="st">a;6.6;20.9;100.1;600.0;5.2</span>
<span class="st">b;20.8;99.8;700.0;3.7;19.2</span>
<span class="st">c;100.0;800.0;6.2;21.4;98.6</span>
<span class="st">d;900;3.3;20.3;101.1;10000&quot;</span>

data &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>data_ori, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>)

data</code></pre></div>
<pre><code>##   Grp_1 Grp_2 Grp_3 Grp_4   Grp_5
## a   6.6  20.9 100.1 600.0     5.2
## b  20.8  99.8 700.0   3.7    19.2
## c 100.0 800.0   6.2  21.4    98.6
## d 900.0   3.3  20.3 101.1 10000.0</code></pre>
<p>获取数据的最大、最小、第一四分位数、中位数、第三四分位数</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data$ID =<span class="st"> </span><span class="kw">rownames</span>(data)
data_m =<span class="st"> </span><span class="kw">melt</span>(data, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;ID&quot;</span>))
summary_v &lt;-<span class="st"> </span><span class="kw">summary</span>(data_m$value)
summary_v</code></pre></div>
<pre><code>##     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
##     3.30    16.05    60.00   681.36   225.82 10000.00</code></pre>
<p>在最小值和第一四分位数之间划出6个区间，第一四分位数和中位数之间划出6个区间，中位数和第三四分位数之间划出5个区间，最后的数划出5个区间</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">break_v &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(<span class="kw">seq</span>(summary_v[<span class="dv">1</span>]*<span class="fl">0.95</span>,summary_v[<span class="dv">2</span>],<span class="dt">length=</span><span class="dv">6</span>), 
        <span class="kw">seq</span>(summary_v[<span class="dv">2</span>],summary_v[<span class="dv">3</span>],<span class="dt">length=</span><span class="dv">6</span>),<span class="kw">seq</span>(summary_v[<span class="dv">3</span>],summary_v[<span class="dv">5</span>],<span class="dt">length=</span><span class="dv">5</span>),
        <span class="kw">seq</span>(summary_v[<span class="dv">5</span>],summary_v[<span class="dv">6</span>]*<span class="fl">1.05</span>,<span class="dt">length=</span><span class="dv">5</span>)))
break_v</code></pre></div>
<pre><code>##  [1]     3.1350     5.7180     8.3010    10.8840    13.4670    16.0500
##  [7]    24.8400    33.6300    42.4200    51.2100    60.0000   101.4562
## [13]   142.9125   184.3687   225.8250  2794.3687  5362.9125  7931.4562
## [19] 10500.0000</code></pre>
<p>按照设定的区间分割数据, 原始数据替换为了其所在的区间的数值</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_m$value &lt;-<span class="st"> </span><span class="kw">cut</span>(data_m$value, <span class="dt">breaks=</span>break_v,<span class="dt">labels=</span>break_v[<span class="dv">2</span>:<span class="kw">length</span>(break_v)])
break_v=<span class="kw">unique</span>(data_m$value)

data_m</code></pre></div>
<pre><code>##    ID variable      value
## 1   a    Grp_1      8.301
## 2   b    Grp_1      24.84
## 3   c    Grp_1  101.45625
## 4   d    Grp_1 2794.36875
## 5   a    Grp_2      24.84
## 6   b    Grp_2  101.45625
## 7   c    Grp_2 2794.36875
## 8   d    Grp_2      5.718
## 9   a    Grp_3  101.45625
## 10  b    Grp_3 2794.36875
## 11  c    Grp_3      8.301
## 12  d    Grp_3      24.84
## 13  a    Grp_4 2794.36875
## 14  b    Grp_4      5.718
## 15  c    Grp_4      24.84
## 16  d    Grp_4  101.45625
## 17  a    Grp_5      5.718
## 18  b    Grp_5      24.84
## 19  c    Grp_5  101.45625
## 20  d    Grp_5      10500</code></pre>
<p>虽然看上去还是数值，但已经不是数字类型了，而是不同的因子了，这样就可以对不同的因子赋予不同的颜色了</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">is.numeric</span>(data_m$value)</code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">is.factor</span>(data_m$value)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">break_v</code></pre></div>
<pre><code>## [1] 8.301      24.84      101.45625  2794.36875 5.718      10500     
## 18 Levels: 5.718 8.301 10.884 13.467 16.05 24.84 33.63 42.42 51.21 ... 10500</code></pre>
<p>产生对应数目的颜色</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gradientC=<span class="kw">c</span>(<span class="st">&#39;green&#39;</span>,<span class="st">&#39;yellow&#39;</span>,<span class="st">&#39;red&#39;</span>)
col &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(gradientC)(<span class="kw">length</span>(break_v))
col</code></pre></div>
<pre><code>## [1] &quot;#00FF00&quot; &quot;#66FF00&quot; &quot;#CCFF00&quot; &quot;#FFCB00&quot; &quot;#FF6500&quot; &quot;#FF0000&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable,<span class="dt">y=</span>ID)) +<span class="st"> </span><span class="kw">xlab</span>(<span class="st">&quot;samples&quot;</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="ot">NULL</span>) +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.key=</span><span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>)) +<span class="st"> </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>value))

<span class="co"># 与上面不同的地方，使用的是scale_fill_manual逐个赋值</span>
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">scale_fill_manual</span>(<span class="dt">values=</span>col)
p
<span class="co">#ggsave(p, filename=&quot;heatmap_nonlinear.pdf&quot;, width=8, height=12, units=c(&quot;cm&quot;),colormodel=&quot;srgb&quot;)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-126-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="section-3.3.5" class="section level3">
<h3><span class="header-section-number">3.3.5</span> 调整行或列的顺序</h3>
<p>如果想保持图中每一行的顺序与输入的数据框一致，需要设置因子的水平。这也是<code>ggplot2</code>中调整图例或横纵轴字符顺序的常用方式。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_rowname &lt;-<span class="st"> </span><span class="kw">rownames</span>(data)
data_rowname &lt;-<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">rownames</span>(data))
data_rownames &lt;-<span class="st"> </span><span class="kw">rev</span>(data_rowname)
data_log_m$ID &lt;-<span class="st"> </span><span class="kw">factor</span>(data_log_m$ID, <span class="dt">levels=</span>data_rownames, <span class="dt">ordered=</span>T)
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_log_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable,<span class="dt">y=</span>ID)) +<span class="st"> </span><span class="kw">xlab</span>(<span class="ot">NULL</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="ot">NULL</span>) +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.key=</span><span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;top&quot;</span>) +<span class="st">  </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>value)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">scale_fill_gradient</span>(<span class="dt">low =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">high =</span> <span class="st">&quot;red&quot;</span>)
p
<span class="co">#ggsave(p, filename=&quot;heatmap_log.pdf&quot;, width=8, height=12, units=c(&quot;cm&quot;),colormodel=&quot;srgb&quot;)</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-127-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>基于ggplot2的heatmap绘制到现在就差不多了，但总是这么画下去也会觉得有点累，有没有办法更简化呢？。</p>
</div>
</div>
<div id="---pheatmap" class="section level2">
<h2><span class="header-section-number">3.4</span> 热图绘制 - pheatmap</h2>
<p>绘制热图除了使用<code>ggplot2</code>，还可以有其它的包或函数，比如<code>pheatmap::pheatmap</code> (pheatmap包中的pheatmap函数)、<code>gplots::heatmap.2</code>等。</p>
<p>相比于<code>ggplot2</code>作heatmap, <code>pheatmap</code>会更为简单一些，一个函数设置不同的参数，可以完成行列聚类、行列注释、<code>Z-score</code>计算、颜色自定义等。那我们来看看效果怎样。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_ori &lt;-<span class="st"> &quot;Grp_1;Grp_2;Grp_3;Grp_4;Grp_5</span>
<span class="st">a;6.6;20.9;100.1;600.0;5.2</span>
<span class="st">b;20.8;99.8;700.0;3.7;19.2</span>
<span class="st">c;100.0;800.0;6.2;21.4;98.6</span>
<span class="st">d;900;3.3;20.3;101.1;10000&quot;</span>

data &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>data_ori, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#pheatmap::pheatmap(data, filename=&quot;pheatmap_1.pdf&quot;)</span>
pheatmap::<span class="kw">pheatmap</span>(data)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-129-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>虽然有点丑，但一步就出来了。</p>
<p>在<code>heatmap美化</code>篇提到的数据前期处理方式，都可以用于<code>pheatmap</code>的画图。此外<code>Z-score</code>计算在<code>pheatmap</code>中只要一个参数就可以实现。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheatmap::<span class="kw">pheatmap</span>(data, <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-130-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>有时可能不需要行或列的聚类，原始展示就可以了。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheatmap::<span class="kw">pheatmap</span>(data, <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>, <span class="dt">cluster_rows=</span><span class="ot">FALSE</span>, <span class="dt">cluster_cols=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-131-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>给矩阵 (<code>data</code>)中行和列不同的分组注释。假如有两个文件，第一个文件为行注释，其第一列与矩阵中的第一列内容相同 (顺序没有关系)，其它列为第一列的不同的标记，如下面示例中(假设行为基因，列为样品)的2,3列对应基因的不同类型 (TF or enzyme)和不同分组。第二个文件为列注释，其第一列与矩阵中第一行内容相同，其它列则为样品的注释。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">row_anno =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">type=</span><span class="kw">c</span>(<span class="st">&quot;TF&quot;</span>,<span class="st">&quot;Enzyme&quot;</span>,<span class="st">&quot;Enzyme&quot;</span>,<span class="st">&quot;TF&quot;</span>), 
        <span class="dt">class=</span><span class="kw">c</span>(<span class="st">&quot;clu1&quot;</span>,<span class="st">&quot;clu1&quot;</span>,<span class="st">&quot;clu2&quot;</span>,<span class="st">&quot;clu2&quot;</span>), <span class="dt">row.names=</span><span class="kw">rownames</span>(data))
row_anno</code></pre></div>
<pre><code>##     type class
## a     TF  clu1
## b Enzyme  clu1
## c Enzyme  clu2
## d     TF  clu2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">col_anno =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">grp=</span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;A&quot;</span>,<span class="st">&quot;A&quot;</span>,<span class="st">&quot;B&quot;</span>,<span class="st">&quot;B&quot;</span>), <span class="dt">size=</span><span class="dv">1</span>:<span class="dv">5</span>, <span class="dt">row.names=</span><span class="kw">colnames</span>(data))
col_anno</code></pre></div>
<pre><code>##       grp size
## Grp_1   A    1
## Grp_2   A    2
## Grp_3   A    3
## Grp_4   B    4
## Grp_5   B    5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheatmap::<span class="kw">pheatmap</span>(data, <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>, 
<span class="dt">cluster_rows=</span><span class="ot">FALSE</span>, <span class="dt">annotation_col=</span>col_anno, <span class="dt">annotation_row=</span>row_anno)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-134-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>自定义下颜色吧。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># &lt;bias&gt; values larger than 1 will give more color for high end. </span>
<span class="co"># Values between 0-1 will give more color for low end.</span>
pheatmap::<span class="kw">pheatmap</span>(data, <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>, <span class="dt">cluster_rows=</span><span class="ot">FALSE</span>, 
<span class="dt">annotation_col=</span>col_anno, <span class="dt">annotation_row=</span>row_anno,
<span class="dt">color=</span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&#39;green&#39;</span>,<span class="st">&#39;yellow&#39;</span>,<span class="st">&#39;red&#39;</span>), <span class="dt">bias=</span><span class="dv">1</span>)(<span class="dv">50</span>))</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-135-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><code>heatmap.2</code>的使用在上一期转录组分析绘制相关性热图时有提到，这次就不介绍了，跟<code>pheatmap</code>有些类似，而且也有不少教程。</p>
</div>
<div id="section-3.5" class="section level2">
<h2><span class="header-section-number">3.5</span> 箱线图</h2>
<p>箱线图是能同时反映数据统计量和整体分布，又很漂亮的展示图。在2014年的Nature Method上有2篇<code>Correspondence</code>论述了使用箱线图的好处和一个在线绘制箱线图的工具。就这样都可以发两篇Nature method，没天理，但也说明了箱线图的重要意义。</p>
<p>下面这张图展示了<code>Bar plot</code>、<code>Box plot</code>、<code>Volin plot</code>和<code>Bean plot</code>对数据分布的反应。从Bar plot上只能看到数据标准差或标准误不同；Box plot可以看到数据分布的集中性不同；Violin plot和Bean plot展示的是数据真正的分布，尤其是对Biomodal数据的展示。</p>
<p>Boxplot从下到上展示的是最小值，第一四分位数 (箱子的下边线)、中位数 (箱子中间的线)、第三四分位数 (箱子上边线)、最大值，具体解读参见 <a href="http://mp.weixin.qq.com/s/t3UTI_qAIi0cy1g6ZmHtwg" class="uri">http://mp.weixin.qq.com/s/t3UTI_qAIi0cy1g6ZmHtwg</a>。</p>
<p><img src="images/boxplot_nm.png" width="100%" style="display: block; margin: auto;" /></p>
<ul>
<li>Nature Method文章 <a href="http://www.nature.com/nmeth/journal/v11/n2/full/nmeth.2811.html" class="uri">http://www.nature.com/nmeth/journal/v11/n2/full/nmeth.2811.html</a></li>
</ul>
<div id="section-3.5.1" class="section level3">
<h3><span class="header-section-number">3.5.1</span> 一步步解析箱线图绘制</h3>
<p>假设有这么一个基因表达矩阵，第一列为基因名字，第一行为样品名字，想绘制样品中基因表达的整体分布。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">profile=<span class="st">&quot;Name;2cell_1;2cell_2;2cell_3;4cell_1;4cell_2;4cell_3;zygote_1;zygote_2;zygote_3</span>
<span class="st">A;4;6;7;3.2;5.2;5.6;2;4;3</span>
<span class="st">B;6;8;9;5.2;7.2;7.6;4;6;5</span>
<span class="st">C;8;10;11;7.2;9.2;9.6;6;8;7</span>
<span class="st">D;10;12;13;9.2;11.2;11.6;8;10;9</span>
<span class="st">E;12;14;15;11.2;13.2;13.6;10;12;11</span>
<span class="st">F;14;16;17;13.2;15.2;15.6;12;14;13</span>
<span class="st">G;15;17;18;14.2;16.2;16.6;13;15;14</span>
<span class="st">H;16;18;19;15.2;17.2;17.6;14;16;15</span>
<span class="st">I;17;19;20;16.2;18.2;18.6;15;17;16</span>
<span class="st">J;18;20;21;17.2;19.2;19.6;16;18;17</span>
<span class="st">L;19;21;22;18.2;20.2;20.6;17;19;18</span>
<span class="st">M;20;22;23;19.2;21.2;21.6;18;20;19</span>
<span class="st">N;21;23;24;20.2;22.2;22.6;19;21;20</span>
<span class="st">O;22;24;25;21.2;23.2;23.6;20;22;21&quot;</span></code></pre></div>
<p>读入数据并转换为ggplot2需要的长数据表格式，好好体会下这个格式，虽然多占用了不少空间，但是确实很方便。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">profile_text &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>profile, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">check.names=</span>F)
<span class="co"># 在melt时保留位置信息</span>
<span class="co"># melt格式是ggplot2画图最喜欢的格式</span>
<span class="co"># </span>

<span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(reshape2)
data_m &lt;-<span class="st"> </span><span class="kw">melt</span>(profile_text)</code></pre></div>
<pre><code>## No id variables; using all as measure variables</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(data_m)</code></pre></div>
<pre><code>##   variable value
## 1  2cell_1     4
## 2  2cell_1     6
## 3  2cell_1     8
## 4  2cell_1    10
## 5  2cell_1    12
## 6  2cell_1    14</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(data_m)</code></pre></div>
<pre><code>##     variable      value      
##  2cell_1:14   Min.   : 2.00  
##  2cell_2:14   1st Qu.:10.25  
##  2cell_3:14   Median :16.00  
##  4cell_1:14   Mean   :14.87  
##  4cell_2:14   3rd Qu.:19.20  
##  4cell_3:14   Max.   :25.00  
##  (Other):42</code></pre>
<p><code>variable</code>和<code>value</code>为矩阵<code>melt</code>后的两列的名字，内部变量, 可以通过<code>?melt</code>查看如何修改。<code>variable</code>代表了点线的属性，<code>value</code>代表对应的值。 像往常一样，就可以直接画图了。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value,<span class="dt">color=</span>variable)) +<span class="st"> </span>
<span class="kw">geom_boxplot</span>() +<span class="st"> </span>
<span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">50</span>,<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">vjust=</span><span class="fl">0.5</span>)) +
<span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>)
p
<span class="co"># 图会存储在当前目录的Rplots.pdf文件中，如果用Rstudio，可以不运行dev.off()</span>
<span class="co"># dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-140-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>箱线图出来了，看上去还可以，再加点色彩 (<code>fill</code>)。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># variable和value为矩阵melt后的两列的名字，内部变量, variable代表了点线的属性，value代表对应的值。</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value)) +<span class="st"> </span>
<span class="kw">geom_boxplot</span>(<span class="kw">aes</span>(<span class="dt">fill=</span><span class="kw">factor</span>(variable))) +<span class="st"> </span>
<span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">50</span>,<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">vjust=</span><span class="fl">0.5</span>)) +
<span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>)
p
<span class="co"># 图会存储在当前目录的Rplots.pdf文件中，如果用Rstudio，可以不运行dev.off()</span>
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-141-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>再看看<code>Violin plot</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># variable和value为矩阵melt后的两列的名字，内部变量, variable代表了点线的属性，value代表对应的值。</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value)) +<span class="st"> </span>
<span class="kw">geom_violin</span>(<span class="kw">aes</span>(<span class="dt">fill=</span><span class="kw">factor</span>(variable))) +<span class="st"> </span>
<span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">50</span>,<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">vjust=</span><span class="fl">0.5</span>)) +
<span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>)
p
<span class="co"># 图会存储在当前目录的Rplots.pdf文件中，如果用Rstudio，可以不运行dev.off()</span>
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-142-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># variable和value为矩阵melt后的两列的名字，内部变量, variable代表了点线的属性，value代表对应的值。</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value)) +<span class="st"> </span>
<span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">color=</span><span class="kw">factor</span>(variable))) +<span class="st"> </span>
<span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">50</span>,<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">vjust=</span><span class="fl">0.5</span>)) +
<span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>)
p
<span class="co"># 图会存储在当前目录的Rplots.pdf文件中，如果用Rstudio，可以不运行dev.off()</span>
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-143-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>还有Jitter plot (这里使用的是ggbeeswarm包)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggbeeswarm)
<span class="co"># 为了更好的效果，只保留其中一个样品的数据</span>
<span class="co"># grepl类似于Linux的grep命令，获取特定模式的字符串</span>

data_m2 &lt;-<span class="st"> </span>data_m[<span class="kw">grepl</span>(<span class="st">&quot;_3&quot;</span>, data_m$variable),]

<span class="co"># variable和value为矩阵melt后的两列的名字，内部变量, </span>
<span class="co"># variable代表了点线的属性，value代表对应的值。</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m2, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value),<span class="dt">color=</span>variable) +<span class="st"> </span>
<span class="kw">geom_quasirandom</span>(<span class="kw">aes</span>(<span class="dt">colour=</span><span class="kw">factor</span>(variable)), <span class="dt">groupOnX=</span><span class="ot">FALSE</span>) +<span class="st"> </span>
<span class="kw">theme_bw</span>() +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(),
<span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>(), <span class="dt">legend.key=</span><span class="kw">element_blank</span>()) +
<span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>)
p
<span class="co">#ggsave(p, filename=&quot;jitterplot.pdf&quot;, width=14, height=8, units=c(&quot;cm&quot;))</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-144-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><img src="images/boxplot_4.png" width="100%" style="display: block; margin: auto;" /></p>
<p>也可以用<code>geom_jitter(aes(colour=factor(variable)))</code>代替<code>geom_quasirandom(aes(colour=factor(variable)))</code>绘制抖动图，但个人认为geom_quasirandom给出的结果更有特色。</p>
</div>
<div id="-a" class="section level3">
<h3><span class="header-section-number">3.5.2</span> 绘制单个基因 (A)的箱线图</h3>
<p>为了更好的展示效果，下面的矩阵增加了样品数量和样品的分组信息。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#profile=&quot;Name;2cell_1;2cell_2;2cell_3;2cell_4;2cell_5;2cell_6;4cell_1;4cell_2;4cell_3;\</span>
<span class="co">#4cell_4;4cell_5;4cell_6;zygote_1;zygote_2;zygote_3;zygote_4;zygote_5;zygote_6</span>
<span class="co">#A;4;6;7;5;8;6;3.2;5.2;5.6;3.6;7.6;4.8;2;4;3;2;4;2.5</span>
<span class="co">#B;6;8;9;7;10;8;5.2;7.2;7.6;5.6;9.6;6.8;4;6;5;4;6;4.5&quot;</span>

profile_text &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/boxplot_singleGene.data&quot;</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>, 
        <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">check.names=</span>F)

data_m =<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">t</span>(profile_text[<span class="st">&#39;A&#39;</span>,]))
data_m$sample =<span class="st"> </span><span class="kw">rownames</span>(data_m)
<span class="co"># 只挑选显示部分</span>
<span class="co"># grepl前面已经讲过用于匹配</span>
data_m[<span class="kw">grepl</span>(<span class="st">&#39;_[123]&#39;</span>, data_m$sample),]</code></pre></div>
<pre><code>##            A   sample
## 2cell_1  4.0  2cell_1
## 2cell_2  6.0  2cell_2
## 2cell_3  7.0  2cell_3
## 4cell_1  3.2  4cell_1
## 4cell_2  5.2  4cell_2
## 4cell_3  5.6  4cell_3
## zygote_1 2.0 zygote_1
## zygote_2 4.0 zygote_2
## zygote_3 3.0 zygote_3</code></pre>
<p>获得样品分组信息 (这个例子比较特殊，样品的分组信息就是样品名字下划线前面的部分)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 可以利用strsplit分割，取出其前面的字符串</span>
<span class="co"># R中复杂的输出结果多数以列表的形式体现，在之前的矩阵操作教程中</span>
<span class="co"># 提到过用str函数来查看复杂结果的结构，并从中获取信息</span>
group =<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(<span class="kw">strsplit</span>(data_m$sample,<span class="st">&quot;_&quot;</span>), function(x) x[<span class="dv">1</span>]))
data_m$group =<span class="st"> </span>group
data_m[<span class="kw">grepl</span>(<span class="st">&#39;_[123]&#39;</span>, data_m$sample),]</code></pre></div>
<pre><code>##            A   sample  group
## 2cell_1  4.0  2cell_1  2cell
## 2cell_2  6.0  2cell_2  2cell
## 2cell_3  7.0  2cell_3  2cell
## 4cell_1  3.2  4cell_1  4cell
## 4cell_2  5.2  4cell_2  4cell
## 4cell_3  5.6  4cell_3  4cell
## zygote_1 2.0 zygote_1 zygote
## zygote_2 4.0 zygote_2 zygote
## zygote_3 3.0 zygote_3 zygote</code></pre>
<p>如果没有这个规律，也可以提到类似于下面的文件，指定样品所属的组的信息。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sampleGroup_text=<span class="st">&quot;Sample;Group</span>
<span class="st">zygote_1;zygote</span>
<span class="st">zygote_2;zygote</span>
<span class="st">zygote_3;zygote</span>
<span class="st">zygote_4;zygote</span>
<span class="st">zygote_5;zygote</span>
<span class="st">zygote_6;zygote</span>
<span class="st">2cell_1;2cell</span>
<span class="st">2cell_2;2cell</span>
<span class="st">2cell_3;2cell</span>
<span class="st">2cell_4;2cell</span>
<span class="st">2cell_5;2cell</span>
<span class="st">2cell_6;2cell</span>
<span class="st">4cell_1;4cell</span>
<span class="st">4cell_2;4cell</span>
<span class="st">4cell_3;4cell</span>
<span class="st">4cell_4;4cell</span>
<span class="st">4cell_5;4cell</span>
<span class="st">4cell_6;4cell&quot;</span>

<span class="co">#sampleGroup = read.table(text=sampleGroup_text,sep=&quot;\t&quot;,header=1,check.names=F,row.names=1)</span>

<span class="co">#data_m &lt;- merge(data_m, sampleGroup, by=&quot;row.names&quot;)</span>

<span class="co"># 会获得相同的结果，脚本注释掉了以免重复执行引起问题。</span></code></pre></div>
<p>矩阵准备好了，开始画图了 (小提琴图做例子，其它类似)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 调整下样品出现的顺序</span>
data_m$group &lt;-<span class="st"> </span><span class="kw">factor</span>(data_m$group, <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&quot;zygote&quot;</span>,<span class="st">&quot;2cell&quot;</span>,<span class="st">&quot;4cell&quot;</span>))
<span class="co"># group和A为矩阵中两列的名字，group代表了值的属性，A代表基因A对应的表达值。</span>
<span class="co"># 注意看修改了的地方</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>group, <span class="dt">y=</span>A)) +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_violin</span>(<span class="kw">aes</span>(<span class="dt">fill=</span><span class="kw">factor</span>(group))) +<span class="st"> </span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">50</span>,<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">vjust=</span><span class="fl">0.5</span>)) +
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>)
p
<span class="co"># 图会存储在当前目录的Rplots.pdf文件中，如果用Rstudio，可以不运行dev.off()</span>
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-149-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="section-3.5.3" class="section level3">
<h3><span class="header-section-number">3.5.3</span> 长矩阵绘制箱线图</h3>
<p>常规矩阵绘制箱线图要求必须是个方正的矩阵输入，而有时想比较的几个组里面检测的值数目不同。比如有三个组，GrpA组检测了6个病人，GrpB组检测了10个病人，GrpC组是12个正常人的检测数据。这时就很难形成一个行为检测值，列为样品的矩阵，就需要长表格模式解决这一问题。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">long_table &lt;-<span class="st"> &quot;Grp;Value</span>
<span class="st">GrpA;10</span>
<span class="st">GrpA;11</span>
<span class="st">GrpA;11</span>
<span class="st">GrpA;11</span>
<span class="st">GrpA;12</span>
<span class="st">GrpA;11</span>
<span class="st">GrpA;15</span>
<span class="st">GrpB;5</span>
<span class="st">GrpB;4</span>
<span class="st">GrpB;3</span>
<span class="st">GrpB;2</span>
<span class="st">GrpB;4</span>
<span class="st">GrpB;3</span>
<span class="st">GrpB;2</span>
<span class="st">GrpB;3</span>
<span class="st">GrpB;3.1</span>
<span class="st">GrpC;2</span>
<span class="st">GrpC;1</span>
<span class="st">GrpC;1</span>
<span class="st">GrpC;1.1</span>
<span class="st">GrpC;1.5</span>
<span class="st">GrpC;1.1</span>
<span class="st">GrpC;1.5</span>
<span class="st">GrpC;1.8</span>
<span class="st">GrpC;2&quot;</span>

long_table &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>long_table,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>,<span class="dt">header=</span>T,<span class="dt">check.names=</span>F)

p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(long_table, <span class="kw">aes</span>(<span class="dt">x=</span>Grp, <span class="dt">y=</span>Value)) +<span class="st"> </span>
<span class="kw">geom_violin</span>(<span class="kw">aes</span>(<span class="dt">fill=</span><span class="kw">factor</span>(Grp))) +<span class="st"> </span>
<span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">50</span>,<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">vjust=</span><span class="fl">0.5</span>)) +
<span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-150-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a =<span class="st"> </span><span class="kw">c</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&#39;F&#39;</span>,<span class="st">&quot;M&quot;</span>),<span class="dv">12</span>), <span class="st">&#39;F&#39;</span>)
long_table$gender &lt;-<span class="st"> </span>a

p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(long_table, <span class="kw">aes</span>(<span class="dt">x=</span>Grp, <span class="dt">y=</span>Value)) +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_boxplot</span>(<span class="kw">aes</span>(<span class="dt">fill=</span><span class="kw">factor</span>(gender))) +
<span class="st">  </span><span class="kw">geom_quasirandom</span>(<span class="dt">groupOnX=</span><span class="ot">FALSE</span>)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-151-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>长表格形式自身就是常规矩阵melt后的格式，这种用来绘制箱线图就很简单了，就不举例子了。</p>
</div>
</div>
<div id="section-3.6" class="section level2">
<h2><span class="header-section-number">3.6</span> 线图</h2>
<p>线图是反映趋势变化的一种方式，其输入数据一般也是一个矩阵。</p>
<div id="section-3.6.1" class="section level3">
<h3><span class="header-section-number">3.6.1</span> 单线图</h3>
<p>假设有这么一个矩阵，第一列为转录起始位点及其上下游<code>5 kb</code>的区域，第二列为<code>H3K27ac</code>修饰在这些区域的丰度，想绘制一张线图展示。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">profile=<span class="st">&quot;Pos;H3K27ac</span>
<span class="st">-5000;8.7</span>
<span class="st">-4000;8.4</span>
<span class="st">-3000;8.3</span>
<span class="st">-2000;7.2</span>
<span class="st">-1000;3.6</span>
<span class="st">0;3.6</span>
<span class="st">1000;7.1</span>
<span class="st">2000;8.2</span>
<span class="st">3000;8.4</span>
<span class="st">4000;8.5</span>
<span class="st">5000;8.5&quot;</span></code></pre></div>
<p>读入数据</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">profile_text &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>profile, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="ot">NULL</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>)
profile_text</code></pre></div>
<pre><code>##      Pos H3K27ac
## 1  -5000     8.7
## 2  -4000     8.4
## 3  -3000     8.3
## 4  -2000     7.2
## 5  -1000     3.6
## 6      0     3.6
## 7   1000     7.1
## 8   2000     8.2
## 9   3000     8.4
## 10  4000     8.5
## 11  5000     8.5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(profile_text$Pos, profile_text$H3K27ac,<span class="dt">type =</span> <span class="st">&quot;l&quot;</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-154-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(profile_text, <span class="kw">aes</span>(<span class="dt">x=</span>Pos, <span class="dt">y=</span>H3K27ac)) +<span class="st"> </span><span class="kw">geom_line</span>()</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-155-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">profile_text &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>profile, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>)
profile_text</code></pre></div>
<pre><code>##       H3K27ac
## -5000     8.7
## -4000     8.4
## -3000     8.3
## -2000     7.2
## -1000     3.6
## 0         3.6
## 1000      7.1
## 2000      8.2
## 3000      8.4
## 4000      8.5
## 5000      8.5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 在melt时保留位置信息</span>
<span class="co"># melt格式是ggplot2画图最喜欢的格式</span>
<span class="co"># 好好体会下这个格式，虽然多占用了不少空间，但是确实很方便</span>

<span class="co"># 这里可以用 `xvariable`，也可以是其它字符串，但需要保证后面与这里的一致</span>
<span class="co"># 因为这一列是要在X轴显示，所以起名为`xvariable`。</span>
profile_text$xvariable =<span class="st"> </span><span class="kw">rownames</span>(profile_text)
<span class="co">#library(ggplot2)</span>
<span class="co">#library(reshape2)</span>
data_m &lt;-<span class="st"> </span><span class="kw">melt</span>(profile_text, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;xvariable&quot;</span>))
data_m</code></pre></div>
<pre><code>##    xvariable variable value
## 1      -5000  H3K27ac   8.7
## 2      -4000  H3K27ac   8.4
## 3      -3000  H3K27ac   8.3
## 4      -2000  H3K27ac   7.2
## 5      -1000  H3K27ac   3.6
## 6          0  H3K27ac   3.6
## 7       1000  H3K27ac   7.1
## 8       2000  H3K27ac   8.2
## 9       3000  H3K27ac   8.4
## 10      4000  H3K27ac   8.5
## 11      5000  H3K27ac   8.5</code></pre>
<p>然后开始画图，与上面画heatmap一样。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># variable和value为矩阵melt后的两列的名字，内部变量, variable代表了点线的属性，value代表对应的值。</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>xvariable, <span class="dt">y=</span>value,<span class="dt">color=</span>variable)) +<span class="st"> </span><span class="kw">geom_line</span>()
p</code></pre></div>
<pre><code>## geom_path: Each group consists of only one observation. Do you need to
## adjust the group aesthetic?</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 图会存储在当前目录的Rplots.pdf文件中，如果用Rstudio，可以不运行dev.off()</span>
<span class="co"># dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-158-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>满心期待一个倒钟形曲线，结果，什么也没有。</p>
<p>仔细看，出来一段提示</p>
<pre><code>geom_path: Each group consists of only one observation. 
Do you need to adjust the group aesthetic?</code></pre>
<p>原来默认ggplot2把每个点都视作了一个分组，什么都没画出来。而<code>data_m</code>中的数据都来源于一个分组<code>H3K27ac</code>，分组的名字为<code>variable</code>，修改下脚本，看看效果。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>xvariable, <span class="dt">y=</span>value,<span class="dt">color=</span>variable,<span class="dt">group=</span>variable)) +<span class="st"> </span>
<span class="st">     </span><span class="kw">geom_line</span>() +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="kw">c</span>(<span class="fl">0.1</span>,<span class="fl">0.9</span>))
p
<span class="co"># dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-159-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>图出来了，一条线，看一眼没问题；再仔细看，不对了，怎么还不是倒钟形，原来横坐标错位了。</p>
<p>检查下数据格式</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(data_m)</code></pre></div>
<pre><code>##   xvariable            variable      value      
##  Length:11          H3K27ac:11   Min.   :3.600  
##  Class :character                1st Qu.:7.150  
##  Mode  :character                Median :8.300  
##                                  Mean   :7.318  
##                                  3rd Qu.:8.450  
##                                  Max.   :8.700</code></pre>
<p>问题来了，<code>xvariable</code>虽然看上去数字，但存储的实际是字符串 (因为是作为行名字读取的)，需要转换为数字。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_m$xvariable &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(data_m$xvariable)

<span class="co">#再检验下</span>
<span class="kw">summary</span>(data_m)</code></pre></div>
<pre><code>##    xvariable        variable      value      
##  Min.   :-5000   H3K27ac:11   Min.   :3.600  
##  1st Qu.:-2500                1st Qu.:7.150  
##  Median :    0                Median :8.300  
##  Mean   :    0                Mean   :7.318  
##  3rd Qu.: 2500                3rd Qu.:8.450  
##  Max.   : 5000                Max.   :8.700</code></pre>
<p>好了，继续画图。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 注意断行时，加号在行尾，不能放在行首</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>xvariable, <span class="dt">y=</span>value,<span class="dt">color=</span>variable,<span class="dt">group=</span>variable)) +
<span class="st">     </span><span class="kw">geom_line</span>() +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="kw">c</span>(<span class="fl">0.07</span>,<span class="fl">0.7</span>))
p
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-162-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>图终于出来了，调了下legend的位置，看上去有点意思了。</p>
<p>有点难看，如果平滑下，会不会好一些，<code>stat_smooth</code>可以对绘制的线进行局部拟合。在不影响变化趋势的情况下，可以使用 (但慎用)。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>xvariable, <span class="dt">y=</span>value,<span class="dt">color=</span>variable,<span class="dt">group=</span>variable)) +<span class="st"> </span>
<span class="st">     </span><span class="kw">geom_line</span>() +<span class="st"> </span><span class="kw">stat_smooth</span>(<span class="dt">method=</span><span class="st">&quot;auto&quot;</span>, <span class="dt">se=</span><span class="ot">FALSE</span>) +<span class="st"> </span>
<span class="st">     </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="kw">c</span>(<span class="fl">0.06</span>,<span class="fl">0.5</span>))
p</code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-163-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>从图中看，趋势还是一致的，线条更优美了。另外一个方式是增加区间的数量，线也会好些，而且更真实。</p>
<p><code>stat_smooth</code>和<code>geom_line</code>各绘制了一条线，只保留一条就好。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>xvariable, <span class="dt">y=</span>value,<span class="dt">color=</span>variable,<span class="dt">group=</span>variable)) +<span class="st"> </span>
<span class="st">     </span><span class="kw">stat_smooth</span>(<span class="dt">method=</span><span class="st">&quot;auto&quot;</span>, <span class="dt">se=</span><span class="ot">FALSE</span>) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="kw">c</span>(<span class="fl">0.06</span>,<span class="fl">0.5</span>))
p</code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-164-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>好了，终于完成了单条线图的绘制。</p>
</div>
<div id="section-3.6.2" class="section level3">
<h3><span class="header-section-number">3.6.2</span> 多线图</h3>
<p>那么再来一个多线图的例子吧，只要给之前的数据矩阵多加几列就好了。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">profile =<span class="st"> &quot;Pos;h3k27ac;ctcf;enhancer;h3k4me3;polII</span>
<span class="st">-5000;8.7;10.7;11.7;10;8.3</span>
<span class="st">-4000;8.4;10.8;11.8;9.8;7.8</span>
<span class="st">-3000;8.3;10.5;12.2;9.4;7</span>
<span class="st">-2000;7.2;10.9;12.7;8.4;4.8</span>
<span class="st">-1000;3.6;8.5;12.8;4.8;1.3</span>
<span class="st">0;3.6;8.5;13.4;5.2;1.5</span>
<span class="st">1000;7.1;10.9;12.4;8.1;4.9</span>
<span class="st">2000;8.2;10.7;12.4;9.5;7.7</span>
<span class="st">3000;8.4;10.4;12;9.8;7.9</span>
<span class="st">4000;8.5;10.6;11.7;9.7;8.2</span>
<span class="st">5000;8.5;10.6;11.7;10;8.2&quot;</span>

profile_text &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>profile, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>)

profile_text$xvariable =<span class="st"> </span><span class="kw">rownames</span>(profile_text)
data_m &lt;-<span class="st"> </span><span class="kw">melt</span>(profile_text, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;xvariable&quot;</span>))
data_m$xvariable &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(data_m$xvariable)

<span class="co"># 这里group=variable，而不是group=1 (如果上面你用的是1的话)</span>
<span class="co"># variable和value为矩阵melt后的两列的名字，内部变量, </span>
<span class="co"># variable代表了点线的属性，value代表对应的值。</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>xvariable, <span class="dt">y=</span>value,<span class="dt">color=</span>variable,<span class="dt">group=</span>variable)) +<span class="st"> </span>
<span class="st">     </span><span class="kw">stat_smooth</span>(<span class="dt">method=</span><span class="st">&quot;auto&quot;</span>, <span class="dt">se=</span><span class="ot">FALSE</span>) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="kw">c</span>(<span class="fl">0.85</span>,<span class="fl">0.2</span>))
p</code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-165-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="section-3.6.3" class="section level3">
<h3><span class="header-section-number">3.6.3</span> 横轴文本线图</h3>
<p>如果横轴是文本，又该怎么调整顺序呢？还记得之前热图旁的行或列的顺序调整吗？重新设置变量的<code>factor</code>水平就可以控制其顺序。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">profile =<span class="st"> &quot;Pos;h3k27ac;ctcf;enhancer;h3k4me3;polII</span>
<span class="st">-5000;8.7;10.7;11.7;10;8.3</span>
<span class="st">-4000;8.4;10.8;11.8;9.8;7.8</span>
<span class="st">-3000;8.3;10.5;12.2;9.4;7</span>
<span class="st">-2000;7.2;10.9;12.7;8.4;4.8</span>
<span class="st">-1000;3.6;8.5;12.8;4.8;1.3</span>
<span class="st">0;3.6;8.5;13.4;5.2;1.5</span>
<span class="st">1000;7.1;10.9;12.4;8.1;4.9</span>
<span class="st">2000;8.2;10.7;12.4;9.5;7.7</span>
<span class="st">3000;8.4;10.4;12;9.8;7.9</span>
<span class="st">4000;8.5;10.6;11.7;9.7;8.2</span>
<span class="st">5000;8.5;10.6;11.7;10;8.2&quot;</span>

profile_text &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>profile, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>)

profile_text_rownames &lt;-<span class="st"> </span><span class="kw">row.names</span>(profile_text)

profile_text$xvariable =<span class="st"> </span><span class="kw">rownames</span>(profile_text)
data_m &lt;-<span class="st"> </span><span class="kw">melt</span>(profile_text, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;xvariable&quot;</span>))
data_m$xvariable &lt;-<span class="st"> </span><span class="kw">factor</span>(data_m$xvariable, <span class="dt">levels=</span>profile_text_rownames, <span class="dt">ordered=</span>T)

<span class="co"># geom_line设置线的粗细和透明度</span>
p1 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>xvariable, <span class="dt">y=</span>value,<span class="dt">color=</span>variable,<span class="dt">group=</span>variable)) +<span class="st"> </span>
<span class="st">     </span><span class="kw">geom_line</span>(<span class="dt">size=</span><span class="dv">1</span>, <span class="dt">alpha=</span><span class="fl">0.9</span>) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="kw">c</span>(<span class="fl">0.8</span>,<span class="fl">0.25</span>)) +
<span class="st">     </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>))

<span class="co"># stat_smooth</span>
p2 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>xvariable, <span class="dt">y=</span>value,<span class="dt">color=</span>variable,<span class="dt">group=</span>variable)) +<span class="st"> </span>
<span class="st">     </span><span class="kw">stat_smooth</span>(<span class="dt">method=</span><span class="st">&quot;auto&quot;</span>, <span class="dt">se=</span><span class="ot">FALSE</span>) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="kw">c</span>(<span class="fl">0.8</span>,<span class="fl">0.25</span>)) +
<span class="st">     </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust=</span><span class="dv">1</span>))
     
<span class="kw">library</span>(<span class="st">&quot;cowplot&quot;</span>)
<span class="kw">plot_grid</span>(p1,p2, <span class="dt">labels=</span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;B&quot;</span>), <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">nrow=</span><span class="dv">1</span>)</code></pre></div>
<pre><code>## `geom_smooth()` using method = &#39;loess&#39;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/fig-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>比较下位置信息做为数字(前面的线图)和位置信息横轴的差别。当为数值时，ggplot2会选择合适的几个刻度做标记，当为文本时，会全部标记。另外文本横轴，<code>smooth</code>效果不明显。</p>
</div>
</div>
<div id="section-3.7" class="section level2">
<h2><span class="header-section-number">3.7</span> 散点图</h2>
<p>散点图在生物信息分析中是应用比较广的一个图，常见的差异基因火山图、功能富集分析泡泡图、相关性分析散点图、抖动图、PCA样品分类图等。凡是想展示分布状态的都可以用散点图。</p>
<div id="section-3.7.1" class="section level3">
<h3><span class="header-section-number">3.7.1</span> 横纵轴都为数字的散点图解析</h3>
<p>绘制散点图的输入一般都是规规矩矩的矩阵，可以让不同的列分别代表X轴、Y轴、点的大小、颜色、形状、名称等。</p>
<div id="-" class="section level4">
<h4><span class="header-section-number">3.7.1.1</span> 输入数据格式 (使用火山图的输入数据为例)</h4>
<p>火山图用于展示基因表达差异的分布，横轴为<code>Log2 Fold Change</code>，越偏离中心差异倍数越大；纵轴为<code>(-1)*Log10 P_adjust</code>，值越大差异越显著。一般横轴越偏离中心的点其纵轴值也会比较大，因此呈现火山喷发的形状。</p>
<p>火山图需要的数据格式如下</p>
<ul>
<li>id: 不是必须的，但一般的软件输出结果中都会包含，表示基因名字。</li>
<li>log2FoldChange: 差异倍数的对数，一般的差异分析输出结果中也会给出对数处理的值, 因此程序没有提供这一步的计算操作。</li>
<li>padj: 多重假设检验矫正过的差异显著性P值；一般的差异分析输出结果为原始值，程序提供一个参数对其求取负对数。</li>
<li>significant: 可选列，标记哪些基因是上调、下调、无差异；若无此列或未在参数中指定此列，默认程序会根据<code>padj</code>列和<code>log2FoldChange</code>列根据给定的阈值自动计算差异基因，并作出不同颜色的标记。</li>
<li>label: 可选列，一般用于在图中标记出感兴趣的基因的名字。非<code>-</code>行的字符串都会标记在图上。</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">volcano =<span class="st"> &quot;id;log2FoldChange;padj;significant;label</span>
<span class="st">E00007;4.28238;0;EHBIO_UP;A</span>
<span class="st">E00008;-1.1036;0.476466843393901;Unchanged;-</span>
<span class="st">E00009;-0.274368;1;Unchanged;-</span>
<span class="st">E00010;4.62347;7.37606076333335e-103;EHBIO_UP;-</span>
<span class="st">E00012;0.973987;0.482982440163204;Unchanged;-</span>
<span class="st">E00017;-1.30205;0.000555693857439792;Baodian_UP;B</span>
<span class="st">E00024;0.617636;2.78047837287061e-13;Unchanged;-</span>
<span class="st">E00033;1.48669;2.56000581595275e-60;EHBIO_UP;-</span>
<span class="st">E00034;-0.783716;0.00341521725291801;Unchanged;-</span>
<span class="st">E00036;2.01592;6.03136656016401e-06;EHBIO_UP;C</span>
<span class="st">E00040;-1.89657;4.73663890849056e-21;Baodian_UP;-</span>
<span class="st">E00041;-0.268168;0.563429434558031;Unchanged;-</span>
<span class="st">E00042;0.0861048;0.367700939634328;Unchanged;-</span>
<span class="st">E00043;-1.19328;1.42673872027352e-153;Baodian_UP;-</span>
<span class="st">E00044;-0.887981;2.43067804654905e-26;Unchanged;-</span>
<span class="st">E00047;-0.610941;5.51696648645932e-57;Unchanged;-&quot;</span>

<span class="co"># 数据的读取之前的R语言统计和绘图系列都已解释过，不再赘述</span>
<span class="co"># 文末也有链接可直达之前的文章，新学者建议从头开始</span>
<span class="co">#volcanoData &lt;- read.table(text=volcano, sep=&quot;;&quot;, header=T, quote=&quot;&quot;, check.names=F)</span>
volcanoData &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/volcano.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header=</span>T, 
                          <span class="dt">row.names=</span><span class="ot">NULL</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>, <span class="dt">check.names=</span>F)
<span class="kw">head</span>(volcanoData)</code></pre></div>
<pre><code>##       id log2FoldChange          padj significant label
## 1 E00007       4.282380  0.000000e+00    EHBIO_UP     A
## 2 E00008      -1.103600  4.764668e-01   Unchanged     -
## 3 E00009      -0.274368  1.000000e+00   Unchanged     -
## 4 E00010       4.623470 7.376061e-103    EHBIO_UP     -
## 5 E00012       0.973987  4.829824e-01   Unchanged     -
## 6 E00017      -1.302050  5.556939e-04  Baodian_UP     B</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(volcanoData)</code></pre></div>
<pre><code>## [1] 1985    5</code></pre>
<p>绘制散点图，只需要指定X轴和Y轴，再加上<code>geom_point</code>即可。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#library(ggplot2)</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(volcanoData, <span class="kw">aes</span>(<span class="dt">x=</span>log2FoldChange, <span class="dt">y=</span>padj))
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">geom_point</span>()
<span class="co"># 前面是给p不断添加图层的过程</span>
<span class="co"># 单输入一个p是真正作图</span>
<span class="co"># 前面有人说，上面都输完了，怎么没出图</span>
<span class="co"># 就因为差了一个p</span>
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-168-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>说好的火山图的例子，但怎么也看不出喷发的态势。</p>
<p>对数据坐下预处理，差异大的基因<code>padj</code>小，先对其求取负对数，所谓负负得正，差异大的基因就会处于图的上方了。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 从示例数据中看到，最小的padj值为0，求取负对数为正无穷。</span>
<span class="co"># 实际上padj值小到一个点对我们来讲就是个数</span>
<span class="co"># 所以可以给所有小于1e-6的padj都让其等于1e-6，再小也没意义</span>
<span class="co"># </span>
volcanoData[volcanoData$padj&lt;<span class="fl">1e-6</span>, <span class="st">&quot;padj&quot;</span>] &lt;-<span class="st"> </span><span class="fl">1e-6</span>
volcanoData$padj &lt;-<span class="st"> </span>(-<span class="dv">1</span>)*<span class="st"> </span><span class="kw">log10</span>(volcanoData$padj)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(volcanoData)</code></pre></div>
<pre><code>##        id       log2FoldChange          padj            significant  
##  E00007 :   1   Min.   :-5.07612   Min.   :0.0000   Baodian_UP: 201  
##  E00008 :   1   1st Qu.:-0.52292   1st Qu.:0.9619   EHBIO_UP  : 270  
##  E00009 :   1   Median : 0.03515   Median :5.4514   Unchanged :1514  
##  E00010 :   1   Mean   : 0.10928   Mean   :3.7426                    
##  E00012 :   1   3rd Qu.: 0.58246   3rd Qu.:6.0000                    
##  E00017 :   1   Max.   : 8.33085   Max.   :6.0000                    
##  (Other):1979                                                        
##  label   
##  -:1981  
##  A:   1  
##  B:   1  
##  C:   1  
##  D:   1  
##          
## </code></pre>
<p>数据中基因的上调倍数远高于下调倍数，使得出来的图是偏的，这次画图时调整下X轴的区间使图对称。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">log2fc_max_abs =<span class="st"> </span><span class="kw">max</span>(<span class="kw">abs</span>(volcanoData$log2FoldChange)) +<span class="st"> </span><span class="fl">0.1</span>
padj_max =<span class="st"> </span><span class="kw">max</span>(volcanoData$padj) +<span class="st"> </span><span class="fl">0.1</span> 
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(volcanoData,  <span class="kw">aes</span>(<span class="dt">x=</span>log2FoldChange,  <span class="dt">y=</span>padj)) +
<span class="st">     </span><span class="kw">geom_point</span>() +<span class="st"> </span><span class="kw">xlim</span>(-<span class="dv">1</span>*log2fc_max_abs, log2fc_max_abs) +<span class="st"> </span><span class="kw">ylim</span>(<span class="dv">0</span>,padj_max)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-171-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>有点意思了，数据太少不明显，下一步加上颜色看看。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(volcanoData,  <span class="kw">aes</span>(<span class="dt">x=</span>log2FoldChange,  <span class="dt">y=</span>padj)) +
<span class="st">     </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color=</span>significant)) +<span class="st">  </span>
<span class="st">    </span><span class="kw">xlim</span>(-<span class="dv">1</span>*log2fc_max_abs, log2fc_max_abs) +<span class="st"> </span><span class="kw">ylim</span>(<span class="dv">0</span>,padj_max)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-172-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">volcanoData[volcanoData$label==<span class="st">&quot;-&quot;</span>, <span class="st">&quot;label&quot;</span>] =<span class="st">  </span><span class="ot">NA</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(volcanoData,  <span class="kw">aes</span>(<span class="dt">x=</span>log2FoldChange,  <span class="dt">y=</span>padj)) +
<span class="st">     </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color=</span>significant)) +<span class="st">  </span><span class="kw">xlim</span>(-<span class="dv">1</span>*log2fc_max_abs, log2fc_max_abs) +<span class="st"> </span>
<span class="st">     </span><span class="kw">ylim</span>(<span class="dv">0</span>,padj_max) +<span class="st"> </span>
<span class="st">     </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>label))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-173-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>利用现有的数据，基本上就是这个样子了。虽然还不太像，原理都已经都点到了。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">volcanoData$sig &lt;-<span class="st"> </span><span class="kw">ifelse</span>(
    volcanoData$padj&gt;<span class="fl">1.30103</span>, 
    <span class="kw">ifelse</span>(
      volcanoData$log2FoldChange&gt;=<span class="dv">1</span>, 
      <span class="st">&quot;UP&quot;</span>, 
      <span class="kw">ifelse</span>(
        volcanoData$log2FoldChange&lt;=-<span class="dv">1</span>, 
        <span class="st">&quot;DW&quot;</span>, 
        <span class="st">&quot;NoDiff&quot;</span>)
    ), 
    <span class="st">&quot;NoDiff&quot;</span>)
volcanoData$sig &lt;-<span class="st"> </span><span class="kw">factor</span>(volcanoData$sig, <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&quot;UP&quot;</span>, <span class="st">&quot;DW&quot;</span>, <span class="st">&quot;NoDiff&quot;</span>))
<span class="kw">summary</span>(volcanoData)</code></pre></div>
<pre><code>##        id       log2FoldChange          padj            significant  
##  E00007 :   1   Min.   :-5.07612   Min.   :0.0000   Baodian_UP: 201  
##  E00008 :   1   1st Qu.:-0.52292   1st Qu.:0.9619   EHBIO_UP  : 270  
##  E00009 :   1   Median : 0.03515   Median :5.4514   Unchanged :1514  
##  E00010 :   1   Mean   : 0.10928   Mean   :3.7426                    
##  E00012 :   1   3rd Qu.: 0.58246   3rd Qu.:6.0000                    
##  E00017 :   1   Max.   : 8.33085   Max.   :6.0000                    
##  (Other):1979                                                        
##   label          sig      
##  -   :   0   UP    : 271  
##  A   :   1   DW    : 201  
##  B   :   1   NoDiff:1513  
##  C   :   1                
##  D   :   1                
##  NA&#39;s:1981                
## </code></pre>
</div>
</div>
<div id="section-3.7.2" class="section level3">
<h3><span class="header-section-number">3.7.2</span> 横纵轴都为字符串的散点图展示</h3>
<div id="section-3.7.2.1" class="section level4">
<h4><span class="header-section-number">3.7.2.1</span> 输入数据格式如下</h4>
<p>这个数据是FASTQC结果总结中的直观的查看所有样品测序碱基质量和GC含量的散点图的示例数据。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fastqc&lt;-<span class="st">&quot;ID;GC_quality;Base_quality</span>
<span class="st">ehbio_1_1;PASS;PASS</span>
<span class="st">ehbio_1_2;PASS;PASS</span>
<span class="st">ehbio_2_1;WARN;PASS</span>
<span class="st">ehbio_2_2;WARN;PASS</span>
<span class="st">Other_1_1;FAIL;FAIL</span>
<span class="st">Other_1_2;FAIL;FAIL&quot;</span>

fastqc_data &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>fastqc, <span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">header=</span>T)
<span class="co"># 就不查看了</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(fastqc_data, <span class="kw">aes</span>(<span class="dt">x=</span>GC_quality, <span class="dt">y=</span>Base_quality)) +<span class="st"> </span><span class="kw">geom_point</span>()
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-176-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(fastqc_data, <span class="kw">aes</span>(<span class="dt">x=</span>GC_quality, <span class="dt">y=</span>Base_quality)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_jitter</span>() +<span class="st"> </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>ID), <span class="dt">position=</span><span class="st">&quot;jitter&quot;</span>)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-177-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>六个点少了只剩下了3个，重叠在一起了，而且也不知道哪个点代表什么样品。这时需要把点抖动下，用到一个包<code>ggbeeswarm</code>，抖动图的神器。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggbeeswarm)
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(fastqc_data, <span class="kw">aes</span>(<span class="dt">x=</span>GC_quality, <span class="dt">y=</span>Base_quality)) +<span class="st"> </span>
<span class="st">            </span><span class="kw">geom_quasirandom</span>(<span class="dt">groupOnX=</span><span class="ot">FALSE</span>)
<span class="co"># 使用geom_text增加点的标记</span>
<span class="co"># label表示标记哪一列的数值</span>
<span class="co"># position_quasirandom获取点偏移后的位置</span>
<span class="co"># xjust调整对齐方式; hjust是水平的对齐方式，0为左，1为右，0.5居中，0-1之间可以取任意值。</span>
<span class="co"># vjust是垂直对齐方式，0底对齐，1为顶对齐，0.5居中，0-1之间可以取任意值。</span>
<span class="co"># check_overlap检查名字在图上是否重叠，若有重叠，只显示一个</span>
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>ID), <span class="dt">position=</span><span class="kw">position_quasirandom</span>(<span class="dt">groupOnX=</span><span class="ot">FALSE</span>), 
        <span class="dt">hjust=</span><span class="fl">1.1</span>,<span class="dt">check_overlap=</span>T)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-178-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="section-3.8" class="section level2">
<h2><span class="header-section-number">3.8</span> 功能富集泡泡图</h2>
<p>功能富集分析用来展示某一组基因(一般是单个样品上调或下调的基因)倾向参与哪些功能调控通路，对从整体理解变化了的基因的功能和潜在的调控意义具有指导作用，也是文章发表中一个有意义的美图。通常会用柱状图、泡泡图和热图进行展示。热图的画法之前已经介绍过，这次介绍下富集分析泡泡图, 其展示的信息是最为全面的，也是比较抓人眼球的。</p>
<p>假设有这么一个富集分析结果矩阵 (文件名为GOenrichement.xls) 存储了EHBIO样品和Baodian样品中各自上调的基因富集的通路。</p>
<p><a href="http://omicslab.genetics.ac.cn/GOEAST" class="uri">http://omicslab.genetics.ac.cn/GOEAST</a></p>
<ul>
<li>Description 为GO通路的描述，也可以是KEGG通路。</li>
<li>GeneRatio 为对应通路差异基因占总差异基因的比例，本列可以用分数或小数表示，都可以处理。</li>
<li>qvalue 表示对应通路富集的显著性程度，可以是log处理过的，也可以是原始的。</li>
<li>Count 为对应通路差异基因数目。</li>
<li>Type 这个矩阵合并了EHBIO样品和Baodian样品中各自上调的基因富集的通路，用Type列做区分。如果只有一个样品可不要。</li>
</ul>
<pre><code>Description GeneRatio   qvalue  Count   Type
ERBB signaling pathway  7/320   0.001836081 7   EHBIO_up
regulation of ERBB signaling pathway    5/320   0.003886659 5   EHBIO_up
negative regulation of cell cycle G1/S phase transition 4/320   0.016153254 4   EHBIO_up
Wnt signaling pathway   13/320  0.01680096  13  EHBIO_up
cell-cell signaling by wnt  13/320  0.0171473   13  EHBIO_up
negative regulation of cell cycle process   8/320   0.019453085 8   EHBIO_up
extrinsic apoptotic signaling pathway   9/320   0.024164034 9   EHBIO_up
positive regulation of extrinsic apoptotic signaling pathway    4/320   0.025708228 4   EHBIO_up
cell cycle G1/S phase transition    7/320   0.035797856 7   EHBIO_up
negative regulation of apoptotic signaling pathway  8/320   0.038684745 8   EHBIO_up
regulation of Notch signaling pathway   4/320   0.041592045 4   EHBIO_up
regulation of cell cycle G1/S phase transition  5/320   0.047407619 5   EHBIO_up
negative regulation of BMP signaling pathway    3/320   0.049460847 3   EHBIO_up
regulation of ERK1 and ERK2 cascade 14/342  0.000629602 14  Baodian_up
positive regulation of cell adhesion    17/342  0.000827275 17  Baodian_up
ERK1 and ERK2 cascade   14/342  0.001086508 14  Baodian_up
regulation of cell growth   17/342  0.002228511 17  Baodian_up
positive regulation of cytoskeleton organization    10/342  0.004406867 10  Baodian_up
regulation of cell-cell adhesion    15/342  0.005075219 15  Baodian_up
regulation of cytoskeleton organization 15/342  0.019685646 15  Baodian_up
negative regulation of Notch signaling pathway  3/342   0.020578211 3   Baodian_up
neuron apoptotic process    10/342  0.040284925 10  Baodian_up</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enrichment =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/GOenrichement.xls&quot;</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="ot">NULL</span>, 
    <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>)
<span class="kw">head</span>(enrichment)</code></pre></div>
<pre><code>##                                               Description GeneRatio
## 1                                  ERBB signaling pathway     7/320
## 2                    regulation of ERBB signaling pathway     5/320
## 3 negative regulation of cell cycle G1/S phase transition     4/320
## 4                                   Wnt signaling pathway    13/320
## 5                              cell-cell signaling by wnt    13/320
## 6               negative regulation of cell cycle process     8/320
##        qvalue Count     Type
## 1 0.001836081     7 EHBIO_up
## 2 0.003886659     5 EHBIO_up
## 3 0.016153254     4 EHBIO_up
## 4 0.016800960    13 EHBIO_up
## 5 0.017147300    13 EHBIO_up
## 6 0.019453085     8 EHBIO_up</code></pre>
<div id="section-3.8.1" class="section level3">
<h3><span class="header-section-number">3.8.1</span> 单样品分开绘制</h3>
<p>示例矩阵中包含两个样品上调基因的富集通路，现在先取出一个样品绘制。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enrichment_sxbd =<span class="st"> </span><span class="kw">droplevels</span>(enrichment[enrichment$Type==<span class="st">&quot;Baodian_up&quot;</span>, ])</code></pre></div>
<p>构造一个函数，转换分数为小数。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(plyr)
<span class="kw">library</span>(stringr)
<span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(grid)

mixedToFloat &lt;-<span class="st"> </span>function(x){
  x &lt;-<span class="st"> </span><span class="kw">sapply</span>(x, as.character)
  is.integer  &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^-?</span><span class="ch">\\</span><span class="st">d+$&quot;</span>, x)
  is.fraction &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^-?</span><span class="ch">\\</span><span class="st">d+</span><span class="ch">\\</span><span class="st">/</span><span class="ch">\\</span><span class="st">d+$&quot;</span>, x)
  is.float &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^-?</span><span class="ch">\\</span><span class="st">d+</span><span class="ch">\\</span><span class="st">.</span><span class="ch">\\</span><span class="st">d+$&quot;</span>, x)
  is.mixed    &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^-?</span><span class="ch">\\</span><span class="st">d+ </span><span class="ch">\\</span><span class="st">d+</span><span class="ch">\\</span><span class="st">/</span><span class="ch">\\</span><span class="st">d+$&quot;</span>, x)
  <span class="kw">stopifnot</span>(<span class="kw">all</span>(is.integer |<span class="st"> </span>is.fraction |<span class="st"> </span>is.float |<span class="st"> </span>is.mixed))

  numbers &lt;-<span class="st"> </span><span class="kw">strsplit</span>(x, <span class="st">&quot;[ /]&quot;</span>)

  <span class="kw">ifelse</span>(is.integer,  <span class="kw">as.numeric</span>(<span class="kw">sapply</span>(numbers, <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">1</span>)),
  <span class="kw">ifelse</span>(is.float,    <span class="kw">as.numeric</span>(<span class="kw">sapply</span>(numbers, <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">1</span>)),
  <span class="kw">ifelse</span>(is.fraction, <span class="kw">as.numeric</span>(<span class="kw">sapply</span>(numbers, <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">1</span>)) /
<span class="st">                      </span><span class="kw">as.numeric</span>(<span class="kw">sapply</span>(numbers, <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">2</span>)),
                      <span class="kw">as.numeric</span>(<span class="kw">sapply</span>(numbers, <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">1</span>)) +
<span class="st">                      </span><span class="kw">as.numeric</span>(<span class="kw">sapply</span>(numbers, <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">2</span>)) /
<span class="st">                      </span><span class="kw">as.numeric</span>(<span class="kw">sapply</span>(numbers, <span class="st">`</span><span class="dt">[</span><span class="st">`</span>, <span class="dv">3</span>)))))
  
}

<span class="kw">mixedToFloat</span>(<span class="kw">c</span>(<span class="st">&#39;1 1/2&#39;</span>, <span class="st">&#39;2 3/4&#39;</span>, <span class="st">&#39;2/3&#39;</span>, <span class="st">&#39;11 1/4&#39;</span>, <span class="st">&#39;1&#39;</span>))</code></pre></div>
<pre><code>## [1]  1.5000000  2.7500000  0.6666667 11.2500000  1.0000000</code></pre>
<p>转换数据列为小数或整数</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enrichment_sxbd$GeneRatio =<span class="st"> </span><span class="kw">mixedToFloat</span>(enrichment_sxbd$GeneRatio)
enrichment_sxbd$Count =<span class="st"> </span><span class="kw">mixedToFloat</span>(enrichment_sxbd$Count)</code></pre></div>
<p>qvalue转换负对数，并作为新的一列</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># qvalue转换</span>
log_name =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;negLog10_&quot;</span>, <span class="st">&quot;qvalue&quot;</span>)
col_name_enrichment_sxbd &lt;-<span class="st"> </span><span class="kw">colnames</span>(enrichment_sxbd)
col_name_enrichment_sxbd &lt;-<span class="st"> </span><span class="kw">c</span>(col_name_enrichment_sxbd, log_name)
enrichment_sxbd$log_name &lt;-<span class="st"> </span><span class="kw">log10</span>(enrichment_sxbd$qvalue) *<span class="st"> </span>(-<span class="dv">1</span>)
<span class="kw">colnames</span>(enrichment_sxbd) &lt;-<span class="st"> </span>col_name_enrichment_sxbd</code></pre></div>
<p>Term排序</p>
<ol style="list-style-type: decimal">
<li>获取Term出现的次数</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 计算每个Term出现的顺序，用于排序，出现次数最多的排在前面</span>
enrichment_sxbd_freq &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">table</span>(enrichment_sxbd$Description))
<span class="kw">colnames</span>(enrichment_sxbd_freq) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Description&quot;</span>, <span class="st">&quot;IDctct&quot;</span>)
<span class="kw">head</span>(enrichment_sxbd_freq)</code></pre></div>
<pre><code>##                                        Description IDctct
## 1                            ERK1 and ERK2 cascade      1
## 2   negative regulation of Notch signaling pathway      1
## 3                         neuron apoptotic process      1
## 4             positive regulation of cell adhesion      1
## 5 positive regulation of cytoskeleton organization      1
## 6                        regulation of cell growth      1</code></pre>
<p>根据出现次数、GeneRatio、-log10(qvalue)排序</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enrichment_sxbd2 &lt;-<span class="st"> </span><span class="kw">merge</span>(enrichment_sxbd, enrichment_sxbd_freq, <span class="dt">by=</span><span class="st">&quot;Description&quot;</span>)

<span class="co"># 首先根据出现次数排序、然后根据GeneRatio、然后根据-log10(qvalue)</span>
enrichment_sxbd3 &lt;-<span class="st"> </span>enrichment_sxbd2[<span class="kw">order</span>(enrichment_sxbd2$IDctct, 
    enrichment_sxbd2$GeneRatio, enrichment_sxbd2$negLog10_qvalue), ]

term_order &lt;-<span class="st"> </span><span class="kw">unique</span>(enrichment_sxbd3$Description)

<span class="co"># 设置排序顺序</span>
enrichment_sxbd$Description &lt;-<span class="st"> </span><span class="kw">factor</span>(enrichment_sxbd$Description, 
    <span class="dt">levels=</span>term_order, <span class="dt">ordered=</span>T)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">color_v &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>)

<span class="co"># 指定x,y</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(enrichment_sxbd, <span class="kw">aes</span>(<span class="dt">x=</span>GeneRatio,<span class="dt">y=</span>Description)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x=</span><span class="st">&quot;GeneRatio&quot;</span>, <span class="dt">y=</span><span class="st">&quot;GO description&quot;</span>) +<span class="st"> </span><span class="kw">labs</span>(<span class="dt">title=</span><span class="st">&quot;&quot;</span>)

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">size=</span>Count, <span class="dt">color=</span>negLog10_qvalue )) +<span class="st"> </span>
<span class="st">    </span><span class="kw">scale_colour_gradient</span>(<span class="dt">low=</span>color_v[<span class="dv">1</span>], <span class="dt">high=</span>color_v[<span class="dv">2</span>], <span class="dt">name=</span><span class="st">&quot;negLog10_qvalue&quot;</span>)

<span class="co"># Term单行长度不超过60字符</span>
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">scale_y_discrete</span>(<span class="dt">labels=</span>function(x) <span class="kw">str_wrap</span>(x, <span class="dt">width=</span><span class="dv">60</span>))

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(), <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>())


top=<span class="st">&#39;top&#39;</span>
bottom=<span class="st">&#39;bottom&#39;</span>
left=<span class="st">&#39;left&#39;</span>
right=<span class="st">&#39;right&#39;</span>
none=<span class="st">&#39;none&#39;</span>
legend_pos_par &lt;-<span class="st"> </span>right


uwid =<span class="st"> </span><span class="dv">0</span>
vhig =<span class="st"> </span><span class="dv">12</span>

<span class="co"># 自动估算图形长宽</span>
if (uwid ==<span class="st"> </span><span class="dv">0</span> ||<span class="st"> </span>vhig ==<span class="st"> </span><span class="dv">0</span>) {
    x_len =<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(enrichment_sxbd$Description))
    if(x_len&lt;<span class="dv">10</span>){
        vhig =<span class="st"> </span><span class="dv">10</span>
    } else if(x_len&lt;<span class="dv">20</span>) {
        vhig =<span class="st"> </span><span class="dv">10</span> +<span class="st"> </span>(x_len<span class="dv">-10</span>)/<span class="dv">3</span>
    } else if(x_len&lt;<span class="dv">100</span>) {
        vhig =<span class="st"> </span><span class="dv">13</span> +<span class="st"> </span>(x_len<span class="dv">-20</span>)/<span class="dv">5</span>
    } else {
        vhig =<span class="st"> </span><span class="dv">40</span>
    }
    uwid =<span class="st"> </span>vhig 
    if(legend_pos_par %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;left&quot;</span>, <span class="st">&quot;right&quot;</span>)){
        uwid =<span class="st"> </span><span class="fl">1.5</span> *<span class="st"> </span>uwid
    }
}

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span>legend_pos_par)

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme</span>( <span class="dt">panel.grid =</span> <span class="kw">element_blank</span>(), <span class="dt">panel.border=</span><span class="kw">element_blank</span>(), 
    <span class="dt">legend.background =</span> <span class="kw">element_blank</span>(),
    <span class="dt">axis.line.x=</span><span class="kw">element_line</span>(<span class="dt">size=</span><span class="fl">0.4</span>, <span class="dt">colour=</span><span class="st">&quot;black&quot;</span>, <span class="dt">linetype=</span><span class="st">&#39;solid&#39;</span>),
    <span class="dt">axis.line.y=</span><span class="kw">element_line</span>(<span class="dt">size=</span><span class="fl">0.4</span>, <span class="dt">colour=</span><span class="st">&quot;black&quot;</span>, <span class="dt">linetype=</span><span class="st">&#39;solid&#39;</span>),
    <span class="dt">axis.ticks =</span> <span class="kw">element_line</span>(<span class="dt">size=</span><span class="fl">0.4</span>)
    )


<span class="co">#ggsave(p, filename=&quot;GOenrichement.ehbio.xls.scatterplot.dv.pdf&quot;, dpi=300, width=uwid,</span>
<span class="co">#height=vhig, units=c(&quot;cm&quot;))</span>
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-186-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="section-3.8.2" class="section level3">
<h3><span class="header-section-number">3.8.2</span> 多样品合并绘制</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#enrichment$Type &lt;- factor(enrichment$Type, levels=sample_ho, ordered=T)</span>

<span class="co"># First order by Term, then order by Sample</span>
enrichment &lt;-<span class="st"> </span>enrichment[<span class="kw">order</span>(enrichment$Description, enrichment$Type), ]

enrichment$GeneRatio =<span class="st"> </span><span class="kw">mixedToFloat</span>(enrichment$GeneRatio)
enrichment$Count =<span class="st"> </span><span class="kw">mixedToFloat</span>(enrichment$Count)


log_name =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;negLog10_&quot;</span>, <span class="st">&quot;qvalue&quot;</span>)
col_name_enrichment &lt;-<span class="st"> </span><span class="kw">colnames</span>(enrichment)
col_name_enrichment &lt;-<span class="st"> </span><span class="kw">c</span>(col_name_enrichment, log_name)
enrichment$log_name &lt;-<span class="st"> </span><span class="kw">log10</span>(enrichment$qvalue) *<span class="st"> </span>(-<span class="dv">1</span>)
<span class="kw">colnames</span>(enrichment) &lt;-<span class="st"> </span>col_name_enrichment

<span class="co"># Get the count of each unique Term</span>
enrichment_freq &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">table</span>(enrichment$Description))
<span class="kw">colnames</span>(enrichment_freq) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Description&quot;</span>, <span class="st">&quot;IDctct&quot;</span>)
enrichment2 &lt;-<span class="st"> </span><span class="kw">merge</span>(enrichment, enrichment_freq, <span class="dt">by=</span><span class="st">&quot;Description&quot;</span>)

<span class="co"># 增加一列，样品信息用于排序</span>
enrichment_samp &lt;-<span class="st"> </span><span class="kw">ddply</span>(enrichment2, <span class="st">&quot;Description&quot;</span>, summarize,
    <span class="dt">sam_ct_ct_ct=</span><span class="kw">paste</span>(Type, <span class="dt">collapse=</span><span class="st">&quot;_&quot;</span>))

enrichment2 &lt;-<span class="st"> </span><span class="kw">merge</span>(enrichment2, enrichment_samp, <span class="dt">by=</span><span class="st">&quot;Description&quot;</span>)

<span class="co"># 排序与上面相同，但增加了按样品组合排序</span>
enrichment3 &lt;-<span class="st"> </span>enrichment2[<span class="kw">order</span>(enrichment2$IDctct, enrichment2$sam_ct_ct_ct, 
    enrichment2$Type, enrichment2$GeneRatio, enrichment2$negLog10_qvalue), ]
<span class="co">#print(enrichment3)</span>

term_order &lt;-<span class="st"> </span><span class="kw">unique</span>(enrichment3$Description)

enrichment$Description &lt;-<span class="st"> </span><span class="kw">factor</span>(enrichment$Description, <span class="dt">levels=</span>term_order, <span class="dt">ordered=</span>T)

<span class="co">#print(enrichment)</span>
<span class="kw">rm</span>(enrichment_freq, enrichment2, enrichment3)

color_v &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>)

p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(enrichment, <span class="kw">aes</span>(<span class="dt">x=</span>GeneRatio,<span class="dt">y=</span>Description)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x=</span><span class="st">&quot;GeneRatio&quot;</span>, <span class="dt">y=</span><span class="st">&quot;GO description&quot;</span>) +<span class="st"> </span><span class="kw">labs</span>(<span class="dt">title=</span><span class="st">&quot;&quot;</span>)

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">size=</span>Count, <span class="dt">color=</span>negLog10_qvalue, <span class="dt">shape=</span>Type)) +
<span class="st">    </span><span class="kw">scale_colour_gradient</span>(<span class="dt">low=</span>color_v[<span class="dv">1</span>], <span class="dt">high=</span>color_v[<span class="dv">2</span>], <span class="dt">name=</span><span class="st">&quot;negLog10_qvalue&quot;</span>)

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">scale_y_discrete</span>(<span class="dt">labels=</span>function(x) <span class="kw">str_wrap</span>(x, <span class="dt">width=</span><span class="dv">60</span>))
 
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(), 
    <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>())

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">45</span>,<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">vjust=</span><span class="dv">1</span>))

top=<span class="st">&#39;top&#39;</span>
bottom=<span class="st">&#39;bottom&#39;</span>
left=<span class="st">&#39;left&#39;</span>
right=<span class="st">&#39;right&#39;</span>
none=<span class="st">&#39;none&#39;</span>
legend_pos_par &lt;-<span class="st"> </span>right


uwid =<span class="st"> </span><span class="dv">0</span>
vhig =<span class="st"> </span><span class="dv">12</span>

if (uwid ==<span class="st"> </span><span class="dv">0</span> ||<span class="st"> </span>vhig ==<span class="st"> </span><span class="dv">0</span>) {
    x_len =<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(enrichment$Description))
    if(x_len&lt;<span class="dv">10</span>){
        vhig =<span class="st"> </span><span class="dv">10</span>
    } else if(x_len&lt;<span class="dv">20</span>) {
        vhig =<span class="st"> </span><span class="dv">10</span> +<span class="st"> </span>(x_len<span class="dv">-10</span>)/<span class="dv">3</span>
    } else if(x_len&lt;<span class="dv">100</span>) {
        vhig =<span class="st"> </span><span class="dv">13</span> +<span class="st"> </span>(x_len<span class="dv">-20</span>)/<span class="dv">5</span>
    } else {
        vhig =<span class="st"> </span><span class="dv">40</span>
    }
    uwid =<span class="st"> </span>vhig 
    if(legend_pos_par %in%<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;left&quot;</span>, <span class="st">&quot;right&quot;</span>)){
        uwid =<span class="st"> </span><span class="fl">1.5</span> *<span class="st"> </span>uwid
    }
}

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span>legend_pos_par)

p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">theme</span>( <span class="dt">panel.grid =</span> <span class="kw">element_blank</span>(), <span class="dt">panel.border=</span><span class="kw">element_blank</span>(), 
    <span class="dt">legend.background =</span> <span class="kw">element_blank</span>(),
    <span class="dt">axis.line.x=</span><span class="kw">element_line</span>(<span class="dt">size=</span><span class="fl">0.4</span>, <span class="dt">colour=</span><span class="st">&quot;black&quot;</span>, <span class="dt">linetype=</span><span class="st">&#39;solid&#39;</span>),
    <span class="dt">axis.line.y=</span><span class="kw">element_line</span>(<span class="dt">size=</span><span class="fl">0.4</span>, <span class="dt">colour=</span><span class="st">&quot;black&quot;</span>, <span class="dt">linetype=</span><span class="st">&#39;solid&#39;</span>),
    <span class="dt">axis.ticks =</span> <span class="kw">element_line</span>(<span class="dt">size=</span><span class="fl">0.4</span>)
    )


<span class="co">#ggsave(p, filename=&quot;GOenrichement.xls.scatterplot.dv.pdf&quot;, dpi=300, width=uwid,</span>
<span class="co">#height=vhig, units=c(&quot;cm&quot;))</span>
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-187-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>通过这张图解释下，富集分析的结果怎么解读。富集分析实际是查找哪些通路里面包含的差异基因占总差异基因的比例显著高于通路中总基因占所有已经注释的基因的比例。这一显著性通常用多重假设检验矫正过的<code>pvalue</code>(又称qvalue, FDR或p.adjust)来表示。在图中体现为点的颜色。从绿到红富集显著性逐渐增高。点的大小表示对应通路中包含的差异基因的数目。点的形状代表了不同类型的基因，如EHBIO中上调的基因和Baodian中上调的基因。横轴表示对应通路包含的差异基因占总的差异基因的比例, 本图中最高不过5%, 这个值越大说明通路被影响的越多。</p>
</div>
</div>
<div id="section-3.9" class="section level2">
<h2><span class="header-section-number">3.9</span> 韦恩图</h2>
<p>维恩图是用来反映不同集合之间的交集和并集情况的展示图。一般用于展示2-5个集合之间的交并关系。集合数目更多时，将会比较难分辨，更多集合的展示方式一般使用<code>upSetView</code>。</p>
<p>较早的文章列举了多个在线工具<a href="http://mp.weixin.qq.com/s/zn654JqG9OeO71rJUTDr2Q" class="uri">http://mp.weixin.qq.com/s/zn654JqG9OeO71rJUTDr2Q</a>。</p>
<div id="section-3.9.1" class="section level3">
<h3><span class="header-section-number">3.9.1</span> 韦恩图三个圈</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(VennDiagram)
list1 =<span class="st"> </span><span class="kw">sample</span>(letters,<span class="dv">20</span>)
list2 =<span class="st"> </span><span class="kw">sample</span>(letters,<span class="dv">20</span>)
list3 =<span class="st"> </span><span class="kw">sample</span>(letters,<span class="dv">20</span>)

color_v &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;dodgerblue&quot;</span>, <span class="st">&quot;goldenrod1&quot;</span>, <span class="st">&quot;darkorange1&quot;</span>)


label_size =<span class="st"> </span><span class="dv">1</span>
margin =<span class="st"> </span><span class="fl">0.1</span>
p &lt;-<span class="st"> </span><span class="kw">venn.diagram</span>( 
    <span class="dt">x =</span> <span class="kw">list</span>(<span class="dt">ehbio1=</span>list1, <span class="dt">ehbio2=</span>list2, <span class="dt">ehbio3=</span>list3),
    <span class="dt">filename =</span> <span class="ot">NULL</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>, <span class="dt">lwd =</span> <span class="dv">1</span>, 
    <span class="dt">fill =</span> color_v, <span class="dt">alpha =</span> <span class="fl">0.50</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>, 
    <span class="dt">label.col =</span> <span class="kw">c</span>(<span class="st">&quot;black&quot;</span>), <span class="dt">cex =</span> <span class="dv">1</span>, <span class="dt">fontfamily =</span> <span class="st">&quot;Helvetica&quot;</span>,
    <span class="dt">cat.col =</span> color_v,<span class="dt">cat.cex =</span> label_size, 
    <span class="dt">margin=</span>margin, 
    <span class="dt">cat.fontfamily =</span> <span class="st">&quot;Helvetica&quot;</span>
)
<span class="kw">grid.draw</span>(p)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-188-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="section-3.9.2" class="section level3">
<h3><span class="header-section-number">3.9.2</span> 韦恩图五个圈</h3>
<p>假设有这么一个矩阵，第一列为不同集合中的ID，第二列为集合的名字，无标题行，存储为<code>venn.txt</code>。</p>
<pre><code>a   ehbio1
b   ehbio1
c   ehbio1
d   ehbio1
e   ehbio1
f   ehbio1
g   ehbio1
h   ehbio2
i   ehbio2
j   ehbio2
k   ehbio2
e   ehbio2
f   ehbio2
g   ehbio2
a   ehbio3
b   ehbio3
h   ehbio3
j   ehbio3
i   ehbio3
f   ehbio3
g   ehbio3
a   ehbio4
b   ehbio4
h   ehbio4
d   ehbio5
e   ehbio5
y   ehbio5
x   ehbio5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(VennDiagram)

<span class="co">#pdf(file=&quot;venn.txt.vennDiagram.pdf&quot;, onefile=FALSE, paper=&quot;special&quot;)</span>

data &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">file=</span><span class="st">&quot;data/venn.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>)

num &lt;-<span class="st"> </span><span class="dv">0</span>

ehbio1 &lt;-<span class="st"> </span>data[<span class="kw">grepl</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">&lt;ehbio1</span><span class="ch">\\</span><span class="st">&gt;&quot;</span>,data[,<span class="dv">2</span>]),<span class="dv">1</span>]
num &lt;-<span class="st"> </span>num +<span class="st"> </span><span class="dv">1</span>

ehbio2 &lt;-<span class="st"> </span>data[<span class="kw">grepl</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">&lt;ehbio2</span><span class="ch">\\</span><span class="st">&gt;&quot;</span>,data[,<span class="dv">2</span>]),<span class="dv">1</span>]
num &lt;-<span class="st"> </span>num +<span class="st"> </span><span class="dv">1</span>

ehbio3 &lt;-<span class="st"> </span>data[<span class="kw">grepl</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">&lt;ehbio3</span><span class="ch">\\</span><span class="st">&gt;&quot;</span>,data[,<span class="dv">2</span>]),<span class="dv">1</span>]
num &lt;-<span class="st"> </span>num +<span class="st"> </span><span class="dv">1</span>

ehbio4 &lt;-<span class="st"> </span>data[<span class="kw">grepl</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">&lt;ehbio4</span><span class="ch">\\</span><span class="st">&gt;&quot;</span>,data[,<span class="dv">2</span>]),<span class="dv">1</span>]
num &lt;-<span class="st"> </span>num +<span class="st"> </span><span class="dv">1</span>

ehbio5 &lt;-<span class="st"> </span>data[<span class="kw">grepl</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">&lt;ehbio5</span><span class="ch">\\</span><span class="st">&gt;&quot;</span>,data[,<span class="dv">2</span>]),<span class="dv">1</span>]
num &lt;-<span class="st"> </span>num +<span class="st"> </span><span class="dv">1</span>

color_v &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;dodgerblue&quot;</span>, <span class="st">&quot;goldenrod1&quot;</span>, <span class="st">&quot;darkorange1&quot;</span>, <span class="st">&quot;seagreen3&quot;</span>, <span class="st">&quot;orchid3&quot;</span>)[<span class="dv">1</span>:num]
<span class="co"># label.col = c(&quot;orange&quot;, &quot;white&quot;, &quot;darkorchid4&quot;, &quot;white&quot;, &quot;white&quot;, &quot;white&quot;, &quot;white&quot;, </span>
    <span class="co">#&quot;white&quot;, &quot;darkblue&quot;, &quot;white&quot;, &quot;white&quot;, &quot;white&quot;, &quot;white&quot;, &quot;darkgreen&quot;, &quot;white&quot;),</span>

label_size =<span class="st"> </span><span class="fl">0.8</span>
margin =<span class="st"> </span><span class="fl">0.3</span>
p &lt;-<span class="st"> </span><span class="kw">venn.diagram</span>( 
    <span class="dt">x =</span> <span class="kw">list</span>(<span class="dt">ehbio1=</span>ehbio1, <span class="dt">ehbio4=</span>ehbio4,
    <span class="dt">ehbio5=</span>ehbio5, <span class="dt">ehbio2=</span>ehbio2,
    <span class="dt">ehbio3=</span>ehbio3),
    <span class="dt">filename =</span> <span class="ot">NULL</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>, <span class="dt">lwd =</span> <span class="dv">1</span>, 
    <span class="dt">fill =</span> color_v,
    <span class="dt">alpha =</span> <span class="fl">0.50</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>, 
    <span class="dt">label.col =</span> <span class="kw">c</span>(<span class="st">&quot;black&quot;</span>),
    <span class="dt">cex =</span> <span class="dv">1</span>, <span class="dt">fontfamily =</span> <span class="st">&quot;Helvetica&quot;</span>,
    <span class="dt">cat.col =</span> color_v,<span class="dt">cat.cex =</span> label_size, 
    <span class="dt">margin=</span>margin, 
    <span class="dt">cat.fontfamily =</span> <span class="st">&quot;Helvetica&quot;</span>
)
<span class="kw">grid.draw</span>(p)
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-189-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="upsetview" class="section level3">
<h3><span class="header-section-number">3.9.3</span> UpSetView展示</h3>
<p>对于集合比较多的时候，包括上面提到的5个集合的交并集情况，如果只是为了展示个炫图，还可以，但如果想解释结果，就会比较头疼，难判断区域的归属。</p>
<p>因此对于这种多集合情况，推荐使用<code>UpSetView</code>展示，看效果如下。</p>
<p>测试数据，存储为<code>upsetview.txt</code> （第一行为集合的名，每个集合一列；每一行为一个ID，如果对应ID出现在这个集合则标记1，否则标记0）：</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">pattern</span> ehbio1  ehbio2  ehbio3  ehbio4  ehbio5
<span class="kw">a</span>   1   0   1   1   0
<span class="kw">b</span>   1   0   1   1   0
<span class="kw">c</span>   1   0   0   0   0
<span class="kw">d</span>   1   0   0   0   1
<span class="kw">e</span>   1   1   0   0   1
<span class="kw">f</span>   1   1   1   0   0
<span class="kw">g</span>   1   1   1   0   0
<span class="kw">h</span>   0   1   1   1   0
<span class="kw">i</span>   0   1   1   0   0
<span class="kw">j</span>   0   1   1   0   0
<span class="kw">k</span>   0   1   0   0   0
<span class="kw">x</span>   0   0   0   0   1
<span class="kw">y</span>   0   0   0   0   1</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(UpSetR)
matrix =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/upsetview.txt&quot;</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="ot">NULL</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
nsets =<span class="st"> </span><span class="kw">dim</span>(matrix)[<span class="dv">2</span>]-<span class="dv">1</span>

<span class="co">#pdf(file=&quot;upsetview.txt.upsetV.pdf&quot;, onefile=FALSE, paper=&quot;special&quot;, width=10, </span>
<span class="co">#height=5, bg=&quot;white&quot;, pointsize=12)</span>
<span class="kw">upset</span>(matrix, <span class="dt">nsets=</span>nsets, <span class="dt">sets.bar.color =</span> <span class="st">&quot;#56B4E9&quot;</span>, <span class="dt">order.by =</span> <span class="st">&quot;freq&quot;</span>, 
    <span class="dt">empty.intersections =</span> <span class="st">&quot;on&quot;</span>)
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-190-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>UpSetR: <a href="http://www.caleydo.org/tools/upset/" class="uri">http://www.caleydo.org/tools/upset/</a> 采用连线的方式展示不同的组合之间共有的和特有的项目，对于特别多的组合尤其适用。</p>
<p>单个点表示特有，连起来的点表示共有，相当于venn图中重叠的部分。</p>
<p>垂直的柱子代表的是Venn图中的数字，看连接的点判断归属。</p>
<p>水平的柱子代表对应样品中Item的总数。</p>
</div>
</div>
<div id="section-3.10" class="section level2">
<h2><span class="header-section-number">3.10</span> 柱状图绘制</h2>
<p>柱状图也是较为常见的一种数据展示方式，可以展示基因的表达量，也可以展示GO富集分析结果，基因注释数据等。<a href="http://mp.weixin.qq.com/s?__biz=MzI5MTcwNjA4NQ==&amp;mid=2247484106&amp;idx=1&amp;sn=687a0def51f6ea91a335754eb3dc9ca9&amp;chksm=ec0dc740db7a4e564e5b1e93a36e5d9447581e262eec9c2983d1d4e76788d673c9c07dec8f8e#rd">39个转录组分析工具，120种组合评估(转录组分析工具哪家强-导读版)</a>中提到了较多<strong>堆积柱状图</strong>的使用。下面就详细介绍下怎么绘制。</p>
<div id="section-3.10.1" class="section level3">
<h3><span class="header-section-number">3.10.1</span> 常规矩阵柱状图绘制</h3>
<p>有如下4个基因在5组样品中的表达值</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_ori &lt;-<span class="st"> &quot;Grp_1;Grp_2;Grp_3;Grp_4;Grp_5</span>
<span class="st">a;2.6;2.9;2.1;2.0;2.2</span>
<span class="st">b;20.8;9.8;7.0;3.7;19.2</span>
<span class="st">c;10.0;11.0;9.2;12.4;9.6</span>
<span class="st">d;9;3.3;10.3;11.1;10&quot;</span>

data &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>data_ori, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>)
data</code></pre></div>
<pre><code>##   Grp_1 Grp_2 Grp_3 Grp_4 Grp_5
## a   2.6   2.9   2.1   2.0   2.2
## b  20.8   9.8   7.0   3.7  19.2
## c  10.0  11.0   9.2  12.4   9.6
## d   9.0   3.3  10.3  11.1  10.0</code></pre>
<p>整理数据格式，保留基因名字信息</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(reshape2)
<span class="kw">library</span>(dplyr)
data_rownames &lt;-<span class="st"> </span><span class="kw">rownames</span>(data)
data_colnames &lt;-<span class="st"> </span><span class="kw">colnames</span>(data)
data$gene &lt;-<span class="st"> </span>data_rownames
data_m &lt;-<span class="st"> </span><span class="kw">melt</span>(data, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;gene&quot;</span>))
data_m</code></pre></div>
<pre><code>##    gene variable value
## 1     a    Grp_1   2.6
## 2     b    Grp_1  20.8
## 3     c    Grp_1  10.0
## 4     d    Grp_1   9.0
## 5     a    Grp_2   2.9
## 6     b    Grp_2   9.8
## 7     c    Grp_2  11.0
## 8     d    Grp_2   3.3
## 9     a    Grp_3   2.1
## 10    b    Grp_3   7.0
## 11    c    Grp_3   9.2
## 12    d    Grp_3  10.3
## 13    a    Grp_4   2.0
## 14    b    Grp_4   3.7
## 15    c    Grp_4  12.4
## 16    d    Grp_4  11.1
## 17    a    Grp_5   2.2
## 18    b    Grp_5  19.2
## 19    c    Grp_5   9.6
## 20    d    Grp_5  10.0</code></pre>
<p>首先看下每个基因在不同组的表达情况</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 给定数据，和x轴、y轴所在列名字</span>
<span class="co"># 直接使用geom_bar就可以绘制柱状图</span>
<span class="co"># position: dodge: 柱子并排放置</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>gene, <span class="dt">y=</span>value))
p +<span class="st"> </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;dodge&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>variable))

<span class="co"># 如果没有图形界面，运行下面的语句把图存在工作目录下的Rplots.pdf文件中</span>
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-193-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>柱子有点多，也可以利用mean±SD的形式展现</p>
<p>首先计算平均值和标准差，使用<code>group_by</code>按<code>gene</code>分组，对每组做<code>summarize</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 获取平均值和标准差</span>
data_m_sd_mean &lt;-<span class="st"> </span>data_m %&gt;%<span class="st"> </span><span class="kw">group_by</span>(gene) %&gt;%<span class="st"> </span>
<span class="st">    </span>dplyr::<span class="kw">summarise</span>(<span class="dt">sd=</span><span class="kw">sd</span>(value), <span class="dt">mean_value=</span><span class="kw">mean</span>(value))
data_m_sd_mean &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(data_m_sd_mean)
data_m_sd_mean</code></pre></div>
<pre><code>##   gene        sd mean_value
## 1    a 0.3781534       2.36
## 2    b 7.5491721      12.10
## 3    c 1.2837445      10.44
## 4    d 3.1325708       8.74</code></pre>
<p>使用<code>geom_errorbar</code>添加误差线</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m_sd_mean, <span class="kw">aes</span>(<span class="dt">x=</span>gene, <span class="dt">y=</span>mean_value)) +<span class="st"> </span>
<span class="st">       </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>) +
<span class="st">     </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin=</span>mean_value-sd, <span class="dt">ymax=</span>mean_value+sd))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-195-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>设置误差线的宽度和位置</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m_sd_mean, <span class="kw">aes</span>(<span class="dt">x=</span>gene, <span class="dt">y=</span>mean_value)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>gene)) +
<span class="st">    </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin=</span>mean_value-sd, <span class="dt">ymax=</span>mean_value+sd), <span class="dt">width=</span><span class="fl">0.2</span>, 
        <span class="dt">position=</span><span class="kw">position_dodge</span>(<span class="dt">width=</span><span class="fl">0.75</span>))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-196-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>每个基因的原始表达值堆积柱状图 (只需要修改<code>positon=stack</code>)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># position=&quot;fill&quot; 展示的是堆积柱状图各部分的相对比例</span>
<span class="co"># position=&quot;stack&quot; 展示的是堆积柱状图的原始值</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value)) +
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;stack&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>gene)) +
<span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>value), <span class="dt">position=</span><span class="kw">position_stack</span>(<span class="dt">vjust=</span><span class="fl">0.5</span>))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-197-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>堆积柱状图显示没问题，但文本标记错位了</p>
<p>指定下分组信息，位置计算就正确了</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># position=&quot;fill&quot; 展示的是堆积柱状图各部分的相对比例</span>
<span class="co"># position=&quot;stack&quot; 展示的是堆积柱状图的原始值</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value, <span class="dt">group=</span>gene)) +
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;stack&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>gene)) +
<span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>value), <span class="dt">position=</span><span class="kw">position_stack</span>(<span class="dt">vjust=</span><span class="fl">0.5</span>))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-198-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>比较每组各个基因的相对表达 (<code>position=fill</code>)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># position=&quot;fill&quot; 展示的是堆积柱状图各部分的相对比例</span>
<span class="co"># position=&quot;stack&quot; 展示的是堆积柱状图的原始值，可以自己体现下看卡差别</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value)) +
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;fill&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>gene))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-199-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>纵轴的显示改为百分比</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value)) +
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;fill&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>gene)) +
<span class="st">    </span><span class="kw">scale_y_continuous</span>(<span class="dt">labels =</span> scales::percent)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-200-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>在柱子中标记百分比值</p>
<p>首先计算百分比，同样是<code>group_by</code> (按照给定的变量分组，然后按组操作)和<code>mutate</code>两个函数(在当前数据表增加新变量)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># group_by: 按照给定的变量分组，然后按组操作</span>
<span class="co"># mutate: 在当前数据表增加新变量</span>
<span class="co"># 第一步增加每个组的加和，第二步计算比例</span>
data_m &lt;-<span class="st"> </span>data_m %&gt;%<span class="st"> </span><span class="kw">group_by</span>(variable) %&gt;%<span class="st"> </span><span class="kw">mutate</span>(<span class="dt">count=</span><span class="kw">sum</span>(value)) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">freq=</span><span class="kw">round</span>(<span class="dv">100</span>*value/count,<span class="dv">2</span>))
<span class="kw">head</span>(data_m)</code></pre></div>
<pre><code>## # A tibble: 6 x 5
## # Groups:   variable [2]
##   gene  variable value count  freq
##   &lt;chr&gt; &lt;fct&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1 a     Grp_1     2.60  42.4  6.13
## 2 b     Grp_1    20.8   42.4 49.1 
## 3 c     Grp_1    10.0   42.4 23.6 
## 4 d     Grp_1     9.00  42.4 21.2 
## 5 a     Grp_2     2.90  27.0 10.7 
## 6 b     Grp_2     9.80  27.0 36.3</code></pre>
<p>再标记相对比例信息</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>variable, <span class="dt">y=</span>value, <span class="dt">group=</span>gene)) +
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;fill&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>gene)) +
<span class="st">    </span><span class="kw">scale_y_continuous</span>(<span class="dt">labels =</span> scales::percent) +
<span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>freq), <span class="dt">position=</span><span class="kw">position_fill</span>(<span class="dt">vjust=</span><span class="fl">0.5</span>))
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-202-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="section-3.10.2" class="section level3">
<h3><span class="header-section-number">3.10.2</span> 长矩阵分面绘制</h3>
<p>再复杂一些的矩阵 (除了有不同时间点的信息，再增加对照和处理的信息)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(reshape2)
<span class="kw">library</span>(dplyr)

data_ori &lt;-<span class="st"> &quot;Gene;Group;Expr;Condition</span>
<span class="st">a;T1;2.6;Control</span>
<span class="st">b;T1;20.8;Control</span>
<span class="st">c;T1;10;Control</span>
<span class="st">d;T1;9;Control</span>
<span class="st">a;T2;2.9;Control</span>
<span class="st">b;T2;9.8;Control</span>
<span class="st">c;T2;11;Control</span>
<span class="st">d;T2;3.3;Control</span>
<span class="st">a;T3;2.1;Control</span>
<span class="st">b;T3;7;Control</span>
<span class="st">c;T3;9.2;Control</span>
<span class="st">d;T3;10.3;Control</span>
<span class="st">a;T4;2;Control</span>
<span class="st">b;T4;3.7;Control</span>
<span class="st">c;T4;12.4;Control</span>
<span class="st">d;T4;11.1;Control</span>
<span class="st">a;T5;2.2;Control</span>
<span class="st">b;T5;19.2;Control</span>
<span class="st">c;T5;9.6;Control</span>
<span class="st">d;T5;10;Control</span>
<span class="st">d;T1;2.6;Treatment</span>
<span class="st">b;T1;20.8;Treatment</span>
<span class="st">c;T1;10;Treatment</span>
<span class="st">a;T1;9;Treatment</span>
<span class="st">d;T2;2.9;Treatment</span>
<span class="st">b;T2;9.8;Treatment</span>
<span class="st">c;T2;11;Treatment</span>
<span class="st">a;T2;3.3;Treatment</span>
<span class="st">a;T3;2.1;Treatment</span>
<span class="st">c;T3;7;Treatment</span>
<span class="st">b;T3;9.2;Treatment</span>
<span class="st">d;T3;10.3;Treatment</span>
<span class="st">a;T4;2;Treatment</span>
<span class="st">c;T4;3.7;Treatment</span>
<span class="st">b;T4;12.4;Treatment</span>
<span class="st">d;T4;11.1;Treatment</span>
<span class="st">a;T5;2.2;Treatment</span>
<span class="st">d;T5;19.2;Treatment</span>
<span class="st">c;T5;9.6;Treatment</span>
<span class="st">b;T5;10;Treatment&quot;</span>

data_m &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>data_ori, <span class="dt">header=</span>T, <span class="dt">sep=</span><span class="st">&quot;;&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>)
<span class="kw">head</span>(data_m)</code></pre></div>
<pre><code>##   Gene Group Expr Condition
## 1    a    T1  2.6   Control
## 2    b    T1 20.8   Control
## 3    c    T1 10.0   Control
## 4    d    T1  9.0   Control
## 5    a    T2  2.9   Control
## 6    b    T2  9.8   Control</code></pre>
<p>首先看下每个基因在不同组的表达情况, <code>facet_grid</code>和<code>facet_wrap</code>可以对图形分面显示。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># scales: free_y 表示不同子图之间使用独立的Y轴信息</span>
<span class="co">#         但x轴使用同样的信息。</span>
<span class="co">#         其它可选参数有free_x, free, fixed</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>Gene, <span class="dt">y=</span>Expr)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;dodge&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>Group)) +
<span class="st">    </span><span class="kw">facet_grid</span>(Condition~., <span class="dt">scales=</span><span class="st">&quot;free_y&quot;</span>)
p
<span class="co"># 如果没有图形界面，运行下面的语句把图存在工作目录下的Rplots.pdf文件中</span>
<span class="co">#dev.off()</span></code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-204-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>柱子有点多，也可以利用<code>mean±SD</code>的形式展现</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 获取平均值和标准差</span>
<span class="co"># 分组时不只Gene一个变量了，还需要考虑Condition</span>
data_m_sd_mean &lt;-<span class="st"> </span>data_m %&gt;%<span class="st"> </span><span class="kw">group_by</span>(Gene, Condition) %&gt;%<span class="st"> </span>
<span class="st">    </span>dplyr::<span class="kw">summarise</span>(<span class="dt">sd=</span><span class="kw">sd</span>(Expr), <span class="dt">value=</span><span class="kw">mean</span>(Expr))
data_m_sd_mean &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(data_m_sd_mean)
data_m_sd_mean</code></pre></div>
<pre><code>##   Gene Condition        sd value
## 1    a   Control 0.3781534  2.36
## 2    a Treatment 2.9978326  3.72
## 3    b   Control 7.5491721 12.10
## 4    b Treatment 4.8299068 12.44
## 5    c   Control 1.2837445 10.44
## 6    c Treatment 2.9458445  8.26
## 7    d   Control 3.1325708  8.74
## 8    d Treatment 6.8568943  9.22</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m_sd_mean, <span class="kw">aes</span>(<span class="dt">x=</span>Gene, <span class="dt">y=</span>value)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>Gene)) +
<span class="st">    </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin=</span>value-sd, <span class="dt">ymax=</span>value+sd), <span class="dt">width=</span><span class="fl">0.2</span>, 
            <span class="dt">position=</span><span class="kw">position_dodge</span>(<span class="dt">width=</span><span class="fl">0.75</span>)) +
<span class="st">    </span><span class="kw">facet_wrap</span>(~Condition, <span class="dt">ncol=</span><span class="dv">1</span>)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-206-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>每组里面各个基因的相对表达, 纵轴的显示改为百分比</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># position=&quot;fill&quot; 展示的是堆积柱状图各部分的相对比例</span>
<span class="co"># position=&quot;stack&quot; 展示的是堆积柱状图的原始值，可以自己体现下看卡差别</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>Group, <span class="dt">y=</span>Expr)) +
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;fill&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>Gene)) +
<span class="st">    </span><span class="kw">scale_y_continuous</span>(<span class="dt">labels =</span> scales::percent) +
<span class="st">    </span><span class="kw">facet_wrap</span>(~Condition, <span class="dt">ncol=</span><span class="dv">1</span>)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-207-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><code>facet</code>后，显示正常，不需要做特别的修改</p>
<p>在柱子中标记百分比值 (计算百分比值需要注意了, 文本显示位置还是跟之前一致)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># group_by: 按照给定的变量分组，然后按组操作</span>
<span class="co"># mutate: 在当前数据表增加新变量</span>
<span class="co"># 第一步增加每个组 (Group和Condition共同定义分组)的加和，第二步计算比例</span>
data_m &lt;-<span class="st"> </span>data_m %&gt;%<span class="st"> </span><span class="kw">group_by</span>(Group, Condition) %&gt;%<span class="st"> </span><span class="kw">mutate</span>(<span class="dt">count=</span><span class="kw">sum</span>(Expr)) %&gt;%
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">freq=</span><span class="kw">round</span>(<span class="dv">100</span>*Expr/count,<span class="dv">2</span>))
<span class="kw">head</span>(data_m)</code></pre></div>
<pre><code>## # A tibble: 6 x 6
## # Groups:   Group, Condition [2]
##   Gene  Group  Expr Condition count  freq
##   &lt;fct&gt; &lt;fct&gt; &lt;dbl&gt; &lt;fct&gt;     &lt;dbl&gt; &lt;dbl&gt;
## 1 a     T1     2.60 Control    42.4  6.13
## 2 b     T1    20.8  Control    42.4 49.1 
## 3 c     T1    10.0  Control    42.4 23.6 
## 4 d     T1     9.00 Control    42.4 21.2 
## 5 a     T2     2.90 Control    27.0 10.7 
## 6 b     T2     9.80 Control    27.0 36.3</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>Group, <span class="dt">y=</span>Expr, <span class="dt">group=</span>Group)) +
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;fill&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>Gene)) +
<span class="st">    </span><span class="kw">scale_y_continuous</span>(<span class="dt">labels =</span> scales::percent) +
<span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>freq), <span class="dt">position=</span><span class="kw">position_fill</span>(<span class="dt">vjust=</span><span class="fl">0.5</span>)) +
<span class="st">    </span><span class="kw">facet_wrap</span>(~Condition, <span class="dt">ncol=</span><span class="dv">1</span>)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-209-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>文本显示位置没有问题，但柱子的位置有些奇怪，使得两组之间不可比。</p>
<p>先对数据做下排序，然后再标记文本 (这部分可以看下视频，或者自己输出下数据看看为什么要排序。底层原因可能是这种新的数据结构的问题。)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># with: 产生一个由data_m组成的局部环境，再这个环境里，列名字可以直接使用</span>
data_m &lt;-<span class="st"> </span>data_m[<span class="kw">with</span>(data_m, <span class="kw">order</span>(Condition, Group, Gene)),] 
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(data_m, <span class="kw">aes</span>(<span class="dt">x=</span>Group, <span class="dt">y=</span>Expr, <span class="dt">group=</span>Group)) +
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;fill&quot;</span>, <span class="kw">aes</span>(<span class="dt">fill=</span>Gene)) +
<span class="st">    </span><span class="kw">scale_y_continuous</span>(<span class="dt">labels =</span> scales::percent) +
<span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>freq), <span class="dt">position=</span><span class="kw">position_fill</span>(<span class="dt">vjust=</span><span class="fl">0.5</span>)) +
<span class="st">    </span><span class="kw">facet_wrap</span>(~Condition, <span class="dt">ncol=</span><span class="dv">1</span>)
p</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-210-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>这样两种条件下的比较更容易了。</p>
</div>
</div>
<div id="section-3.11" class="section level2">
<h2><span class="header-section-number">3.11</span> 图形支持中文字体</h2>
<div id="section-3.11.1" class="section level3">
<h3><span class="header-section-number">3.11.1</span> 修改图形的字体</h3>
<p>ggplot2中修改图形字体。</p>
<pre><code># 修改坐标轴和legend、标题的字体
theme(text=element_text(family=&quot;Arial&quot;))
# 或者
theme_bw(base_family=&quot;Arial&quot;)

# 修改geom_text的字体
geom_text(family=&quot;Arial&quot;)</code></pre>
</div>
<div id="ggplot2pdf" class="section level3">
<h3><span class="header-section-number">3.11.2</span> ggplot2支持中文字体输出PDF</h3>
<p><code>showtext</code>包可给定字体文件，加载到R环境中，生成新的字体家族名字，后期调用这个名字设定字体，并且支持中文写入pdf不乱码</p>
<pre><code>library(showtext)
showtext.auto(enable=TRUE)

font_path = &quot;FZSTK.TTF&quot;
font_name = tools::file_path_sans_ext(basename(font_path))
font.add(font_name, font_path)

# 修改坐标轴和legend、标题的字体
theme(text=element_text(family=font_name))

# 修改geom_text的字体
geom_text(family=font_name)</code></pre>
</div>
<div id="section-3.11.3" class="section level3">
<h3><span class="header-section-number">3.11.3</span> 系统可用字体</h3>
<ul>
<li><p>Linux字体一般在 <code>/usr/share/fonts</code>下，也可以使用<code>fc-list</code>列出所以加载的字体。</p></li>
<li><p>Windows字体在 <code>C:\Windows\Fonts\</code>下，直接可以看到，也可以拷贝到Linux下使用。</p></li>
</ul>
</div>
<div id="section-3.11.4" class="section level3">
<h3><span class="header-section-number">3.11.4</span> 合并字体支持中英文</h3>
<p>通常情况下，作图的字体都是英文，ggplot2默认的或按需求加载一种字体就可以了。但如果中英文混合出现时，单个字体只能支持一种文字，最好的方式是合并两种字体，类似于Word中设置中英文分别使用不同的字体。</p>
<p>软件<a href="https://github.com/fontforge/fontforge">FontForge</a>可以方便的合并中英文字体，其安装也比较简单，直接 <code>yum install fontforge.x86_64</code>。</p>
<p>假如需要合并<code>FZSTK.TTF</code> (windows下获取)和<code>Schoolbell-Regular.ttf</code> (谷歌下载)，这两个都是手写字体。按如下，把字体文件和程序脚本<code>mergefont.pe</code>放在同一目录下，运行<code>fontforge -script mergefont.pe</code>即可获得合并后的字体<code>FZ_School.ttf</code>。</p>
<pre><code>ct@ehbio $ ls
FZSTK.TTF mergefont.pe Schoolbell-Regular.ttf
ct@ehbio $ cat mergefont.pe
Open(&quot;FZSTK.TTF&quot;)
SelectAll()
ScaleToEm(1024)
Generate(&quot;temp.ttf&quot;, &quot;&quot;, 0x14)
Close()

# Open English font and merge to the Chinese font
Open(&quot;Schoolbell-Regular.ttf&quot;)
SelectAll()
ScaleToEm(1024)

MergeFonts(&quot;temp.ttf&quot;)
SetFontNames(&quot;FZ_School&quot;, &quot;FZST&quot;, &quot;Schoolbel&quot;, &quot;Regular&quot;, &quot;&quot;)
Generate(&quot;FZ_School.ttf&quot;, &quot;&quot;, 0x14)
Close()

ct@ehbio $ fontforge -script mergefont.pe
ct@ehbio $ ls
FZ_School.ttf FZSTK.TTF mergefont.pe Schoolbell-Regular.ttf</code></pre>
<p>然后安装前面的介绍使用<code>showtext</code>导入即可使用。</p>
</div>
<div id="section-3.11.5" class="section level3">
<h3><span class="header-section-number">3.11.5</span> 一个示例</h3>
<p>字体文件自己从Windows获取，School bell从Google fonts获取。</p>
<pre><code>library(showtext)
## Add fonts that are available on current path

# 方正字体+schoole bell (中英混合)
font.add(&quot;FZ_School&quot;, &quot;font/FZ_School.ttf&quot;)
# 黑体
font.add(&quot;simhei&quot;, &quot;font/simhei.ttf&quot;)
font.add(&quot;Arial&quot;,&quot;font/arial.ttf&quot;)

# 黑体和Arial的合体
font.add(&quot;HeiArial&quot;, &quot;font/HeiArial.ttf&quot;)
showtext.auto()  ## automatically use showtext for new devices

library(ggplot2)

p = ggplot(NULL, aes(x = 1:10, y = 2^(1:10), group=1)) + geom_line() +
  theme(axis.title.y=element_text(family=&quot;Arial&quot;), 
        axis.title.x=element_text(family=&quot;HeiArial&quot;), 
        plot.title=element_text(family=&quot;simhei&quot;)) +
  xlab(&quot;Days spent on 生信宝典&quot;) + 
  ylab(&quot;Things you have learned&quot;) +
  ggtitle(&quot;生信宝典，换个角度学生信&quot;) + 
  annotate(&quot;text&quot;, 7, 300, family = &quot;FZ_School&quot;, size = 8,
           label = &quot;收获曲线 (Harvest curve)&quot;, angle=15) 

# annotate指定的是文字的中间部分的位置

#ggsave(p, filename=&quot;example-SXBD.pdf&quot;, width = 7, height = 4)  ## PDF device
p</code></pre>
<p><img src="images/sxbd_font.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="pca" class="section level2">
<h2><span class="header-section-number">3.12</span> PCA</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(knitr)
<span class="kw">library</span>(psych)
<span class="kw">library</span>(reshape2)
<span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(ggbeeswarm)
<span class="kw">library</span>(scatterplot3d)
<span class="kw">library</span>(useful)
<span class="kw">library</span>(ggfortify)</code></pre></div>
<div id="section-3.12.1" class="section level3">
<h3><span class="header-section-number">3.12.1</span> 主成分分析简介</h3>
<p>主成分分析 (PCA, principal component analysis)是一种数学降维方法, 利用正交变换 (orthogonal transformation)把一系列可能<strong>线性相关的变量</strong>转换为一组<strong>线性不相关的新变量</strong>，也称为主成分，从而利用新变量在更小的维度下展示数据的特征。</p>
<p>主成分是原有变量的线性组合，其数目不多于原始变量。组合之后，相当于我们获得了一批新的观测数据，这些数据的含义不同于原有数据，但包含了之前数据的大部分特征，并且有着较低的维度，便于进一步的分析。</p>
<p>在空间上，PCA可以理解为把原始数据投射到一个新的坐标系统，第一主成分为第一坐标轴，它的含义代表了原始数据中多个变量经过某种变换得到的新变量的变化区间；第二成分为第二坐标轴，代表了原始数据中多个变量经过某种变换得到的第二个新变量的变化区间。这样我们把利用原始数据解释样品的差异转变为利用新变量解释样品的差异。</p>
<p>这种投射方式会有很多，为了最大限度保留对原始数据的解释，一般会用最大方差理论或最小损失理论，使得第一主成分有着最大的方差或变异数 (就是说其能尽量多的解释原始数据的差异)；随后的每一个主成分都与前面的主成分正交，且有着仅次于前一主成分的最大方差 (正交简单的理解就是两个主成分空间夹角为90°，两者之间无线性关联，从而完成去冗余操作)。</p>
</div>
<div id="section-3.12.2" class="section level3">
<h3><span class="header-section-number">3.12.2</span> 主成分分析的意义</h3>
<ol style="list-style-type: decimal">
<li><p>简化运算。</p>
<p>在问题研究中，为了全面系统地分析问题，我们通常会收集众多的影响因素也就是众多的变量。这样会使得研究更丰富，通常也会带来较多的冗余数据和复杂的计算量。</p>
<p>比如我们我们测序了100种样品的基因表达谱借以通过分子表达水平的差异对这100种样品进行分类。在这个问题中，研究的变量就是不同的基因。每个基因的表达都可以在一定程度上反应样品之间的差异，但某些基因之间却有着调控、协同或拮抗的关系，表现为它们的表达值存在一些相关性，这就造成了统计数据所反映的信息存在一定程度的冗余。另外假如某些基因如持家基因在所有样本中表达都一样，它们对于解释样本的差异也没有意义。这么多的变量在后续统计分析中会增大运算量和计算复杂度，应用PCA就可以在尽量多的保持变量所包含的信息又能维持尽量少的变量数目，帮助简化运算和结果解释。</p></li>
<li><p>去除数据噪音。</p>
<p>比如说我们在样品的制备过程中，由于不完全一致的操作，导致样品的状态有细微的改变，从而造成一些持家基因也发生了相应的变化，但变化幅度远小于核心基因 (一般认为噪音的方差小于信息的方差）。而PCA在降维的过程中滤去了这些变化幅度较小的噪音变化，增大了数据的信噪比。</p></li>
<li><p>利用散点图实现多维数据可视化。</p>
<p>在上面的表达谱分析中，假如我们有1个基因，可以在线性层面对样本进行分类；如果我们有2个基因，可以在一个平面对样本进行分类；如果我们有3个基因，可以在一个立体空间对样本进行分类；如果有更多的基因，比如说n个，那么每个样品就是n维空间的一个点，则很难在图形上展示样品的分类关系。利用PCA分析，我们可以选取贡献最大的2个或3个主成分作为数据代表用以可视化。这比直接选取三个表达变化最大的基因更能反映样品之间的差异。（利用Pearson相关系数对样品进行聚类在样品数目比较少时是一个解决办法）</p></li>
<li><p>发现隐性相关变量。</p>
<p>我们在合并冗余原始变量得到主成分过程中，会发现某些原始变量对同一主成分有着相似的贡献，也就是说这些变量之间存在着某种相关性，为相关变量。同时也可以获得这些变量对主成分的贡献程度。对基因表达数据可以理解为发现了存在协同或拮抗关系的基因。</p></li>
</ol>
</div>
<div id="section-3.12.3" class="section level3">
<h3><span class="header-section-number">3.12.3</span> 示例展示原始变量对样品的分类</h3>
<p>假设有一套数据集，包含100个样品中某一基因的表达量。如下所示，每一行为一个样品，每一列为基因的表达值。这也是做PCA分析的基本数据组织方式，每一行代表一个样品，每一列代表一组观察数据即一个变量。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">count &lt;-<span class="st"> </span><span class="dv">50</span>
Gene1_a &lt;-<span class="st"> </span><span class="kw">rnorm</span>(count,<span class="dv">5</span>,<span class="fl">0.5</span>)
Gene1_b &lt;-<span class="st"> </span><span class="kw">rnorm</span>(count,<span class="dv">20</span>,<span class="fl">0.5</span>)
grp_a &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="st">&#39;a&#39;</span>, count)
grp_b &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="st">&#39;b&#39;</span>, count)
cy_data &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Gene1 =</span> <span class="kw">c</span>(Gene1_a, Gene1_b), <span class="dt">Group=</span><span class="kw">c</span>(grp_a, grp_b))
cy_data &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(cy_data)
label &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste0</span>(grp_a, <span class="dv">1</span>:count), <span class="kw">paste0</span>(grp_b, <span class="dv">1</span>:count))
<span class="kw">row.names</span>(cy_data) &lt;-<span class="st"> </span>label
<span class="kw">library</span>(knitr)
<span class="kw">library</span>(psych)
<span class="co"># Add additional column to data only for plotting</span>
cy_data$Y &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>,count*<span class="dv">2</span>)
<span class="kw">kable</span>(<span class="kw">headTail</span>(cy_data), <span class="dt">booktabs=</span><span class="ot">TRUE</span>, <span class="dt">caption=</span><span class="st">&quot;Expression profile for Gene1 in 100 samples&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:datainit1">Table 3.1: </span>Expression profile for Gene1 in 100 samples</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">Gene1</th>
<th align="left">Group</th>
<th align="left">Y</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">a1</td>
<td align="left">4.71</td>
<td align="left">a</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td align="left">a2</td>
<td align="left">4.66</td>
<td align="left">a</td>
<td align="left">0</td>
</tr>
<tr class="odd">
<td align="left">a3</td>
<td align="left">4.74</td>
<td align="left">a</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td align="left">a4</td>
<td align="left">4.66</td>
<td align="left">a</td>
<td align="left">0</td>
</tr>
<tr class="odd">
<td align="left">…</td>
<td align="left">…</td>
<td align="left">NA</td>
<td align="left">…</td>
</tr>
<tr class="even">
<td align="left">b47</td>
<td align="left">20.11</td>
<td align="left">b</td>
<td align="left">0</td>
</tr>
<tr class="odd">
<td align="left">b48</td>
<td align="left">19.44</td>
<td align="left">b</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td align="left">b49</td>
<td align="left">19.39</td>
<td align="left">b</td>
<td align="left">0</td>
</tr>
<tr class="odd">
<td align="left">b50</td>
<td align="left">20.08</td>
<td align="left">b</td>
<td align="left">0</td>
</tr>
</tbody>
</table>
<p>从下图可以看出，100个样品根据<code>Gene1</code>表达量的不同在横轴上被被分为了2类，可以看做是在线性水平的分类。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;ggplot2&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;ggbeeswarm&quot;</span>)

<span class="co"># geom_quasirandom:用于画Jitter Plot</span>
<span class="co"># theme(axis.*.y): 去除Y轴</span>
<span class="co"># xlim, ylim设定坐标轴的区间</span>
<span class="kw">ggplot</span>(cy_data,<span class="kw">aes</span>(Gene1, Y))+<span class="kw">geom_quasirandom</span>(<span class="kw">aes</span>(<span class="dt">color=</span><span class="kw">factor</span>(Group)))+
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">0.7</span>)) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.title=</span><span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">scale_fill_discrete</span>(<span class="dt">name=</span><span class="st">&quot;Group&quot;</span>) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.line.y=</span><span class="kw">element_blank</span>(), 
    <span class="dt">axis.text.y=</span><span class="kw">element_blank</span>(), <span class="dt">axis.ticks.y=</span><span class="kw">element_blank</span>(), 
    <span class="dt">axis.title.y=</span><span class="kw">element_blank</span>()) +<span class="st"> </span><span class="kw">ylim</span>(-<span class="fl">0.5</span>,<span class="dv">5</span>) +<span class="st"> </span><span class="kw">xlim</span>(<span class="dv">0</span>,<span class="dv">25</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/jitteroneDdata-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>那么如果有2个基因呢？</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">count &lt;-<span class="st"> </span><span class="dv">50</span>
Gene2_a &lt;-<span class="st"> </span><span class="kw">rnorm</span>(count,<span class="dv">5</span>,<span class="fl">0.2</span>)
Gene2_b &lt;-<span class="st"> </span><span class="kw">rnorm</span>(count,<span class="dv">5</span>,<span class="fl">0.2</span>)

cy_data2 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Gene1 =</span> <span class="kw">c</span>(Gene1_a, Gene1_b), <span class="dt">Gene2 =</span> <span class="kw">c</span>(Gene2_a, Gene2_b), 
        <span class="dt">Group=</span><span class="kw">c</span>(grp_a, grp_b))
cy_data2 &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(cy_data2)

<span class="kw">row.names</span>(cy_data2) &lt;-<span class="st"> </span>label

<span class="kw">kable</span>(<span class="kw">headTail</span>(cy_data2), <span class="dt">booktabs=</span>T, 
        <span class="dt">caption=</span><span class="st">&quot;Expression profile for Gene1 and Gene2 in 100 samples&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:datainit2">Table 3.2: </span>Expression profile for Gene1 and Gene2 in 100 samples</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">Gene1</th>
<th align="left">Gene2</th>
<th align="left">Group</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">a1</td>
<td align="left">4.71</td>
<td align="left">5.01</td>
<td align="left">a</td>
</tr>
<tr class="even">
<td align="left">a2</td>
<td align="left">4.66</td>
<td align="left">4.82</td>
<td align="left">a</td>
</tr>
<tr class="odd">
<td align="left">a3</td>
<td align="left">4.74</td>
<td align="left">4.84</td>
<td align="left">a</td>
</tr>
<tr class="even">
<td align="left">a4</td>
<td align="left">4.66</td>
<td align="left">5.02</td>
<td align="left">a</td>
</tr>
<tr class="odd">
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">b47</td>
<td align="left">20.11</td>
<td align="left">4.84</td>
<td align="left">b</td>
</tr>
<tr class="odd">
<td align="left">b48</td>
<td align="left">19.44</td>
<td align="left">5.07</td>
<td align="left">b</td>
</tr>
<tr class="even">
<td align="left">b49</td>
<td align="left">19.39</td>
<td align="left">5.3</td>
<td align="left">b</td>
</tr>
<tr class="odd">
<td align="left">b50</td>
<td align="left">20.08</td>
<td align="left">4.95</td>
<td align="left">b</td>
</tr>
</tbody>
</table>
<p>从下图可以看出，100个样品根据<code>Gene1</code>和<code>Gene2</code>的表达量的不同在坐标轴上被被分为了2类，可以看做是在平面水平的分类。而且在这个例子中，我们可以很容易的看出<code>Gene1</code>对样品分类的贡献要比<code>Gene2</code>大，因为<code>Gene1</code>在样品间的表达差异大。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(cy_data2,<span class="kw">aes</span>(Gene1, Gene2))+<span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color=</span><span class="kw">factor</span>(Group)))+
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">0.9</span>)) +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.title=</span><span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">ylim</span>(<span class="dv">0</span>,<span class="dv">10</span>) +<span class="st"> </span><span class="kw">xlim</span>(<span class="dv">0</span>,<span class="dv">25</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/jitter2Ddata-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>如果有3个基因呢？</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">count &lt;-<span class="st"> </span><span class="dv">50</span>
Gene3_a &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rnorm</span>(count/<span class="dv">2</span>,<span class="dv">5</span>,<span class="fl">0.2</span>), <span class="kw">rnorm</span>(count/<span class="dv">2</span>,<span class="dv">15</span>,<span class="fl">0.2</span>))
Gene3_b &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">rnorm</span>(count/<span class="dv">2</span>,<span class="dv">15</span>,<span class="fl">0.2</span>), <span class="kw">rnorm</span>(count/<span class="dv">2</span>,<span class="dv">5</span>,<span class="fl">0.2</span>))

data3 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Gene1 =</span> <span class="kw">c</span>(Gene1_a, Gene1_b), <span class="dt">Gene2 =</span> <span class="kw">c</span>(Gene2_a, Gene2_b), 
        <span class="dt">Gene3 =</span> <span class="kw">c</span>(Gene3_a, Gene3_b), <span class="dt">Group=</span><span class="kw">c</span>(grp_a, grp_b))
data3 &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(data3)

<span class="kw">row.names</span>(data3) &lt;-<span class="st"> </span>label

<span class="kw">kable</span>(<span class="kw">headTail</span>(data3), <span class="dt">booktabs=</span>T, <span class="dt">caption=</span><span class="st">&quot;Expression profile for 3 genes in 100 samples&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:datainit3">Table 3.3: </span>Expression profile for 3 genes in 100 samples</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">Gene1</th>
<th align="left">Gene2</th>
<th align="left">Gene3</th>
<th align="left">Group</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">a1</td>
<td align="left">4.71</td>
<td align="left">5.01</td>
<td align="left">4.64</td>
<td align="left">a</td>
</tr>
<tr class="even">
<td align="left">a2</td>
<td align="left">4.66</td>
<td align="left">4.82</td>
<td align="left">5.03</td>
<td align="left">a</td>
</tr>
<tr class="odd">
<td align="left">a3</td>
<td align="left">4.74</td>
<td align="left">4.84</td>
<td align="left">4.93</td>
<td align="left">a</td>
</tr>
<tr class="even">
<td align="left">a4</td>
<td align="left">4.66</td>
<td align="left">5.02</td>
<td align="left">5.26</td>
<td align="left">a</td>
</tr>
<tr class="odd">
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">b47</td>
<td align="left">20.11</td>
<td align="left">4.84</td>
<td align="left">4.83</td>
<td align="left">b</td>
</tr>
<tr class="odd">
<td align="left">b48</td>
<td align="left">19.44</td>
<td align="left">5.07</td>
<td align="left">4.99</td>
<td align="left">b</td>
</tr>
<tr class="even">
<td align="left">b49</td>
<td align="left">19.39</td>
<td align="left">5.3</td>
<td align="left">5.04</td>
<td align="left">b</td>
</tr>
<tr class="odd">
<td align="left">b50</td>
<td align="left">20.08</td>
<td align="left">4.95</td>
<td align="left">5.23</td>
<td align="left">b</td>
</tr>
</tbody>
</table>
<p>从下图可以看出，100个样品根据<code>Gene1</code>、<code>Gene2</code>和<code>Gene3</code>的表达量的不同在坐标轴上被被分为了4类，可以看做是立体空间的分类。而且在这个例子中，我们可以很容易的看出<code>Gene1</code>和<code>Gene3</code>对样品分类的贡献要比<code>Gene2</code>大。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scatterplot3d)
colorl &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#E69F00&quot;</span>, <span class="st">&quot;#56B4E9&quot;</span>)
<span class="co"># Extract same number of colors as the Group and same Group would have same color.</span>
colors &lt;-<span class="st"> </span>colorl[<span class="kw">as.numeric</span>(data3$Group)]
<span class="kw">scatterplot3d</span>(data3[,<span class="dv">1</span>:<span class="dv">3</span>], <span class="dt">color=</span>colors, <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">25</span>), <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">25</span>), <span class="dt">zlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">25</span>), 
    <span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">pch=</span><span class="dv">16</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)</code></pre></div>
<p><img src="R_course_files/figure-html/jitter3Ddata1-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>当我们向由<code>Gene1</code>和<code>Gene2</code>构成的<code>X-Y</code>平面做垂线时，可以很明显的看出，<code>Gene2</code>所在的轴对样品的分类没有贡献。因为投射到<code>X-Y</code>屏幕上的点在<code>Y</code>轴几乎处于同一位置。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scatterplot3d)
colorl &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#E69F00&quot;</span>, <span class="st">&quot;#56B4E9&quot;</span>)
colors &lt;-<span class="st"> </span>colorl[<span class="kw">as.numeric</span>(data3$Group)]
<span class="kw">scatterplot3d</span>(data3[,<span class="dv">1</span>:<span class="dv">3</span>], <span class="dt">color=</span>colors, <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">25</span>), <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">25</span>), <span class="dt">zlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">25</span>),<span class="dt">type=</span><span class="st">&#39;h&#39;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)</code></pre></div>
<p><img src="R_course_files/figure-html/jitter3Ddata2-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>我们把坐标轴做一个转换，可以看到在由<code>Gene1</code>和<code>Gene3</code>构成的<code>X-Y</code>平面上，样品被分为了4类。<code>Gene2</code>对样品的分类几乎没有贡献，因为几乎所有样品在Gene2维度上的值都一样。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scatterplot3d)
colorl &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#E69F00&quot;</span>, <span class="st">&quot;#56B4E9&quot;</span>)
colors &lt;-<span class="st"> </span>colorl[<span class="kw">as.numeric</span>(data3$Group)]
<span class="kw">scatterplot3d</span>(<span class="dt">x=</span>data3$Gene1, <span class="dt">y=</span> data3$Gene3, <span class="dt">z=</span> data3$Gene2, <span class="dt">color=</span>colors, 
        <span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">25</span>), <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">25</span>), <span class="dt">zlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">25</span>),<span class="dt">type=</span><span class="st">&#39;h&#39;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)</code></pre></div>
<p><img src="R_course_files/figure-html/jitter3Ddata3-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>在上述例子中，我们可以很容易的区分出<code>Gene1</code>和<code>Gene3</code>可以作为分类的主成分，而<code>Gene2</code>则对分类没有帮助，可以在计算中去除。</p>
<p>但是如果我们测序了几万个基因的表达时，就很难通过肉眼去看，或者作出一个图供我们筛选哪些基因对样本分类贡献大。这时我们应该怎么做呢？</p>
<p>其中有一个方法是，在这个基因表达矩阵中选出3个变化最大的基因做为3个主成分对样品进行分类。我们试验下效果怎么样。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 数据集来源于 http://satijalab.org/seurat/old-get-started/</span>
<span class="co"># 原始下载链接 http://www.broadinstitute.org/~rahuls/seurat/seurat_files_nbt.zip</span>
<span class="co"># 为了保证文章的使用，文末附有数据的新下载链接，以防原链接失效</span>
data4 &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/HiSeq301-RSEM-linear_values.txt&quot;</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">dim</span>(data4)</code></pre></div>
<pre><code>## [1] 23730   301</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(useful)
<span class="kw">kable</span>(<span class="kw">corner</span>(data4,<span class="dt">r=</span><span class="dv">15</span>,<span class="dt">c=</span><span class="dv">8</span>), <span class="dt">booktabs=</span>T, <span class="dt">caption=</span><span class="st">&quot;Gene expression matrix&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:nbthiseqdata">Table 3.4: </span>Gene expression matrix</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Hi_2338_1</th>
<th align="right">Hi_2338_2</th>
<th align="right">Hi_2338_3</th>
<th align="right">Hi_2338_4</th>
<th align="right">Hi_2338_5</th>
<th align="right">Hi_2338_6</th>
<th align="right">Hi_2338_7</th>
<th align="right">Hi_2338_8</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">A1BG</td>
<td align="right">9.08</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">1.75</td>
<td align="right">0.00</td>
<td align="right">0.40</td>
<td align="right">0.00</td>
<td align="right">0.78</td>
</tr>
<tr class="even">
<td align="left">A1BG-AS1</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">3.47</td>
<td align="right">0.36</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="left">A1CF</td>
<td align="right">0.00</td>
<td align="right">0.05</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td align="left">A2LD1</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.29</td>
<td align="right">0.00</td>
<td align="right">9.19</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="left">A2M</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td align="left">A2M-AS1</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="left">A2ML1</td>
<td align="right">0.10</td>
<td align="right">0.00</td>
<td align="right">0.14</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td align="left">A2MP1</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="left">A4GALT</td>
<td align="right">0.57</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.35</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td align="left">A4GNT</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="left">AA06</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td align="left">AAAS</td>
<td align="right">38.95</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">4.44</td>
<td align="right">0.00</td>
<td align="right">32.90</td>
<td align="right">0.00</td>
<td align="right">5.58</td>
</tr>
<tr class="odd">
<td align="left">AACS</td>
<td align="right">0.12</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.58</td>
<td align="right">1.03</td>
<td align="right">2.16</td>
<td align="right">65.74</td>
</tr>
<tr class="even">
<td align="left">AACSP1</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="left">AADAC</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
</tbody>
</table>
<p>我们筛选变异系数最大的3个基因。在这之前我们先剔除在少于5个样品中表达的基因和少于1000个表达的基因样品 （这里我们把表达值不小于1的基因视为表达的基因），并把所有基因根据其在不同样品中表达值的变异系数排序。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#去除表达值全为0的行</span>
<span class="co">#data4_nonzero &lt;- data4[rowSums(data4)!=0,]</span>

<span class="co">#筛选符合要求的表达的行和列</span>
<span class="co">#data4_use &lt;- data4[apply(data4,1,function(row) sum(row&gt;=1)&gt;=5),]</span>
<span class="co">#data4_use &lt;- data4[,apply(data4,2,function(col) sum(col&gt;=1)&gt;=1000),]</span>
data4_use &lt;-<span class="st"> </span>data4[<span class="kw">rowSums</span>(data4&gt;=<span class="dv">1</span>)&gt;<span class="dv">5</span>,<span class="kw">colSums</span>(data4&gt;=<span class="dv">1</span>)&gt;<span class="dv">1000</span>]

<span class="co"># 对于表达谱数据，因为涉及到PCR的指数扩增，一般会取log处理</span>
<span class="co"># 其它数据log处理会降低数据之间的差异，不一定适用</span>
data4_use_log2 &lt;-<span class="st"> </span><span class="kw">log2</span>(data4_use<span class="dv">+1</span>)

<span class="kw">dim</span>(data4_use_log2)</code></pre></div>
<pre><code>## [1] 16482   301</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 计算变异系数(标准差除以平均值)度量基因表达变化幅度</span>
<span class="co">#cv &lt;- apply(data4_use_log2,1,sd)/rowMeans(data4_use_log2)</span>
<span class="co"># 根据变异系数排序</span>
<span class="co">#data4_use_log2 &lt;- data4_use_log2[order(cv,decreasing = T),]</span>

<span class="co"># 计算中值绝对偏差 (MAD, median absolute deviation)度量基因表达变化幅度</span>
<span class="co"># 在基因表达中，尽管某些基因很小的变化会导致重要的生物学意义，</span>
<span class="co"># 但是很小的观察值会引入很大的背景噪音，因此也意义不大。</span>
mads &lt;-<span class="st"> </span><span class="kw">apply</span>(data4_use_log2, <span class="dv">1</span>, mad)
data4_use_log2 &lt;-<span class="st"> </span>data4_use_log2[<span class="kw">rev</span>(<span class="kw">order</span>(mads)),]

<span class="co">#筛选前3行</span>
data_var3 &lt;-<span class="st"> </span>data4_use_log2[<span class="dv">1</span>:<span class="dv">3</span>,]

<span class="co"># 转置矩阵使得每一行为一个样品，每一列为一组变量</span>
data_var3_forPCA &lt;-<span class="st"> </span><span class="kw">t</span>(data_var3)

<span class="kw">dim</span>(data_var3_forPCA)</code></pre></div>
<pre><code>## [1] 301   3</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">kable</span>(<span class="kw">corner</span>(data_var3_forPCA, <span class="dt">r=</span><span class="dv">10</span>,<span class="dt">c=</span><span class="dv">5</span>), <span class="dt">booktabs=</span><span class="ot">TRUE</span>, 
        <span class="dt">caption=</span><span class="st">&quot;A table of the 3 most variable genes&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:varsel">Table 3.5: </span>A table of the 3 most variable genes</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">MT2A</th>
<th align="right">ANXA1</th>
<th align="right">ARHGDIB</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Hi_2338_1</td>
<td align="right">12.211493</td>
<td align="right">11.837198</td>
<td align="right">9.543283</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_2</td>
<td align="right">11.306216</td>
<td align="right">8.098769</td>
<td align="right">8.071623</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_3</td>
<td align="right">11.926226</td>
<td align="right">10.688626</td>
<td align="right">9.720535</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_4</td>
<td align="right">10.974207</td>
<td align="right">9.386574</td>
<td align="right">9.883376</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_5</td>
<td align="right">14.603994</td>
<td align="right">10.375072</td>
<td align="right">9.970379</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_6</td>
<td align="right">6.904604</td>
<td align="right">11.155349</td>
<td align="right">10.093510</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_7</td>
<td align="right">12.436719</td>
<td align="right">10.852249</td>
<td align="right">7.742882</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_8</td>
<td align="right">9.798375</td>
<td align="right">9.783227</td>
<td align="right">6.270716</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_9</td>
<td align="right">11.743673</td>
<td align="right">9.626476</td>
<td align="right">9.250251</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_10</td>
<td align="right">11.240016</td>
<td align="right">11.303056</td>
<td align="right">0.000000</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 获得样品分组信息</span>
sample &lt;-<span class="st"> </span><span class="kw">rownames</span>(data_var3_forPCA)

<span class="co"># 把样品名字按 &lt;_&gt; 分割，取出其第二部分作为样品的组名</span>
<span class="co"># lapply(X, FUC) 对列表或向量中每个元素执行FUC操作，FUNC为自定义或R自带的函数</span>
## One better way to generate group
group &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(<span class="kw">strsplit</span>(sample, <span class="st">&quot;_&quot;</span>), function(x) x[<span class="dv">2</span>]))

##One way to generate group
<span class="co">#sample_split &lt;- strsplit(sample,&quot;_&quot;)</span>
<span class="co">#group &lt;- matrix(unlist(sample_split), ncol=3, byrow=T)[,2]</span>
<span class="kw">print</span>(sample[<span class="dv">1</span>:<span class="dv">4</span>])</code></pre></div>
<pre><code>## [1] &quot;Hi_2338_1&quot; &quot;Hi_2338_2&quot; &quot;Hi_2338_3&quot; &quot;Hi_2338_4&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(group[<span class="dv">1</span>:<span class="dv">4</span>])</code></pre></div>
<pre><code>## [1] &quot;2338&quot; &quot;2338&quot; &quot;2338&quot; &quot;2338&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_var3_scatter &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(data_var3_forPCA)
data_var3_scatter$group &lt;-<span class="st"> </span>group
<span class="kw">kable</span>(<span class="kw">corner</span>(data_var3_scatter, <span class="dt">r=</span><span class="dv">10</span>,<span class="dt">c=</span><span class="dv">5</span>), <span class="dt">booktabs=</span><span class="ot">TRUE</span>, 
    <span class="dt">caption=</span><span class="st">&quot;A table of the 3 most variable genes&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:addgrp">Table 3.6: </span>A table of the 3 most variable genes</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">MT2A</th>
<th align="right">ANXA1</th>
<th align="right">ARHGDIB</th>
<th align="left">group</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Hi_2338_1</td>
<td align="right">12.211493</td>
<td align="right">11.837198</td>
<td align="right">9.543283</td>
<td align="left">2338</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_2</td>
<td align="right">11.306216</td>
<td align="right">8.098769</td>
<td align="right">8.071623</td>
<td align="left">2338</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_3</td>
<td align="right">11.926226</td>
<td align="right">10.688626</td>
<td align="right">9.720535</td>
<td align="left">2338</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_4</td>
<td align="right">10.974207</td>
<td align="right">9.386574</td>
<td align="right">9.883376</td>
<td align="left">2338</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_5</td>
<td align="right">14.603994</td>
<td align="right">10.375072</td>
<td align="right">9.970379</td>
<td align="left">2338</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_6</td>
<td align="right">6.904604</td>
<td align="right">11.155349</td>
<td align="right">10.093510</td>
<td align="left">2338</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_7</td>
<td align="right">12.436719</td>
<td align="right">10.852249</td>
<td align="right">7.742882</td>
<td align="left">2338</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_8</td>
<td align="right">9.798375</td>
<td align="right">9.783227</td>
<td align="right">6.270716</td>
<td align="left">2338</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_9</td>
<td align="right">11.743673</td>
<td align="right">9.626476</td>
<td align="right">9.250251</td>
<td align="left">2338</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_10</td>
<td align="right">11.240016</td>
<td align="right">11.303056</td>
<td align="right">0.000000</td>
<td align="left">2338</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(reshape2)
<span class="kw">library</span>(ggplot2)
data_var3_melt &lt;-<span class="st"> </span><span class="kw">melt</span>(data_var3_scatter, <span class="dt">id.vars=</span><span class="kw">c</span>(<span class="st">&quot;group&quot;</span>))
<span class="kw">kable</span>(<span class="kw">corner</span>(data_var3_melt, <span class="dt">r=</span><span class="dv">10</span>,<span class="dt">c=</span><span class="dv">5</span>), <span class="dt">booktabs=</span><span class="ot">TRUE</span>, 
        <span class="dt">caption=</span><span class="st">&quot;A table of the 3 most variable genes in melted format&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:quasimPLot">Table 3.7: </span>A table of the 3 most variable genes in melted format</caption>
<thead>
<tr class="header">
<th align="left">group</th>
<th align="left">variable</th>
<th align="right">value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">12.211493</td>
</tr>
<tr class="even">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">11.306216</td>
</tr>
<tr class="odd">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">11.926226</td>
</tr>
<tr class="even">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">10.974207</td>
</tr>
<tr class="odd">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">14.603994</td>
</tr>
<tr class="even">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">6.904604</td>
</tr>
<tr class="odd">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">12.436719</td>
</tr>
<tr class="even">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">9.798375</td>
</tr>
<tr class="odd">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">11.743673</td>
</tr>
<tr class="even">
<td align="left">2338</td>
<td align="left">MT2A</td>
<td align="right">11.240016</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(data_var3_melt, <span class="kw">aes</span>(<span class="kw">factor</span>(variable),value))+<span class="kw">ylab</span>(<span class="st">&quot;Gene expression&quot;</span>)+
<span class="st">    </span><span class="kw">geom_violin</span>(<span class="kw">aes</span>(<span class="dt">fill=</span><span class="kw">factor</span>(group)), <span class="dt">stat=</span><span class="st">&quot;ydensity&quot;</span>, <span class="dt">position=</span><span class="st">&quot;dodge&quot;</span>,
            <span class="dt">scale=</span><span class="st">&quot;width&quot;</span>, <span class="dt">trim=</span><span class="ot">TRUE</span>) +<span class="kw">xlab</span>(<span class="ot">NULL</span>)

<span class="co">#ggplot(data_var3_melt, aes(factor(variable),value))+ylab(&quot;Gene expression&quot;)+ </span>
    <span class="co">#geom_quasirandom(aes(color=factor(group))) +xlab(NULL)</span></code></pre></div>
<p><img src="R_course_files/figure-html/quasimPLot-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 根据分组数目确定颜色变量</span>
colorA &lt;-<span class="st"> </span><span class="kw">rainbow</span>(<span class="kw">length</span>(<span class="kw">unique</span>(group)))

<span class="co"># 根据每个样品的分组信息获取对应的颜色变量</span>
colors &lt;-<span class="st"> </span>colorA[<span class="kw">as.factor</span>(group)]

<span class="co"># 根据样品分组信息获得legend的颜色</span>
colorl &lt;-<span class="st"> </span>colorA[<span class="kw">as.factor</span>(<span class="kw">unique</span>(group))]

<span class="co"># 获得PCH symbol列表</span>
pch_l &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">as.factor</span>(<span class="kw">unique</span>(group)))
<span class="co"># 产生每个样品的pch symbol</span>
pch &lt;-<span class="st"> </span>pch_l[<span class="kw">as.factor</span>(group)]

<span class="kw">scatterplot3d</span>(data_var3_forPCA[,<span class="dv">1</span>:<span class="dv">3</span>], <span class="dt">color=</span>colors, <span class="dt">pch=</span>pch)
<span class="kw">legend</span>(<span class="dv">0</span>,<span class="dv">10</span>, <span class="dt">legend=</span><span class="kw">levels</span>(<span class="kw">as.factor</span>(group)), <span class="dt">col=</span>colorl, <span class="dt">pch=</span>pch_l, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>F, <span class="dt">ncol=</span><span class="dv">6</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/p3dvar1-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>我们看到图中的样品并没有按照预先设定的标签完全分开。当然我们也可以通过其他方法筛选变异最大的三个基因，最终的分类效果不会相差很大。因为不管怎么筛选，我们都只用到了3个基因的表达量。</p>
<p>假如我们把这个数据用PCA来分类，结果是怎样的呢？</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Pay attention to the format of PCA input </span>
<span class="co"># Rows are samples and columns are variables</span>
data4_use_log2_t &lt;-<span class="st"> </span><span class="kw">t</span>(data4_use_log2)

<span class="co"># Add group column for plotting</span>
data4_use_log2_label &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(data4_use_log2_t)
data4_use_log2_label$group &lt;-<span class="st"> </span>group

<span class="co"># By default, prcomp will centralized the data using mean.</span>
<span class="co"># Normalize data for PCA by dividing each data by column standard deviation.</span>
<span class="co"># Often, we would normalize data.</span>
<span class="co"># Only when we care about the real number changes other than the trends,</span>
<span class="co"># `scale` can be set to TRUE. </span>
<span class="co"># We will show the differences of scaling and un-scaling effects.</span>
pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data4_use_log2_t, <span class="dt">scale=</span>T)

<span class="co"># sdev: standard deviation of the principle components.</span>
<span class="co"># Square to get variance</span>
percentVar &lt;-<span class="st"> </span>pca$sdev^<span class="dv">2</span> /<span class="st"> </span><span class="kw">sum</span>( pca$sdev^<span class="dv">2</span>)

<span class="co"># To check what&#39;s in pca</span>
<span class="kw">print</span>(<span class="kw">str</span>(pca))</code></pre></div>
<pre><code>## List of 5
##  $ sdev    : num [1:301] 36.6 30.4 23.3 21.6 19.9 ...
##  $ rotation: num [1:16482, 1:301] -0.01133 -0.01955 -0.00199 -0.00569 -0.0204 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:16482] &quot;MT2A&quot; &quot;ANXA1&quot; &quot;ARHGDIB&quot; &quot;RND3&quot; ...
##   .. ..$ : chr [1:301] &quot;PC1&quot; &quot;PC2&quot; &quot;PC3&quot; &quot;PC4&quot; ...
##  $ center  : Named num [1:16482] 6.5 5.47 5.43 4.23 4.1 ...
##   ..- attr(*, &quot;names&quot;)= chr [1:16482] &quot;MT2A&quot; &quot;ANXA1&quot; &quot;ARHGDIB&quot; &quot;RND3&quot; ...
##  $ scale   : Named num [1:16482] 5.62 4.96 4.78 4.13 3.98 ...
##   ..- attr(*, &quot;names&quot;)= chr [1:16482] &quot;MT2A&quot; &quot;ANXA1&quot; &quot;ARHGDIB&quot; &quot;RND3&quot; ...
##  $ x       : num [1:301, 1:301] -37.6 -28.6 -30.6 -43.7 -11.4 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:301] &quot;Hi_2338_1&quot; &quot;Hi_2338_2&quot; &quot;Hi_2338_3&quot; &quot;Hi_2338_4&quot; ...
##   .. ..$ : chr [1:301] &quot;PC1&quot; &quot;PC2&quot; &quot;PC3&quot; &quot;PC4&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;prcomp&quot;
## NULL</code></pre>
<p>从图中可以看到，数据呈现了一定的分类模式 (当然这个分类结果也不理想，我们随后再进一步优化)。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggfortify)
<span class="kw">autoplot</span>(pca, <span class="dt">data=</span>data4_use_log2_label, <span class="dt">colour=</span><span class="st">&quot;group&quot;</span>) +<span class="st"> </span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC1 (&quot;</span>, <span class="kw">round</span>(percentVar[<span class="dv">1</span>]*<span class="dv">100</span>), <span class="st">&quot;% variance)&quot;</span>)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC2 (&quot;</span>, <span class="kw">round</span>(percentVar[<span class="dv">2</span>]*<span class="dv">100</span>), <span class="st">&quot;% variance)&quot;</span>)) +<span class="st"> </span><span class="kw">theme_bw</span>() +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(), <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>()) +<span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;right&quot;</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/pcaclusterplot-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>采用3个主成分获得的分类效果优于2个主成分，因为这样保留的原始信息更多。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 根据分组数目确定颜色变量</span>
colorA &lt;-<span class="st"> </span><span class="kw">rainbow</span>(<span class="kw">length</span>(<span class="kw">unique</span>(group)))

<span class="co"># 根据每个样品的分组信息获取对应的颜色变量</span>
colors &lt;-<span class="st"> </span>colorA[<span class="kw">as.factor</span>(group)]

<span class="co"># 根据样品分组信息获得legend的颜色</span>
colorl &lt;-<span class="st"> </span>colorA[<span class="kw">as.factor</span>(<span class="kw">unique</span>(group))]

<span class="co"># 获得PCH symbol列表</span>
pch_l &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">as.factor</span>(<span class="kw">unique</span>(group)))
<span class="co"># 产生每个样品的pch symbol</span>
pch &lt;-<span class="st"> </span>pch_l[<span class="kw">as.factor</span>(group)]

pc &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(pca$x)
<span class="kw">scatterplot3d</span>(<span class="dt">x=</span>pc$PC1, <span class="dt">y=</span>pc$PC2, <span class="dt">z=</span>pc$PC3, <span class="dt">pch=</span>pch, <span class="dt">color=</span>colors, 
    <span class="dt">xlab=</span><span class="kw">paste0</span>(<span class="st">&quot;PC1 (&quot;</span>, <span class="kw">round</span>(percentVar[<span class="dv">1</span>]*<span class="dv">100</span>), <span class="st">&quot;% variance)&quot;</span>), 
    <span class="dt">ylab=</span><span class="kw">paste0</span>(<span class="st">&quot;PC2 (&quot;</span>, <span class="kw">round</span>(percentVar[<span class="dv">2</span>]*<span class="dv">100</span>), <span class="st">&quot;% variance)&quot;</span>), 
    <span class="dt">zlab=</span><span class="kw">paste0</span>(<span class="st">&quot;PC3 (&quot;</span>, <span class="kw">round</span>(percentVar[<span class="dv">3</span>]*<span class="dv">100</span>), <span class="st">&quot;% variance)&quot;</span>))

<span class="kw">legend</span>(-<span class="dv">3</span>,<span class="dv">8</span>, <span class="dt">legend=</span><span class="kw">levels</span>(<span class="kw">as.factor</span>(group)), <span class="dt">col=</span>colorl, <span class="dt">pch=</span>pch_l, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>F, <span class="dt">ncol=</span><span class="dv">6</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/pcaclusterplot3d-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="pca" class="section level3">
<h3><span class="header-section-number">3.12.4</span> PCA的实现原理</h3>
<p>在上面的例子中，PCA分析不是简单地选取2个或3个变化最大的基因，而是先把原始的变量做一个评估，计算各个变量各自的变异度(方差)和两两变量的相关度（协方差），得到一个协方差矩阵。在这个协方差矩阵中，对角线的值为每一个变量的方差，其它值为每两个变量的协方差。随后对原变量的协方差矩阵对角化处理，即求解其特征值和特征向量。原变量与特征向量的乘积（对原始变量的线性组合）即为新变量（回顾下线性代数中的矩阵乘法）；新变量的协方差矩阵为对角协方差矩阵且对角线上的方差由大到小排列；然后从新变量中选择信息最丰富也就是方差最大的的前2个或前3个新变量也就是主成分用以可视化。下面我们一步步阐释这是怎么做的。</p>
<p>我们先回忆两个数学概念，方差和协方差。<code>方差</code>用来表示一组一维数据的离散程度。<code>协方差</code>表示2组一维数据的相关性。当协方差为0时，表示两组数据完全独立。当协方差为正时，表示一组数据增加时另外一组也会增加；当协方差为负时表示一组数据增加时另外一组数据会降低 （与相关系数类似）。如果我们有很多组一维数据，比如很多基因的表达数据，就会得到很多协方差，这就构成了协方差矩阵。</p>
<p>方差和协方差的计算公式如下：</p>
<p><span class="math display">\[
方差 Var(X) = \frac{\sum_{i=1}^{n}(X_i-\bar X)^2}{n-1}
\]</span></p>
<p><span class="math display">\[
协方差 cov(X,Y) = \frac{\sum_{i=1}^{n}(X_i-\bar X)(Y_i-\bar Y)}{n-1}
\]</span></p>
<p>如果数据的均值为0，这个公式可以进一步简化。简化后的公式把计算协方差转变为了矩阵乘法运算。这也是为什么PCA需要中心化数据。</p>
<p><span class="math display">\[
方差 Var(X) = \frac{\sum_{i=1}^{n}{X_i}^2}{n-1}
\]</span></p>
<p><span class="math display">\[
协方差 cov(X,Y) = \frac{\sum_{i=1}^{n}X_iY_i}{n-1}
\]</span></p>
<p><span class="math display">\[
协方差矩阵 cov(X,Y) = \frac{X_{n,m}^T Y_{n,m}}{n-1}
\]</span></p>
<p>假如我们有一个矩阵如下，</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">20</span>,<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">nrow=</span><span class="dv">4</span>))
<span class="kw">colnames</span>(mat) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;Gene_&quot;</span>, letters[<span class="dv">1</span>:<span class="dv">5</span>])
<span class="kw">rownames</span>(mat) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;Samp_&quot;</span>, <span class="dv">1</span>:<span class="dv">4</span>)
mat</code></pre></div>
<pre><code>##             Gene_a     Gene_b     Gene_c      Gene_d      Gene_e
## Samp_1 -0.95854605 -1.4641695  1.4118119 -1.87124035  2.26763879
## Samp_2 -1.07372321  0.2325534  0.3048656 -1.22271308 -0.02195949
## Samp_3  0.27186238 -0.4106860  0.9816736  0.09329151 -0.17824977
## Samp_4  0.05494982  0.1682696 -1.6447469 -0.73149523  0.62384560</code></pre>
<p>平均值中心化 (mean centering)：中心化数据使其平均值为0</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># mean-centering data for columns</span>
<span class="co"># Get mean-value matrix first</span>
mat_mean_norm &lt;-<span class="st"> </span>mat -<span class="st"> </span><span class="kw">rep</span>(<span class="kw">colMeans</span>(mat),<span class="kw">rep.int</span>(<span class="kw">nrow</span>(mat),<span class="kw">ncol</span>(mat)))
mat_mean_norm</code></pre></div>
<pre><code>##            Gene_a      Gene_b      Gene_c     Gene_d      Gene_e
## Samp_1 -0.5321818 -1.09566138  1.14841086 -0.9382011  1.59482000
## Samp_2 -0.6473589  0.60106152  0.04146455 -0.2896738 -0.69477827
## Samp_3  0.6982266 -0.04217784  0.71827253  1.0263308 -0.85106855
## Samp_4  0.4813141  0.53677770 -1.90814794  0.2015441 -0.04897318</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># mean-centering using scale for columns</span>
<span class="kw">scale</span>(mat, <span class="dt">center=</span>T, <span class="dt">scale=</span>F)</code></pre></div>
<pre><code>##            Gene_a      Gene_b      Gene_c     Gene_d      Gene_e
## Samp_1 -0.5321818 -1.09566138  1.14841086 -0.9382011  1.59482000
## Samp_2 -0.6473589  0.60106152  0.04146455 -0.2896738 -0.69477827
## Samp_3  0.6982266 -0.04217784  0.71827253  1.0263308 -0.85106855
## Samp_4  0.4813141  0.53677770 -1.90814794  0.2015441 -0.04897318
## attr(,&quot;scaled:center&quot;)
##     Gene_a     Gene_b     Gene_c     Gene_d     Gene_e 
## -0.4263643 -0.3685081  0.2634011 -0.9330393  0.6728188</code></pre>
<p>中位数中心化 (median centering)：如果数据变换范围很大或有异常值，中位数标准化效果会更好。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># median-centering data for columns</span>
mat_median_norm &lt;-<span class="st"> </span>mat -<span class="st"> </span><span class="kw">rep</span>(<span class="kw">apply</span>(mat,<span class="dv">2</span>,median),<span class="kw">rep.int</span>(<span class="kw">nrow</span>(mat),<span class="kw">ncol</span>(mat)))
mat_mean_norm</code></pre></div>
<pre><code>##            Gene_a      Gene_b      Gene_c     Gene_d      Gene_e
## Samp_1 -0.5321818 -1.09566138  1.14841086 -0.9382011  1.59482000
## Samp_2 -0.6473589  0.60106152  0.04146455 -0.2896738 -0.69477827
## Samp_3  0.6982266 -0.04217784  0.71827253  1.0263308 -0.85106855
## Samp_4  0.4813141  0.53677770 -1.90814794  0.2015441 -0.04897318</code></pre>
<p>我们可以计算<code>Gene_a</code>的方差为 0.4738249 (<code>var(mat$Gene_a)</code>)；<code>Gene_a</code>和<code>Gene_b</code>的协方差为0.1409658。</p>
<p>mat中5组基因的表达值的方差计算如下：</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">apply</span>(mat,<span class="dv">2</span>,var)</code></pre></div>
<pre><code>##    Gene_a    Gene_b    Gene_c    Gene_d    Gene_e 
## 0.4738249 0.6172194 1.8258369 0.6860357 1.2509613</code></pre>
<p>mat中5组基因表达值的协方差计算如下：</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cov</span>(mat)</code></pre></div>
<pre><code>##            Gene_a     Gene_b     Gene_c     Gene_d     Gene_e
## Gene_a  0.4738249  0.1409658 -0.3516357  0.5001446 -0.3389245
## Gene_b  0.1409658  0.6172194 -0.7626310  0.3062449 -0.7184596
## Gene_c -0.3516357 -0.7626310  1.8258369 -0.2456140  0.4282830
## Gene_d  0.5001446  0.3062449 -0.2456140  0.6860357 -0.7261170
## Gene_e -0.3389245 -0.7184596  0.4282830 -0.7261170  1.2509613</code></pre>
<p>如果均值为0，数值矩阵的协方差矩阵为矩阵的乘积 （实际上是矩阵的转置与其本身的乘积除以变量的维数减1）。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Covariance matrix for Mean normalized matrix</span>
<span class="kw">cov</span>(mat_mean_norm)</code></pre></div>
<pre><code>##            Gene_a     Gene_b     Gene_c     Gene_d     Gene_e
## Gene_a  0.4738249  0.1409658 -0.3516357  0.5001446 -0.3389245
## Gene_b  0.1409658  0.6172194 -0.7626310  0.3062449 -0.7184596
## Gene_c -0.3516357 -0.7626310  1.8258369 -0.2456140  0.4282830
## Gene_d  0.5001446  0.3062449 -0.2456140  0.6860357 -0.7261170
## Gene_e -0.3389245 -0.7184596  0.4282830 -0.7261170  1.2509613</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Covariance matrix for Mean normalized matrix </span>
<span class="co"># crossprod: matrix multiplication</span>
<span class="kw">crossprod</span>(<span class="kw">as.matrix</span>(mat_mean_norm)) /<span class="st"> </span>(<span class="kw">nrow</span>(mat_mean_norm)-<span class="dv">1</span>)</code></pre></div>
<pre><code>##            Gene_a     Gene_b     Gene_c     Gene_d     Gene_e
## Gene_a  0.4738249  0.1409658 -0.3516357  0.5001446 -0.3389245
## Gene_b  0.1409658  0.6172194 -0.7626310  0.3062449 -0.7184596
## Gene_c -0.3516357 -0.7626310  1.8258369 -0.2456140  0.4282830
## Gene_d  0.5001446  0.3062449 -0.2456140  0.6860357 -0.7261170
## Gene_e -0.3389245 -0.7184596  0.4282830 -0.7261170  1.2509613</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Use %*% for matrix multiplication (slower)</span>
<span class="kw">t</span>(<span class="kw">as.matrix</span>(mat_mean_norm)) %*%<span class="st"> </span><span class="kw">as.matrix</span>(mat_mean_norm) /<span class="st"> </span>(<span class="kw">nrow</span>(mat_mean_norm)-<span class="dv">1</span>)</code></pre></div>
<pre><code>##            Gene_a     Gene_b     Gene_c     Gene_d     Gene_e
## Gene_a  0.4738249  0.1409658 -0.3516357  0.5001446 -0.3389245
## Gene_b  0.1409658  0.6172194 -0.7626310  0.3062449 -0.7184596
## Gene_c -0.3516357 -0.7626310  1.8258369 -0.2456140  0.4282830
## Gene_d  0.5001446  0.3062449 -0.2456140  0.6860357 -0.7261170
## Gene_e -0.3389245 -0.7184596  0.4282830 -0.7261170  1.2509613</code></pre>
<p>用矩阵形式书写如下，便于理解</p>
<p><span class="math display">\[
\mathbf{cov(mat)} = \frac{1}{3} \left[\begin{array} {rrrr} -0.53 &amp; -0.65 &amp; 0.7 &amp; 0.48 \\
 -1.1 &amp; 0.6 &amp; -0.04 &amp; 0.54 \\
 1.15 &amp; 0.04 &amp; 0.72 &amp; -1.91 \\
 -0.94 &amp; -0.29 &amp; 1.03 &amp; 0.2 \\
 1.59 &amp; -0.69 &amp; -0.85 &amp; -0.05 \\
 \end{array}\right] \left[\begin{array} {rrrrr} -0.53 &amp; -1.1 &amp; 1.15 &amp; -0.94 &amp; 1.59 \\
 -0.65 &amp; 0.6 &amp; 0.04 &amp; -0.29 &amp; -0.69 \\
 0.7 &amp; -0.04 &amp; 0.72 &amp; 1.03 &amp; -0.85 \\
 0.48 &amp; 0.54 &amp; -1.91 &amp; 0.2 &amp; -0.05 \\
 \end{array}\right] = \left[\begin{array} {rrrrr} 0.47 &amp; 0.14 &amp; -0.35 &amp; 0.5 &amp; -0.34 \\
 0.14 &amp; 0.62 &amp; -0.76 &amp; 0.31 &amp; -0.72 \\
 -0.35 &amp; -0.76 &amp; 1.83 &amp; -0.25 &amp; 0.43 \\
 0.5 &amp; 0.31 &amp; -0.25 &amp; 0.69 &amp; -0.73 \\
 -0.34 &amp; -0.72 &amp; 0.43 &amp; -0.73 &amp; 1.25 \\
 \end{array}\right] 
\]</span></p>
<p>根据前面的描述，原始变量的协方差矩阵表示原始变量自身的方差（协方差矩阵的主对角线位置）和原始变量之间的相关程度(非主对角线位置)。如果从这些数据中筛选主成分，则要选择方差大(主对角线值大)，且与其它已选变量之间相关性最小的变量（非主对角线值很小）。如果这些原始变量之间毫不相关，则它们的协方差矩阵在除主对角线处外其它地方的值都为0，这种矩阵成为对角矩阵。</p>
<p>而做PCA分析就是产生一组新的变量，使得新变量的协方差矩阵为对角阵，满足上面的要求。从而达到去冗余的目的。然后再选取方差大的变量，实现降维和去噪。</p>
<p>如果正向推导，这种组合可能会有很多种，一一计算会比较麻烦。那反过来看呢？ 我们不去寻找这种组合，而是计算如何使原变量的协方差矩阵变为对角阵。</p>
<p>数学推导中谨记的两个概念：</p>
<ol style="list-style-type: decimal">
<li><strong>假设</strong>: 把未求解到的变量假设出来，用符号代替；这样有助于思考和演算</li>
<li><strong>逆向</strong>：如果正向推导求不出，不妨倒着来；尽量多的利用已有信息</li>
</ol>
<p>前面提到，新变量(<span class="math inline">\(Y_{m,k}\)</span>)是原始变量(<span class="math inline">\(X_{m,n}\)</span>)(原始变量的协方差矩阵为(<span class="math inline">\(C_{n,n}\)</span>))的线性组合，那么<strong>假设</strong>我们找到了这么一个线性组合(命名为特征矩阵(<span class="math inline">\(P_{n,k}\)</span>))，得到一组新变量<span class="math inline">\(Y_{m,k}=X_{m,n}P_{n,k}\)</span>，并且新变量的协方差矩阵(<span class="math inline">\(D_{k,k}\)</span>)为对角阵。那么这个特征矩阵(<span class="math inline">\(P_{n,k}\)</span>)需要符合什么条件呢？</p>
<p>从矩阵运算可以看出，最终的特征矩阵(<span class="math inline">\(P_{n,k}\)</span>)需要把原变量协方差矩阵(<span class="math inline">\(C_{n,n}\)</span>)转换为对角阵(因为新变量的协方差矩阵(<span class="math inline">\(D_{k,k}\)</span>)为对角阵)，并且对角元素从大到小排列（保证每个主成分的贡献度依次降低）。</p>
<p>现在就把求解新变量的任务转变为了求解原变量协方差矩阵的对角化问题了。在线性代数中，矩阵对角化的问题就是求解矩阵的特征值和特征向量的问题。</p>
<p>我们举一个例子讲述怎么求解特征值和特征向量。</p>
<p>假设<span class="math inline">\(A_{n,n}\)</span>为n阶对称阵，如存在<span class="math inline">\(\lambda\)</span>和非零向量<span class="math inline">\(x\)</span>，使得<span class="math inline">\(Ax=\lambda x\)</span>，则称<span class="math inline">\(\lambda\)</span>为矩阵<span class="math inline">\(A_{n,n}\)</span>的特征值，非零向量<span class="math inline">\(x\)</span>为为矩阵<span class="math inline">\(A_{n,n}\)</span>对应于特征值<span class="math inline">\(\lambda\)</span>的特征向量。</p>
<p>根据这个定义可以得出<span class="math inline">\((A_{n,n} - \lambda E)x = 0\)</span>，由于<span class="math inline">\(x\)</span>为非零向量，所以行列式<span class="math inline">\(|A-\lambda E| = 0\)</span>。</p>
<p>由此求解出n个根<span class="math inline">\(\lambda_{1}, \lambda_{2}, ..., \lambda_{3}\)</span>就是矩阵<span class="math inline">\(A\)</span>的特征值。</p>
<p>回顾下行列式的计算：</p>
<ul>
<li>行列式的值为行列式第一列的每一个数乘以它的余子式（余子式是行列式中除去当前元素所在行和列之后剩下的行列式）。</li>
<li>当行列式中存在线性相关的行或列或者有一行或一列元素全为0时，行列式的值为0。</li>
<li>上三角形行列式的值为其主对角线上元素的乘积。</li>
<li>互换行列式的两行或两列，行列式变号。</li>
<li>行列式的某一列（行）乘以同意书加到另一列（列）对应元素上去，行列式不变。</li>
</ul>
<p>假如我们有一个矩阵 <span class="math inline">\(\mathbf{A} = \left[\begin{array} {cc} 3 &amp; -1 \\ -1 &amp; 3 \end{array}\right]\)</span>，如何计算它的特征值和特征向量呢？</p>
<p>则<span class="math inline">\(\lambda\)</span>的值为2或4。</p>
<p>对<span class="math inline">\(\lambda_{1}=2\)</span>时，求解<span class="math inline">\((A-2E)x=\left|\begin{array} {cc} 1 &amp; -1 \\ -1 &amp; 1 \end{array}\right| x= 0\)</span>，得<span class="math inline">\(x=k\left|\begin{array}{c} 1 \\ 1 \end{array}\right|\)</span>，则对应于<span class="math inline">\(\lambda_{1}=2\)</span>时的特征向量<span class="math inline">\(p_{1}=\left|\begin{array}{c} 1 \\ 1 \end{array}\right|\)</span></p>
<p>对<span class="math inline">\(\lambda_{2}=4\)</span>时，求解<span class="math inline">\((A-2E)x=\left|\begin{array} {cc} -1 &amp; -1 \\ -1 &amp; -1 \end{array}\right| x= 0\)</span>，得<span class="math inline">\(x=k\left|\begin{array}{c} 1 \\ -1 \end{array}\right|\)</span>，则对应于<span class="math inline">\(\lambda_{2}=4\)</span>时的特征向量<span class="math inline">\(p_{2}=\left|\begin{array}{c} 1 \\ -1 \end{array}\right|\)</span></p>
<p>以上就完成了PCA的数学推导。</p>
</div>
<div id="pca" class="section level3">
<h3><span class="header-section-number">3.12.5</span> 简单的PCA实现</h3>
<p>我们使用前面用到的数据<code>data3</code>来演示下如何用R函数实现PCA的计算，并与R中自带的<code>prcomp</code>做个比较。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(knitr)
<span class="kw">kable</span>(<span class="kw">headTail</span>(data3), <span class="dt">booktabs=</span>T, <span class="dt">caption=</span><span class="st">&quot;Expression profile for 3 genes in 100 samples&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:showdata3">Table 3.8: </span>Expression profile for 3 genes in 100 samples</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">Gene1</th>
<th align="left">Gene2</th>
<th align="left">Gene3</th>
<th align="left">Group</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">a1</td>
<td align="left">4.71</td>
<td align="left">5.01</td>
<td align="left">4.64</td>
<td align="left">a</td>
</tr>
<tr class="even">
<td align="left">a2</td>
<td align="left">4.66</td>
<td align="left">4.82</td>
<td align="left">5.03</td>
<td align="left">a</td>
</tr>
<tr class="odd">
<td align="left">a3</td>
<td align="left">4.74</td>
<td align="left">4.84</td>
<td align="left">4.93</td>
<td align="left">a</td>
</tr>
<tr class="even">
<td align="left">a4</td>
<td align="left">4.66</td>
<td align="left">5.02</td>
<td align="left">5.26</td>
<td align="left">a</td>
</tr>
<tr class="odd">
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">NA</td>
</tr>
<tr class="even">
<td align="left">b47</td>
<td align="left">20.11</td>
<td align="left">4.84</td>
<td align="left">4.83</td>
<td align="left">b</td>
</tr>
<tr class="odd">
<td align="left">b48</td>
<td align="left">19.44</td>
<td align="left">5.07</td>
<td align="left">4.99</td>
<td align="left">b</td>
</tr>
<tr class="even">
<td align="left">b49</td>
<td align="left">19.39</td>
<td align="left">5.3</td>
<td align="left">5.04</td>
<td align="left">b</td>
</tr>
<tr class="odd">
<td align="left">b50</td>
<td align="left">20.08</td>
<td align="left">4.95</td>
<td align="left">5.23</td>
<td align="left">b</td>
</tr>
</tbody>
</table>
<p>标准化数据</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data3_center_scale &lt;-<span class="st"> </span><span class="kw">scale</span>(data3[,<span class="dv">1</span>:<span class="dv">3</span>], <span class="dt">center=</span>T, <span class="dt">scale=</span>T)
<span class="kw">kable</span>(<span class="kw">headTail</span>(data3_center_scale), <span class="dt">booktabs=</span>T, 
        <span class="dt">caption=</span><span class="st">&quot;Normalized expression for 3 genes in 100 samples&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:scaledata3">Table 3.9: </span>Normalized expression for 3 genes in 100 samples</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">Gene1</th>
<th align="left">Gene2</th>
<th align="left">Gene3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">a1</td>
<td align="left">-1.04</td>
<td align="left">0.14</td>
<td align="left">-1.06</td>
</tr>
<tr class="even">
<td align="left">a2</td>
<td align="left">-1.04</td>
<td align="left">-0.8</td>
<td align="left">-0.98</td>
</tr>
<tr class="odd">
<td align="left">a3</td>
<td align="left">-1.03</td>
<td align="left">-0.73</td>
<td align="left">-1</td>
</tr>
<tr class="even">
<td align="left">a4</td>
<td align="left">-1.04</td>
<td align="left">0.19</td>
<td align="left">-0.93</td>
</tr>
<tr class="odd">
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
</tr>
<tr class="even">
<td align="left">b47</td>
<td align="left">1.02</td>
<td align="left">-0.68</td>
<td align="left">-1.02</td>
</tr>
<tr class="odd">
<td align="left">b48</td>
<td align="left">0.93</td>
<td align="left">0.4</td>
<td align="left">-0.99</td>
</tr>
<tr class="even">
<td align="left">b49</td>
<td align="left">0.92</td>
<td align="left">1.53</td>
<td align="left">-0.98</td>
</tr>
<tr class="odd">
<td align="left">b50</td>
<td align="left">1.02</td>
<td align="left">-0.2</td>
<td align="left">-0.94</td>
</tr>
</tbody>
</table>
<p>计算协方差矩阵</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data3_center_scale_cov &lt;-<span class="st"> </span><span class="kw">cov</span>(data3_center_scale)
<span class="kw">kable</span>(data3_center_scale_cov, <span class="dt">booktabs=</span>T, 
        <span class="dt">caption=</span><span class="st">&quot;Covariance matrix for 3 genes in 100 samples&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:covdata3">Table 3.10: </span>Covariance matrix for 3 genes in 100 samples</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Gene1</th>
<th align="right">Gene2</th>
<th align="right">Gene3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Gene1</td>
<td align="right">1.0000000</td>
<td align="right">-0.0640112</td>
<td align="right">-0.0004707</td>
</tr>
<tr class="even">
<td align="left">Gene2</td>
<td align="right">-0.0640112</td>
<td align="right">1.0000000</td>
<td align="right">0.1878111</td>
</tr>
<tr class="odd">
<td align="left">Gene3</td>
<td align="right">-0.0004707</td>
<td align="right">0.1878111</td>
<td align="right">1.0000000</td>
</tr>
</tbody>
</table>
<p>求解特征值和特征向量</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data3_center_scale_cov_eigen &lt;-<span class="st"> </span><span class="kw">eigen</span>(data3_center_scale_cov)

<span class="co"># 特征值，从大到小排序</span>
data3_center_scale_cov_eigen$values</code></pre></div>
<pre><code>## [1] 1.1985640 0.9997125 0.8017235</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 特征向量, 每一列为对应特征值的特征向量</span>
data3_center_scale_cov_eigen$vectors</code></pre></div>
<pre><code>##            [,1]        [,2]       [,3]
## [1,]  0.2294536 0.946533046 -0.2267736
## [2,] -0.7068494 0.001878625 -0.7073616
## [3,] -0.6691151 0.322601400  0.6694873</code></pre>
<p>产生新的矩阵</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pc_select =<span class="st"> </span><span class="dv">3</span>
label =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;PC&quot;</span>,<span class="kw">c</span>(<span class="dv">1</span>:pc_select))
data3_new &lt;-<span class="st"> </span>data3_center_scale %*%<span class="st"> </span>data3_center_scale_cov_eigen$vectors[,<span class="dv">1</span>:pc_select]
<span class="kw">colnames</span>(data3_new) &lt;-<span class="st"> </span>label
<span class="kw">kable</span>(<span class="kw">headTail</span>(data3_new), <span class="dt">booktabs=</span>T, 
        <span class="dt">caption=</span><span class="st">&quot;PCA generated matrix for the expression of 3 genes in 100 samples&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:newdata3">Table 3.11: </span>PCA generated matrix for the expression of 3 genes in 100 samples</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">PC1</th>
<th align="left">PC2</th>
<th align="left">PC3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">a1</td>
<td align="left">0.37</td>
<td align="left">-1.32</td>
<td align="left">-0.57</td>
</tr>
<tr class="even">
<td align="left">a2</td>
<td align="left">0.98</td>
<td align="left">-1.3</td>
<td align="left">0.15</td>
</tr>
<tr class="odd">
<td align="left">a3</td>
<td align="left">0.95</td>
<td align="left">-1.3</td>
<td align="left">0.08</td>
</tr>
<tr class="even">
<td align="left">a4</td>
<td align="left">0.25</td>
<td align="left">-1.29</td>
<td align="left">-0.52</td>
</tr>
<tr class="odd">
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
</tr>
<tr class="even">
<td align="left">b47</td>
<td align="left">1.4</td>
<td align="left">0.64</td>
<td align="left">-0.43</td>
</tr>
<tr class="odd">
<td align="left">b48</td>
<td align="left">0.59</td>
<td align="left">0.56</td>
<td align="left">-1.15</td>
</tr>
<tr class="even">
<td align="left">b49</td>
<td align="left">-0.22</td>
<td align="left">0.56</td>
<td align="left">-1.94</td>
</tr>
<tr class="odd">
<td align="left">b50</td>
<td align="left">1</td>
<td align="left">0.66</td>
<td align="left">-0.72</td>
</tr>
</tbody>
</table>
<p>比较原始数据和新产生的主成分对样品的聚类</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#library(scatterplot3d)</span>
colorl &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#E69F00&quot;</span>, <span class="st">&quot;#56B4E9&quot;</span>)
<span class="co"># Extract same number of colors as the Group and same Group would have same color.</span>
colors &lt;-<span class="st"> </span>colorl[<span class="kw">as.numeric</span>(data3$Group)]

<span class="co"># 1 row 2 columns</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))

<span class="kw">scatterplot3d</span>(data3[,<span class="dv">1</span>:<span class="dv">3</span>], <span class="dt">color=</span>colors, <span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">main=</span><span class="st">&quot;Original data&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)


<span class="kw">scatterplot3d</span>(data3_new, <span class="dt">color=</span>colors,<span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">main=</span><span class="st">&quot;Principle components&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)

<span class="co">#par(mfrow=c(1,1))</span></code></pre></div>
<p><img src="R_course_files/figure-html/data3oldplot-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>利用<code>prcomp</code>进行主成分分析</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca_data3 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data3[,<span class="dv">1</span>:<span class="dv">3</span>], <span class="dt">center=</span><span class="ot">TRUE</span>, <span class="dt">scale=</span><span class="ot">TRUE</span>)

<span class="co">#Show whats in the result returned by prcomp</span>
<span class="kw">str</span>(pca_data3)</code></pre></div>
<pre><code>## List of 5
##  $ sdev    : num [1:3] 1.095 1 0.895
##  $ rotation: num [1:3, 1:3] 0.22945 -0.70685 -0.66912 -0.94653 -0.00188 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:3] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot;
##   .. ..$ : chr [1:3] &quot;PC1&quot; &quot;PC2&quot; &quot;PC3&quot;
##  $ center  : Named num [1:3] 12.47 4.99 9.98
##   ..- attr(*, &quot;names&quot;)= chr [1:3] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot;
##  $ scale   : Named num [1:3] 7.495 0.207 5.063
##   ..- attr(*, &quot;names&quot;)= chr [1:3] &quot;Gene1&quot; &quot;Gene2&quot; &quot;Gene3&quot;
##  $ x       : num [1:100, 1:3] 0.373 0.98 0.946 0.255 1.109 ...
##   ..- attr(*, &quot;dimnames&quot;)=List of 2
##   .. ..$ : chr [1:100] &quot;a1&quot; &quot;a2&quot; &quot;a3&quot; &quot;a4&quot; ...
##   .. ..$ : chr [1:3] &quot;PC1&quot; &quot;PC2&quot; &quot;PC3&quot;
##  - attr(*, &quot;class&quot;)= chr &quot;prcomp&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 新的数据，与前面计算的抑制</span>
data3_pca_new &lt;-<span class="st"> </span>pca_data3$x
<span class="kw">kable</span>(<span class="kw">headTail</span>(data3_pca_new), <span class="dt">booktabs=</span>T, 
    <span class="dt">caption=</span><span class="st">&quot;PCA generated matrix usig princomp for the expression of 3 genes in 100 samples&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:data3prcomp">Table 3.12: </span>PCA generated matrix usig princomp for the expression of 3 genes in 100 samples</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">PC1</th>
<th align="left">PC2</th>
<th align="left">PC3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">a1</td>
<td align="left">0.37</td>
<td align="left">1.32</td>
<td align="left">-0.57</td>
</tr>
<tr class="even">
<td align="left">a2</td>
<td align="left">0.98</td>
<td align="left">1.3</td>
<td align="left">0.15</td>
</tr>
<tr class="odd">
<td align="left">a3</td>
<td align="left">0.95</td>
<td align="left">1.3</td>
<td align="left">0.08</td>
</tr>
<tr class="even">
<td align="left">a4</td>
<td align="left">0.25</td>
<td align="left">1.29</td>
<td align="left">-0.52</td>
</tr>
<tr class="odd">
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
<td align="left">…</td>
</tr>
<tr class="even">
<td align="left">b47</td>
<td align="left">1.4</td>
<td align="left">-0.64</td>
<td align="left">-0.43</td>
</tr>
<tr class="odd">
<td align="left">b48</td>
<td align="left">0.59</td>
<td align="left">-0.56</td>
<td align="left">-1.15</td>
</tr>
<tr class="even">
<td align="left">b49</td>
<td align="left">-0.22</td>
<td align="left">-0.56</td>
<td align="left">-1.94</td>
</tr>
<tr class="odd">
<td align="left">b50</td>
<td align="left">1</td>
<td align="left">-0.66</td>
<td align="left">-0.72</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 特征向量，与我们前面计算的一致(特征向量的符号是任意的)</span>
pca_data3$rotation</code></pre></div>
<pre><code>##              PC1          PC2        PC3
## Gene1  0.2294536 -0.946533046 -0.2267736
## Gene2 -0.7068494 -0.001878625 -0.7073616
## Gene3 -0.6691151 -0.322601400  0.6694873</code></pre>
<p>比较手动实现的PCA与<code>prcomp</code>实现的PCA的聚类结果</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#library(scatterplot3d)</span>
colorl &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#E69F00&quot;</span>, <span class="st">&quot;#56B4E9&quot;</span>)
<span class="co"># Extract same number of colors as the Group and same Group would have same color.</span>
colors &lt;-<span class="st"> </span>colorl[<span class="kw">as.numeric</span>(data3$Group)]

<span class="co"># 1 row 2 columns</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))

<span class="kw">scatterplot3d</span>(data3_new, <span class="dt">color=</span>colors,<span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">main=</span><span class="st">&quot;PCA by steps&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)

<span class="kw">scatterplot3d</span>(data3_pca_new, <span class="dt">color=</span>colors,<span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">main=</span><span class="st">&quot;PCA by prcomp&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)

<span class="co">#par(mfrow=c(1,1))</span></code></pre></div>
<p><img src="R_course_files/figure-html/pcacompare-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>自定义PCA计算函数</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ct_PCA &lt;-<span class="st"> </span>function(data, <span class="dt">center=</span><span class="ot">TRUE</span>, <span class="dt">scale=</span><span class="ot">TRUE</span>){
  data_norm &lt;-<span class="st"> </span><span class="kw">scale</span>(data, <span class="dt">center=</span>center, <span class="dt">scale=</span>scale)
  data_norm_cov &lt;-<span class="st"> </span><span class="kw">crossprod</span>(<span class="kw">as.matrix</span>(data_norm)) /<span class="st"> </span>(<span class="kw">nrow</span>(data_norm)-<span class="dv">1</span>)
  data_eigen &lt;-<span class="st"> </span><span class="kw">eigen</span>(data_norm_cov)

  rotation &lt;-<span class="st"> </span>data_eigen$vectors
  label &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;PC&#39;</span>, <span class="kw">c</span>(<span class="dv">1</span>:<span class="kw">ncol</span>(rotation)))
  <span class="kw">colnames</span>(rotation) &lt;-<span class="st"> </span>label
  sdev &lt;-<span class="st"> </span><span class="kw">sqrt</span>(data_eigen$values)
  data_new &lt;-<span class="st"> </span>data_norm %*%<span class="st"> </span>rotation
  <span class="kw">colnames</span>(data_new) &lt;-<span class="st"> </span>label
  ct_pca &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&#39;rotation&#39;</span>=rotation, <span class="st">&#39;x&#39;</span>=data_new, <span class="st">&#39;sdev&#39;</span>=sdev)
  <span class="kw">return</span>(ct_pca)
}</code></pre></div>
<p>比较有无<code>scale</code>对聚类的影响，从图中可以看到，如果不对数据进行<code>scale</code>处理，样品的聚类结果更像原始数据，本身数值大的基因对主成分的贡献会大。如果关注的是每个变量自身的实际方差对样品分类的贡献，则不应该<code>SCALE</code>；如果关注的是变量的相对大小对样品分类的贡献，则应该<code>SCALE</code>，以防数值高的变量导入的大方差引入的偏见。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data3_pca_noscale_step =<span class="st"> </span><span class="kw">ct_PCA</span>(data3[,<span class="dv">1</span>:<span class="dv">3</span>], <span class="dt">center=</span><span class="ot">TRUE</span>, <span class="dt">scale=</span><span class="ot">FALSE</span>)

<span class="co"># 特征向量</span>
data3_pca_noscale_step$rotation</code></pre></div>
<pre><code>##                PC1         PC2          PC3
## [1,]  0.9999982499 0.000609949 -0.001768688
## [2,] -0.0017733305 0.007697251 -0.999968803
## [3,] -0.0005963159 0.999970190  0.007698319</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 新变量</span>
data3_pca_noscale_pc &lt;-<span class="st"> </span>data3_pca_noscale_step$x</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#library(scatterplot3d)</span>
colorl &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#E69F00&quot;</span>, <span class="st">&quot;#56B4E9&quot;</span>)
<span class="co"># Extract same number of colors as the Group and same Group would have same color.</span>
colors &lt;-<span class="st"> </span>colorl[<span class="kw">as.numeric</span>(data3$Group)]

<span class="co"># 1 row 2 columns</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))

<span class="kw">scatterplot3d</span>(data3[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">2</span>)], <span class="dt">color=</span>colors, <span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">main=</span><span class="st">&quot;Original data&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)

<span class="kw">scatterplot3d</span>(data3_pca_noscale_pc, <span class="dt">color=</span>colors,<span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">main=</span><span class="st">&quot;PCA (no scale)&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)

<span class="kw">scatterplot3d</span>(data3_center_scale[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">2</span>)], <span class="dt">color=</span>colors, <span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">pch=</span><span class="dv">16</span>, 
        <span class="dt">main=</span><span class="st">&quot;Original data (scale)&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)

<span class="kw">scatterplot3d</span>(data3_new, <span class="dt">color=</span>colors,<span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">main=</span><span class="st">&quot;PCA (scale)&quot;</span>)
<span class="kw">legend</span>(<span class="st">&quot;top&quot;</span>, <span class="dt">legend=</span><span class="kw">levels</span>(data3$Group), <span class="dt">col=</span>colorl, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xpd=</span>T, <span class="dt">horiz=</span>T)

<span class="co">#par(mfrow=c(1,1))</span></code></pre></div>
<p><img src="R_course_files/figure-html/comparescale-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="pca" class="section level3">
<h3><span class="header-section-number">3.12.6</span> PCA结果解释</h3>
<p><code>prcomp</code>函数会返回主成分的标准差、特征向量和主成分构成的新矩阵。接下来，探索下不同主成分对数据差异的贡献和主成分与原始变量的关系。</p>
<ul>
<li>主成分的平方为为特征值,其含义为每个主成分可以解释的数据差异，计算方式为 <code>eigenvalues = (pca$sdev)^2</code></li>
<li>每个主成分可以解释的数据差异的比例为 <code>percent_var = eigenvalues*100/sum(eigenvalues)</code></li>
<li>可以使用<code>summary(pca)</code>获取以上两条信息。</li>
</ul>
<p>这两个信息可以判断主成分分析的质量：</p>
<ul>
<li>成功的降维需要保证在前几个为数不多的主成分对数据差异的解释可以达到80-90%。</li>
</ul>
<p>指导选择主成分的数目：</p>
<ul>
<li>选择的主成分足以解释的总方差大于80% (方差比例碎石图)</li>
<li>从前面的协方差矩阵可以看到，自动定标(scale)的变量的方差为1 (协方差矩阵对角线的值)。待选择的主成分应该是那些方差大于1的主成分，即其解释的方差大于原始变量（特征值碎石图，方差大于1，特征值也会大于1，反之亦然）。</li>
</ul>
<p>鉴定核心变量和变量间的隐性关系:</p>
<ul>
<li>原始变量与主成分的相关性 <code>Variable correlation with PCs (var.cor) = loadings * sdev</code></li>
<li>原始数据对主成分的贡献度 <code>var.cor^2 / (total var.cor^2)</code></li>
</ul>
<p>在测试数据中，<code>scale</code>后，三个主成分对数据差异的贡献度大都在30%左右，而未<code>scale</code>的数据，三个主成分对数据差异的贡献度相差很大。这是因为三个基因由于自身表达量级所引起的方差的差异导致它们各自对数据的权重差异，从而使主成分偏向于数值大的变量。</p>
</div>
<div id="pca" class="section level3">
<h3><span class="header-section-number">3.12.7</span> PCA应用于测试数据</h3>
<p>前面用到一组比较大的测试数据集，并做了PCA分析，现在测试不同的处理对结果的影响。</p>
<p>首先回顾下我们用到的数据。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;gridExtra&quot;</span>)
<span class="co"># Pay attention to the format of PCA input </span>
<span class="co"># Rows are samples and columns are variables</span>
<span class="co"># data4_use_log2_t &lt;- t(data4_use_log2)</span>

<span class="co"># Add group column for plotting</span>
<span class="co"># data4_use_log2_label &lt;- as.data.frame(data4_use_log2_t)</span>
<span class="co"># data4_use_log2_label$group &lt;- group</span>

<span class="kw">kable</span>(<span class="kw">corner</span>(data4_use_log2_label), <span class="dt">digits=</span><span class="dv">3</span>, <span class="dt">caption=</span><span class="st">&quot;Single cell gene expression data&quot;</span>)</code></pre></div>
<table>
<caption><span id="tab:pcaclustershowdata">Table 3.13: </span>Single cell gene expression data</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">MT2A</th>
<th align="right">ANXA1</th>
<th align="right">ARHGDIB</th>
<th align="right">RND3</th>
<th align="right">BLVRB</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Hi_2338_1</td>
<td align="right">12.211</td>
<td align="right">11.837</td>
<td align="right">9.543</td>
<td align="right">9.762</td>
<td align="right">9.162</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_2</td>
<td align="right">11.306</td>
<td align="right">8.099</td>
<td align="right">8.072</td>
<td align="right">10.107</td>
<td align="right">9.423</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_3</td>
<td align="right">11.926</td>
<td align="right">10.689</td>
<td align="right">9.721</td>
<td align="right">9.388</td>
<td align="right">9.694</td>
</tr>
<tr class="even">
<td align="left">Hi_2338_4</td>
<td align="right">10.974</td>
<td align="right">9.387</td>
<td align="right">9.883</td>
<td align="right">8.792</td>
<td align="right">9.767</td>
</tr>
<tr class="odd">
<td align="left">Hi_2338_5</td>
<td align="right">14.604</td>
<td align="right">10.375</td>
<td align="right">9.970</td>
<td align="right">8.815</td>
<td align="right">9.350</td>
</tr>
</tbody>
</table>
<p>比较对数运算和<code>scale</code>对样品分类的影响。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ct_pca_2d_plot &lt;-<span class="st"> </span>function(pca, data_with_label, <span class="dt">labelName=</span><span class="st">&#39;group&#39;</span>, <span class="dt">title=</span><span class="st">&#39;PCA&#39;</span>) {
  <span class="co"># sdev: standard deviation of the principle components.</span>
  <span class="co"># Square to get variance</span>
  percentVar &lt;-<span class="st"> </span>pca$sdev^<span class="dv">2</span> /<span class="st"> </span><span class="kw">sum</span>( pca$sdev^<span class="dv">2</span>)
  <span class="co">#data &lt;- data_with_label</span>
  <span class="co">#data[labelName] &lt;- factor(unlist(data[labelName]))</span>
  level &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(<span class="kw">unlist</span>(data_with_label[labelName])))
  shapes =<span class="st"> </span>(<span class="dv">1</span>:level)%%<span class="dv">30</span>  <span class="co"># maximum allow 30 types of symbols</span>
  p =<span class="st"> </span><span class="kw">autoplot</span>(pca, <span class="dt">data=</span>data_with_label, <span class="dt">colour=</span>labelName, <span class="dt">shape=</span>labelName) +<span class="st"> </span>
<span class="st">      </span><span class="kw">scale_shape_manual</span>(<span class="dt">values=</span>shapes) +
<span class="st">      </span><span class="kw">xlab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC1 (&quot;</span>, <span class="kw">round</span>(percentVar[<span class="dv">1</span>]*<span class="dv">100</span>), <span class="st">&quot;% variance)&quot;</span>)) +<span class="st"> </span>
<span class="st">      </span><span class="kw">ylab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC2 (&quot;</span>, <span class="kw">round</span>(percentVar[<span class="dv">2</span>]*<span class="dv">100</span>), <span class="st">&quot;% variance)&quot;</span>)) +<span class="st"> </span>
<span class="st">      </span><span class="kw">theme_bw</span>() +<span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;right&quot;</span>) +<span class="st"> </span><span class="kw">labs</span>(<span class="dt">title=</span>title) +
<span class="st">      </span><span class="kw">theme</span>(<span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(), <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>())
  <span class="kw">return</span>(p)
}</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># By default, prcomp will centralized the data using mean.</span>
<span class="co"># Normalize data for PCA by dividing each data by column standard deviation.</span>
<span class="co"># Often, we would normalize data.</span>
<span class="co"># Only when we care about the real number changes other than the trends,</span>
<span class="co"># `scale` can be set to TRUE. </span>
<span class="co"># We will show the differences of scaling and un-scaling effects.</span>
data4_use_t &lt;-<span class="st"> </span><span class="kw">t</span>(data4_use)
ori_scale_pca_test &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data4_use_t, <span class="dt">scale=</span>T)
ori_no_scale_pca_test &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data4_use_t, <span class="dt">scale=</span>F)
log_scale_pca_test &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data4_use_log2_t, <span class="dt">scale=</span>T)
log_no_scale_pca_test &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data4_use_log2_t, <span class="dt">scale=</span>F)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ori_scale_pca_plot =<span class="st"> </span><span class="kw">ct_pca_2d_plot</span>(ori_scale_pca_test, data4_use_log2_label, 
        <span class="dt">title=</span><span class="st">&quot;Scaled original data&quot;</span>)
ori_no_scale_pca_plot =<span class="st"> </span><span class="kw">ct_pca_2d_plot</span>(ori_no_scale_pca_test, data4_use_log2_label, 
        <span class="dt">title=</span><span class="st">&quot;Un-scaled original data&quot;</span>)
log_scale_pca_plot =<span class="st"> </span><span class="kw">ct_pca_2d_plot</span>(log_scale_pca_test, data4_use_log2_label, 
        <span class="dt">title=</span><span class="st">&quot;Scaled log transformed data&quot;</span>)
log_no_scale_pca_plot =<span class="st"> </span><span class="kw">ct_pca_2d_plot</span>(log_no_scale_pca_test, data4_use_log2_label, 
        <span class="dt">title=</span><span class="st">&quot;Un-scaled log transformed data&quot;</span>)

a__ &lt;-<span class="st"> </span><span class="kw">grid.arrange</span>(ori_scale_pca_plot, ori_no_scale_pca_plot, log_scale_pca_plot, 
        log_no_scale_pca_plot, <span class="dt">ncol=</span><span class="dv">2</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/comparepcascalelogplot-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>如果首先提取500个变化最大的基因，再执行PCA分析会怎样呢？</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data4_use_mad &lt;-<span class="st"> </span><span class="kw">apply</span>(data4_use, <span class="dv">1</span>, mad)
data4_use_mad_top500 &lt;-<span class="st"> </span><span class="kw">t</span>(data4_use[<span class="kw">rev</span>(<span class="kw">order</span>(data4_use_mad))[<span class="dv">1</span>:<span class="dv">500</span>],])

data4_use_log2_mad &lt;-<span class="st"> </span><span class="kw">apply</span>(data4_use_log2, <span class="dv">1</span>, mad)
data4_use_log2_mad_top500 &lt;-<span class="st"> </span><span class="kw">t</span>(data4_use_log2[<span class="kw">rev</span>(<span class="kw">order</span>(data4_use_log2_mad))[<span class="dv">1</span>:<span class="dv">500</span>],])

ori_scale_pca_top500 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data4_use_mad_top500, <span class="dt">scale=</span>T)
ori_no_scale_pca_top500 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data4_use_mad_top500, <span class="dt">scale=</span>F)
log_scale_pca_top500 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data4_use_log2_mad_top500, <span class="dt">scale=</span>T)
log_no_scale_pca_top500 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(data4_use_log2_mad_top500, <span class="dt">scale=</span>F)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ori_scale_pca_plot_t5 =<span class="st"> </span><span class="kw">ct_pca_2d_plot</span>(ori_scale_pca_top500, data4_use_log2_label, 
        <span class="dt">title=</span><span class="st">&quot;Scaled original data&quot;</span>)
ori_no_scale_pca_plot_t5 =<span class="st"> </span><span class="kw">ct_pca_2d_plot</span>(ori_no_scale_pca_top500, 
        data4_use_log2_label, <span class="dt">title=</span><span class="st">&quot;Un-scaled original data&quot;</span>)
log_scale_pca_plot_t5 =<span class="st"> </span><span class="kw">ct_pca_2d_plot</span>(log_scale_pca_top500, 
        data4_use_log2_label, <span class="dt">title=</span><span class="st">&quot;Scaled log transformed data&quot;</span>)
log_no_scale_pca_plot_t5 =<span class="st"> </span><span class="kw">ct_pca_2d_plot</span>(log_no_scale_pca_top500, 
        data4_use_log2_label, <span class="dt">title=</span><span class="st">&quot;Un-scaled log transformed data&quot;</span>)

a__ &lt;-<span class="st"> </span><span class="kw">grid.arrange</span>(ori_scale_pca_plot_t5, ori_no_scale_pca_plot_t5, 
                    log_scale_pca_plot_t5, log_no_scale_pca_plot_t5, <span class="dt">ncol=</span><span class="dv">2</span>)</code></pre></div>
<p><img src="R_course_files/figure-html/comparevargenepcascalelogplot-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="pca" class="section level3">
<h3><span class="header-section-number">3.12.8</span> PCA注意事项</h3>
<ol style="list-style-type: decimal">
<li><p>一般说来，在PCA之前原始数据需要中心化（centering，数值减去平均值）。中心化的方法很多，除了平均值中心化（mean-centering）外，还包括其它更稳健的方法，比如中位数中心化等。</p></li>
<li><p>除了中心化以外，定标 (Scale, 数值除以标准差) 也是数据前处理中需要考虑的一点。如果数据没有定标，则原始数据中方差大的变量对主成分的贡献会很大。数据的方差与其量级成指数关系，比如一组数据<code>(1,2,3,4)</code>的方差是<code>1.67</code>，而<code>(10,20,30,40)</code>的方差就是<code>167</code>,数据变大10倍，方差放大了100倍。</p></li>
<li><p>但是定标(scale)可能会有一些负面效果，因为定标后变量之间的权重就是变得相同。如果我们的变量中有噪音的话，我们就在无形中把噪音和信息的权重变得相同，但PCA本身无法区分信号和噪音。在这样的情形下，我们就不必做定标。</p></li>
<li><p>一般而言，对于度量单位不同的指标或是取值范围彼此差异非常大的指标不直接由其协方差矩阵出发进行主成分分析，而应该考虑对数据的标准化。比如度量单位不同，有万人、万吨、万元、亿元，而数据间的差异性也非常大，小则几十大则几万，因此在用协方差矩阵求解主成分时存在<strong>协方差矩阵中数据的差异性很大</strong>。在后面提取主成分时发现，<strong>只提取了一个主成分</strong>，而此时并不能将所有的变量都解释到，这就没有真正起到<strong>降维</strong>的作用。此时就需要对数据进行定标(scale)，这样提取的主成分可以覆盖更多的变量，这就实现主成分分析的最终目的。但是对原始数据进行标准化后更倾向于使得各个指标的作用在主成分分析构成中相等。对于数据取值范围不大或是度量单位相同的指标进行标准化处理后，其主成分分析的结果与仍由协方差矩阵出发求得的结果有较大区别。这是因为对数据标准化的过程实际上就是<strong>抹杀原有变量离散程度差异的过程</strong>，标准化后方差均为1，而实际上方差是对数据信息的重要概括形式，也就是说，对原始数据进行标准化后抹杀了一部分重要信息，因此才使得标准化后各变量在主成分构成中的作用趋于相等。因此，<strong>对同度量或是取值范围在同量级的数据</strong>还是直接使用非定标数据求解主成分为宜。</p></li>
<li><p>中心化和定标都会受数据中离群值（outliers）或者数据不均匀（比如数据被分为若干个小组）的影响，应该用更稳健的中心化和定标方法。</p></li>
<li><p>PCA也存在一些限制，例如它可以很好的解除线性相关，但是对于高阶相关性就没有办法了，对于存在高阶相关性的数据，可以考虑Kernel PCA，通过Kernel函数将非线性相关转为线性相关，关于这点就不展开讨论了。另外，PCA假设数据各主特征是分布在正交方向上，如果在非正交方向上存在几个方差较大的方向，PCA的效果就大打折扣了。</p></li>
</ol>
</div>
<div id="section-3.12.9" class="section level3">
<h3><span class="header-section-number">3.12.9</span> 参考资料</h3>
<ul>
<li><a href="https://www.zhihu.com/question/20998460" class="uri">https://www.zhihu.com/question/20998460</a></li>
<li><a href="http://blog.csdn.net/zhongkelee/article/details/44064401">PCA 教程1</a></li>
<li><a href="http://www.xiaolingzi.com/?p=963">PCA 文字化描述</a></li>
<li><p><a href="http://wenku.baidu.com/link?url=hsnzR5gUvsPBwkwwcWU4T3aTSC_fsZDxAmaGGBPfumsIW_I0TJAdJEWhFyiQgw7uA58DKukR-9g5x0DyzE97kHddMaXOxk_iZBjoIdbdB6e">pca1</a></p></li>
<li><p><a href="http://stackoverflow.com/questions/22092220/plot-only-y-axis-but-nothing-else">ggplot2 axis</a></p></li>
<li><p><a href="http://www.sthda.com/english/wiki/scatterplot3d-3d-graphics-r-software-and-data-visualization">scatterplot3D</a></p></li>
<li><p><a href="http://dong.farbox.com/32">稳健PCA</a></p></li>
<li><p><a href="http://www.nlpca.org/pca_principal_component_analysis.html" class="uri">http://www.nlpca.org/pca_principal_component_analysis.html</a></p></li>
<li><p><a href="http://gastonsanchez.com/how-to/2014/01/15/Center-data-in-R/">Data centering</a></p></li>
<li><p><a href="http://www.statpower.net/Content/310/R%20Stuff/SampleMarkdown.html">Sample R markdown</a></p></li>
<li><p><a href="http://www2.edu-edu.com.cn/lesson_crs78/self/02198/resource/contents/ch_05/ch_05.html">矩阵特征值，对称矩阵的对角化</a></p></li>
<li><p><a href="http://www.sthda.com/english/wiki/principal-component-analysis-in-r-prcomp-vs-princomp-r-software-and-data-mining">Detail usage and visualization of prcomp result</a></p></li>
<li><p><a href="http://stackoverflow.com/questions/1249548/side-by-side-plots-with-ggplot2">ggplot2 side by side plot</a></p></li>
</ul>
</div>
</div>
<div id="section-3.13" class="section level2">
<h2><span class="header-section-number">3.13</span> 表达聚类分析</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#</span></code></pre></div>
</div>
<div id="section-3.14" class="section level2">
<h2><span class="header-section-number">3.14</span> 生存分析</h2>
<p>生存分析指根据试验或调查得到的数据对生物或人的生存时间进行分析和推断，研究生存时间和结局与众多影响因素间关系及其程度大小的方法，也称生存率分析或存活率分析。常用于肿瘤等疾病的标志物筛选、疗效及预后的考核。</p>
<p><mark>简单地说，比较两组或多组人群随着时间的延续，存活个体的比例变化趋势。活着的个体越少的组危险性越大，对应的基因对疾病影响越大，对应的药物治疗效果越差。</mark></p>
<p>生存分析适合于处理时间-事件数据，如下</p>
<p><img src="images/surv_input_data.png" width="100%" style="display: block; margin: auto;" /></p>
<p>生存时间数据有两种类型：</p>
<ul>
<li>完全数据 (complete data)指被观测对象从观察起点到出现终点事件所经历的时间; 一般用状态值1或TRUE表示。</li>
<li>截尾数据 (consored data)或删失数据，指在出现终点事件前，被观测对象的观测过程终止了。由于被观测对象所提供的信息是不完全的，只知道他们的生存事件超过了截尾时间。截尾主要由于失访、退出和终止产生。一般用状态值0或FALSE表示。</li>
<li>TCGA中的临床数据标记也符合这个规律，在下面软件运行时也可修改状态值的含义, 但一般遵循这个规律。</li>
</ul>
<p>生存概率 (survival probability)指某段时间开始时存活的个体至该时间结束时仍然存活的可能性大小。</p>
<p><code>生存概率=某人群活过某段时间例数/该人群同时间段期初观察例数。</code></p>
<p>生存率 (Survival rate)，用<code>S(t)</code>表示，指经历<code>t</code>个单位时间后仍存活的概率，若无删失数据，则为活过了<code>t</code>时刻仍然存活的例数/观察开始的总例数。如果有删失数据，分母则需要按时段进行校正。</p>
<p>生存分析一个常用的方法是寿命表法。</p>
<p>寿命表是描述一段时间内生存状况、终点事件和生存概率的表格，需计算累积生存概率即每一步生存概率的乘积 (也可能是原始生存概率)，可完成对病例随访资料在任意指定时点的生存状况评价。</p>
<p><img src="images/surv_life_table.png" width="100%" style="display: block; margin: auto;" /></p>
<div id="r" class="section level3">
<h3><span class="header-section-number">3.14.1</span> R做生存分析</h3>
<p>R中做生存分析需要用到包<code>survival</code>和<code>survminer</code>。输入数据至少两列，<code>存活时间</code>和<code>生存状态</code>，也就是测试数据中的<code>Days.survial</code>和<code>vital_status</code>列。如果需要比较不同组之间的差异，也需要提供个体的分组信息，如测试数据中的<code>PAM50</code>列。对应TCGA的数据，一般根据某个基因的表达量或突变有无对个体进行分组。</p>
<p>读入数据</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(survival)
BRCA &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&#39;data/BRCA.tsv&#39;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header=</span>T)
<span class="kw">head</span>(BRCA)</code></pre></div>
<pre><code>##                ID SampleType PAM50Call_RNAseq Days.survival
## 1 TCGA-E9-A2JT-01 Tumor_type             LumA           288
## 2 TCGA-BH-A0W4-01 Tumor_type             LumA           759
## 3 TCGA-BH-A0B5-01 Tumor_type             LumA          2136
## 4 TCGA-AC-A3TM-01 Tumor_type          Unknown           762
## 5 TCGA-E9-A5FL-01 Tumor_type          Unknown            24
## 6 TCGA-AC-A3TN-01 Tumor_type          Unknown           456
##   pathologic_stage vital_status
## 1        stage iia            0
## 2        stage iia            0
## 3       stage iiia            0
## 4       stage iiia            0
## 5        stage iib            0
## 6        stage iib            0</code></pre>
<p>简单地看下每一列都有什么内容，方便对数据整体有个了解，比如有无特殊值。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(BRCA)</code></pre></div>
<pre><code>##                ID            SampleType   PAM50Call_RNAseq
##  TCGA-3C-AAAU-01:   1   Tumor_type:1090   Basal  :138     
##  TCGA-3C-AALI-01:   1                     Her2   : 65     
##  TCGA-3C-AALJ-01:   1                     LumA   :415     
##  TCGA-3C-AALK-01:   1                     LumB   :194     
##  TCGA-4H-AAAK-01:   1                     Normal : 24     
##  TCGA-5L-AAT0-01:   1                     Unknown:254     
##  (Other)        :1084                                     
##  Days.survival      pathologic_stage  vital_status   
##  Min.   :   0.0   stage iia :359     Min.   :0.0000  
##  1st Qu.: 450.2   stage iib :259     1st Qu.:0.0000  
##  Median : 848.0   stage iiia:156     Median :0.0000  
##  Mean   :1247.0   stage i   : 90     Mean   :0.1394  
##  3rd Qu.:1682.8   stage ia  : 85     3rd Qu.:0.0000  
##  Max.   :8605.0   stage iiic: 67     Max.   :1.0000  
##                   (Other)   : 74</code></pre>
<p>计算寿命表</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Days.survival：跟踪到的存活时间</span>
<span class="co"># vital_status: 跟踪到的存活状态</span>
<span class="co"># ~1表示不进行分组</span>
fit &lt;-<span class="st"> </span><span class="kw">survfit</span>(<span class="kw">Surv</span>(Days.survival, vital_status)~<span class="dv">1</span>, <span class="dt">data=</span>BRCA)

<span class="co"># 获得的survial列就是生存率 </span>
<span class="kw">summary</span>(fit)</code></pre></div>
<pre><code>## Call: survfit(formula = Surv(Days.survival, vital_status) ~ 1, data = BRCA)
## 
##  time n.risk n.event survival  std.err lower 95% CI upper 95% CI
##   116   1021       1    0.999 0.000979        0.997        1.000
##   158   1017       1    0.998 0.001386        0.995        1.000
##   160   1016       1    0.997 0.001697        0.994        1.000
##   172   1010       1    0.996 0.001962        0.992        1.000
##   174   1008       1    0.995 0.002195        0.991        0.999
##   197   1003       1    0.994 0.002406        0.989        0.999
##   224    993       1    0.993 0.002604        0.988        0.998
##   227    990       1    0.992 0.002788        0.987        0.998
##   239    987       1    0.991 0.002961        0.985        0.997
##   255    981       1    0.990 0.003125        0.984        0.996
##   266    978       1    0.989 0.003282        0.983        0.996
##   295    965       1    0.988 0.003435        0.981        0.995
##   302    962       1    0.987 0.003581        0.980        0.994
##   304    958       1    0.986 0.003723        0.979        0.993
##   320    948       1    0.985 0.003862        0.977        0.993
##   322    946       1    0.984 0.003995        0.976        0.992
##   336    938       1    0.983 0.004127        0.975        0.991
##   348    929       1    0.982 0.004256        0.973        0.990
##   362    921       1    0.981 0.004382        0.972        0.989
##   365    918       1    0.980 0.004506        0.971        0.989
##   377    896       1    0.979 0.004632        0.970        0.988
##   385    887       2    0.976 0.004877        0.967        0.986
##   426    845       1    0.975 0.005006        0.965        0.985
##   446    826       1    0.974 0.005137        0.964        0.984
##   524    765       1    0.973 0.005286        0.962        0.983
##   538    751       1    0.971 0.005435        0.961        0.982
##   548    743       1    0.970 0.005583        0.959        0.981
##   558    732       1    0.969 0.005731        0.958        0.980
##   563    729       1    0.967 0.005875        0.956        0.979
##   571    723       1    0.966 0.006017        0.954        0.978
##   573    721       1    0.965 0.006156        0.953        0.977
##   584    708       1    0.963 0.006297        0.951        0.976
##   612    684       1    0.962 0.006443        0.949        0.975
##   614    680       1    0.961 0.006587        0.948        0.974
##   616    678       1    0.959 0.006728        0.946        0.972
##   639    660       1    0.958 0.006873        0.944        0.971
##   678    632       1    0.956 0.007027        0.943        0.970
##   723    608       1    0.955 0.007189        0.941        0.969
##   749    592       1    0.953 0.007356        0.939        0.968
##   754    589       1    0.951 0.007519        0.937        0.966
##   763    577       1    0.950 0.007685        0.935        0.965
##   785    570       1    0.948 0.007850        0.933        0.964
##   786    568       1    0.946 0.008012        0.931        0.962
##   792    564       2    0.943 0.008327        0.927        0.960
##   811    557       1    0.941 0.008483        0.925        0.958
##   821    552       1    0.940 0.008637        0.923        0.957
##   825    550       1    0.938 0.008789        0.921        0.955
##   860    541       1    0.936 0.008942        0.919        0.954
##   879    533       1    0.934 0.009096        0.917        0.952
##   883    531       1    0.933 0.009248        0.915        0.951
##   904    526       1    0.931 0.009399        0.913        0.950
##   912    522       1    0.929 0.009548        0.911        0.948
##   921    517       1    0.927 0.009697        0.909        0.947
##   943    511       1    0.926 0.009847        0.906        0.945
##   959    505       1    0.924 0.009996        0.904        0.944
##   967    502       1    0.922 0.010144        0.902        0.942
##   976    494       1    0.920 0.010294        0.900        0.940
##   991    489       2    0.916 0.010590        0.896        0.937
##  1004    480       1    0.914 0.010739        0.894        0.936
##  1009    473       1    0.912 0.010889        0.891        0.934
##  1032    464       1    0.910 0.011042        0.889        0.932
##  1034    463       2    0.907 0.011339        0.885        0.929
##  1048    453       1    0.905 0.011489        0.882        0.927
##  1072    444       1    0.902 0.011642        0.880        0.926
##  1093    439       1    0.900 0.011796        0.878        0.924
##  1104    434       1    0.898 0.011950        0.875        0.922
##  1127    426       1    0.896 0.012106        0.873        0.920
##  1142    419       1    0.894 0.012265        0.870        0.918
##  1148    418       1    0.892 0.012421        0.868        0.917
##  1152    414       1    0.890 0.012576        0.866        0.915
##  1174    404       1    0.888 0.012736        0.863        0.913
##  1272    371       1    0.885 0.012925        0.860        0.911
##  1275    370       1    0.883 0.013109        0.858        0.909
##  1286    367       1    0.880 0.013293        0.855        0.907
##  1324    356       1    0.878 0.013483        0.852        0.905
##  1365    347       1    0.875 0.013680        0.849        0.903
##  1388    343       1    0.873 0.013876        0.846        0.900
##  1411    342       1    0.870 0.014068        0.843        0.898
##  1430    338       1    0.868 0.014260        0.840        0.896
##  1439    335       1    0.865 0.014451        0.837        0.894
##  1508    322       1    0.862 0.014654        0.834        0.892
##  1542    313       1    0.860 0.014864        0.831        0.889
##  1556    305       2    0.854 0.015291        0.825        0.885
##  1563    302       1    0.851 0.015500        0.821        0.882
##  1642    283       1    0.848 0.015735        0.818        0.880
##  1649    279       1    0.845 0.015969        0.814        0.877
##  1673    277       1    0.842 0.016200        0.811        0.874
##  1688    271       1    0.839 0.016436        0.807        0.872
##  1692    269       1    0.836 0.016668        0.804        0.869
##  1694    267       1    0.833 0.016897        0.800        0.867
##  1699    266       1    0.830 0.017121        0.797        0.864
##  1759    260       1    0.826 0.017350        0.793        0.861
##  1781    258       1    0.823 0.017576        0.790        0.858
##  1793    256       1    0.820 0.017799        0.786        0.856
##  1812    254       1    0.817 0.018020        0.782        0.853
##  1884    241       1    0.813 0.018261        0.778        0.850
##  1900    239       1    0.810 0.018499        0.775        0.847
##  1920    236       1    0.807 0.018736        0.771        0.844
##  1927    233       1    0.803 0.018973        0.767        0.841
##  1993    225       1    0.800 0.019221        0.763        0.838
##  2009    223       1    0.796 0.019467        0.759        0.835
##  2097    212       1    0.792 0.019734        0.754        0.832
##  2127    208       1    0.788 0.020003        0.750        0.829
##  2192    195       1    0.784 0.020305        0.746        0.825
##  2207    192       1    0.780 0.020606        0.741        0.822
##  2273    181       2    0.772 0.021261        0.731        0.814
##  2296    174       1    0.767 0.021596        0.726        0.811
##  2348    168       1    0.763 0.021945        0.721        0.807
##  2361    167       1    0.758 0.022284        0.716        0.803
##  2373    161       1    0.753 0.022638        0.710        0.799
##  2417    156       1    0.749 0.023002        0.705        0.795
##  2469    152       1    0.744 0.023372        0.699        0.791
##  2483    150       1    0.739 0.023736        0.694        0.787
##  2520    144       1    0.734 0.024119        0.688        0.782
##  2534    143       1    0.728 0.024490        0.682        0.778
##  2551    140       1    0.723 0.024861        0.676        0.774
##  2573    137       1    0.718 0.025234        0.670        0.769
##  2636    127       1    0.712 0.025661        0.664        0.764
##  2712    118       1    0.706 0.026144        0.657        0.759
##  2763    115       1    0.700 0.026628        0.650        0.754
##  2798    112       1    0.694 0.027114        0.643        0.749
##  2854    109       1    0.687 0.027602        0.635        0.744
##  2866    107       1    0.681 0.028082        0.628        0.738
##  2911    105       1    0.675 0.028554        0.621        0.733
##  2965    102       1    0.668 0.029030        0.613        0.727
##  3063     86       1    0.660 0.029713        0.604        0.721
##  3262     66       1    0.650 0.030901        0.592        0.714
##  3409     54       1    0.638 0.032590        0.577        0.705
##  3418     53       1    0.626 0.034127        0.563        0.697
##  3461     50       1    0.614 0.035668        0.548        0.688
##  3462     49       2    0.589 0.038357        0.518        0.669
##  3472     47       1    0.576 0.039532        0.504        0.659
##  3492     46       1    0.563 0.040608        0.489        0.649
##  3669     40       1    0.549 0.041965        0.473        0.638
##  3736     38       1    0.535 0.043280        0.457        0.627
##  3873     36       1    0.520 0.044555        0.440        0.615
##  3926     35       1    0.505 0.045693        0.423        0.603
##  3941     34       1    0.490 0.046703        0.407        0.591
##  3945     33       1    0.476 0.047593        0.391        0.579
##  3959     31       1    0.460 0.048467        0.374        0.566
##  4267     24       1    0.441 0.050097        0.353        0.551
##  4456     19       1    0.418 0.052562        0.326        0.535
##  6456     10       1    0.376 0.061715        0.273        0.519
##  6593      9       1    0.334 0.067535        0.225        0.497
##  7455      6       1    0.279 0.075850        0.163        0.475</code></pre>
<p>绘制生存曲线，横轴表示生存时间，纵轴表示生存概率，为一条梯形下降的曲线。下降幅度越大，表示生存率越低或生存时间越短。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(survminer)
<span class="co"># conf.int：是否显示置信区间</span>
<span class="co"># risk.table: 对应时间存活个体总结表格</span>
<span class="kw">ggsurvplot</span>(fit, <span class="dt">conf.int=</span>T,<span class="dt">risk.table=</span>T)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-218-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><code>PAM50</code>是通过50个基因的表达量把乳腺癌分为四种类型 (Luminal A, Luminal B, HER2-enriched, and Basal-like)作为预后的标志。根据<code>PAM50</code>属性对病人进行分组，评估比较两组之间生存率的差别。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 这三步不是必须的，只是为了方便，选择其中的4个确定了的分组进行分析</span>
<span class="co"># 同时为了简化图例，给列重命名一下，使得列名不那么长</span>
BRCA_PAM50 &lt;-<span class="st"> </span>BRCA[<span class="kw">grepl</span>(<span class="st">&quot;Basal|Her2|LumA|LumB&quot;</span>,BRCA$PAM50Call_RNAseq),]
BRCA_PAM50 &lt;-<span class="st"> </span><span class="kw">droplevels</span>(BRCA_PAM50)
<span class="kw">colnames</span>(BRCA_PAM50)[<span class="kw">colnames</span>(BRCA_PAM50)==<span class="st">&quot;PAM50Call_RNAseq&quot;</span>] &lt;-<span class="st"> &#39;PAM50&#39;</span>
<span class="co"># 按PAM50分组</span>
fit &lt;-<span class="st"> </span><span class="kw">survfit</span>(<span class="kw">Surv</span>(Days.survival, vital_status)~PAM50, <span class="dt">data=</span>BRCA_PAM50)
<span class="co"># 绘制曲线</span>
<span class="kw">ggsurvplot</span>(fit, <span class="dt">conf.int=</span>F,<span class="dt">risk.table=</span>T, <span class="dt">risk.table.col=</span><span class="st">&quot;strata&quot;</span>, <span class="dt">pval=</span>T)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-219-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>简化<code>Stage</code>信息，先只查看大的阶段</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BRCA_PAM50$pathologic_stage &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;(i+v*).*&#39;</span>, <span class="st">&quot;</span><span class="ch">\\</span><span class="st">1&quot;</span>, BRCA_PAM50$pathologic_stage)
BRCA_PAM50$pathologic_stage &lt;-<span class="st"> </span><span class="kw">as.factor</span>(BRCA_PAM50$pathologic_stage)
<span class="kw">colnames</span>(BRCA_PAM50)[<span class="kw">colnames</span>(BRCA_PAM50)==<span class="st">&quot;pathologic_stage&quot;</span>] &lt;-<span class="st"> &#39;PS&#39;</span>
fit &lt;-<span class="st"> </span><span class="kw">survfit</span>(<span class="kw">Surv</span>(Days.survival, vital_status)~PS, <span class="dt">data=</span>BRCA_PAM50)
<span class="co"># 绘制曲线</span>
<span class="kw">ggsurvplot</span>(fit, <span class="dt">conf.int=</span>F,<span class="dt">risk.table=</span>T, <span class="dt">risk.table.col=</span><span class="st">&quot;strata&quot;</span>, <span class="dt">pval=</span>T)</code></pre></div>
<p><img src="R_course_files/figure-html/unnamed-chunk-220-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="section-3.15" class="section level2">
<h2><span class="header-section-number">3.15</span> 一步作图的优势</h2>
<p>一步作图相比于直接写R代码有什么好处呢？</p>
<ol style="list-style-type: decimal">
<li><p>模块化好。数据处理和可视化分开；一步作图，只是作图，不做无关的处理，更随意。</p></li>
<li><p>易用性强。敲代码，总不如给改数来得快。</p></li>
<li><p>重用性强。假如我要做十几个分开的基因集合呢？一段段复制代码？改错了或忘记改某个地方了怎么办？</p></li>
<li><p>快速出图。先快速出个原型，再接着调整。</p></li>
</ol>
<p>获取地址 <a href="https://github.com/Tong-Chen/s-plot" class="uri">https://github.com/Tong-Chen/s-plot</a>.</p>
<p>在线绘图 <a href="http://www.ehbio.com/ImageGP/" class="uri">http://www.ehbio.com/ImageGP/</a></p>
<p><code>s-plot</code>可以绘制的图的类型还有一些，列举如下；</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">Usage</span>:

<span class="kw">s-plot</span> options

<span class="kw">Function</span>:

<span class="kw">This</span> software is designed to simply the process of plotting and help
<span class="kw">researchers</span> focus more on data rather than technology.

<span class="kw">Currently</span>, the following types of plot are supported.

<span class="co">#### Bars</span>
<span class="kw">s-plot</span> barPlot

<span class="co">#### Lines</span>
<span class="kw">s-plot</span> lines

<span class="co">#### Dots</span>
<span class="kw">s-plot</span> pca
<span class="kw">s-plot</span> scatterplot3d
<span class="kw">s-plot</span> scatterplot2
<span class="kw">s-plot</span> scatterplotColor
<span class="kw">s-plot</span> scatterplotContour
<span class="kw">s-plot</span> scatterplotLotsData
<span class="kw">s-plot</span> scatterplotMatrix
<span class="kw">s-plot</span> scatterplotDoubleVariable
<span class="kw">s-plot</span> contourPlot
<span class="kw">s-plot</span> density2d

<span class="co">#### Distribution</span>
<span class="kw">s-plot</span> areaplot
<span class="kw">s-plot</span> boxplot
<span class="kw">s-plot</span> densityHistPlot

<span class="co">#### Cluster</span>
<span class="kw">s-plot</span> hcluster_gg (latest)

<span class="co">#### Heatmap</span>
<span class="kw">s-plot</span> heatmapS
<span class="kw">s-plot</span> heatmapM
<span class="kw">s-plot</span> heatmap.2
<span class="kw">s-plot</span> pheatmap
<span class="kw">s-plot</span> prettyHeatmap

<span class="co">#### Others</span>
<span class="kw">s-plot</span> volcano
<span class="kw">s-plot</span> vennDiagram
<span class="kw">s-plot</span> upsetView</code></pre></div>
</div>
<div id="section-3.16" class="section level2">
<h2><span class="header-section-number">3.16</span> 不改脚本的热图绘制</h2>
<p>绘图时通常会碰到两个头疼的问题： 1. 需要画很多的图，唯一的不同就是输出文件，其它都不需要修改。如果用R脚本，需要反复替换文件名，繁琐又容易出错。</p>
<ol start="2" style="list-style-type: decimal">
<li>每次绘图都需要不断的调整参数，时间久了不用，就忘记参数放哪了；或者调整次数过多，有了很多版本，最后不知道用哪个了。</li>
</ol>
<p>为了简化绘图、维持脚本的一致，我用<code>bash</code>对<code>R</code>做了一个封装，然后就可以通过修改命令好参数绘制不同的图了。</p>
<p>先看一看怎么使用</p>
<p>首先把测试数据存储到文件中方便调用。数据矩阵存储在<code>heatmap_data.xls</code>文件中；行注释存储在<code>heatmap_row_anno.xls</code>文件中；列注释存储在<code>heatmap_col_anno.xls</code>文件中。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># tab键分割，每列不加引号</span>
<span class="kw">write.table</span>(data, <span class="dt">file=</span><span class="st">&quot;heatmap_data.xls&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span>T, <span class="dt">col.names=</span>T,<span class="dt">quote=</span>F)
<span class="co"># 如果看着第一行少了ID列不爽，可以填补下</span>
<span class="kw">system</span>(<span class="st">&quot;sed -i &#39;1 s/^/ID</span><span class="ch">\t</span><span class="st">/&#39; heatmap_data.xls&quot;</span>)

<span class="kw">write.table</span>(row_anno, <span class="dt">file=</span><span class="st">&quot;heatmap_row_anno.xls&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span>T, 
        <span class="dt">col.names=</span>T,<span class="dt">quote=</span>F)
<span class="kw">write.table</span>(col_anno, <span class="dt">file=</span><span class="st">&quot;heatmap_col_anno.xls&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span>T, 
        <span class="dt">col.names=</span>T,<span class="dt">quote=</span>F)</code></pre></div>
<p>然后用程序<code>sp_pheatmap.sh</code>绘图。</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># -f: 指定输入的矩阵文件</span>
<span class="co"># -d：指定是否计算Z-score，&lt;none&gt; (否), &lt;row&gt; (按行算), &lt;col&gt; (按列算)</span>
<span class="co"># -P: 行注释文件</span>
<span class="co"># -Q: 列注释文件</span>
<span class="kw">ct@ehbio</span>:~/$ sp_pheatmap.sh -f heatmap_data.xls -d row -P heatmap_row_anno.xls \
        -Q heatmap_col_anno.xls</code></pre></div>
<p><img src="images/pheatmap_6.png" width="100%" style="display: block; margin: auto;" /></p>
<p>字有点小，是因为图太大了，把图的宽和高缩小下试试。</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># -f: 指定输入的矩阵文件</span>
<span class="co"># -d：指定是否计算Z-score，&lt;none&gt; (否), &lt;row&gt; (按行算), &lt;col&gt; (按列算)</span>
<span class="co"># -P: 行注释文件</span>
<span class="co"># -Q: 列注释文件</span>
<span class="co"># -u: 设置宽度，单位是inch</span>
<span class="co"># -v: 设置高度，单位是inch</span>
<span class="kw">ct@ehbio</span>:~/$ sp_pheatmap.sh -f heatmap_data.xls -d row -P heatmap_row_anno.xls \
        -Q heatmap_col_anno.xls -u 8 -v 12</code></pre></div>
<p><img src="images/pheatmap_7.png" width="100%" style="display: block; margin: auto;" /></p>
<p>横轴的标记水平放置</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># -A: 0, X轴标签选择0度</span>
<span class="co"># -C: 自定义颜色，注意引号的使用，最外层引号与内层引号不同，引号之间无交叉</span>
<span class="co"># -T: 指定给定的颜色的类型；如果给的是vector (如下面的例子), 则-T需要指定为vector; </span>
<span class="co">#     否则结果会很怪异，只有俩颜色。</span>
<span class="co"># -t: 指定图形的题目，注意引号的使用；参数中包含空格或特殊字符等都要用引号引起来作为一个整体。</span>
<span class="kw">ct@ehbio</span>:~/$ sp_pheatmap.sh -f heatmap_data.xls -d row -P heatmap_row_anno.xls \
    -Q heatmap_col_anno.xls -u 8 -v 12 -A 0 -C <span class="st">&#39;c(&quot;white&quot;, &quot;blue&quot;)&#39;</span> -T vector \
    -t <span class="st">&quot;Heatmap of gene expression profile&quot;</span> </code></pre></div>
<p><code>sp_pheatmap.sh</code>的参数还有一些，可以完成前面讲述过的所有热图的绘制，具体如下：</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">***CREATED</span> BY Chen Tong (chentong_biology@163.com)<span class="kw">***</span>

<span class="kw">----Matrix</span> file--------------
<span class="kw">Name</span>    T0_1    T0_2    T0_3    T4_1    T4_2
<span class="kw">TR19267|c0_g1|CYP703A2</span>  1.431   0.77    1.309   1.247   0.485
<span class="kw">TR19612|c1_g3|CYP707A1</span>  0.72    0.161   0.301   2.457   2.794
<span class="kw">TR60337|c4_g9|CYP707A1</span>  0.056   0.09    0.038   7.643   15.379
<span class="kw">TR19612|c0_g1|CYP707A3</span>  2.011   0.689   1.29    0   0
<span class="kw">TR35761|c0_g1|CYP707A4</span>  1.946   1.575   1.892   1.019   0.999
<span class="kw">TR58054|c0_g2|CYP707A4</span>  12.338  10.016  9.387   0.782   0.563
<span class="kw">TR14082|c7_g4|CYP707A4</span>  10.505  8.709   7.212   4.395   6.103
<span class="kw">TR60509|c0_g1|CYP707A7</span>  3.527   3.348   2.128   3.257   2.338
<span class="kw">TR26914|c0_g1|CYP710A1</span>  1.899   1.54    0.998   0.255   0.427
<span class="kw">----Matrix</span> file--------------

<span class="kw">----Row</span> annorarion file --------------
<span class="kw">------1.</span> At least two columns--------------
<span class="kw">------2.</span> The first column should be the same as the first column in
         <span class="kw">matrix</span> (order does not matter)<span class="kw">--------------</span>
<span class="kw">Name</span>    Clan    Family
<span class="kw">TR19267|c0_g1|CYP703A2</span>  CYP71   CYP703
<span class="kw">TR19612|c1_g3|CYP707A1</span>  CYP85   CYP707
<span class="kw">TR60337|c4_g9|CYP707A1</span>  CYP85   CYP707
<span class="kw">TR19612|c0_g1|CYP707A3</span>  CYP85   CYP707
<span class="kw">TR35761|c0_g1|CYP707A4</span>  CYP85   CYP707
<span class="kw">TR58054|c0_g2|CYP707A4</span>  CYP85   CYP707
<span class="kw">TR14082|c7_g4|CYP707A4</span>  CYP85   CYP707
<span class="kw">TR60509|c0_g1|CYP707A7</span>  CYP85   CYP707
<span class="kw">TR26914|c0_g1|CYP710A1</span>  CYP710  CYP710
<span class="kw">----Row</span> annorarion file --------------

<span class="kw">----Column</span> annorarion file --------------
<span class="kw">------1.</span> At least two columns--------------
<span class="kw">------2.</span> The first column should be the same as the first row in
<span class="kw">---------matrix</span> (order does not matter)<span class="kw">--------------</span>
<span class="kw">Name</span>    Sample
<span class="kw">T0_1</span>    T0
<span class="kw">T0_2</span>    T0
<span class="kw">T0_3</span>    T0
<span class="kw">T4_1</span>    T4
<span class="kw">T4_2</span>    T4
<span class="kw">----Column</span> annorarion file --------------


<span class="kw">Usage</span>:

<span class="kw">sp_pheatmap.sh</span> options

<span class="kw">Function</span>:

<span class="kw">This</span> script is used to do heatmap using package pheatmap.

<span class="kw">The</span> parameters for logical variable are either TRUE or FALSE.

<span class="kw">OPTIONS</span>:
    <span class="kw">-f</span>  Data file (with header line, the first column is the
        <span class="kw">rowname</span>, tab seperated. Colnames must be unique unless you
        <span class="kw">know</span> what you are doing.)[<span class="kw">NECESSARY</span>]
    <span class="kw">-t</span>  Title of picture[Default empty title]
        [<span class="st">&quot;Heatmap of gene expression profile&quot;</span>]
    <span class="kw">-a</span>  Display xtics. [Default TRUE]
    <span class="kw">-A</span>  Rotation angle for x-axis value (anti clockwise)
        [<span class="kw">Default</span> 90]
    <span class="kw">-b</span>  Display ytics. [Default TRUE]
    <span class="kw">-H</span>  Hieratical cluster for columns.
        <span class="kw">Default</span> FALSE, accept TRUE
    <span class="kw">-R</span>  Hieratical cluster for rows.
        <span class="kw">Default</span> TRUE, accept FALSE
    <span class="kw">-c</span>  Clustering method, Default <span class="st">&quot;complete&quot;</span>. 
        <span class="kw">Accept</span> <span class="st">&quot;ward.D&quot;</span>, <span class="st">&quot;ward.D2&quot;</span>,<span class="st">&quot;single&quot;</span>, <span class="st">&quot;average&quot;</span> (=UPGMA), 
        <span class="st">&quot;mcquitty&quot;</span> <span class="kw">(</span>=<span class="kw">WPGMA)</span>, <span class="st">&quot;median&quot;</span> <span class="kw">(</span>=<span class="kw">WPGMC)</span> <span class="kw">or</span> <span class="st">&quot;centroid&quot;</span> (=UPGMC)
    <span class="kw">-C</span>  Color vector. 
        <span class="kw">Default</span> pheatmap_default. 
        <span class="kw">Aceept</span> a vector containing multiple colors such as 
        <span class="kw">&lt;</span><span class="st">&#39;c(&quot;white&quot;, &quot;blue&quot;)&#39;</span><span class="kw">&gt;</span> <span class="kw">will</span> be transferred 
        <span class="kw">to</span> <span class="kw">&lt;</span>colorRampPalette(c(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;blue&quot;</span>), <span class="ot">bias=</span>1)<span class="kw">(30)&gt;</span>
        <span class="kw">or</span> an R function 
        <span class="kw">&lt;colorRampPalette</span>(rev(brewer.pal(n=7, name=<span class="st">&quot;RdYlBu&quot;</span>)))<span class="kw">(100)&gt;</span>
        <span class="kw">generating</span> a list of colors.
        
    <span class="kw">-T</span>  Color type, a vetcor which will be transferred as described in <span class="kw">&lt;</span>-C<span class="kw">&gt;</span> [vector] or
        <span class="kw">a</span> raw vector [direct vector] or a function [function (default)]<span class="kw">.</span>
    <span class="kw">-B</span>  A positive number. Default 1. Values larger than 1 will give more color
        <span class="kw">for</span> <span class="kw">high</span> end. Values between 0-1 will give more color for low end.  
    <span class="kw">-D</span>  Clustering distance method for rows.
        <span class="kw">Default</span> <span class="st">&#39;correlation&#39;</span>, accept <span class="st">&#39;euclidean&#39;</span>, 
        <span class="st">&quot;manhattan&quot;</span>, <span class="st">&quot;maximum&quot;</span>, <span class="st">&quot;canberra&quot;</span>, <span class="st">&quot;binary&quot;</span>, <span class="st">&quot;minkowski&quot;</span><span class="kw">.</span> 
    <span class="kw">-I</span>  Clustering distance method for cols.
        <span class="kw">Default</span> <span class="st">&#39;correlation&#39;</span>, accept <span class="st">&#39;euclidean&#39;</span>, 
        <span class="st">&quot;manhattan&quot;</span>, <span class="st">&quot;maximum&quot;</span>, <span class="st">&quot;canberra&quot;</span>, <span class="st">&quot;binary&quot;</span>, <span class="st">&quot;minkowski&quot;</span><span class="kw">.</span> 
    <span class="kw">-L</span>  First get log-value, then do other analysis.
        <span class="kw">Accept</span> an R function log2 or log10. 
        [<span class="kw">Default</span> FALSE]
    <span class="kw">-d</span>  Scale the data or not for clustering and visualization.
        [<span class="kw">Default</span> <span class="st">&#39;none&#39;</span> means no scale, accept <span class="st">&#39;row&#39;</span>, <span class="st">&#39;column&#39;</span> to 
        <span class="kw">scale</span> by row or column.]
    <span class="kw">-m</span>  The maximum value you want to keep, any number larger willl
        <span class="kw">be</span> taken as this given maximum value.
        [<span class="kw">Default</span> Inf, Optional] 
    <span class="kw">-s</span>  The smallest value you want to keep, any number smaller will
        <span class="kw">be</span> taken as this given minimum value.
        [<span class="kw">Default</span> -Inf, Optional]  
    <span class="kw">-k</span>  Aggregate the rows using kmeans clustering. 
        <span class="kw">This</span> is advisable if number of rows is so big that R cannot 
        <span class="kw">handle</span> their hierarchical clustering anymore, roughly more than 1000.
        <span class="kw">Instead</span> of showing all the rows separately one can cluster the
        <span class="kw">rows</span> in advance and show only the cluster centers. The number
        <span class="kw">of</span> clusters can be tuned here.
        [<span class="kw">Default</span> <span class="st">&#39;NA&#39;</span> which means no
        <span class="kw">cluster</span>, other positive interger is accepted for executing
        <span class="kw">kmeans</span> cluster, also the parameter represents the number of
        <span class="kw">expected</span> clusters.]
    <span class="kw">-P</span>  A file to specify row-annotation with format described above.
        [<span class="kw">Default</span> NA]
    <span class="kw">-Q</span>  A file to specify col-annotation with format described above.
        [<span class="kw">Default</span> NA]
    <span class="kw">-u</span>  The width of output picture.[Default 20]
    <span class="kw">-v</span>  The height of output picture.[Default 20] 
    <span class="kw">-E</span>  The type of output figures.[Default pdf, accept
        <span class="kw">eps</span>/<span class="kw">ps</span>, tex (pictex), <span class="kw">png</span>, jpeg, tiff, bmp, svg and wmf)]
    <span class="kw">-r</span>  The resolution of output picture.[Default 300 ppi]
    <span class="kw">-F</span>  Font size [Default 14]
    <span class="kw">-p</span>  Preprocess data matrix to avoid <span class="st">&#39;STDERR 0 in cor(t(mat))&#39;</span>.
        <span class="kw">Lowercase</span> <span class="kw">&lt;</span>p<span class="kw">&gt;</span>.
        [<span class="kw">Default</span> TRUE]
    <span class="kw">-e</span>  Execute script (Default) <span class="kw">or</span> just output the script.
        [<span class="kw">Default</span> TRUE]
    <span class="kw">-i</span>  Install the required packages. Normmaly should be TRUE if this is 
        <span class="kw">your</span> first time run s-plot.[Default FALSE]</code></pre></div>
<div id="---" class="section level3">
<h3><span class="header-section-number">3.16.1</span> 箱线图 - 一步绘制</h3>
<p>首先把测试数据存储到文件中方便调用。数据矩阵存储在<code>boxplot.normal.data</code>、<code>sampleGroup</code>和<code>boxplot.melt.data</code>文件中 (TAB键分割，内容在文档最后。如果你手上有自己的数据，也可以拿来用)。</p>
<p>使用正常矩阵默认参数绘制箱线图</p>
<pre><code># -f: 指定输入的矩阵文件，第一列为行名字，第一行为header
      列数不限，列名字不限；行数不限，行名字默认为文本
sp_boxplot.sh -f boxplot.normal.data</code></pre>
<p>箱线图出来了，但有点小乱。</p>
<p><img src="images/boxplot_6.png" width="100%" style="display: block; margin: auto;" /></p>
<pre><code># -f: 指定输入的矩阵文件，第一列为行名字，第一行为header
      列数不限，列名字不限；行数不限，行名字默认为文本
# -P: none, 去掉legend (uppercase P)
# -b: X-axis旋转45度
# -V: TRUE 绘制小提琴图
sp_boxplot.sh -f boxplot.normal.data -P none -b 45 -V TRUE</code></pre>
<p><img src="images/boxplot_7.png" width="100%" style="display: block; margin: auto;" /></p>
<p>绘制单个基因的小提琴图加抖动图</p>
<pre><code># -q: 指定某一行的名字，此处为基因名，绘制基因A的表达图谱
# -Q: 指定样本分组，绘制基因A在不同样品组的表达趋势
# -F Group: sampleGroup中第二列的名字，指代分组信息，根据需要修改
# -J TRUE: 绘制抖动图 jitter plot
# -L: 设置X轴样品组顺序
# -c TRUE -C &quot;&#39;red&#39;, &#39;pink&#39;, &#39;blue&#39;&quot;: 指定每个箱线图的颜色
sp_boxplot.sh -f boxplot.normal.data -q A -Q sampleGroup -F Group -V TRUE -J TRUE \
        -L &quot;&#39;zygote&#39;,&#39;2cell&#39;,&#39;4cell&#39;&quot; -c TRUE -C &quot;&#39;red&#39;, &#39;pink&#39;, &#39;blue&#39;&quot; -P none</code></pre>
<p><img src="images/boxplot_8.png" width="100%" style="display: block; margin: auto;" /></p>
<p>使用melted矩阵默认参数绘箱线图</p>
<pre><code># -f: 指定输入文件
# -m TRUE: 指定输入的矩阵为melted format
# -d Expr：指定表达值所在的列
# -F Rep: 指定子类所在列，也就是legend 
# -a Group：指定X轴分组信息
# -j TRUE: jitter plot
sp_boxplot.sh -f boxplot.melt.data -m TRUE -d Expr -F Rep -a Group  -j TRUE</code></pre>
<p><img src="images/boxplot_9.png" width="100%" style="display: block; margin: auto;" /></p>
<pre><code># 如果没有子类，则-a和-F指定为同一值
# -R TRUE: 旋转boxplot
sp_boxplot.sh -f boxplot.melt.data -m TRUE -d Expr -a Group -F Group -J TRUE -R TRUE</code></pre>
<p><img src="images/boxplot_10.png" width="100%" style="display: block; margin: auto;" /></p>
<p>参数中最需要注意的是<strong>引号</strong>的使用：</p>
<ul>
<li>外层引号与内层引号不能相同</li>
<li>凡参数值中包括了<code>空格</code>，<code>括号</code>，<code>逗号</code>等都用引号括起来作为一个整体。</li>
</ul>
<pre><code>#boxplot.normal.data
Name    2cell_1 2cell_2 2cell_3 2cell_4 2cell_5 2cell_6 4cell_1 4cell_2 4cell_3 4cell_4 4cell_5 4cell_6 zygote_1    zygote_2    zygote_3    zygote_4    zygote_5    zygote_6
A   4   6   7   5   8   6   3.2 5.2 5.6 3.6 7.6 4.8 2   4   3   2   4   2.5
B   6   8   9   7   10  8   5.2 7.2 7.6 5.6 9.6 6.8 4   6   5   4   6   4.5
C   8   10  11  9   12  10  7.2 9.2 9.6 7.6 11.6    8.8 6   8   7   6   8   6.5
D   10  12  13  11  14  12  9.2 11.2    11.6    9.6 13.6    10.8    8   10  9   8   10  8.5
E   12  14  15  13  16  14  11.2    13.2    13.6    11.6    15.6    12.8    10  12  11  10  12  10.5
F   14  16  17  15  18  16  13.2    15.2    15.6    13.6    17.6    14.8    12  14  13  12  14  12.5
G   15  17  18  16  19  17  14.2    16.2    16.6    14.6    18.6    15.8    13  15  14  13  15  13.5
H   16  18  19  17  20  18  15.2    17.2    17.6    15.6    19.6    16.8    14  16  15  14  16  14.5
I   17  19  20  18  21  19  16.2    18.2    18.6    16.6    20.6    17.8    15  17  16  15  17  15.5
J   18  20  21  19  22  20  17.2    19.2    19.6    17.6    21.6    18.8    16  18  17  16  18  16.5
L   19  21  22  20  23  21  18.2    20.2    20.6    18.6    22.6    19.8    17  19  18  17  19  17.5
M   20  22  23  21  24  22  19.2    21.2    21.6    19.6    23.6    20.8    18  20  19  18  20  18.5
N   21  23  24  22  25  23  20.2    22.2    22.6    20.6    24.6    21.8    19  21  20  19  21  19.5
O   22  24  25  23  26  24  21.2    23.2    23.6    21.6    25.6    22.8    20  22  21  20  22  20.5</code></pre>
<pre><code>#boxplot.melt.data

Gene    Sample  Group   Expr    Rep
A   zygote_1    zygote  2   1
A   zygote_2    zygote  4   2
A   zygote_3    zygote  3   3
A   zygote_4    zygote  2   4
A   zygote_5    zygote  4   5
A   zygote_6    zygote  2.5 6
A   2cell_1 2cell   4   1
A   2cell_2 2cell   6   2
A   2cell_3 2cell   7   3
A   2cell_4 2cell   5   4
A   2cell_5 2cell   8   5
A   2cell_6 2cell   6   6
A   4cell_1 4cell   3.2 1
A   4cell_2 4cell   5.2 2
A   4cell_3 4cell   5.6 3
A   4cell_4 4cell   3.6 4
A   4cell_5 4cell   7.6 5
A   4cell_6 4cell   4.8 6</code></pre>
<pre><code>#sampleGroup
Sample  Group
zygote_1    zygote
zygote_2    zygote
zygote_3    zygote
zygote_4    zygote
zygote_5    zygote
zygote_6    zygote
2cell_1 2cell
2cell_2 2cell
2cell_3 2cell
2cell_4 2cell
2cell_5 2cell
2cell_6 2cell
4cell_1 4cell
4cell_2 4cell
4cell_3 4cell
4cell_4 4cell
4cell_5 4cell
4cell_6 4cell</code></pre>
</div>
<div id="---" class="section level3">
<h3><span class="header-section-number">3.16.2</span> 线图 - 一步绘制</h3>
<p>首先把测试数据存储到文件中方便调用。数据矩阵存储在<code>line_data.xls</code>和<code>line_data_melt.xls</code>文件中 (直接拷贝到文件中也可以，这里这么操作只是为了随文章提供个测试文件，方便使用。如果你手上有自己的数据，也可以拿来用)。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">profile =<span class="st"> &quot;Pos;H3K27ac;CTCF;Enhancer;H3K4me3;polII</span>
<span class="st">-5000;8.7;10.7;11.7;10;8.3</span>
<span class="st">-4000;8.4;10.8;11.8;9.8;7.8</span>
<span class="st">-3000;8.3;10.5;12.2;9.4;7</span>
<span class="st">-2000;7.2;10.9;12.7;8.4;4.8</span>
<span class="st">-1000;3.6;8.5;12.8;4.8;1.3</span>
<span class="st">0;3.6;8.5;13.4;5.2;1.5</span>
<span class="st">1000;7.1;10.9;12.4;8.1;4.9</span>
<span class="st">2000;8.2;10.7;12.4;9.5;7.7</span>
<span class="st">3000;8.4;10.4;12;9.8;7.9</span>
<span class="st">4000;8.5;10.6;11.7;9.7;8.2</span>
<span class="st">5000;8.5;10.6;11.7;10;8.2&quot;</span>

profile_text &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>profile, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>)
<span class="co"># tab键分割，每列不加引号</span>
<span class="kw">write.table</span>(profile_text, <span class="dt">file=</span><span class="st">&quot;line_data.xls&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span>T, <span class="dt">col.names=</span>T,<span class="dt">quote=</span>F)
<span class="co"># 如果看着第一行少了ID列不爽，可以填补下</span>
<span class="co">#system(&quot;sed -i &#39;1 s/^/ID\t/&#39; line_data.xls&quot;)</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">profile =<span class="st"> &quot;Pos;variable;value;set</span>
<span class="st">-5000;H3K27ac;8.71298;A</span>
<span class="st">-4000;H3K27ac;8.43246;A</span>
<span class="st">-3000;H3K27ac;8.25497;A</span>
<span class="st">-2000;H3K27ac;7.16265;A</span>
<span class="st">-1000;H3K27ac;3.55341;A</span>
<span class="st">0;H3K27ac;3.5503;A</span>
<span class="st">1000;H3K27ac;7.07502;A</span>
<span class="st">2000;H3K27ac;8.24328;A</span>
<span class="st">3000;H3K27ac;8.43869;A</span>
<span class="st">4000;H3K27ac;8.48877;A</span>
<span class="st">-5000;CTCF;10.6913;A</span>
<span class="st">-4000;CTCF;10.7668;A</span>
<span class="st">-3000;CTCF;10.5441;A</span>
<span class="st">-2000;CTCF;10.8635;A</span>
<span class="st">-1000;CTCF;8.45751;A</span>
<span class="st">0;CTCF;8.50316;A</span>
<span class="st">1000;CTCF;10.9143;A</span>
<span class="st">2000;CTCF;10.7022;A</span>
<span class="st">3000;CTCF;10.4101;A</span>
<span class="st">4000;CTCF;10.5757;A</span>
<span class="st">-5000;H3K27ac;8.71298;B</span>
<span class="st">-4000;H3K27ac;8.43246;B</span>
<span class="st">-3000;H3K27ac;8.25497;B</span>
<span class="st">-2000;H3K27ac;7.16265;B</span>
<span class="st">-1000;H3K27ac;3.55341;B</span>
<span class="st">0;H3K27ac;3.5503;B</span>
<span class="st">1000;H3K27ac;7.07502;B</span>
<span class="st">2000;H3K27ac;8.24328;B</span>
<span class="st">3000;H3K27ac;8.43869;B</span>
<span class="st">4000;H3K27ac;8.48877;B</span>
<span class="st">-5000;CTCF;10.6913;B</span>
<span class="st">-4000;CTCF;10.7668;B</span>
<span class="st">-3000;CTCF;10.5441;B</span>
<span class="st">-2000;CTCF;10.8635;B</span>
<span class="st">-1000;CTCF;8.45751;B</span>
<span class="st">0;CTCF;8.50316;B</span>
<span class="st">1000;CTCF;10.9143;B</span>
<span class="st">2000;CTCF;10.7022;B</span>
<span class="st">3000;CTCF;10.4101;B</span>
<span class="st">4000;CTCF;10.5757;B&quot;</span>

profile_text &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">text=</span>profile, <span class="dt">header=</span>T, <span class="dt">quote=</span><span class="st">&quot;&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;;&quot;</span>)
<span class="co"># tab键分割，每列不加引号</span>
<span class="kw">write.table</span>(profile_text, <span class="dt">file=</span><span class="st">&quot;line_data_melt.xls&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span>T, <span class="dt">col.names=</span>T,<span class="dt">quote=</span>F)
<span class="co"># 如果看着第一行少了ID列不爽，可以填补下</span>
<span class="co"># system(&quot;sed -i &#39;1 s/^/ID\t/&#39; line_data_melt.xls&quot;)</span></code></pre></div>
<p>使用正常矩阵默认参数绘制个线图</p>
<pre><code># -f: 指定输入的矩阵文件，第一列为行名字，第一行为header
      列数不限，列名字不限；行数不限，行名字默认为文本
# -A FALSE: 指定行名为数字
sp_lines.sh -f line_data.xls -A FALSE</code></pre>
<p><img src="images/line_9.png" width="100%" style="display: block; margin: auto;" /></p>
<pre><code># -l: 设定图例的顺序
# -o TRUE: 局部拟合获得平滑曲线
# -A FALSE: 指定行名为数字
# -P: 设置legend位置，相对于原点的坐标
# -x, -y指定横纵轴标记
sp_lines.sh -f line_data.xls -l &quot;&#39;CTCF&#39;,&#39;Enhancer&#39;,&#39;polII&#39;,&#39;H3K4me3&#39;,&#39;H3K27ac&#39;&quot; \
    -P &#39;c(0.8,0.3)&#39; -o TRUE -A FALSE -x &#39;Up and down 5 kb of TSS&#39; -y &#39;Relative density&#39;</code></pre>
<p><img src="images/line_10.png" width="100%" style="display: block; margin: auto;" /></p>
<pre><code># -A FALSE: 指定行名为数字
# -V &#39;c(-1000, 500)&#39;: 设置垂线的位置
# -D: 设置垂线的文本标记，参数为引号引起来的vector，注意引号的嵌套
# -I: 设置横轴的标记的位置
# -b: 设置横轴标记的文字
sp_lines.sh -f line_data.xls -A FALSE -V &#39;c(-1000,500)&#39; -D &quot;c(&#39;+1 kb&#39;,&#39;-0.5 kb&#39;)&quot; \
    -I &quot;c(-5000,0,5000)&quot; -b &quot;c(&#39;-5 kb&#39;, &#39;TSS&#39;, &#39;+5 kb&#39;)&quot;</code></pre>
<p><img src="images/line_11.png" width="100%" style="display: block; margin: auto;" /></p>
<p>使用melted矩阵默认参数绘制个线图 (除需要改变文件格式，指定<code>-m TRUE</code>, <code>-a xvariable</code>外其它与正常矩阵一样)</p>
<pre><code># -f: 指定输入文件
# -m TRUE: 指定输入的矩阵为melted format, 三列，第一列为Pos (给-a)
#          第二列为variable (给-H，-H默认即为variable)
#          第三列为value，名字不可修改
# -A FALSE: 指定行名为数字
# -P &#39;c(0.8,0.2)&#39;: 设置legend位置，相对于原点的坐标
sp_lines.sh -f line_data_melt.xls -a Pos -m TRUE -A FALSE -P &#39;c(0.8,0.2)&#39;</code></pre>
<p><img src="images/line_12.png" width="100%" style="display: block; margin: auto;" /></p>
<p>完整的图</p>
<pre><code># -C: 自定义线的颜色
sp_lines.sh -f line_data_melt.xls -a Pos -m TRUE -A FALSE -P &#39;c(0.8,0.2)&#39; -o TRUE \
    -V &#39;c(-1000,500)&#39; -D &quot;c(&#39;+1 kb&#39;,&#39;-0.5 kb&#39;)&quot; \
    -I &quot;c(-5000,0,4000)&quot; -b &quot;c(&#39;-5 kb&#39;, &#39;TSS&#39;, &#39;+4 kb&#39;)&quot; \
    -x &#39;Up 5 kb and down 4 kb of TSS&#39; -y &#39;Relative density&#39; -C &quot;&#39;pink&#39;, &#39;blue&#39;&quot;</code></pre>
<p><img src="images/line_13.png" width="100%" style="display: block; margin: auto;" /></p>
<p>参数中最需要注意的是<strong>引号</strong>的使用：</p>
<ul>
<li>外层引号与内层引号不能相同</li>
<li>凡参数值中包括了<code>空格</code>，<code>括号</code>，<code>逗号</code>等都用引号括起来作为一个整体。</li>
</ul>
<p>完整参数列表如下：</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">ct@ehbio</span>:~ <span class="ot">$sp_lines</span>.sh

<span class="kw">***CREATED</span> BY Chen Tong (chentong_biology@163.com)<span class="kw">***</span>

<span class="kw">Usage</span>:

<span class="kw">/MPATHB/self/s-plot/sp_lines.sh</span> options

<span class="kw">Function</span>:

<span class="kw">This</span> script is used to draw a line or multiple lines using ggplot2.
<span class="kw">You</span> can specify whether or not smooth your line or lines.

<span class="kw">Two</span> types of input files are supported, normal matrix or melted matrix format. Column separator for both types of input files is **tab**. 

<span class="kw">Here</span> is an example of normal matrix format. The first column will be treated as X-axis variables and other columns represents each type of lines. The number of columns is unlimited and names of columns is unlimited.

<span class="kw">**Set**</span> column is not needed. If given, <span class="kw">&lt;</span>facet_plot<span class="kw">&gt;</span> (multiple plots in one page) <span class="kw">could</span> be displayed. 

<span class="kw">------------------------------------------------------------</span>
<span class="kw">Pos</span> H3K27ac CTCF    Enhancer    H3K4me3 polII
<span class="kw">-5000</span>   8.71298 10.69130    11.7359 10.02510    8.26866
<span class="kw">-4000</span>   8.43246 10.76680    11.8442 9.76927 7.78358
<span class="kw">-3000</span>   8.25497 10.54410    12.2470 9.40346 6.96859
<span class="kw">-2000</span>   7.16265 10.86350    12.6889 8.35070 4.84365
<span class="kw">-1000</span>   3.55341 8.45751 12.8372 4.84680 1.26110
<span class="kw">0</span>   3.55030 8.50316 13.4152 5.17401 1.50022
<span class="kw">1000</span>    7.07502 10.91430    12.3588 8.13909 4.88096
<span class="kw">2000</span>    8.24328 10.70220    12.3888 9.47255 7.67968
<span class="kw">3000</span>    8.43869 10.41010    11.9760 9.80665 7.94148
<span class="kw">4000</span>    8.48877 10.57570    11.6562 9.71986 8.17849
<span class="kw">------------------------------------------------------</span>

<span class="kw">------------With</span> SET------------------------------------------
<span class="kw">Pos</span> H3K27ac CTCF    Enhancer    H3K4me3 polII   Set
<span class="kw">-5000</span>   8.71298 10.69130    11.7359 10.02510    8.26866 1
<span class="kw">-4000</span>   8.43246 10.76680    11.8442 9.76927 7.78358 1
<span class="kw">-3000</span>   8.25497 10.54410    12.2470 9.40346 6.96859 1
<span class="kw">-2000</span>   7.16265 10.86350    12.6889 8.35070 4.84365 1
<span class="kw">-1000</span>   3.55341 8.45751 12.8372 4.84680 1.26110 1
<span class="kw">0</span>   3.55030 8.50316 13.4152 5.17401 1.50022 1
<span class="kw">1000</span>    7.07502 10.91430    12.3588 8.13909 4.88096 1
<span class="kw">2000</span>    8.24328 10.70220    12.3888 9.47255 7.67968 1
<span class="kw">3000</span>    8.43869 10.41010    11.9760 9.80665 7.94148 1
<span class="kw">4000</span>    8.48877 10.57570    11.6562 9.71986 8.17849 1
<span class="kw">-5000</span>   8.71298 10.69130    11.7359 10.02510    8.26866 2
<span class="kw">-4000</span>   8.43246 10.76680    11.8442 9.76927 7.78358 2
<span class="kw">-3000</span>   8.25497 10.54410    12.2470 9.40346 6.96859 2
<span class="kw">-2000</span>   7.16265 10.86350    12.6889 8.35070 4.84365 2
<span class="kw">-1000</span>   3.55341 8.45751 12.8372 4.84680 1.26110 2
<span class="kw">0</span>   3.55030 8.50316 13.4152 5.17401 1.50022 2
<span class="kw">1000</span>    7.07502 10.91430    12.3588 8.13909 4.88096 2
<span class="kw">2000</span>    8.24328 10.70220    12.3888 9.47255 7.67968 2
<span class="kw">3000</span>    8.43869 10.41010    11.9760 9.80665 7.94148 2
<span class="kw">4000</span>    8.48877 10.57570    11.6562 9.71986 8.17849 2
<span class="kw">-------------------------------------------------------------</span>

<span class="kw">For</span> matrix format, example command lines include:

<span class="kw">*</span> Attribute of X-axis value (first column of matrix) <span class="kw">is</span> <span class="kw">&lt;</span>number<span class="kw">&gt;</span>

    <span class="kw">s-plot</span> lines -f matrix.file -A FALSE

<span class="kw">*</span> Attribute of X-axis value (first column of matrix) <span class="kw">is</span> <span class="kw">&lt;</span>text<span class="kw">&gt;</span>
    <span class="kw">s-plot</span> lines -f matrix.file 

<span class="kw">*</span> Attribute of X-axis value (first column of matrix) <span class="kw">is</span> numbers, change legned order (default alphabet order)

    <span class="kw">s-plot</span> lines -f matrix.file -l <span class="st">&quot;&#39;polII&#39;, &#39;CTCF&#39;, &#39;Enhancer&#39;, &#39;H3K27ac&#39;, &#39;H3K4me3&#39;&quot;</span>

<span class="kw">*</span> Attribute of X-axis value (first column of matrix) <span class="kw">is</span> numbers, change legned order (default alphabet order), <span class="kw">smooth</span> lines to look better (Pay attention to whether this will change the data trend)

    <span class="kw">s-plot</span> lines -f matrix.file -l <span class="st">&quot;&#39;polII&#39;, &#39;CTCF&#39;, &#39;Enhancer&#39;, &#39;H3K27ac&#39;, &#39;H3K4me3&#39;&quot;</span> -o TRUE

<span class="kw">*</span> Attribute of X-axis value (first column of matrix) <span class="kw">is</span> numbers, with <span class="kw">&lt;</span>Set<span class="kw">&gt;</span> (Set is column name) <span class="kw">column</span>
    
    <span class="kw">s-plot</span> lines -f matrix.file -F <span class="st">&quot;+facet_grid(Set ~ ., scale=&#39;free_y&#39;)&quot;</span>


<span class="kw">FILEFORMAT</span> when -m is true
<span class="co">#The name &quot;value&quot; shoud **not** be altered.</span>
<span class="co">#variable can be altered using -H</span>
<span class="co">#Actually this format is the melted result of last format.</span>
<span class="kw">--------------------------------------------------------------</span>
<span class="kw">Pos</span> variable    value
<span class="kw">-5000</span>   H3K27ac 8.71298
<span class="kw">-4000</span>   H3K27ac 8.43246
<span class="kw">-3000</span>   H3K27ac 8.25497
<span class="kw">-2000</span>   H3K27ac 7.16265
<span class="kw">-1000</span>   H3K27ac 3.55341
<span class="kw">0</span>   H3K27ac 3.55030
<span class="kw">1000</span>    H3K27ac 7.07502
<span class="kw">2000</span>    H3K27ac 8.24328
<span class="kw">3000</span>    H3K27ac 8.43869
<span class="kw">4000</span>    H3K27ac 8.48877
<span class="kw">-5000</span>   CTCF    10.69130
<span class="kw">-4000</span>   CTCF    10.76680
<span class="kw">-3000</span>   CTCF    10.54410
<span class="kw">-2000</span>   CTCF    10.86350
<span class="kw">-1000</span>   CTCF    8.45751
<span class="kw">0</span>   CTCF    8.50316
<span class="kw">1000</span>    CTCF    10.91430
<span class="kw">2000</span>    CTCF    10.70220
<span class="kw">3000</span>    CTCF    10.41010
<span class="kw">4000</span>    CTCF    10.57570
<span class="kw">-------------------------------------------------------------</span>

<span class="kw">*</span> Attribute of X-axis value (melt format) <span class="kw">is</span> <span class="kw">&lt;</span>number<span class="kw">&gt;</span>

    <span class="kw">s-plot</span> lines -f matrix.file -m TRUE -a Pos -A FALSE

<span class="kw">*</span> Attribute of X-axis value (first column of matrix) <span class="kw">is</span> <span class="kw">&lt;</span>text<span class="kw">&gt;</span>

    <span class="kw">s-plot</span> lines -f matrix.file -m TRUE -a Pos

<span class="kw">*</span> If the name of the second column is <span class="kw">&lt;</span>type<span class="kw">&gt;</span> not <span class="kw">&lt;</span>variable<span class="kw">&gt;</span>, one should specify with <span class="kw">&lt;</span>-H<span class="kw">&gt;</span>. 
    
    <span class="kw">s-plot</span> lines -f matrix.file -A FALSE -m TRUE -a Pos -H type

<span class="kw">*</span> Attribute of X-axis value (first column of matrix) <span class="kw">is</span> numbers, change legned order (default alphabet order)

    <span class="kw">s-plot</span> lines -f matrix.file -m TRUE -a Pos -l <span class="st">&quot;&#39;polII&#39;, &#39;CTCF&#39;, &#39;Enhancer&#39;, &#39;H3K27ac&#39;, &#39;H3K4me3&#39;&quot;</span>

<span class="kw">*</span> Attribute of X-axis value (first column of matrix) <span class="kw">is</span> numbers, change legned order (default alphabet order), <span class="kw">smooth</span> lines to look better (Pay attention to whether this will change the data trend)

    <span class="kw">s-plot</span> lines -f matrix.file -m TRUE -a Pos -l <span class="st">&quot;&#39;polII&#39;, &#39;CTCF&#39;, &#39;Enhancer&#39;, &#39;H3K27ac&#39;, &#39;H3K4me3&#39;&quot;</span> -o TRUE

<span class="kw">*</span> Attribute of X-axis value (first column of matrix) <span class="kw">is</span> numbers, with <span class="kw">&lt;</span>Set<span class="kw">&gt;</span> (Set is column name) <span class="kw">column</span>
    
    <span class="kw">s-plot</span> lines -f matrix.file -F <span class="st">&quot;+facet_grid(Set ~ ., scale=&#39;free_y&#39;)&quot;</span>


<span class="kw">OPTIONS</span>:
    <span class="kw">-f</span>  Data file (with header line, the first column would be be treated as rownames for
        <span class="kw">normal</span> matrix. No rownames for melted format. Columns are tab seperated)
        [<span class="kw">NECESSARY</span>]
    <span class="kw">-m</span>  When true, it will skip melt preprocesses. But the format must be
        <span class="kw">the</span> same as listed before.
        [<span class="kw">Default</span> FALSE, accept TRUE]
    <span class="kw">-a</span>  Name for x-axis variable
        [<span class="kw">Only</span> needed when <span class="kw">&lt;</span>-m<span class="kw">&gt;</span> is <span class="kw">&lt;</span>TRUE<span class="kw">&gt;</span>.  
        <span class="kw">For</span> the melted data, <span class="st">&#39;Pos&#39;</span> should be given here. 
        <span class="kw">For</span> normal matrix,  default the first column will be used,
        <span class="kw">program</span> will assign an value <span class="st">&#39;xvariable&#39;</span> to represent it.
        ]]
    <span class="kw">-A</span>  Are x-axis variables numbers.
        [<span class="kw">Default</span> <span class="kw">&lt;</span>TRUE<span class="kw">&gt;</span>, meaning X-axis label is <span class="kw">&lt;</span>text<span class="kw">&gt;</span>.
        <span class="kw">&lt;FALSE&gt;</span> means X-axis label is <span class="kw">&lt;</span>numerical<span class="kw">&gt;</span>.]
    <span class="kw">-H</span>  Name for legend variable.
        [<span class="kw">Default</span> variable, this should only be set when -m is TRUE]
    <span class="kw">-J</span>  Name for color variable.
        [<span class="kw">Default</span> same as -H, this should only be set when -m is TRUE]
    <span class="kw">-l</span>  Set orders of legend variable.
        [<span class="kw">Default</span> column order for normal matrix, accept a string like
        <span class="st">&quot;&#39;CTCF&#39;,&#39;H3K27ac&#39;,&#39;Enhancer&#39;&quot;</span> <span class="kw">to</span> set your own order. 
        <span class="kw">Pay</span> attention to the usage of two types of quotes. 
        <span class="kw">***When</span> -m is TRUE, default order would be alphabet order.********* 
        ]
    <span class="kw">-P</span>  Legend position[Default right. Accept
        <span class="kw">top</span>, bottom, left, none, or <span class="st">&#39;c(0.08,0.8)&#39;</span>.]
    <span class="kw">-L</span>  Levels for x-axis variable, suitable when x-axis is not treated as numerical. 
        [<span class="kw">Default</span> the order of first column for normal matrix. 
        <span class="kw">Accept</span> a string like <span class="st">&quot;&#39;g&#39;,&#39;a&#39;,&#39;j&#39;,&#39;x&#39;,&#39;s&#39;,&#39;c&#39;,&#39;o&#39;,&#39;u&#39;&quot;</span> to set your own oder.
        <span class="kw">This</span> will only be considered when -A is TRUE.
        <span class="kw">***When</span> -m is used, this default order would be alphabet order.********* 
        ]
    <span class="kw">-o</span>  Smooth lines or not.
        [<span class="kw">Default</span> FALSE means no smooth. Accept TRUE to smooth lines.]
    <span class="kw">-O</span>  The smooth method you want to use.
        [<span class="kw">smoothing</span> method (function) <span class="kw">to</span> use,  eg. lm, glm, gam, loess,rlm.
        <span class="kw">For</span> datasets with n <span class="kw">&lt;</span> 1000 default is <span class="st">&#39;loess&#39;</span>. 
        <span class="kw">For</span> datasets with 1000 or more observations defaults to <span class="st">&#39;gam&#39;</span>.
        ]
    <span class="kw">-V</span>  Add vertical lines.[Default FALSE, accept a series of
        <span class="kw">numbers</span> in following format <span class="st">&quot;c(1,2,3,4,5)&quot;</span> or other
        <span class="kw">R</span> code that can generate a vector.]
    <span class="kw">-D</span>  Add labels to vlines.
        [<span class="kw">Default</span> same as -V.
        <span class="kw">Accept</span> a series of numbers in following format <span class="st">&quot;c(1,2,3,4,5)&quot;</span> or other R code
        <span class="kw">that</span> can generate a vector as labels.
        <span class="kw">Or</span> one can give <span class="st">&#39;1&#39;</span> to disallow labels]
    <span class="kw">-j</span>  Add horizontal lines.[Default FALSE, accept a series of
        <span class="kw">numbers</span> in following format <span class="st">&quot;c(1,2,3,4,5)&quot;</span> or other
        <span class="kw">R</span> code that can generate a vector]
    <span class="kw">-d</span>  Add labels to hline.
        [<span class="kw">Default</span> same as -j
        <span class="kw">Accept</span> a series of numbers in following format <span class="st">&quot;c(1,2,3,4,5)&quot;</span> or other R code
        <span class="kw">that</span> can generate a vector as labels.
        <span class="kw">Or</span> one can give <span class="st">&#39;1&#39;</span> to disallow labels]
    <span class="kw">-I</span>  Manually set the position of xtics.
        [<span class="kw">Default</span> FALSE,  accept a series of
        <span class="kw">numbers</span> in following format <span class="st">&quot;c(1,2,3,4,5)&quot;</span> or other R code
        <span class="kw">that</span> can generate a vector to set the position of xtics]
    <span class="kw">-b</span>  Manually set the value of xtics when -I is specified.
        [<span class="kw">Default</span> the content of -I when -I is specified, 
        <span class="kw">accept</span> a series of numbers in following format <span class="st">&quot;c(1,2,3,4,5)&quot;</span> or other R code
        <span class="kw">that</span> can generate a vector to set the position of xtics]
    <span class="kw">-X</span>  Display xtics. [Default TRUE]
    <span class="kw">-Y</span>  Display ytics. [Default TRUE]
    <span class="kw">-R</span>  Rotation angle for x-axis labels (anti clockwise)
        [<span class="kw">Default</span> 0]
    <span class="kw">-B</span>  line size. [Default 1. Accept a number.]
    <span class="kw">-t</span>  Title of picture[Default empty title]
    <span class="kw">-x</span>  xlab of picture[Default empty xlab]
    <span class="kw">-y</span>  ylab of picture[Default empty ylab]
    <span class="kw">-c</span>  Manually set colors for each line.[Default FALSE, meaning using ggplot2 default.]
    <span class="kw">-C</span>  Color for each line.
        <span class="kw">When</span> -c is TRUE, one has two options:

        <span class="kw">1.</span> Supplying a function to generate colors, 
        <span class="kw">like</span> <span class="st">&quot;rainbow(11)&quot;</span> or <span class="st">&quot;rainbow(11, alpha=0.6)&quot;</span>, 
            <span class="kw">rainbow</span> is an R color palletes, 
            <span class="kw">11</span> is the number of colors you want to get, 
            <span class="kw">0.6</span> is the alpha value.
        <span class="kw">The</span> R palletes include <span class="kw">&lt;</span>heat.colors<span class="kw">&gt;</span>, <span class="kw">&lt;</span>terrain.colors<span class="kw">&gt;</span>,
            <span class="kw">&lt;topo.colors&gt;</span>, <span class="kw">&lt;</span>cm.colors<span class="kw">&gt;</span>.

        <span class="kw">2.</span> Supplying a list of colors in given format, 
        <span class="kw">the</span> number of colors should be equal to the number of
        <span class="kw">bars</span> like <span class="st">&quot;&#39;red&#39;,&#39;pink&#39;,&#39;blue&#39;,&#39;cyan&#39;,&#39;green&#39;,&#39;yellow&#39;&quot;</span> or
        <span class="st">&quot;rgb(255/255,0/255,0/255),rgb(255/255,0/255,255/255),</span>
<span class="st">         rgb(0/255,0/255,255/255),rgb(0/255,255/255,255/255),</span>
<span class="st">         rgb(0/255,255/255,0/255),rgb(255/255,255/255,0/255)&quot;</span>
        

        <span class="kw">One</span> can use R fucntion <span class="kw">&lt;</span>colors()<span class="kw">&gt;</span> <span class="kw">to</span> list all available colors.
    <span class="kw">-s</span>  Scale y axis
        [<span class="kw">Default</span> null. Accept TRUE. This function is depleted. 
        <span class="kw">But</span> if the supplied number after -S is not 0, this parameter will be set to TRUE]
    <span class="kw">-F</span>  The formula for facets.[Default no facets, 
        <span class="st">&quot;+facet_grid(level ~ .)&quot;</span> <span class="kw">means</span> divide by levels of <span class="st">&#39;level&#39;</span> vertically.
        <span class="st">&quot;+facet_grid(. ~ level)&quot;</span> <span class="kw">means</span> divide by levels of <span class="st">&#39;level&#39;</span> horizontally.
        <span class="st">&quot;+facet_grid(lev1 ~ lev2)&quot;</span> <span class="kw">means</span> divide by lev1 vertically and lev2 horizontally.
        <span class="st">&quot;+facet_wrap(~level, ncol=2)&quot;</span> <span class="kw">means</span> wrap horizontally with 2 columns.

        <span class="co">#Pay attention to the single quote for parameters in function for scale.</span>
        <span class="kw">Example</span>: <span class="st">&quot;+facet_wrap(~Size,ncol=6,scale=&#39;free&#39;)&quot;</span>
        <span class="kw">Example</span>: <span class="st">&quot;+facet_grid(Size ~ .,scale=&#39;free_y&#39;)&quot;</span>
        ]
    <span class="kw">-G</span>  If facet is given, you may want to specifize the order of
        <span class="kw">variable</span> in your facet, default alphabetical order.
        [<span class="kw">Accept</span> sth like (one level one sentence, separate by<span class="st">&#39;;&#39;</span>) 
        <span class="st">&#39;data$size &lt;- factor(data$size, levels=c(&quot;l1&quot;, &quot;l2&quot;,...,&quot;l10&quot;), ordered=T)&#39;</span> ]
    <span class="kw">-v</span>  If scale is TRUE, give the following <span class="st">&#39;scale_y_log10()&#39;</span>[default], <span class="st">&#39;coord_trans(y=&quot;log10&quot;)&#39;</span>, 
        <span class="kw">or</span> other legal command for ggplot2 or simply <span class="st">&#39;log2&#39;</span>.]
    <span class="kw">-S</span>  A number to add if scale is used.
        [<span class="kw">Default</span> 0. If a non-zero number is given, -s would be set to TRUE.]    
    <span class="kw">-p</span>  Other legal R codes for gggplot2 could be given here.
        [<span class="kw">Begin</span> with <span class="st">&#39;+&#39;</span> ]
    <span class="kw">-w</span>  The width of output picture (cm)<span class="kw">.</span>[Default 20]
    <span class="kw">-u</span>  The height of output picture (cm)<span class="kw">.</span>[Default 12] 
    <span class="kw">-E</span>  The type of output figures.[Default pdf, accept
        <span class="kw">eps</span>/<span class="kw">ps</span>, tex (pictex), <span class="kw">png</span>, jpeg, tiff, bmp, svg and wmf)]
    <span class="kw">-r</span>  The resolution of output picture.[Default 300 ppi]
    <span class="kw">-z</span>  Is there a header. Must be TRUE. [Default TRUE]
    <span class="kw">-e</span>  Execute or not[Default TRUE]
    <span class="kw">-i</span>  Install depended packages[Default FALSE]</code></pre></div>
</div>
<div id="section-3.16.3" class="section level3">
<h3><span class="header-section-number">3.16.3</span> 一网打进散点图绘制</h3>
<p>假如有一个输入数据如下所示(存储于文件scatterplot.xls中)</p>
<pre><code>Samp    Gene1   Gene2   Color   Size    GC_quality  Base_quality
a   1   1   grp1    10  PASS    PASS
b   2   2   grp1    10  PASS    PASS
c   1   3   grp1    10  WARN    PASS
d   3   1   grp2    15  WARN    WARN
e   2   2   grp2    15  PASS    WARN
f   3   3   grp3    5   PASS    PASS
g   2   1   grp3    5   WARN    PASS</code></pre>
<p>想绘制样品在这两个Gene为轴的空间的分布，并标记样品的属性，只需要运行如下命令</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># -f: 指定输入文件，列数不限，顺序不限; 第一行为列名字，第一列无特殊要求，必选</span>
<span class="co"># -X: 指定哪一列为X轴信息，必选</span>
<span class="co"># -Y: 指定哪一列为Y轴信息，必选</span>
<span class="co"># -c: 指定用哪一列标记颜色，可选</span>
<span class="co"># -s: 指定哪一列标记大小，一般为数字列，可选</span>
<span class="co"># -S: 指定哪一列标记形状，可选</span>
<span class="co"># -L: 指定哪一列用来作为文本标记</span>
<span class="co"># -w, -u: 指定图的长宽</span>
<span class="kw">sp_scatterplot2.sh</span> -f scatterplot.xls -X Gene1 -Y Gene2 -c Color -s Size \
        -S GC_quality -L Samp -w 10 -u 10</code></pre></div>
<p><img src="images/scatterplot6.png" width="100%" style="display: block; margin: auto;" /></p>
<p>如果横纵轴为字符串，且有重复, 则需指定参数<code>-J TRUE</code>以错开重叠的点，具体如下</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># -O: 指定X轴变量的顺序, 默认是字母顺序</span>
<span class="co"># 其它列或其它属性的顺序也可以用相应的方式指示，具体看程序的帮助提示</span>
<span class="co"># -c Gene1: 用特定基因的表达对点着色，单细胞分析图中常用</span>
<span class="co"># -J TRUE: 见上</span>
<span class="co"># -Z FALSE：默认使用geom_text_repel添加点的标记，及其智能，不会出现标签过多覆盖的情况</span>
<span class="co"># 但对jitterplot，会有些冲突，所以在`-J TRUE`且出来的图中点的标签不符合预期时，设定</span>
<span class="co"># 次参数为FALSE，使用geom_text标记点。</span>

<span class="kw">sp_scatterplot2.sh</span> -f scatterplot.xls -X GC_quality -Y Base_quality \
        -O <span class="st">&quot;&#39;WARN&#39;, &#39;PASS&#39;&quot;</span> -c Gene1 -w 10 -u 10 -J TRUE -L Samp -Z FALSE</code></pre></div>
<p><img src="images/scatterplot7.png" width="100%" style="display: block; margin: auto;" /></p>
<p>只有想不到，没有做不到，<code>sp_scatterplot2.sh</code>还可以完成更多你想做的散点图，而且只需调参数，无需改代码，简单可重用。</p>
</div>
</div>
<div id="-1" class="section level2">
<h2><span class="header-section-number">3.17</span> 参考资料</h2>
<ul>
<li><a href="http://rpubs.com/xuefliang/153247" class="uri">http://rpubs.com/xuefliang/153247</a></li>
<li><a href="http://www.sthda.com/english/wiki/survminer-r-package-survival-data-analysis-and-visualization" class="uri">http://www.sthda.com/english/wiki/survminer-r-package-survival-data-analysis-and-visualization</a></li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="Rbasic.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="section-4.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["R_course.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
